<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Statistical methods for epigenetics</title>
    <meta charset="utf-8" />
    <meta name="author" content="  Keegan Korthauer " />
    <meta name="date" content="2022-03-09" />
    <script src="libs/header-attrs-2.11/header-attrs.js"></script>
    <link rel="stylesheet" href="xaringan-themer.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# <big>Statistical methods for epigenetics</big>
### <br><font size=6> Keegan Korthauer </font>
### <font size=6>9 March 2022 </font>

---






&lt;style type="text/css"&gt;
pre {
  white-space: pre-wrap;
}
.remark-code {
  background: #f8f8f8;
}
.remark-inline-code {
  background: "white";
}
.remark-code {
  font-size: 22px;
}
.huge .remark-code { /*Change made here*/
  font-size: 200% !important;
}
.tiny .remark-code { /*Change made here*/
  font-size: 60% !important;
}
.smaller .remark-code { /*Change made here*/
  font-size: 90% !important;
}
.smaller2 .remark-code { /*Change made here*/
  font-size: 80% !important;
}
.smaller3 .remark-code { /*Change made here*/
  font-size: 70% !important;
}
&lt;/style&gt;

&lt;style&gt;
div.blue { background-color:#e8f2f6; border-radius: 5px; padding: 20px;}
&lt;/style&gt;

# Learning objectives for today

* Be familiar with the types of epigenetic measurements that involve **"peak-calling"**, and what a peak represents in each

* Understand the main statistical test performed by the widely used peak-caller MACS2

* Explain the similarities and differences (and understand their implications) between analysis of **differential expression** and **differential methylation**, with regard to:

  * continuous models of microarray data vs count models of sequencing data (similarity)
  
  * methylation level represents a proportion, expression level unbounded (difference)
  
  * grouping of sites into differentially methylated regions (difference)

* We won't have time to cover preprocessing / normalization steps

---

# Epigenetics

&lt;img src="https://i1.abocdn.com/ca9b7843e0579ece3b87bf220d1672f202b00ac9/resources/images/blog/epigenetics-schema.jpeg" width="1000" style="display: block; margin: auto;" /&gt;

&lt;small&gt;[Image source](https://www.antibodies-online.com/areas/epigenetics/)

Recall basics of DNA-binding and DNA methylation assays from [lecture 2](https://github.com/STAT540-UBC/lectures/raw/main/lect02-methods.pdf)

---

# Open chromatin &amp; nucleosome positioning

&lt;img src="img/atac.jpeg" width="800" style="display: block; margin: auto;" /&gt;

&lt;small&gt;[Image source](https://rockefelleruniversity.github.io/RU_ATAC_Workshop.html)

---

# Peak calling 

* Sequencing after one of these selection/digestion techniques (IP, DNase, MNase, transposase) generally requires a peak-finding analysis

* One of the most widely used methods is [MACS2](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2008-9-9-r137):

&lt;img src="img/macs.png" width="600" style="display: block; margin: auto;" /&gt;

---

# MACS2 overview

.pull-left[
* Shift reads to remove strand asymmetry: estimate `\(d\)` from 1,000 regions
&lt;img src="https://hbctraining.github.io/Intro-to-ChIPseq/img/plos_chipseq_arrow.png" width="700" style="display: block; margin: auto;" /&gt;
]
.pull-right[
* Let `\(\lambda_{BG}\)` be the expected number of reads in each window
* Let `\(\lambda_{local} = max(\lambda_{BG}, \lambda_{1K}, \lambda_{5K}, \lambda_{10K})\)` be the expected number of reads in a window
* P-value for `\(k\)` observed reads = `\(e^{-\lambda_{local}} \sum_{i=0}^k \frac{\lambda_{local}^i}{i!}\)`  
&lt;img src="https://hbctraining.github.io/Intro-to-ChIPseq/img/lambda.png" width="450" style="display: block; margin: auto;" /&gt;
]

---

# DNA Methylation (5mC) 

- chemical modification to an individual DNA base (C = Cytosine), referred to as 5mC (5-methylcytosine)
- doesn't change coding sequence
- modification can happen at any C, but in mammals occurs almost* exclusively at CpGs 

&lt;img src="./img/5mC.jpg" width="375px" style="display: block; margin: auto;" /&gt;

&lt;small&gt;
image source: https://www.epigentek.com
&lt;p&gt;
*other contexts observed in neuronal cells

---

# DNA Methylation at CpGs

&lt;img src="./img/5thbase.jpg" width="750px" style="display: block; margin: auto;" /&gt;

~30 Million CpGs in human

---

# The 5th base?

* heritable upon cell division 
     - 5mC added to complementary strand by methyltransferases
* dynamic
     - changes during development and with age
     - can be altered by environmental exposure

&lt;img src="./img/copymeth.jpg" width="725px" style="display: block; margin: auto;" /&gt;

&lt;small&gt;&lt;small&gt;
image source: https://slideplayer.com/slide/10719924/

---

# Role in gene regulation 

&lt;img src="./img/genereg.jpg" width="720px" style="display: block; margin: auto;" /&gt;

---

# Evolution of methylation assays

&lt;img src="./img/methassays.jpg" width="670px" style="display: block; margin: auto;" /&gt;

&lt;small&gt;
Laird 2010 (https://doi.org/10.1038/nrg2732)

---

## Distinguishing methylated and unmethylated sites

.pull-left[
#### 1. Methylated DNA immunoprecipitation (MeDIP)
* treat fragmented DNA with antibody targeting 5mC
* wash away unbound DNA
* measure extent to which each cytosine is still present (methylated) with **microarray or sequencing**

&lt;small&gt; image source: https://en.wikipedia.org/wiki/Methylated_DNA_immunoprecipitation
]
.pull-right[
&lt;img src="./img/medip.jpg" width="300px" style="display: block; margin: auto;" /&gt;
]

---

## Distinguishing methylated and unmethylated sites

.pull-left[
#### 2. Bisulfite conversion
* 5mC `\(\rightarrow\)` stays 5mC
* C (not methylated) `\(\rightarrow\)` converted to U
* Sequencing:
    * 5mC read as C
    * C read as T 
* For each C measure fraction of signal corresponding to C vs T with **microarray or sequencing** 
* *Note that can't distinguish between C `\(\rightarrow\)` T mutation and unmethylated C*
]
.pull-right[
&lt;img src="./img/bisulfite-conversion.jpg" width="450px" style="display: block; margin: auto;" /&gt;

&lt;small&gt; image: [https://www.diagenode.com](https://www.diagenode.com/en/categories/bisulfite-conversion)
]

---

## Analysis of MeDIP

.pull-left[
* MeDIP targets **only** methylated DNA
* **MeDIP-chip**: much like gene expression microarrays
* **MeDIP-seq**: differential abundance of counts (similar to RNA-seq)
  * Complication: instead of genes or transcripts, need to define genomic "bins" to count reads over 
  * One option is to use sliding windows: [MEDIPS R package](http://bioconductor.org/packages/MEDIPS/) performs preprocessing and implements `edgeR` to test for differential abundance of each window

&lt;br&gt;
&lt;small&gt;
image source: https://en.wikipedia.org/wiki/Methylated_DNA_immunoprecipitation
]

.pull-right[
&lt;img src="./img/medip.jpg" width="300px" style="display: block; margin: auto;" /&gt;
]

---

## Bisulfite conversion arrays (e.g. Infinum 450K, Epic 850K)

.pull-left[
* Most commonly used human methylation arrays use bisulfite conversion
  * 450K / 850K refer to the number of sites probed
* Analysis has similarities to expression microarray analysis
  * Signal is **continuous** (intensity)
  * Moderate throughput: 1-16 samples per array
* Difference: focus on **relative** intensity of methylated vs unmethylated binding affinity
]

.pull-right[

&lt;img src="./img/microarray.jpg" width="450px" style="display: block; margin: auto;" /&gt;

&lt;small&gt;
image: https://en.wikipedia.org/wiki/Illumina_Methylation_Assay
]
    
---

class: middle

## Beta values and M values

Need some way to summarize the two relative measures for each probe:
1. methylated intensity
2. unmethylated intensity

---

## Beta `\((\beta)\)` values 

* Combine two measurements for each probe, and measure level of methylation: `$$\beta = \frac{\text{methylated intensity}}{(\text{methylated intensity} + \text{unmeth intensity})}$$`
* Range from 0 to 1
* Easily interpretable as a methylation proportion
* Best used for visualization

---

## M-values

* Transformed `\(\beta\)` values: `$$M= log_2\big(\frac{\text{methylated intensity}}{\text{unmethylated intensity}}\big) = log_2(\beta/(1-\beta))$$`
* This type of transformation is referred to as the *logit*
  * maps a proportion onto real line
* alleviate heteroscedasticity of `\(\beta\)` values (recall that Binomial(n,p) variance = np(1-p))
   
&lt;img src="./img/betavsmvals.png" width="650px" style="display: block; margin: auto;" /&gt;

&lt;small&gt;
[Du et al. (2010)](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-11-587)
    
---

# Methylation array analysis

* Analysis strategy: linear regression on M-values 
* R Packages: 
   * [minfi](http://bioconductor.org/packages/minfi/) for preprocessing and normalization
   * [limma](http://bioconductor.org/packages/limma/) for individual cytosine analysis
   * [bumphunter](http://bioconductor.org/packages/bumphunter/) for region-level analysis
   * Workflow that combines these: [methylationArrayAnalysis]( 	https://bioconductor.org/packages/methylationArrayAnalysis/) - explore in Seminar 10
    
---

## Example - DNA Methylation in sorted T cells

Data from [Zhang et al. (2013)](https://www.ncbi.nlm.nih.gov/pubmed/23974203)


.smaller[

```r
mSet
```

```
## class: GenomicRatioSet 
## dim: 439918 10 
## metadata(0):
## assays(2): M CN
## rownames(439918): cg13869341 cg24669183 ... cg08265308 cg14273923
## rowData names(0):
## colnames(10): naive.1 rTreg.2 ... act_naive.9 act_rTreg.10
## colData names(13): Sample_Name Sample_Well ... yMed predictedSex
## Annotation
##   array: IlluminaHumanMethylation450k
##   annotation: ilmn12.hg19
## Preprocessing
##   Method: Raw (no normalization or bg correction)
##   minfi version: 1.33.0
##   Manifest version: 0.4.0
```
]
---

## Meta data values


```r
head(targets[c(1:5)])
```

```
##   Sample_Name Sample_Well Sample_Source Sample_Group Sample_Label
## 1           1          A1           M28        naive        naive
## 2           2          B1           M28        rTreg        rTreg
## 3           3          C1           M28    act_naive    act_naive
## 4           4          D1           M29        naive        naive
## 5           5          E1           M29    act_naive    act_naive
## 6           6          F1           M29    act_rTreg    act_rTreg
```

```r
table(targets$Sample_Label)
```

```
## 
## act_naive act_rTreg     naive     rTreg 
##         3         2         3         2
```

---

## Beta values


```r
bVals &lt;- getBeta(mSet)
dim(bVals)
```

```
## [1] 439918     10
```

```r
head(bVals[,1:5])
```

```
##               naive.1    rTreg.2 act_naive.3    naive.4 act_naive.5
## cg13869341 0.84267937 0.85118462   0.8177504 0.82987650  0.81186174
## cg24669183 0.81812908 0.82489238   0.8293297 0.75610281  0.81967323
## cg15560884 0.77219626 0.74903910   0.7516263 0.77417882  0.77266205
## cg01014490 0.08098986 0.06590459   0.0233755 0.04127262  0.04842397
## cg17505339 0.89439216 0.93822870   0.9471357 0.90520570  0.92641305
## cg11954957 0.74495496 0.79008516   0.7681146 0.84450764  0.75431167
```

---

## M values


```r
mVals &lt;- getM(mSet)
dim(mVals)
```

```
## [1] 439918     10
```

```r
head(mVals[,1:5])
```

```
##              naive.1   rTreg.2 act_naive.3   naive.4 act_naive.5
## cg13869341  2.421276  2.515948    2.165745  2.286314    2.109441
## cg24669183  2.169414  2.235964    2.280734  1.632309    2.184435
## cg15560884  1.761176  1.577578    1.597503  1.777486    1.764999
## cg01014490 -3.504268 -3.825119   -5.384735 -4.537864   -4.296526
## cg17505339  3.082191  3.924931    4.163206  3.255373    3.654134
## cg11954957  1.546401  1.912204    1.727910  2.441267    1.618331
```

---

## Comparing beta and M values

&lt;img src="lect16-methylation_files/figure-html/unnamed-chunk-23-1.png" width="1080" /&gt;
---

## Comparing beta and M values

&lt;img src="lect16-methylation_files/figure-html/unnamed-chunk-24-1.png" width="360" style="display: block; margin: auto;" /&gt;

--

#### Why do we have beta values far from 0 or 1 if methylation is either present or not at each CpG?

---

## Differentially methylated cytosines (DMC)

#### Which cytosines/probes are differentially methylated between naive &amp; activated naive T cells?

.pull-left[
.tiny[

```r
# this is the factor of interest
cellType &lt;- factor(targets$Sample_Group)
# this is the individual effect that we need to account for
individual &lt;- factor(targets$Sample_Source) 

# use the above to create a design matrix
design &lt;- model.matrix(~cellType+individual, 
                       data=targets)
colnames(design) &lt;- c(levels(cellType),
                      levels(individual)[-1])
```
]]
.pull-right[
.tiny[

```r
design
```

```
##    act_naive act_rTreg naive rTreg M29 M30
## 1          1         0     1     0   0   0
## 2          1         0     0     1   0   0
## 3          1         0     0     0   0   0
## 4          1         0     1     0   1   0
## 5          1         0     0     0   1   0
## 6          1         1     0     0   1   0
## 7          1         0     1     0   0   1
## 8          1         0     0     1   0   1
## 9          1         0     0     0   0   1
## 10         1         1     0     0   0   1
## attr(,"assign")
## [1] 0 1 1 1 2 2
## attr(,"contrasts")
## attr(,"contrasts")$cellType
## [1] "contr.treatment"
## 
## attr(,"contrasts")$individual
## [1] "contr.treatment"
```
]]

---

## Differential methylation: limma on M-values

Here, the reference is `act_naive` cells


```r
# fit the linear model (one for each probe)
mfit &lt;- lmFit(mVals, design)
cfit &lt;- eBayes(mfit)

# look at top DM CpGs between naive and ref (act_naive)
topTable(cfit, coef="naive")
```

```
##                logFC     AveExpr         t      P.Value   adj.P.Val        B
## cg15459165 -3.892074 -1.47895212 -22.09541 1.993958e-08 0.002793543 6.810420
## cg19704755 -4.139885  0.09100138 -20.35779 3.787335e-08 0.002793543 6.594795
## cg13531460  3.484448 -0.63132888  20.23020 3.978221e-08 0.002793543 6.577256
## cg17137500 -3.007052 -2.47635924 -20.21569 4.000601e-08 0.002793543 6.575245
## cg17048073 -3.650550 -0.87247048 -20.04737 4.270889e-08 0.002793543 6.551674
## cg03899643  4.603864  0.02760746  19.86354 4.589787e-08 0.002793543 6.525407
## cg19827923  3.316270  0.15932891  19.82438 4.661126e-08 0.002793543 6.519739
## cg26800893 -3.080184 -2.39958457 -19.60718 5.080116e-08 0.002793543 6.487839
## cg14688905 -3.119261 -1.31909419 -18.94774 6.635412e-08 0.003243375 6.385901
## cg21837189 -3.717982 -1.08764112 -18.00052 9.895908e-08 0.003769603 6.224845
```

---

## Summary for all coefficients 

Significant tests (BH-adjusted p-value less than 0.05) of the null hypotheses: each coefficient equal to zero


```r
# look at the numbers of DM CpGs at FDR &lt; 0.05
summary(decideTests(cfit))
```

```
##        act_naive act_rTreg  naive  rTreg    M29    M30
## Down      184747       919    400    827   1367   1245
## NotSig     14791    438440 439291 438570 437499 437716
## Up        240380       559    227    521   1052    957
```

Note that the first column tests whether the reference group (`act_naive`) has an M value of zero (not usually of interest)

---

## Interpretation of parameters and effect sizes

* Our previous model gives parameter estimates for differences in M values `\(\alpha\)`
* To interpret on the scale of methylation proportions, need to back-transform:
  * Recall: `\(M= log_2\big(\frac{\text{methylated intensity}}{\text{unmethylated intensity}}\big) = log_2(\beta/(1-\beta))\)`
 * Solving for beta gives `\(\beta = \frac{2^M}{(1+2^M)}\)`



```r
(M &lt;- cfit$coefficients["cg15459165", "naive"])
```

```
## [1] -3.892074
```

```r
2^M/(1+2^M)
```

```
## [1] 0.06310445
```
---

## Summary of array DMC analysis

  
* **Just like for limma analysis of gene expression microarray**: Can use additive models, models with interactions, continuous and categorical variables

  * **interpretation of hypothesis tests, contrasts** with `makeContrasts` (for other comparisons not represented by coefficients in our design matrix): follows as before
  
  * **interpretation of parameters/effect sizes**: keep in mind that M-values are *transformed* proportions

--

* Annotation packages (e.g. `IlluminaHumanMethylation450kanno.ilmn12.hg19`) allow for adding information about each CpG 
  * chromosome and position
  * strand
  * whether in CpG Island

---

## Bisulfite sequencing

* Sequencing of bisulfite-converted DNA

* **WGBS** = whole-genome bisulfite sequencing

* **RRBS** = reduced representation bisulfite sequening (targets similar regions to methylation arrays e.g. islands)

* Pros: base-pair resolution; WGBS measures all CpGs

* Con: expensive (WGBS cost is on the order of whole genome DNA sequencing)

 
---

## Bisulfite sequencing

&lt;img src="./img/wgbscounts2.jpg" width="1000px" style="display: block; margin: auto;" /&gt;

---

## Analysis of bisulfite sequencing

.pull-left[
* BS assays (e.g. WGBS, RRBS) yield base-level counts of methylated `\((M \text{counts})\)` and ununmethylated `\((U \text{counts})\)` reads

* Differential methylation analysis quantifies significance of differences in proportion of methylated reads `\((\frac{M}{M+U})\)`

* R Packages: 
    * [bsseq](http://bioconductor.org/packages/bsseq/): "BSmooth" - smoothed t-test
    * [dss](http://bioconductor.org/packages/dss/): approximate beta-binomial regression
    * [dmrseq](http://bioconductor.org/packages/dmrseq/): approximate beta-binomial regression over regions
]
.pull-right[
&lt;img src="./img/bsseq.jpg" width="200px" style="display: block; margin: auto;" /&gt;&lt;img src="./img/dmrseqhex.jpg" width="200px" style="display: block; margin: auto;" /&gt;
]
---

## BSmooth: t-test on smoothed methylation proportions

* **Key idea:** methylation levels of nearby CpGs are correlated

  * use locally weighted polynomial regression values (lowess) to obtain smoothed estimate of methylation difference 
  
  * similarly, obtain smoothed location-dependent standard deviation  
  
* Compute t-test using smoothed quantities


&lt;img src="img/bsmooth.jpg" width="900px" style="display: block; margin: auto;" /&gt;
&lt;small&gt;[Hansen, et al. 2012](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2012-13-10-r83#Sec8)
  - Red line: 35x coverage
  - Black line: subsampled to 5x coverage

---

## Modeling count nature of bisulfite reads 

* For cytosine `\(i\)` and sample `\(j\)` in condition `\(s\)`, we have:

  * `\(M_{ij}\)` reads corresponding to methylation

  * `\(N_{ij}\)` total reads 

* Let `\(p_s\)` be the methylation probability in condition `\(s\)`

--

* Let `\(M_{ij} \sim Binom(N_{ij}, p_{s})\)` for `\(j \in s\)`

--

* How to test whether `\(p_1 = p_2\)` (and optionally adjust for covariates)?

---

## Binomial (Logisitic) regression

- Generalized linear model for probability of success `\(p\)`: `$$logit(p)=log\Big(\frac{p}{1-p}\Big) = \boldsymbol{X\beta}$$`
--

- Link function `\(g(p) = log\big(\frac{p}{1-p}\big)\)` describes relationship between mean parameter of response and linear predictor

--

- No closed form; fit with iterative ML estimation

--

- Interpret coefficients on original scale with inverse link function `\(g^{-1}\)`: $$ p = \frac{e^{\boldsymbol{X\beta}}}{1+e^{\boldsymbol{X\beta}}} $$

--


- In R: `glm(cbind(successes, failures) ~ x, family="binomial")`

---

## Other link functions for GLM

&lt;img src="./img/link.jpg" width="300px" style="display: block; margin: auto;" /&gt;
Thiele &amp; Markussen 2012 (https://doi.org/10.1079/PAVSNNR20127016)

---

## Why logit transformation?

&lt;img src="lect16-methylation_files/figure-html/unnamed-chunk-34-1.png" width="324" style="display: block; margin: auto;" /&gt;

* Same idea as M-value (though M-value uses log base 2 instead of natural log by convention):
    * Transforms methylation proportions to the real line
    * Mitigate heteroscedasticity

---

## Binomial regression with overdispersion

- In ordinary logistic regression, `\(p\)` is assumed to be constant for all samples with in group

--

- To model *overdispersion* (additional biological variability), allow this quantity to vary

--

- For example, let `$$p \sim Beta(\alpha, \beta)$$` `$$M \sim Binom(N, p)$$`

--

- This is the **Beta-Binomial** distribution

  - Mean of M `\(= \frac{n\alpha}{\alpha + \beta} = n\pi\)`
  
  - Variance of M `\(= \pi(1-\pi)(1 + (N - 1)\phi)\)` 

  - where `\(\phi\)` is the overdispersion parameter: `\(\phi = \frac{1}{\alpha + \beta + 1}\)`

--

- In R: `aod::betabin(cbind(successes,failures) ~ x, ~ x)`
 
---


## Pitfalls of binomial regression 

- What happens to logit link function `\(log\big(\frac{p}{1-p}\big)\)` if `\(p=0\)` or `\(1\)`?

--
    - binomial regression unstable for fully methylated or unmethylated cytosines
    
--

- Computationally intensive to fit model at every cytosine

--

- DSS: Park &amp; Wu 2016 (https://doi.org/10.1093/bioinformatics/btw026)
    - Differential methylation under general experimental design
    - Alternate link function: `\(arcsine(2p-1)\)`
    - Approximate fitting with Generalized Least Squares (GLS) 

---


## Generalized Least Squares (GLS) in a nutshell

- Hybrid of **linear regression** and **generalized linear model**

- **Pro**: stable &amp; closed form estimates (fast)

- **Con**: approximate

- **Key idea**: flexible covariance structure allows for specification of approximate beta-binomial error

---

## DMCs vs DMRs

* Depending on the biological question, it might not make sense to only test *individual* CpGs (DMCs)

* If instead differences occur across groups of nearby CpGs, we call them differentially methylated regions (DMRs)

  * Methylation of nearby CpGs is correlated
  
  * Individual CpGs are not thought to act independently (e.g. are regulated in tandem)

---


## Individual cytosine differences

&lt;img src="./img/dmc1.jpg" width="900px" /&gt;

---


## CpG 1 

&lt;img src="./img/dmc2.jpg" width="800px" /&gt;

---


## CpG 2 

&lt;img src="./img/dmc3.jpg" width="800px" /&gt;

---

## Previous approaches: grouping significant CpGs

&lt;img src="./img/fdr2.jpg" width="900px" style="display: block; margin: auto;" /&gt;

Examples: BSmooth ([Hansen et al. 2012](https://doi.org/10.1186/gb-2012-13-10-r83)), DSS ([Park et al. 2016](https://doi.org/10.1093/bioinformatics/btw026))

---


## FDR at the region level

&lt;img src="./img/fdr3.jpg" width="900px" style="display: block; margin: auto;" /&gt;

$$ \text{False Discovery Rate (FDR)} = E\Big[\frac{FP}{FP + TP}\Big]$$
--

- `\(FDR_{CpG} = 2/10 = 0.20\)`
--

- `\(FDR_{DMR} = 1/2 = 0.50\)`

---


## Accurate inference of DMRs

- Korthauer et al. 2019 (https://doi.org/10.1093/biostatistics/kxy007)

- Key ideas: 
    - model methylation signal over region to account for correlated signal
    - permutation to acheive accurate FDR control
    
- Implemented in the [dmrseq](http://bioconductor.org/packages/dmrseq/) Bioconductor package
    
&lt;img src="./img/dmrseqhex.jpg" width="200px" style="display: block; margin: auto;" /&gt;

---

## dmrseq: 2-stage approach

&lt;img src="./img/2stage.jpg" width="800px" style="display: block; margin: auto;" /&gt;

---

## dmrseq output

&lt;img src="./img/exdmr.jpg" width="800px" style="display: block; margin: auto;" /&gt;

- *de novo* regions with FDR estimate
- ranked by effect size of methylation difference
- can adjust for covariates 

---

## What's changing: methylation or cell type composition?

&lt;img src="./img/celltypecomp2.jpg" width="750px" style="display: block; margin: auto;" /&gt;

&lt;small&gt; image: [Jaffe &amp; Irizarry 2014](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r31)

---

## Cell type composition confounding

&lt;img src="./img/celltypecomp.jpg" width="600px" style="display: block; margin: auto;" /&gt;

&lt;small&gt; image: [methylationArrayAnalysis](https://bioconductor.org/packages/methylationArrayAnalysis/)

Can estimate cell type composition with models built on purified cell type methylation data (not always available)
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false,
"ratio": "16:9"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
