<!DOCTYPE html>
<html lang="en"><head>
<script src="lec14-clustering-guest_files/libs/clipboard/clipboard.min.js"></script>
<script src="lec14-clustering-guest_files/libs/quarto-html/tabby.min.js"></script>
<script src="lec14-clustering-guest_files/libs/quarto-html/popper.min.js"></script>
<script src="lec14-clustering-guest_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="lec14-clustering-guest_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="lec14-clustering-guest_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="lec14-clustering-guest_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.2.269">

  <meta name="author" content="Lucy Gao (w/ slide support from Keegan)">
  <meta name="dcterms.date" content="2023-03-14">
  <title>Unsupervised learning in biological data analysis: clustering</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="lec14-clustering-guest_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="lec14-clustering-guest_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1.6em;
      vertical-align: middle;
    }
  </style>
  <link rel="stylesheet" href="lec14-clustering-guest_files/libs/revealjs/dist/theme/quarto.css" id="theme">
  <link href="lec14-clustering-guest_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="lec14-clustering-guest_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="lec14-clustering-guest_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="lec14-clustering-guest_files/libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="lec14-clustering-guest_files/libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="lec14-clustering-guest_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-captioned.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-captioned) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-captioned.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-captioned .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-captioned .callout-caption  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-captioned.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-captioned.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-caption {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-caption {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-caption {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-captioned .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-captioned .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-captioned) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-caption {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-caption {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-caption {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-caption {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-caption {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" data-background-color="#197aa0" data-background-image="https://github.com/STAT540-UBC/stat540-ubc.github.io/raw/main/images/stat540-logo-s.png" data-background-opacity="0.9" data-background-position="95% 90%" data-background-size="12%" class="quarto-title-block center">
  <h1 class="title">Unsupervised learning in biological data analysis: clustering</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Lucy Gao (w/ slide support from Keegan) 
</div>
</div>
</div>

  <p class="date">March 14, 2023</p>
</section>
<section id="hello" class="slide level2">
<h2>Hello!</h2>

<img data-src="img/me.jpeg" width="300" class="r-stretch quarto-figure-center"><ul>
<li>My background: biostatistics by way of math/stats</li>
<li>Unsupervised learning researcher since 2015 (has involved lots of clustering …)</li>
</ul>
</section>
<section id="learning-objectives" class="slide level2">
<h2>Learning objectives</h2>
<div>
<ul>
<li class="fragment"><p>Understand the unsupervised learning setting</p></li>
<li class="fragment"><p>Understand why clustering is powerful, but hard to apply in biological data analysis</p></li>
<li class="fragment"><p>Understand high-level principles behind two well-known clustering algorithms:</p>
<ul>
<li class="fragment">k-means clustering</li>
<li class="fragment">hierarchical clustering</li>
</ul></li>
<li class="fragment"><p>Understand the challenges of doing statistical inference downstream of clustering</p></li>
</ul>
</div>
</section>
<section id="unsupervised-learning-data-setting" class="slide level2">
<h2>Unsupervised learning data setting</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="lec14-clustering-guest_files/figure-revealjs/unnamed-chunk-3-1.png" width="384"></p>
</figure>
</div>
</div>
</div>
<p>NCI-60 Human Tumour Cell Lines microarray data (~7000 genes, only 100 visualized)</p>
</div><div class="column" style="width:50%;">
<p><br></p>
<div>
<ul>
<li class="fragment">We have access to <span class="math inline">\(m\)</span> features <span class="math inline">\(X_1, \ldots, X_m\)</span> measured on <span class="math inline">\(n\)</span> samples</li>
<li class="fragment">Represent as a matrix <span class="math inline">\(X\)</span> with <span class="math inline">\(m\)</span> rows (number of features) and <span class="math inline">\(n\)</span> columns (number of samples)</li>
<li class="fragment"><strong>IMPORTANTLY</strong>: we don’t have access to <em>anything</em> other than this data matrix</li>
</ul>
</div>
</div>
</div>
</section>
<section id="what-we-cant-do-in-the-unsupervised-data-setting" class="slide level2">
<h2>What we can’t do in the unsupervised data setting</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="lec14-clustering-guest_files/figure-revealjs/unnamed-chunk-4-1.png" width="384"></p>
</figure>
</div>
</div>
</div>
<p>NCI-60 Human Tumour Cell Lines microarray data (~7000 genes, only 200 visualized)</p>
</div><div class="column" style="width:50%;">
<div>
<ul>
<li class="fragment">We don’t know the experiment design, sample phenotype, batch, etc.</li>
<li class="fragment">This greatly limits what we can do:
<ul>
<li class="fragment">We can’t quantify the association between <span class="math inline">\(X_1, \ldots, X_p\)</span> and a response variable <span class="math inline">\(Y\)</span></li>
<li class="fragment">We can’t use <span class="math inline">\(X\)</span> to predict a response variable <span class="math inline">\(Y\)</span></li>
<li class="fragment">We can’t (directly) compare populations represented in <span class="math inline">\(X\)</span> with two-sample testing</li>
</ul></li>
</ul>
</div>
</div>
</div>
</section>
<section id="what-we-can-do-in-this-data-setting" class="slide level2">
<h2>What we can do in this data setting</h2>
<h3 id="discover-interesting-things-about-the-measurements-we-have">Discover “interesting things” about the measurements we have</h3>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="lec14-clustering-guest_files/figure-revealjs/unnamed-chunk-5-1.png" width="384"></p>
</figure>
</div>
</div>
</div>
<p>NCI-60 Human Tumour Cell Lines data set (~7000 genes, only 200 visualized)</p>
</div><div class="column" style="width:50%;">
<div>
<ul>
<li class="fragment"><p>Each sample’s measurements (column) are a vector in 7000-dimensional space</p></li>
<li class="fragment"><p>Each feature’s measurements (row) are a vector in 64-dimensional space</p></li>
<li class="fragment"><p>Look for possible “hidden structure” within samples and/or features:</p>
<ul>
<li class="fragment"><strong>Dimension reduction</strong>: project samples/features onto lower-dimensional space</li>
<li class="fragment"><strong>Clustering</strong>: group the samples (columns) and/or group the features (rows)</li>
</ul></li>
</ul>
</div>
</div>
</div>
</section>
<section id="clustering-problem-definition" class="slide level2">
<h2>Clustering problem definition:</h2>
<h3 id="break-a-set-of-objects-up-into-groups-so-that-objects-within-each-group-are-more-similar-to-each-other-than-to-objects-in-other-groups">Break a set of objects up into groups, so that objects <strong>within each group</strong> are <strong>more similar to each other</strong> than to objects in <strong>other groups</strong></h3>

<img data-src="img/cassie-cats-clustered.png" width="400" class="r-stretch quarto-figure-center"><p>(Photo and idea credit: Cassie Kozyrkov)</p>
</section>
<section id="similar-how" class="slide level2">
<h2>Similar how?</h2>
<ul>
<li><p>We will define a numerical vector of <strong>attributes</strong> for each object</p></li>
<li><p>For a set of <span class="math inline">\(n\)</span> objects, we will represent each object <span class="math inline">\(i\)</span> <span class="math inline">\((i=1,...,n)\)</span> as a numeric vector <span class="math inline">\(\boldsymbol{x}_i\)</span> of length <span class="math inline">\(p\)</span> (containing <span class="math inline">\(p\)</span> attributes)</p></li>
<li><p>So each object <span class="math inline">\(i\)</span> is represented by <span class="math inline">\(\boldsymbol{x}_i = (x_{i1}, x_{i2}, ..., x_{ip})\)</span></p></li>
</ul>
<p><br></p>
<p>Now we can define the clustering problem more precisely:</p>
<h3 id="break-a-set-of-objects-up-into-groups-so-that-attributes-of-objects-within-each-group-are-more-similar-to-each-other-than-to-attributes-of-objects-in-other-groups">Break a set of objects up into groups, so that <em>attributes</em> of objects within each group are more similar to each other than to <em>attributes</em> of objects in other groups</h3>
</section>
<section id="attribute-similarity" class="slide level2">
<h2>Attribute similarity</h2>
<p>Is this precise enough?</p>
<h3 id="break-a-set-of-objects-up-into-groups-so-that-attributes-of-objects-within-each-group-are-more-similar-to-each-other-than-to-attributes-of-objects-in-other-groups-1">Break a set of objects up into groups, so that <em>attributes</em> of objects within each group are more similar to each other than to <em>attributes</em> of objects in other groups</h3>
<p><br></p>
<div>
<ul>
<li class="fragment"><p>We also need to define a formula to describe similarity between attribute vectors <span class="math inline">\(\boldsymbol{x}_i\)</span></p>
<ul>
<li class="fragment">We call this a measure of <strong>similarity</strong></li>
</ul></li>
<li class="fragment"><p>Equivalently, we can define a formula to describe dissimilarity between attribute vectors <span class="math inline">\(\boldsymbol{x}_i\)</span></p>
<ul>
<li class="fragment">This is convenient because we can use <strong>distance</strong> metrics (e.g.&nbsp;Euclidean, Manhattan)</li>
</ul></li>
</ul>
</div>
</section>
<section id="eg.-define-small-euclidean-distance-to-mean-similar" class="slide level2">
<h2>Eg. define small Euclidean distance to mean “similar”</h2>
<p>Euclidean distance between two vectors of <span class="math inline">\(p\)</span> attributes <span class="math inline">\(\boldsymbol{x}_1\)</span> and <span class="math inline">\(\boldsymbol{x}_2\)</span>:</p>
<p><span class="math display">\[D_{euc}(\boldsymbol{x}_1, \boldsymbol{x}_2) = || \boldsymbol{x}_1 - \boldsymbol{x}_2 ||_2 = \sqrt{\sum_{j=1}^p (x_{1j} - x_{2j})^2}\]</span> In two dimensions:</p>

<img data-src="https://www.tutorialexample.com/wp-content/uploads/2020/05/Euclidean-distance-in-tensorflow.png" width="400" class="r-stretch quarto-figure-center"></section>
<section id="euclidean-distance-should-we-scale" class="slide level2">
<h2>Euclidean distance: should we scale?</h2>
<p>Euclidean distance between two vectors of <span class="math inline">\(p\)</span> attributes <span class="math inline">\(\boldsymbol{x}_1\)</span> and <span class="math inline">\(\boldsymbol{x}_2\)</span>:</p>
<p><span class="math display">\[D_{euc}(\boldsymbol{x}_1, \boldsymbol{x}_2) = || \boldsymbol{x}_1 - \boldsymbol{x}_2 ||_2 = \sqrt{\sum_{j=1}^p (x_{1j} - x_{2j})^2}\]</span></p>
<div>
<ul>
<li class="fragment"><p>If the <span class="math inline">\(p\)</span> attributes have vastly different ranges/units, then some of the attributes will have a way bigger impact on the Euclidean distance than others</p></li>
<li class="fragment"><p>If we think that’s a bad thing, then we could scale each vector of attributes to have standard deviation 1</p></li>
<li class="fragment"><p>But whether or not it’s a bad thing is highly context dependent</p></li>
</ul>
</div>
</section>
<section id="eg.-define-large-pearson-correlation-to-mean-similar" class="slide level2">
<h2>Eg. define large Pearson correlation to mean “similar”</h2>
<p>Pearson correlation between two vectors of <span class="math inline">\(p\)</span> attributes <span class="math inline">\(\boldsymbol{x}_1\)</span> and <span class="math inline">\(\boldsymbol{x}_2\)</span>:</p>
<p><span class="math display">\[r_{\boldsymbol{x}_1, \boldsymbol{x}_{2}} = Cor(\boldsymbol{x}_1, \boldsymbol{x}_{2}) = \frac{Cov(\boldsymbol{x}_1, \boldsymbol{x}_{2})}{\sqrt{Var(\boldsymbol{x}_1)Var(\boldsymbol{x}_{2})}}\]</span></p>
<p>In the plot below, attribute vectors 1 and 2 have large Pearson similarity (they “move together”)</p>

<img data-src="lec14-clustering-guest_files/figure-revealjs/unnamed-chunk-8-1.png" width="960" class="r-stretch quarto-figure-center"></section>
<section id="euclidean-distance-and-pearson-similarity-for-standardized-attributes" class="slide level2">
<h2>Euclidean distance and Pearson similarity for standardized attributes</h2>
<p>If attribute vectors are <strong>standardized</strong> (centre and scaled), then:</p>
<p><span class="math display">\[r_{\boldsymbol{x}_1, \boldsymbol{x}_{2}} = \sum_{j=1}^p x_{1j}x_{2j}\]</span></p>
<p>We can now see that: <span class="math display">\[D_{euc}^2 (\boldsymbol{x}_1, \boldsymbol{x}_{2}) = \sum_{j=1}^p (x_{1j} - x_{2j})^2 = \sum_{j=1}^p x_{1j}^2 + \sum_{j=1}^p x_{2j}^2 + 2\sum_{j=1}^p x_{1j}x_{2j} = 2(1-r_{\boldsymbol{x}_1, \boldsymbol{x}_{2}})\]</span></p>
</section>
<section id="clustering-for-visualization-and-data-exploration" class="slide level2">
<h2>Clustering for visualization and data exploration</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>NCI-60 data: centered and scaled</p>
<p><br></p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="lec14-clustering-guest_files/figure-revealjs/unnamed-chunk-9-1.png" width="384"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<p>Columns and rows clustered and re-ordered</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="lec14-clustering-guest_files/figure-revealjs/unnamed-chunk-10-1.png" width="441"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="what-do-these-clusters-mean" class="slide level2">
<h2>What do these clusters mean?</h2>
<div class="columns">
<div class="column" style="width:60%;">
<h3 id="at-face-value">At face value</h3>
<div>
<ul>
<li class="fragment">Column clusters:
<ul>
<li class="fragment">Samples with similar feature measurements</li>
<li class="fragment"><strong>Specific</strong> definition of “similar” (e.g.&nbsp;small Euclidean distance)</li>
</ul></li>
<li class="fragment">Row clusters:
<ul>
<li class="fragment">Features that move together</li>
<li class="fragment"><strong>Specific</strong> definition of “moving together” (e.g.&nbsp;high Pearson correlation)</li>
</ul></li>
</ul>
</div>
</div><div class="column" style="width:40%;">
<h3 id="speculating-further">Speculating further</h3>
<div>
<ul>
<li class="fragment">Column clusters:
<ul>
<li class="fragment">Samples from different batches?</li>
<li class="fragment">Samples from “real” underlying populations?</li>
</ul></li>
<li class="fragment">Row clusters:
<ul>
<li class="fragment">Features that really work together or belong together?</li>
</ul></li>
</ul>
</div>
</div>
</div>
</section>
<section id="a-famous-success-story-clustering-samples-in-microarrays" class="slide level2">
<h2>A famous success story: clustering samples in microarrays</h2>
<p><br></p>
<p>Key finding of <a href="https://www.pnas.org/doi/full/10.1073/pnas.191367098">Sørlie et al.&nbsp;2001</a>: “Gene expression patterns of breast cancer carcinomas distinguish tumor subclasses with clinical implications”</p>
<div class="columns">
<div class="column" style="width:60%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/sorlie2001-dendrogram.jpeg" width="1200"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:40%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/sorlie2001-survival.jpeg" width="500"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="a-famous-success-story-clustering-genes-in-microarrays" class="slide level2">
<h2>A famous success story: clustering genes in microarrays</h2>
<p>Key finding of <a href="https://www.pnas.org/content/95/25/14863">Eisen, et al.&nbsp;(1998)</a>: “Clustering gene expression data groups together efficiently genes of known similar function”</p>
<p><br></p>
<div class="columns">
<div class="column" style="width:40%;">
<div>
<ul>
<li class="fragment">Columns combine time courses of gene expression in yeast</li>
<li class="fragment">Row/gene clusters are highlighted</li>
<li class="fragment">Genes in clusters have similar functional annotations - involved in common cellular processes</li>
</ul>
</div>
</div><div class="column" style="width:60%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/eisen1998-fig2.jpeg" width="390"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="but-be-cautious-there-are-many-ways-to-not-succeed" class="slide level2">
<h2>But be cautious: there are many ways to not succeed</h2>
<h3 id="there-might-be-very-little-heterogeneity-in-the-data">There might be very little heterogeneity in the data</h3>
<p>2D toy example: clustering pure noise samples</p>

<img data-src="lec14-clustering-guest_files/figure-revealjs/unnamed-chunk-14-1.png" width="960" class="r-stretch quarto-figure-center"></section>
<section id="if-there-is-heterogeneity-in-the-data" class="slide level2">
<h2>If there is heterogeneity in the data:</h2>
<h3 id="subgroups-might-not-be-the-best-explanation-for-it">Subgroups might not be the best explanation for it</h3>
<p>2D toy example: clustering samples from a simple linear model</p>

<img data-src="lec14-clustering-guest_files/figure-revealjs/unnamed-chunk-15-1.png" width="960" class="r-stretch quarto-figure-center"></section>
<section id="even-if-subgroups-are-a-good-explanation" class="slide level2">
<h2>Even if subgroups are a good explanation:</h2>
<h3 id="our-definition-of-similarity-could-be-problematic">Our definition of similarity could be problematic</h3>
<div class="columns">
<div class="column" style="width:50%;">
<p>Our clustering from before</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/cassie-cats-clustered.png" width="1200"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<p>What I actually care about</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/cassie-cats-clustered-2.png" width="12000"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
<p><br> (Photo and idea credit: Cassie Kozyrkov)</p>
</section>
<section id="the-fundamental-frustration-we-have-no-labelled-data" class="slide level2">
<h2>The fundamental frustration: we have no labelled data</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://imgs.xkcd.com/comics/k_means_clustering.png" width="600"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div>
<ul>
<li class="fragment">This makes it hard to check our work statistically</li>
<li class="fragment">Be cautious when interpreting your clustering results</li>
<li class="fragment">Gather other corroborating information whenever you can</li>
<li class="fragment">Prior information, experience, and domain knowledge are huge assets here!</li>
</ul>
</div>
</div>
</div>
</section>
<section id="clustering-algorithms-inputs-and-outputs" class="slide level2">
<h2>Clustering algorithms: inputs and outputs</h2>
<ul>
<li><p><strong>Input</strong>:</p>
<ul>
<li><p>Data matrix, e.g.&nbsp;<span class="math inline">\(X_{p \times n}\)</span> (features in rows, samples in columns)</p></li>
<li><p>Number of clusters <span class="math inline">\(K\)</span></p></li>
</ul></li>
<li><p><strong>Output</strong>:</p>
<ul>
<li><p>Discrete: Assignment of cluster membership for each object <span class="math inline">\(C_i\)</span>, where <span class="math inline">\(C_i=k\)</span> if object <span class="math inline">\(i\)</span> is assigned to cluster <span class="math inline">\(k\)</span></p></li>
<li><p>Probabilistic: <span class="math inline">\(K\)</span>-length vector of probabilities for each object <span class="math inline">\(\boldsymbol{C}_i\)</span>, were <span class="math inline">\(C_{ik}\in [0,1]\)</span> is the probability that object <span class="math inline">\(i\)</span> belongs to cluster <span class="math inline">\(k\)</span>, and <span class="math inline">\(\sum_{k=1}^K C_{ik} =1\)</span></p></li>
</ul></li>
</ul>
</section>
<section id="there-are-many-clustering-algorithms" class="slide level2">
<h2>There are many clustering algorithms</h2>
<p><br> <br></p>

<img data-src="img/clustmethods.png" width="1000" class="r-stretch quarto-figure-center"></section>
<section id="k-means-clustering" class="slide level2">
<h2>k-means clustering</h2>
<ul>
<li><p>One of the most widely used partition-based clustering approaches</p></li>
<li><p>Partition-based (flat), and discrete</p></li>
</ul>
<h3 id="main-idea-minimize-the-sum-of-within-cluster-variation">Main idea: minimize the sum of within-cluster variation</h3>
<p>Let <span class="math inline">\(G_k\)</span> be a set containing the objects assigned to cluster <span class="math inline">\(k\)</span>, for <span class="math inline">\(k = 1, 2, \ldots K\)</span>.</p>
<p>We want to find non-overlapping sets <span class="math inline">\(G_1, \ldots, G_K\)</span> that minimize:<br>
<span class="math display">\[ \sum \limits_{k=1}^K W(G_k)\]</span></p>
<p>Typically, we define within-cluster variation <span class="math inline">\(W(G_k)\)</span> with squared Euclidean distance: <span class="math display">\[  W(G_k) = \frac{1}{|G_k|} \sum_{i, i' \in G_k} \| \boldsymbol{x}_i - \boldsymbol{x}_{i'}\|_2^2 = \frac{1}{|G_k|} \sum_{i, i' \in G_k} \sum \limits_{j=1}^p (x_{ij} - x_{i'j})^2 \]</span></p>
</section>
<section id="more-on-the-k-means-objective-function" class="slide level2">
<h2>More on the K-means objective function</h2>
<p>The optimization problem that defines k-means clustering is therefore: <span class="math display">\[ \underset{G_1, \ldots, G_K}{\text{minimize}} ~~ \left \{ \sum \limits_{k=1}^K \frac{1}{|G_k|}   \sum_{i, i' \in G_k} \sum \limits_{j=1}^p (x_{ij} - x_{i'j})^2 \right \} \]</span></p>
<p>It turns out that:</p>
<p><span class="math display">\[ \frac{1}{|G_k|}    \sum_{i, i' \in G_k} \sum \limits_{j=1}^p (x_{ij} - x_{i'j})^2 = 2 \sum \limits_{i \in G_k} \sum \limits_{j=1}^p (x_{ij} - \bar{x}_{kj})^2, \]</span> where <span class="math inline">\(\bar{x}_{kj} = \frac{1}{|G_k|} \sum \limits_{i \in G_k} x_{ij}\)</span> is the mean for attribute <span class="math inline">\(j\)</span> in cluster <span class="math inline">\(G_k\)</span>.</p>
<h3 id="so-another-interpretation-of-this-optimization-problem-is-that-we-want-to-minimize-the-sum-of-within-cluster-distances-to-cluster-centres.">So another interpretation of this optimization problem is that we want to minimize the sum of within-cluster distances to cluster centres.</h3>
</section>
<section id="algorithm-k-means-clustering" class="slide level2">
<h2>Algorithm: k-means clustering</h2>
<p>(<a href="https://allisonhorst.com/k-means-clustering">Helpful illustrated example by Allison Horst</a>)</p>
<p><br></p>
<p>Initialize: Pick <span class="math inline">\(K\)</span> random points as initial cluster centres</p>
<ol type="1">
<li>Measure distance (squared Euclidean) between all points and the cluster centres</li>
<li>Assign points to nearest cluster</li>
<li>Update cluster means</li>
</ol>
<p>We repeat until cluster assignments stop changing.</p>
<div class="callout callout-tip callout-captioned callout-style-default">
<div class="callout-body">
<div class="callout-caption">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Tip</strong></p>
</div>
<div class="callout-content">
<p><span style="font-size: 1.2em;">This algorithm does not guarantee that you reach the global minimum value of the k-means objective function! Use multiple random initializations, then pick the clustering result that gives you the smallest k-means objective function.</span></p>
</div>
</div>
</div>
</section>
<section id="when-might-you-want-a-hierarchical-approach-instead" class="slide level2">
<h2>When might you want a hierarchical approach instead?</h2>
<p>Sometimes, a hierarchical organization just makes more sense:</p>

<img data-src="https://community.jmp.com/t5/image/serverpage/image-id/16821iE19A16F9BF548BFF/image-dimensions/343x264?v=v2.png" width="200" class="r-stretch quarto-figure-center"><p>(Picture by Chelsea Parlett-Pelleriti)</p>
<p>More pragmatically, sometimes we don’t want to pre-specify the number of clusters <span class="math inline">\(K\)</span></p>
</section>
<section id="agglomerative-hierarchical-clustering" class="slide level2">
<h2>Agglomerative hierarchical clustering</h2>
<p><strong>Agglomerative</strong>: in the algorithm, each observation starts in its own cluster; going up in hierarchy pairs of clusters are merged</p>

<img data-src="https://ars.els-cdn.com/content/image/3-s2.0-B9780124157811000091-f09-04-9780124157811.jpg" width="400" class="r-stretch quarto-figure-center"><p>Note: vertical red bar ‘cuts’ the tree to define the clusters indicated by label colours</p>
</section>
<section id="how-to-use-a-dendrogram" class="slide level2">
<h2>How to use a dendrogram</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="lec14-clustering-guest_files/figure-revealjs/unnamed-chunk-22-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div>
<ul>
<li class="fragment">Each <em>leaf</em> is one of the 9 objects, and as we move up the tree, leaves <em>fuse</em> into <em>branches</em></li>
<li class="fragment">The <em>height</em> where branches containing two objects is the dissimilarity between them</li>
<li class="fragment">Note: horizontal proximity is meaningless. We use <strong>vertical</strong> proximity to measure similarity</li>
<li class="fragment">Cutting the dendrogram horizontally at a given height yields (hierarchically nested) clusters</li>
</ul>
</div>
</div>
</div>
</section>
<section id="algorithm-hierarchical-agglomerative-clustering" class="slide level2">
<h2>Algorithm: Hierarchical agglomerative clustering</h2>
<p>(<a href="https://allisonhorst.com/agglomerative-hierarchical-clustering">Helpful illustrated example by Allison Horst</a>)</p>
<p><br></p>
<p>Given <span class="math inline">\(n\)</span> objects with <span class="math inline">\(p\)</span> attributes and a distance metric:</p>
<p><strong>Initialize</strong>: Treat each object as its own cluster and compute pairwise distances between all clusters</p>
<p><strong>Iterate</strong>: Repeat steps 1 and 2 until all objects belong to a single cluster</p>
<ol type="1">
<li><p>Find the “closest” pair of clusters, and <strong>merge</strong> them into a single cluster</p></li>
<li><p>Compute new distances between clusters</p></li>
</ol>
<p><br></p>
<p>But what’s the distance between the cluster {1, 2, 3} and the cluster {8, 9} on the last slide?</p>
</section>
<section id="we-need-the-concept-of-linkage" class="slide level2">
<h2>We need the concept of “linkage”</h2>
<p>Linkage extends the notion of distance between two observations (a distance metric, like Euclidean distance) to a notion of distance between <em>groups</em> of observations</p>
<p><br></p>
<h3 id="common-linkages">Common linkages:</h3>
<ul>
<li><p><strong>Single</strong> linkage: distance between two clusters is the <strong>minimum</strong> distance between any pair of elements</p></li>
<li><p><strong>Average</strong> linkage: distance between two clusters is the <strong>average</strong> distance between all pairs of elements</p></li>
<li><p><strong>Complete</strong> linkage: distance between two clusters is the <strong>maximum</strong> distance between any pair of elements</p></li>
</ul>
</section>
<section id="different-linkages-can-give-very-different-results" class="slide level2">
<h2>Different linkages can give very different results</h2>
<div class="columns">
<div class="column" style="width:60%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/linkage.jpeg" width="640"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:40%;">
<p>What’s the distance if we use:</p>
<ul>
<li><p>Single linkage?</p></li>
<li><p>Average linkage?</p></li>
<li><p>Complete linkage?</p></li>
</ul>
<p>(Image and idea credit to Chelsea Parlett-Pelleriti)</p>
</div>
</div>
<p><br></p>
<p>In general, there’s no “right” or “wrong” linkage.</p>
<p>But, a reasonable heuristic is to use average and complete over single linkage, since they tend to give more balanced dendrograms</p>
</section>
<section id="choosing-the-number-of-clusters-k" class="slide level2">
<h2>Choosing the number of clusters <span class="math inline">\(K\)</span></h2>
<p><strong>Note</strong>: picking <span class="math inline">\(K\)</span> to minimize (say) the sum of the within-cluster variations will <strong>not</strong> be informative, as it leads to the solution that each object should be in its own cluster. Therefore, need a method that takes into account the “cost” of adding additional clusters</p>
<div class="columns">
<div class="column" style="width:50%;">
<div>
<ul>
<li class="fragment"><p>Prior knowledge</p></li>
<li class="fragment"><p>“Elbow” method: point of diminishing returns</p></li>
<li class="fragment"><p>Information Criteria (AIC or BIC): likelihood penalized for number of parameters</p></li>
<li class="fragment"><p>Silhouette metric: how similar an object is to its own cluster compared to other clusters</p></li>
<li class="fragment"><p>Gap Statistics: total within-cluster variation compared to that expected under the null</p></li>
</ul>
</div>
</div><div class="column" style="width:50%;">
<p><br></p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/elbow.JPG" width="500"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="pragmatic-advice-try-all-the-things-get-a-fuller-picture" class="slide level2">
<h2>Pragmatic advice: Try all the things, get a fuller picture</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/xkcd-machine-learning.png" width="500"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<ul>
<li>Try using different combinations of attributes</li>
<li>Try using different measures of similarity</li>
<li>Try different ways of standardizing the data</li>
<li>Try using different algorithms</li>
<li>Try using a different number of clusters</li>
<li>Try using different linkages</li>
<li>Try clustering subsets of the objects</li>
</ul>
</div>
</div>
</section>
<section id="inference-after-clustering-its-tempting" class="slide level2">
<h2>Inference after clustering: it’s tempting</h2>
<p>Suppose you clustered samples into two groups.</p>
<p><br></p>
<div class="columns">
<div class="column" style="width:50%;">
<h3 id="follow-up-questions">Follow-up questions:</h3>
<ol type="1">
<li>Are these clusters associated with a feature that wasn’t used for clustering (<span class="math inline">\(Y\)</span>)?</li>
<li>Were these clusters really sampled from different populations?</li>
<li>If so, then which features are actually different across these populations?</li>
</ol>
</div><div class="column" style="width:50%;">
<h3 id="attempts-to-answer-them">Attempts to answer them:</h3>
<ol type="1">
<li><p>Two group comparisons test on measurements of <span class="math inline">\(Y\)</span></p></li>
<li><p>Two group comparisons test on measurements of <span class="math inline">\(X_1, \ldots, X_p\)</span></p></li>
<li><p>For <span class="math inline">\(j = 1, \ldots, p\)</span>, two group comparisons test on measurements of <span class="math inline">\(X_j\)</span></p></li>
</ol>
</div>
</div>
<p><br></p>
<p><strong>Unfortunately</strong>, this can get you in big trouble for questions 2 &amp; 3</p>
</section>
<section id="the-key-difference-circularitydouble-dipping" class="slide level2">
<h2>The key difference: circularity/double-dipping</h2>
<p>Recall how we tried to answer the three questions:</p>
<ol type="1">
<li>Two group comparisons test on measurements of <span class="math inline">\(Y\)</span>, with clusters as groups</li>
<li>Two group comparisons test on measurements of <span class="math inline">\(X_1, \ldots, X_p\)</span>, with clusters as groups</li>
<li>For <span class="math inline">\(j = 1, 2, \ldots, p\)</span>, two group comparisons test on measurements of <span class="math inline">\(X_j\)</span>, with clusters as groups</li>
</ol>
<h3 id="theres-circularity-in-questions-2-and-3">There’s circularity in questions 2 and 3!</h3>
<ul>
<li>In questions 2 and 3, we used the same data set to define groups (<span class="math inline">\(X\)</span>) and to test for a difference between groups (<span class="math inline">\(X\)</span>)</li>
</ul>
<h3 id="compare-to-question-1-no-circularity">Compare to question 1: no circularity</h3>
<ul>
<li>In question 1, we used different data sets to define groups (<span class="math inline">\(X\)</span>) and to test for a difference between groups (<span class="math inline">\(Y\)</span>)</li>
</ul>
</section>
<section id="intuition-why-is-circularitydouble-dipping-bad" class="slide level2">
<h2>Intuition: why is circularity/double-dipping bad?</h2>
<ul>
<li>When we choose a hypothesis to test <strong>after</strong> looking at our data, then we often wind up <strong>cherry-picking</strong> one that seems to be supported by our data</li>
</ul>
<div>
<ul>
<li class="fragment"><p>Recall: roughly speaking, p-values measure how surprised we are to observe such an extreme test statistic</p></li>
<li class="fragment"><p>When we cherry-pick, we will get more extreme test statistic values by chance alone even if the null hypothesis is true, so <strong>it should be harder to surprise us</strong></p></li>
<li class="fragment"><p>Unless this fact is taken into account, we are going to exaggerate the amount of evidence against the null hypothesis (i.e.&nbsp;p-values are going to be too small)</p></li>
</ul>
</div>
</section>
<section id="visual-illustration-of-what-can-go-wrong" class="slide level2">
<h2>Visual illustration of what can go wrong</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="lec14-clustering-guest_files/figure-revealjs/unnamed-chunk-26-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<ol type="1">
<li><p>Sample 50 observations from a single population</p></li>
<li><p>Cluster the observations with k-means clustering</p></li>
<li><p>Two-sample <span class="math inline">\(t\)</span>-test to compare feature 1 across clusters: p-value = 4.28^{-4}</p></li>
<li><p>Two-sample <span class="math inline">\(t\)</span>-test to compare feature 2 across clusters: p-value = 4.03^{-7}</p></li>
</ol>
</div>
</div>
<h3 id="p-values-are-too-small-because-weve-cherry-picked-two-groups-to-compare-via-clustering">p-values are too small, because we’ve cherry-picked two groups to compare via clustering</h3>
</section>
<section id="compare-to-sampling-from-two-populations-with-the-same-mean-for-features-1-and-2" class="slide level2">
<h2>Compare to: sampling from two populations with the same mean for features 1 and 2</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="lec14-clustering-guest_files/figure-revealjs/unnamed-chunk-27-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<ol type="1">
<li><p>Sample 25 observations from a population for group 1</p></li>
<li><p>Sample 25 observations from another population (with the same population means) for group 2</p></li>
<li><p>Two-sample <span class="math inline">\(t\)</span>-test to compare feature 1 across groups: p-value = 0.117</p></li>
<li><p>Two-sample <span class="math inline">\(t\)</span>-test to compare feature 2 across groups: p-value = 0.316</p></li>
</ol>
</div>
</div>
<h3 id="p-values-are-not-too-small-because-we-didnt-do-any-cherry-picking">p-values are not too small, because we didn’t do any cherry-picking</h3>
</section>
<section id="comparing-the-distributions-of-the-mean-differences" class="slide level2">
<h2>Comparing the distributions of the mean differences</h2>
<div class="callout callout-note callout-captioned callout-style-default">
<div class="callout-body">
<div class="callout-caption">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Recall: Numerator of two-sample <span class="math inline">\(t\)</span>-test statistic for feature 1</strong></p>
</div>
<div class="callout-content">
<p><span style="font-size: 1.5em;">Mean of feature 1 in group 1 - Mean of feature 1 in group 2</span></p>
</div>
</div>
</div>

<img data-src="lec14-clustering-guest_files/figure-revealjs/unnamed-chunk-28-1.png" width="1344" class="r-stretch quarto-figure-center"></section>
<section id="an-extreme-exception-where-theres-no-problem" class="slide level2">
<h2>An extreme exception where there’s no problem</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="lec14-clustering-guest_files/figure-revealjs/unnamed-chunk-29-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<ol type="1">
<li><p>Sample 25 observations from two very different populations</p></li>
<li><p>Ignore population labels and cluster all the observations with k-means clustering</p></li>
<li><p>Two-sample <span class="math inline">\(t\)</span>-test to compare feature 1 across clusters: p-value = 2.4^{-33}</p></li>
<li><p>Two-sample <span class="math inline">\(t\)</span>-test to compare feature 2 across clusters: p-value = 2.04^{-35}</p></li>
</ol>
</div>
</div>
<h3 id="these-results-will-look-the-same-as-if-we-used-true-population-labels">These results will look the same as if we used true population labels</h3>
</section>
<section id="double-dipping-is-particularly-relevant-to-scrna-seq" class="slide level2">
<h2>Double-dipping is particularly relevant to scRNA-seq</h2>
<h3 id="marker-gene-detection-and-cell-type-annotation">Marker gene detection and cell type annotation</h3>

<img data-src="img/deg.png" width="500" class="r-stretch quarto-figure-center"></section>
<section id="pragmatic-advice" class="slide level2">
<h2>Pragmatic advice</h2>
<h3 id="ask-yourself-how-strong-and-unambiguous-is-the-signal">Ask yourself: how strong and unambiguous is the signal?</h3>
<ul>
<li><p>If the grouping signal is strong and unambigous enough that clustering output is close to deterministic (i.e.&nbsp;not random), then go ahead and “double-dip”</p></li>
<li><p>Otherwise, double-dipping can get you into trouble</p></li>
<li><p>Without a lot of prior information and domain knowledge, there’s no way of knowing which case you’re in</p></li>
</ul>
<h3 id="my-0.02-if-you-dont-feel-confident-in-prior-informationdomain-knowledge-then-its-safer-to-just-avoid-double-dipping">My $0.02: if you don’t feel confident in prior information/domain knowledge, then it’s safer to just avoid double-dipping</h3>
</section>
<section id="summary-of-unsupervised-learning-and-clustering" class="slide level2">
<h2>Summary of unsupervised learning and clustering</h2>
<ul>
<li><p>Many choices to make when you want to cluster a set of objects:</p>
<ul>
<li>Objective, algorithm, attributes/features, distance metric, number of clusters</li>
</ul></li>
<li><p>There is no “best” method - it all depends on data and goal</p></li>
<li><p>Clustering is very powerful, but reckless application leads to misguided conclusions</p></li>
<li><p>Be especially careful when performing statistical inference with clustering outputs</p></li>
</ul>

<img src="https://github.com/STAT540-UBC/stat540-ubc.github.io/raw/main/images/stat540-logo-s.png" class="slide-logo r-stretch"><div class="footer footer-default">

</div>
</section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="lec14-clustering-guest_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="lec14-clustering-guest_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="lec14-clustering-guest_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="lec14-clustering-guest_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="lec14-clustering-guest_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="lec14-clustering-guest_files/libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="lec14-clustering-guest_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="lec14-clustering-guest_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="lec14-clustering-guest_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="lec14-clustering-guest_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="lec14-clustering-guest_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1600,

        height: 900,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        target: function(trigger) {
          return trigger.previousElementSibling;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'quarto-reveal',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>