<!DOCTYPE html>
<html lang="en"><head>
<script src="geneset_analysis_files/libs/clipboard/clipboard.min.js"></script>
<script src="geneset_analysis_files/libs/quarto-html/tabby.min.js"></script>
<script src="geneset_analysis_files/libs/quarto-html/popper.min.js"></script>
<script src="geneset_analysis_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="geneset_analysis_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="geneset_analysis_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="geneset_analysis_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="geneset_analysis_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.3.450">

  <meta name="author" content="Keegan Korthauer (with slides contributions from Yongjin Park and Paul Pavlidis)">
  <meta name="dcterms.date" content="2024-02-27">
  <title>Gene set analysis</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="geneset_analysis_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="geneset_analysis_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="geneset_analysis_files/libs/revealjs/dist/theme/quarto.css">
  <link href="geneset_analysis_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="geneset_analysis_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="geneset_analysis_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="geneset_analysis_files/libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="geneset_analysis_files/libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="geneset_analysis_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" data-background-color="#197aa0" data-background-image="https://github.com/STAT540-UBC/stat540-ubc.github.io/raw/main/images/stat540-logo-s.png" data-background-opacity="0.9" data-background-position="95% 90%" data-background-size="12%" class="quarto-title-block center">
  <h1 class="title">Gene set analysis</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Keegan Korthauer (with slides contributions from Yongjin Park and Paul Pavlidis) 
</div>
</div>
</div>

  <p class="date">February 27, 2024</p>
</section>
<section id="reminders" class="slide level2">
<h2>Reminders</h2>
<ul>
<li><p>Analysis Assignment due Thursday</p></li>
<li><p>Optional midcourse feedback (bonus point) due Friday</p></li>
</ul>
</section>
<section id="todays-lecture" class="slide level2">
<h2>Today’s lecture</h2>
<ul>
<li><strong>Motivations: What’s next after genomics analysis?</strong>
<ul>
<li>What have we learned?</li>
<li>How do we know that our discovery is meaningful?</li>
</ul></li>
<li><strong>What is Gene set analysis?</strong>
<ul>
<li>Set-based approach: hypergeometric test</li>
<li>Rank-based approach GSEA by KS statistic</li>
</ul></li>
</ul>
</section>
<section id="biological-mechanisms-processes-are-modular" class="slide level2">
<h2>Biological mechanisms / processes are modular</h2>
<div class="columns">
<div class="column">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://media.springernature.com/lw685/springer-static/image/art%3A10.1038%2Fnature14248/MediaObjects/41586_2015_Article_BFnature14248_Fig7_HTML.jpg" style="width:80.0%"></p>
</figure>
</div>
</div><div class="column">
<p><a href="https://doi.org/10.1038/nature14248">Roadmap epigenomics, Nature 2015</a></p>
<p><br> Epigenomic (enhancer element) modules across 111 human tissues</p>
<br><br>
<p style="text-align: center;">
↓
</p>
<p><br><br></p>
<p>Gene Ontology information confirms cell-type or tissue-specific biological pathways</p>
</div>
</div>
</section>
<section id="clusters-of-cells-correspond-to-cell-types" class="slide level2">
<h2>Clusters of cells correspond to ‘cell types’</h2>
<div class="columns">
<div class="column" style="width:60%;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://media.springernature.com/lw685/springer-static/image/art%3A10.1038%2Fs41586-018-0590-4/MediaObjects/41586_2018_590_Fig2_HTML.png" style="width:50.0%"></p>
</figure>
</div>
<blockquote>
<p><a href="https://doi.org/10.1038/s41586-018-0590-4"><em>Tabula Muris</em>, Nature 2018</a>: “Comparison of cell-type determination as performed by unbiased whole-transcriptome comparison versus manual annotation of clusters by organ-specific experts.”</p>
</blockquote>
</div><div class="column" style="width:40%;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://media.springernature.com/lw685/springer-static/image/art%3A10.1038%2Fs41586-018-0590-4/MediaObjects/41586_2018_590_Fig3_HTML.png" style="width:70.0%"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="enrichment-analysis-to-confirm-previous-knowledge" class="slide level2">
<h2>Enrichment analysis to confirm previous knowledge</h2>

<img data-src="https://media.springernature.com/lw685/springer-static/image/art%3A10.1038%2Fs41586-018-0768-9/MediaObjects/41586_2018_768_Fig2_HTML.png" class="r-stretch quarto-figure-center"><p><a href="https://doi.org/10.1038/s41586-018-0768-9">Ishizuka et al., Nature 2019</a></p>
</section>
<section id="what-is-gene-set-analysis" class="slide level2">
<h2>What is Gene Set Analysis?</h2>
<div class="quarto-layout-panel">
<div class="quarto-layout-row quarto-layout-valign-top">
<div id="first-column" class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<h4 id="discrete-gene-set-analysis">(Discrete) Gene Set Analysis</h4>
<p><strong>Input:</strong></p>
<ol type="1">
<li><p>A dictionary of gene sets that map genes to sets (gene-to-set mapping)</p></li>
<li><p>A list of <strong>top</strong> genes identified in our own study (after FDR control)</p></li>
</ol>
</div>
<div id="second-column" class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<h4 id="rank-based-gene-set-enrichment">(Rank-based) Gene Set Enrichment</h4>
<p><strong>Input:</strong></p>
<ol type="1">
<li><p>A dictionary of gene sets that map genes to sets</p></li>
<li><p>A <strong>full</strong> list of gene-level <strong>scores</strong> (e.g., p-values)</p></li>
</ol>
</div>
</div>
</div>
<p><strong>Output:</strong> A table of scores for all the gene sets in the ‘dictionary’</p>
</section>
<section id="where-does-the-dictionary-come-from" class="slide level2">
<h2>Where does the ‘dictionary’ come from?</h2>
<ul>
<li><p>Gene Ontology (GO) is a controlled vocabulary for function</p></li>
<li><p>Began ~25 years ago; over 45K terms today</p></li>
<li><p>Mostly species-agnostic</p></li>
<li><p>Three aspects:</p>
<ol type="1">
<li>Biological Process</li>
<li>Cellular Component</li>
<li>Molecular Function</li>
</ol></li>
</ul>
<p><img data-src="https://geneontology.org/assets/go-logo.large.png" class="absolute" style="top: 120px; right: 0px; width: 30%; "> <img data-src="Vis/gotrack.png" class="absolute" style="bottom: 40px; right: 0px; width: 50%; " alt="gotrack.msl.ubc"></p>
</section>
<section id="section" class="slide level2" data-background-iframe="https://amigo.geneontology.org/amigo/search/ontology?q=cell+cycle&amp;searchtype=ontology">
<h2></h2>
</section>
<section id="go-structure" class="slide level2">
<h2>GO structure</h2>
<ul>
<li>Structure is an acyclic graph, not a tree: terms <strong>commonly</strong> have multiple parents</li>
<li>Annotations in the database don’t separately list out all the parents</li>
<li>Includes a mix of ‘directly annotated’ and many ‘inferred’ terms</li>
<li>Several relationship types</li>
</ul>
<div class="columns">
<div class="column" style="width:0.30;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="./Vis/relationships.png" style="width:50.0%"></p>
</figure>
</div>
</div><div class="column" style="width:0.70;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="./Vis/child.png"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="go-aspects" class="slide level2">
<h2>GO aspects</h2>
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Molectular Function (MF)</strong></p>
</div>
<div class="callout-content">
<ul>
<li><p>“Elemental activities, such as catalysis or binding, describing the actions of a gene product at the molecular level.”</p></li>
<li><p>Primarily enzymatic activities</p></li>
</ul>
</div>
</div>
</div>
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Cellular Component (CC)</strong></p>
</div>
<div class="callout-content">
<ul>
<li><p>Part of a cell/extracellular environment of which a gene product is a component</p></li>
<li><p>Examples: “cell wall”, “actin body”</p></li>
</ul>
</div>
</div>
</div>
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Biological Process (BP)</strong></p>
</div>
<div class="callout-content">
<ul>
<li><p>“A phenomenon marked by changes that lead to a particular result, mediated by one or more gene products.”</p></li>
<li><p>Closest GO concept to “pathways”</p></li>
</ul>
</div>
</div>
</div>
</section>
<section id="limitations-of-go-terms" class="slide level2">
<h2>Limitations of GO terms</h2>
<ul>
<li><p>Incomplete and unstable (continual updates)</p></li>
<li><p>Distribution of annotations per gene is non-uniform (many genes have few/none, others have hundreds)</p></li>
<li><p>Doesn’t give relationship between genes within a term</p></li>
</ul>
</section>
<section id="pathways-another-grouping-of-genes" class="slide level2">
<h2>Pathways: Another grouping of genes</h2>

<img data-src="Vis/kegg.png" style="width:10.0%" class="r-stretch quarto-figure-center"><ul>
<li><p>Example: Kyoto Encyclopedia of Genes and Genomes (KEGG) Pathways database</p></li>
<li><p>Species-specific directed networks of genes and their relationships, curated from the literature</p></li>
<li><p>Pathways are organized into functional categories (not nearly as complex as GO)</p></li>
</ul>
</section>
<section id="section-1" class="slide level2" data-background-iframe="https://www.genome.jp/pathway/hsa04110">
<h2></h2>
</section>
<section id="what-have-we-learned-from-our-differential-expression-analysis" class="slide level2">
<h2>What have we learned from our differential expression analysis?</h2>

<img data-src="./Vis/geneset_1.jpg" class="r-stretch quarto-figure-center"><p><strong>blue:</strong> some biological pathway of interest</p>
</section>
<section id="lets-drop-the-edges-in-biological-networks" class="slide level2">
<h2>Let’s drop the edges in biological networks</h2>

<img data-src="./Vis/geneset_2.jpg" class="r-stretch quarto-figure-center"><p>Prior Knowledge of biological pathways define a set of genes (blue)</p>
</section>
<section id="does-this-pathway-overlap-with-our-deg-list" class="slide level2">
<h2>Does this pathway overlap with our DEG list?</h2>

<img data-src="./Vis/geneset_3.jpg" class="r-stretch quarto-figure-center"><p>Suppose we discovered the 8 DEG outlined (pink)</p>
</section>
<section id="is-our-deg-list-enriched-for-genes-in-this-pathway" class="slide level2">
<h2>Is our DEG list <em>enriched</em> for genes in this pathway?</h2>

<img data-src="./Vis/geneset_4.jpg" class="r-stretch quarto-figure-center"><p>Suppose we found <span class="math inline">\(q\)</span> genes are overlapping with this pathway (green)</p>
</section>
<section id="gene-set-analysis-testing-over-representation-of-degs" class="slide level2">
<h2>Gene Set Analysis testing over-representation of DEGs</h2>
<div class="columns">
<div class="column" style="width:.65;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="./Vis/geneset_5.jpg"></p>
</figure>
</div>
</div><div class="column" style="width:.35;">
<p>What are the relevant numbers?</p>
<ul>
<li><p><span class="math inline">\(N\)</span>: # genes in this universe</p></li>
<li><p><span class="math inline">\(m\)</span>: # genes in this set</p></li>
<li><p><span class="math inline">\(n\)</span>: # genes <em>not</em> in this gene set</p></li>
<li><p><span class="math inline">\(k\)</span>: # DEGs in our analysis</p></li>
<li><p><span class="math inline">\(q\)</span>: # DEGs (of <span class="math inline">\(k\)</span>) overlapping with the set of <span class="math inline">\(m\)</span> genes</p></li>
</ul>
</div>
</div>
</section>
<section id="is-this-overlap-of-q6-of-k8-genes-significant" class="slide level2">
<h2>Is this overlap of <span class="math inline">\(q=6\)</span> of <span class="math inline">\(k=8\)</span> genes <em>significant</em>?</h2>
<div class="columns">
<div class="column" style="width:.65;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="./Vis/geneset_5.jpg"></p>
</figure>
</div>
</div><div class="column" style="width:.35;">
<p>In this example:</p>
<ul>
<li><p><span class="math inline">\(N = 24\)</span> total genes</p></li>
<li><p><span class="math inline">\(m = 9\)</span> genes in this set</p></li>
<li><p><span class="math inline">\(n = N - m = 15\)</span> not in this set</p></li>
<li><p><span class="math inline">\(k = 8\)</span> DEGs</p></li>
<li><p><span class="math inline">\(q=6\)</span> out of <span class="math inline">\(k=8\)</span> overlap</p></li>
</ul>
</div>
</div>
</section>
<section id="is-this-overlap-of-q6-of-k8-genes-significant-1" class="slide level2">
<h2>Is this overlap of <span class="math inline">\(q=6\)</span> of <span class="math inline">\(k=8\)</span> genes significant?</h2>
<div class="columns">
<div class="column" style="width:.65;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="./Vis/geneset_5.jpg"></p>
</figure>
</div>
</div><div class="column" style="width:.35;">
<p>Questions:</p>
<ul>
<li><p>Is it meaningful enough to report?</p></li>
<li><p>Is it surprising enough that we recapitulated 6/9 (~67 %) genes in the set?</p></li>
<li><p>Is it surprising enough that we 6/8 (~75 %) of our DEG are in the set?</p></li>
<li><p>What is the null distribution?</p></li>
<li><p>What is the generative/simulation scheme?</p></li>
</ul>
</div>
</div>
</section>
<section id="how-do-we-find-q-out-of-k-degs-overlapping-with-a-gene-set-of-m-genes" class="slide level2">
<h2>How do we find <span class="math inline">\(q\)</span> out of <span class="math inline">\(k\)</span> DEGs overlapping with a gene set of <span class="math inline">\(m\)</span> genes?</h2>
<div class="columns">
<div class="column" style="width:.55;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="./Vis/geneset_5.jpg"></p>
</figure>
</div>
</div><div class="column" style="width:.45;">
<p>Under the null of hypergeometric distribution</p>
<ol type="1">
<li><p>Randomly sample <span class="math inline">\(k\)</span> DEGs out of <span class="math inline">\(N\)</span> genes</p></li>
<li><p>Of these <span class="math inline">\(k\)</span> genes, <span class="math inline">\(q\)</span> overlap with a gene set consisting of <span class="math inline">\(m\)</span> genes</p></li>
<li><p>The rest <span class="math inline">\(k-q\)</span> genes overlap with genes outside of the gene set <span class="math inline">\(N-m\)</span></p></li>
</ol>
<p>How to generalize?</p>
</div>
</div>
</section>
<section id="binomial-coefficient-lets-see-if-we-can-estimate-the-null-distribution-by-counting" class="slide level2">
<h2><em>Binomial coefficient</em>: Let’s see if we can estimate the null distribution by counting</h2>
<ul>
<li><p>How many possible ways to select <span class="math inline">\(k=8\)</span> out of <span class="math inline">\(N=24\)</span> genes, ignoring the order of <span class="math inline">\(k\)</span> selected genes and <span class="math inline">\(N-k\)</span> <em>not</em> selected genes?</p></li>
<li><p>We can think of this as three steps:</p>
<ol type="1">
<li>enumerating <span class="math inline">\(N\)</span> genes</li>
<li>partition them into the first <span class="math inline">\(k\)</span> genes and the rest</li>
<li>ignore the order within each partition</li>
</ol></li>
</ul>
<p><span class="math display">\[{24 \choose 8} = \frac{{\color{teal}\textsf{\{all possible ways to enumerate }24 \textsf{ genes}\}}}
{{\color{magenta}\{ \textsf{enumerating }8 \textsf{ genes}\}} {\color{blue}\{\textsf{enumerating }16 \textsf{ genes} \}}}
= \frac{{\color{teal}24!}}{{\color{magenta}8!}{\color{blue}16!}}\]</span></p>
</section>
<section id="calculating-the-binomial-coefficient" class="slide level2">
<h2>Calculating the binomial coefficient</h2>
<p>There are a lot of ways to select <span class="math inline">\(k=8\)</span> DEGs out of <span class="math inline">\(N=24\)</span> total genes!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="fu">choose</span>(<span class="dv">24</span>,<span class="dv">8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 735471</code></pre>
</div>
</div>
<p>What is the probability of selecting (at random) this exact set of <span class="math inline">\(k\)</span> DEGs out of <span class="math inline">\(N\)</span> genes?</p>
<div class="fragment">
<p><span class="math display">\[{N \choose k}^{-1}\]</span></p>
</div>
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="dv">1</span><span class="sc">/</span><span class="fu">choose</span>(<span class="dv">24</span>,<span class="dv">8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1.359673e-06</code></pre>
</div>
</div>
</div>
</section>
<section id="enumerating-the-other-possibilities" class="slide level2">
<h2>Enumerating the other possibilities</h2>
<ul>
<li>How many possible ways of finding <span class="math inline">\(q\)</span> DEGs overlapping with <span class="math inline">\(m\)</span> genes in the gene set? <span class="math display">\[{m \choose q}\]</span></li>
<li>How many possible ways of finding the rest <span class="math inline">\(k-q\)</span> DEGs overlapping with (<span class="math inline">\(N-m=n\)</span>) genes in the gene set? <span class="math display">\[{N-m \choose k-q}\]</span></li>
</ul>
</section>
<section id="probability-distribution-to-describe-this-process" class="slide level2">
<h2>Probability distribution to describe this process</h2>
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Hypergeometric distribution</strong></p>
</div>
<div class="callout-content">
<p>A discrete probability distribution that describes the probability of <span class="math inline">\(q\)</span> successes (random draws for which the object drawn has a specified feature) in <span class="math inline">\(k\)</span> draws, without replacement, from a finite population of size <span class="math inline">\(N\)</span> that contains exactly <span class="math inline">\(m\)</span> objects with that feature.</p>
</div>
</div>
</div>

<img data-src="https://upload.wikimedia.org/wikipedia/commons/thumb/c/c1/HypergeometricPDF.png/300px-HypergeometricPDF.png" class="r-stretch quarto-figure-center"><p class="caption">image source: <a href="https://en.wikipedia.org/wiki/Hypergeometric_distribution">en.wikipedia.org/wiki/Hypergeometric_distribution</a></p></section>
<section id="hypergeometric-distribution-1" class="slide level2">
<h2>Hypergeometric distribution</h2>
<p>Probability mass function</p>
<p><span class="math display">\[
\begin{eqnarray*}
  P(x=q|N,m,k) &amp;=&amp;
                     \sum_{\begin{array}{l}
                     {\color{blue}                             
    \textsf{# ways to select }q\textsf{ out of }m } \\
                     {\color{teal}
        \textsf{# ways to select }(k-q)\textsf{ out of }N-m }
        \end{array}}
{\color{magenta}
\begin{array}{l}
  \textsf{the probability} \\
  \textsf{of choosing a set size }k \\
  \textsf{out of total }N
\end{array}} \\
&amp;=&amp;
{
\underbrace{\color{blue} {m \choose q}}_{
\begin{array}{l}
{\color{blue}\textsf{# ways to choose }} \\
{\color{blue}q\, \textsf{ overlap out of }m}
\end{array} }
}
{
    \times
\underbrace{\color{teal} {N-m \choose k - q}}_{
\begin{array}{l}
{\color{teal}\textsf{# ways to choose }} \\
{\color{teal}(k-q) \textsf{ out of }N-m}
\end{array} }
}
{
\times
{\color{magenta} {N \choose k}^{-1} }
}
\end{eqnarray*}
\]</span></p>
<p><img data-src="Vis/geneset_5.jpg" class="absolute" style="left: 0px; bottom: 0px; width: 30%; "></p>
</section>
<section id="what-is-the-probability-of-k-overlapping-degs" class="slide level2">
<h2>What is the probability of <span class="math inline">\(k\)</span> overlapping DEGs?</h2>
<div class="columns">
<div class="column" style="width:.45;">
<p><strong>Hypergeometric PMF</strong></p>
<p><span class="math display">\[P(x=q|N,m,k) = \frac{{m \choose q} {N - m \choose k-q}}{{N \choose k}}\]</span></p>
</div><div class="column" style="width:.45;">
<p><strong>Hypergeometric CDF</strong></p>
<p><span class="math display">\[P(x{\le} q|N,m,k) = \sum_{i = 0}^{q} \frac{{m \choose i} {N - m \choose k-i}}{{N \choose k}}\]</span></p>
</div>
</div>
</section>
<section id="hypergeometric-test-for-testing-significant-overlap" class="slide level2">
<h2>Hypergeometric test for testing significant overlap</h2>
<p><span class="math display">\[H_{0}:\, x \le q \quad \textsf{vs.} \quad H_{1}:\, x &gt; q\]</span></p>
<p>We may observe overlap <span class="math inline">\(q\)</span> genes by random sampling of <span class="math inline">\(k\)</span> genes <strong>without</strong> replacement.</p>
<p>Therefore, we can calculate the p-value:</p>
<p><span class="math display">\[P(x &gt; q|n, m, k) = 1 - \sum_{i = 0}^{q} \frac{{m \choose i} {n \choose k-i}}{{n+m \choose k}}\]</span></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="dv">1</span><span class="sc">-</span><span class="fu">phyper</span>(<span class="at">q =</span> <span class="dv">6</span>, <span class="at">m =</span> <span class="dv">9</span>, <span class="at">n =</span> <span class="dv">15</span>, <span class="at">k =</span> <span class="dv">8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.0007464604</code></pre>
</div>
</div>
</section>
<section id="how-significant-is-q-overlap-in-our-discovery" class="slide level2">
<h2>How significant is <span class="math inline">\(q\)</span> overlap in our discovery?</h2>
<div class="quarto-layout-panel">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="column quarto-layout-cell" style="flex-basis: 33.3%;justify-content: center;">
<p><span class="math inline">\(m=9\)</span>, <span class="math inline">\(n=15\)</span>, &amp; <span class="math inline">\(k=8\)</span>:</p>
<div class="cell">
<div class="cell-output-display">
<p><img data-src="geneset_analysis_files/figure-revealjs/unnamed-chunk-5-1.png" width="480"></p>
</div>
</div>
</div>
<div class="column quarto-layout-cell" style="flex-basis: 33.3%;justify-content: center;">
<div class="fragment">
<p>If <span class="math inline">\(q=3\)</span>:</p>
<div class="cell">
<div class="cell-output-display">
<p><img data-src="geneset_analysis_files/figure-revealjs/unnamed-chunk-6-1.png" width="480"></p>
</div>
</div>
</div>
</div>
<div class="column quarto-layout-cell" style="flex-basis: 33.3%;justify-content: center;">
<div class="fragment">
<p>If <span class="math inline">\(q=6\)</span>:</p>
<div class="cell">
<div class="cell-output-display">
<p><img data-src="geneset_analysis_files/figure-revealjs/unnamed-chunk-7-1.png" width="480"></p>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="not-all-gene-sets-are-the-same-gene-length-bias" class="slide level2">
<h2>Not all gene sets are the same (gene length bias)</h2>
<p>So far, we assumed all genes are independent and equally likely to be DE under the null</p>
<div class="columns">
<div class="column">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="Vis/youngetal2010.png"></p>
</figure>
</div>
</div><div class="column">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://media.springernature.com/full/springer-static/image/art%3A10.1186%2Fgb-2010-11-2-r14/MediaObjects/13059_2009_Article_2307_Fig1_HTML.jpg"></p>
<figcaption><a href="https://doi.org/10.1186/gb-2010-11-2-r14">Young et al.&nbsp;2010, Genome Biology</a></figcaption>
</figure>
</div>
</div>
</div>
<ol type="a">
<li><p>Distribution of average gene lengths in GO categories</p></li>
<li><p>Wilcoxon p-values of median length of genes in 7,873 GO categories compared to overall</p></li>
</ol>
</section>
<section id="longer-genes-tend-to-be-more-differentially-expressed" class="slide level2">
<h2>Longer genes tend to be more differentially expressed</h2>

<img data-src="https://media.springernature.com/full/springer-static/image/art%3A10.1186%2Fgb-2010-11-2-r14/MediaObjects/13059_2009_Article_2307_Fig2_HTML.jpg" class="r-stretch quarto-figure-center"><p class="caption"><a href="https://doi.org/10.1186/gb-2010-11-2-r14">Young et al.&nbsp;2010, Genome Biology</a></p></section>
<section id="goseq-null-distribution-conditions-on-gene-lengths" class="slide level2">
<h2>GOseq null distribution conditions on gene lengths</h2>

<img data-src="https://media.springernature.com/full/springer-static/image/art%3A10.1186%2Fgb-2010-11-2-r14/MediaObjects/13059_2009_Article_2307_Fig4_HTML.jpg" class="r-stretch quarto-figure-center"><p class="caption"><a href="https://doi.org/10.1186/gb-2010-11-2-r14">Young et al.&nbsp;2010, Genome Biology</a></p></section>
<section id="section-2" class="slide level2" data-background-iframe="https://bioconductor.org/packages/release/bioc/html/goseq.html">
<h2></h2>
</section>
<section id="summary-of-discrete-gene-set-analysis" class="slide level2">
<h2>Summary of discrete gene set analysis</h2>
<div class="columns">
<div class="column" style="width:.5;">
<p><strong>When does it apply?</strong></p>
<ul>
<li><p>We have a list of genes after accounting for multiple hypothesis testing</p></li>
<li><p>We want to know whether our list is overrepresented in some gene sets of interest</p></li>
</ul>
</div><div class="column" style="width:.5;">
<p><strong>When does it not apply?</strong></p>
<ul>
<li><p><strong>Don’t</strong> have a good way to select a list of genes:</p>
<ul>
<li>Our discovery data may lack statistical power, i.e., no (or a few) significant genes left after multiple hypothesis correction</li>
</ul></li>
<li><p>There is a hidden factor that can affect two steps: (1) gene set selection (annotations/knowledge) and (2) differential expression calling</p></li>
</ul>
</div>
</div>
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Note</strong></p>
</div>
<div class="callout-content">
<p>Estimating the null distribution of ranks/scores over gene sets of interest is nontrivial</p>
</div>
</div>
</div>
</section>
<section id="rank-based-gene-set-enrichment-analysis-gsea" class="slide level2">
<h2>Rank-based Gene Set Enrichment Analysis (GSEA)</h2>

<img data-src="Vis/gsea.png" class="r-stretch quarto-figure-center"><p><a href="https://www.gsea-msigdb.org/gsea/index.jsp">Image source: gsea-msigdb.org</a></p>
</section>
<section id="rank-based-gsea-subramanian-et-al.-2005" class="slide level2">
<h2>Rank-based GSEA (<a href="https://doi.org/10.1073/pnas.0506580102">Subramanian <em>et al.</em> 2005</a>)</h2>
<ul>
<li><p>A collection of gene-sets: <span class="math inline">\(\mathcal{C}_{1}, \ldots, \mathcal{C}_{K}\)</span></p></li>
<li><p>A vector of gene-level scores (<span class="math inline">\(G\)</span> genes): <span class="math inline">\(z_{1}, \ldots, z_{G}\)</span></p></li>
<li><p>Each <span class="math inline">\(z_{g}\)</span> could come from differential expression analysis</p></li>
</ul>

<img data-src="https://www.gsea-msigdb.org/gsea/doc/GSEAUserGuideTEXT.fld/image009.jpg" class="r-stretch quarto-figure-center"></section>
<section id="gsea-algorithm" class="slide level2">
<h2>GSEA algorithm</h2>
<ul>
<li><p>For each <span class="math inline">\(k\)</span></p>
<ul>
<li><p>Compute a set-level score <span class="math inline">\(S_{k}(\mathbf{z},\mathcal{C}_{k})\)</span></p></li>
<li><p>E.g., Kolmogorov-Smirnov statistic comparing</p></li>
</ul></li>
</ul>
<p><span class="math display">\[\{z_{g}:\,g \in \mathcal{C}_{k}\} \, \textsf{vs.} \, \{z_{g}:\,g \notin \mathcal{C}_{k}\}\]</span></p>
</section>
<section id="rank-based-gsea-subramanian-et-al.-2005-1" class="slide level2">
<h2>Rank-based GSEA (<a href="https://doi.org/10.1073/pnas.0506580102">Subramanian <em>et al.</em> 2005</a>)</h2>
<ul>
<li><p>Construct null distribution of <span class="math inline">\(S_{1}, \ldots, S_{K}\)</span> by sample label (case-control) or gene-to-set membership permutation</p></li>
<li><p>Using null distribution by permutation, estimate p-values and false discovery rates</p></li>
<li><p>If we knew null distribution, we would not need expensive permutations</p></li>
</ul>
<div class="columns">
<div class="column" style="width:.45;">
<p><strong>Good</strong>:</p>
<ul>
<li><p>No cutoff/assumptions needed to estimate null distribution</p></li>
<li><p>Aggregate scores across many genes! (boost the power)</p></li>
</ul>
</div><div class="column" style="width:.45;">
<p><strong>Bad</strong>:</p>
<ul>
<li><p>Need to define an appropriate statistic?</p></li>
<li><p>What do we permute? Samples? Gene sets?</p></li>
<li><p>Permutations are computationally intensive</p></li>
</ul>
</div>
</div>
</section>
<section id="fast-gsea-fgsea" class="slide level2">
<h2>Fast GSEA (FGSEA)</h2>

<img data-src="https://www.biorxiv.org/content/biorxiv/early/2021/02/01/060012/F2.large.jpg" class="r-stretch quarto-figure-center"><p class="caption"><a href="https://doi.org/10.1101/060012">Korotkevich et al.&nbsp;2021</a></p></section>
<section id="section-3" class="slide level2" data-background-iframe="https://bioconductor.org/packages/release/bioc/html/fgsea.html">
<h2></h2>
</section>
<section id="simulation-efron-and-tibshirani-2007" class="slide level2">
<h2>Simulation (<a href="https://doi.org/10.1214/07-AOAS101">Efron and Tibshirani 2007</a>)</h2>
<ol type="1">
<li>Generate basal gene expression (sample <span class="math inline">\(i\)</span>, gene <span class="math inline">\(g\)</span>)</li>
</ol>
<p><span class="math display">\[X_{i,g} \sim \mathcal{N}\!\left(0,1\right)\]</span></p>
<ol start="2" type="1">
<li><p>Sample case vs.&nbsp;control membership (the rows of <span class="math inline">\(X\)</span>) uniformly at random</p></li>
<li><p>Sample gene set membership uniformly at random</p></li>
<li><p>For the first gene set, select a certain fraction of genes to perturb</p></li>
<li><p>For the selected genes <span class="math inline">\(g^{*}\)</span>, add some <span class="math inline">\(\Delta\)</span> value to <span class="math inline">\(X_{i,g^*}\)</span> if the sample <span class="math inline">\(i\)</span> belongs to the control group</p></li>
</ol>
</section>
<section id="lets-simulate-some-gene-set-data-efron-tibshirani-2007" class="slide level2">
<h2>Let’s simulate some gene set data (Efron &amp; Tibshirani 2007)</h2>
<div class="columns">
<div class="column" style="width:.55;">
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>simulate.data <span class="ot">&lt;-</span></span>
<span id="cb7-2"><a href="#cb7-2"></a>    <span class="cf">function</span>(<span class="at">G =</span> <span class="dv">1000</span>,        <span class="co"># genes</span></span>
<span id="cb7-3"><a href="#cb7-3"></a>             <span class="at">K =</span> <span class="dv">150</span>,         <span class="co"># gene sets</span></span>
<span id="cb7-4"><a href="#cb7-4"></a>             <span class="at">n.samp =</span> <span class="dv">100</span>,    <span class="co"># sample size</span></span>
<span id="cb7-5"><a href="#cb7-5"></a>             <span class="at">delta =</span> .<span class="dv">4</span>,      <span class="co"># perturbation</span></span>
<span id="cb7-6"><a href="#cb7-6"></a>             <span class="at">p.perturb =</span> <span class="dv">1</span>) { <span class="co"># Pr of petrub</span></span>
<span id="cb7-7"><a href="#cb7-7"></a>  case.control <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">0</span><span class="sc">:</span><span class="dv">1</span>, n.samp, <span class="cn">TRUE</span>)</span>
<span id="cb7-8"><a href="#cb7-8"></a>  S <span class="ot">&lt;-</span> <span class="fu">sample</span>(K, G, <span class="cn">TRUE</span>)       <span class="co"># gene sets</span></span>
<span id="cb7-9"><a href="#cb7-9"></a>  X <span class="ot">&lt;-</span> <span class="fu">.rnorm</span>(n.samp, G)        <span class="co"># All the other genes</span></span>
<span id="cb7-10"><a href="#cb7-10"></a>  <span class="do">## Perturbation of the first gene set</span></span>
<span id="cb7-11"><a href="#cb7-11"></a>  .genes<span class="fl">.1</span> <span class="ot">&lt;-</span> <span class="fu">which</span>(S <span class="sc">==</span> <span class="dv">1</span>)</span>
<span id="cb7-12"><a href="#cb7-12"></a>  n1 <span class="ot">&lt;-</span> <span class="fu">length</span>(.genes<span class="fl">.1</span>)</span>
<span id="cb7-13"><a href="#cb7-13"></a>  n.perturb <span class="ot">&lt;-</span> <span class="fu">max</span>(<span class="fu">floor</span>(n1 <span class="sc">*</span> p.perturb),<span class="dv">1</span>)</span>
<span id="cb7-14"><a href="#cb7-14"></a>  .genes<span class="fl">.1</span> <span class="ot">&lt;-</span> <span class="fu">sample</span>(.genes<span class="fl">.1</span>, n.perturb)</span>
<span id="cb7-15"><a href="#cb7-15"></a>  .case <span class="ot">&lt;-</span> case.control <span class="sc">==</span> <span class="dv">1</span></span>
<span id="cb7-16"><a href="#cb7-16"></a>  X[.case, .genes<span class="fl">.1</span>] <span class="ot">&lt;-</span> X[.case, .genes<span class="fl">.1</span>] <span class="sc">+</span> delta</span>
<span id="cb7-17"><a href="#cb7-17"></a>  <span class="fu">require</span>(Matrix)</span>
<span id="cb7-18"><a href="#cb7-18"></a>  .membership <span class="ot">&lt;-</span> <span class="fu">sparseMatrix</span>(<span class="at">j=</span><span class="dv">1</span><span class="sc">:</span>G, <span class="at">i=</span>S, <span class="at">x=</span><span class="fu">rep</span>(<span class="dv">1</span>,G))</span>
<span id="cb7-19"><a href="#cb7-19"></a>  <span class="fu">list</span>(<span class="at">X=</span>X, <span class="at">S=</span>.membership, <span class="at">Y=</span>case.control)</span>
<span id="cb7-20"><a href="#cb7-20"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div><div class="column" style="width:.35;">
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb8-2"><a href="#cb8-2"></a>dat <span class="ot">&lt;-</span> <span class="fu">simulate.data</span>()</span>
<span id="cb8-3"><a href="#cb8-3"></a></span>
<span id="cb8-4"><a href="#cb8-4"></a><span class="fu">str</span>(dat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of 3
 $ X: num [1:100, 1:1000] 0.399 -0.262 -0.303 -0.734 -1.382 ...
 $ S:Formal class 'dgCMatrix' [package "Matrix"] with 6 slots
  .. ..@ i       : int [1:1000] 107 47 32 44 148 30 101 144 72 86 ...
  .. ..@ p       : int [1:1001] 0 1 2 3 4 5 6 7 8 9 ...
  .. ..@ Dim     : int [1:2] 150 1000
  .. ..@ Dimnames:List of 2
  .. .. ..$ : NULL
  .. .. ..$ : NULL
  .. ..@ x       : num [1:1000] 1 1 1 1 1 1 1 1 1 1 ...
  .. ..@ factors : list()
 $ Y: int [1:100] 0 1 0 0 1 0 0 0 1 1 ...</code></pre>
</div>
</div>
</div>
</div>
</section>
<section id="wed-like-a-representative-score-for-genes-within-each-set" class="slide level2">
<h2>We’d like a representative score for genes within each set</h2>

<img data-src="geneset_analysis_files/figure-revealjs/unnamed-chunk-10-1.png" width="960" class="r-stretch"></section>
<section id="what-will-be-a-proper-gene-set-score" class="slide level2">
<h2>What will be a proper gene set score?</h2>
<p>Can we simply aggregate gene-level z-scores (or t-statistics) within each set?</p>
<div class="fragment">
<!-- -->
<h3 id="irizarry-et-al.-2009-using-stouffer-z-score"><a href="doi:10.1177/0962280209351908">Irizarry <em>et al.</em> (2009)</a>, using Stouffer Z-score</h3>
<p><span class="math display">\[S_{k} = \sum_{g \in \mathcal{C}_{k}} z_{g} /  \sqrt{|\mathcal{C}_{k}|} \sim \mathcal{N}\!\left(0, 1\right)\]</span> if <span class="math inline">\(Z_{g} \sim \mathcal{N}\!\left(0,1\right),\,\forall g\)</span></p>
</div>
</section>
<section id="aggregating-z-scores-within-a-set" class="slide level2">
<h2>Aggregating z-scores within a set</h2>
<div class="columns">
<div class="column" style="width:.4;">
<p>Function to run a t test for each gene:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>run.t.test <span class="ot">&lt;-</span> <span class="cf">function</span>(X, Y){</span>
<span id="cb10-2"><a href="#cb10-2"></a>  .case <span class="ot">&lt;-</span> Y <span class="sc">==</span> <span class="dv">1</span></span>
<span id="cb10-3"><a href="#cb10-3"></a>  .ctrl <span class="ot">&lt;-</span> Y <span class="sc">==</span> <span class="dv">0</span></span>
<span id="cb10-4"><a href="#cb10-4"></a>  .fun <span class="ot">&lt;-</span> <span class="cf">function</span>(x){</span>
<span id="cb10-5"><a href="#cb10-5"></a>    <span class="fu">t.test</span>(x[.case],</span>
<span id="cb10-6"><a href="#cb10-6"></a>           x[.ctrl])<span class="sc">$</span>statistic</span>
<span id="cb10-7"><a href="#cb10-7"></a>  }</span>
<span id="cb10-8"><a href="#cb10-8"></a>  <span class="fu">apply</span>(X, <span class="dv">2</span>, .fun)</span>
<span id="cb10-9"><a href="#cb10-9"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Function to compute Stouffer Z-score</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>geneset.score <span class="ot">&lt;-</span> <span class="cf">function</span>(X, S, Y) {</span>
<span id="cb11-2"><a href="#cb11-2"></a>    z.genes <span class="ot">&lt;-</span> <span class="fu">run.t.test</span>(X, Y)</span>
<span id="cb11-3"><a href="#cb11-3"></a>    n.sets <span class="ot">&lt;-</span> <span class="fu">apply</span>(S, <span class="dv">1</span>, sum)</span>
<span id="cb11-4"><a href="#cb11-4"></a>    z.sets <span class="ot">&lt;-</span> (S <span class="sc">%*%</span> z.genes <span class="sc">/</span></span>
<span id="cb11-5"><a href="#cb11-5"></a>               <span class="fu">sqrt</span>(n.sets))</span>
<span id="cb11-6"><a href="#cb11-6"></a>}</span>
<span id="cb11-7"><a href="#cb11-7"></a>z.sets <span class="ot">&lt;-</span> <span class="fu">geneset.score</span>(dat<span class="sc">$</span>X, dat<span class="sc">$</span>S, dat<span class="sc">$</span>Y)     </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div><div class="column" style="width:.45;">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="geneset_analysis_files/figure-revealjs/unnamed-chunk-14-1.png" width="528"></p>
</div>
</div>
<div class="cell">
<div class="cell-output-display">
<p><img data-src="geneset_analysis_files/figure-revealjs/unnamed-chunk-16-1.png" width="528"></p>
</div>
</div>
</div>
</div>
</section>
<section id="constructing-the-null-distribution-by-gene-permutation" class="slide level2">
<h2>Constructing the null distribution by gene permutation</h2>
<p>What if we don’t know the distribution of set-wise scores?</p>
<div class="columns">
<div class="column" style="width:.65;">
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>S.perm <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">apply</span>(dat<span class="sc">$</span>S, <span class="dv">1</span>, sample))</span>
<span id="cb12-2"><a href="#cb12-2"></a>z.perm <span class="ot">&lt;-</span> <span class="fu">geneset.score</span>(dat<span class="sc">$</span>X, S.perm, dat<span class="sc">$</span>Y)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output-display">
<p><img data-src="geneset_analysis_files/figure-revealjs/unnamed-chunk-18-1.png" width="960"></p>
</div>
</div>
</div><div class="column" style="width:.35;">
<p>How it goes:</p>
<ul>
<li><p>Repeatedly permuting gene set membership matrix while preserving the number of genes within each set</p></li>
<li><p>Compute set-level z-scores (or a similar kind) and construct null distribution</p></li>
<li><p>Calculate p-values by counting the frequency of observed <span class="math inline">\(S_{k}^{*} &gt; S_{k}^{\textsf{perm}}\)</span></p></li>
</ul>
</div>
</div>
</section>
<section id="constructing-the-null-distribution-by-sample-permutation" class="slide level2">
<h2>Constructing the null distribution by sample permutation</h2>
<div class="columns">
<div class="column" style="width:.65;">
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>Y.perm <span class="ot">&lt;-</span> <span class="fu">sample</span>(dat<span class="sc">$</span>Y)</span>
<span id="cb13-2"><a href="#cb13-2"></a>z.perm <span class="ot">&lt;-</span> <span class="fu">geneset.score</span>(dat<span class="sc">$</span>X, S.perm, Y.perm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output-display">
<p><img data-src="geneset_analysis_files/figure-revealjs/unnamed-chunk-20-1.png" width="960"></p>
</div>
</div>
</div><div class="column" style="width:.35;">
<ul>
<li><p>Repeat the permutation of case-control labels while preserving the same number of cases and controls</p></li>
<li><p>Compute set-level z-scores (or a similar kind) and construct null distribution</p></li>
<li><p>Calculate p-values by counting the frequency of observed <span class="math inline">\(S_{k}^{*} &gt; S_{k}^{\textsf{perm}}\)</span></p></li>
</ul>
</div>
</div>
</section>
<section id="summary-of-rank-based-gene-set-analysis" class="slide level2">
<h2>Summary of rank-based gene set analysis</h2>
<div class="columns">
<div class="column" style="width:.5;">
<p><strong>When does it apply?</strong></p>
<ul>
<li><p>We have ranked all our genes by a score/test statistic (or compute a summary score for every gene set of interest)</p></li>
<li><p>We want to know whether any of our gene sets of interest tend to have more highly ranked genes (or higher summary scores)</p></li>
</ul>
</div><div class="column" style="width:.5;">
<p><strong>When does it not apply?</strong></p>
<ul>
<li>There is a hidden factor that can affect two steps: (1) gene set selection (annotations/knowledge) and (2) gene scoring</li>
</ul>
</div>
</div>
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Note</strong></p>
</div>
<div class="callout-content">
<p>Estimating the null distribution of ranks/scores over gene sets of interest is nontrivial</p>
</div>
</div>
</div>

<img src="https://github.com/STAT540-UBC/stat540-ubc.github.io/raw/main/images/stat540-logo-s.png" class="slide-logo r-stretch"><div class="footer footer-default">

</div>
</section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="geneset_analysis_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="geneset_analysis_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="geneset_analysis_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="geneset_analysis_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="geneset_analysis_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="geneset_analysis_files/libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="geneset_analysis_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="geneset_analysis_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="geneset_analysis_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="geneset_analysis_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="geneset_analysis_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1600,

        height: 900,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>