<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Statistical Inference for RNA-seq - Part II</title>
    <meta charset="utf-8" />
    <meta name="author" content="  Keegan Korthauer " />
    <script src="libs/header-attrs-2.11/header-attrs.js"></script>
    <link rel="stylesheet" href="xaringan-themer.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# <big>Statistical Inference for RNA-seq - Part II</big>
### <br><font size=6> Keegan Korthauer </font>
### <font size=6>14 February 2022 </font> <br><br><font size=4> with slide contributions from Paul Pavlidis </font>

---






&lt;style type="text/css"&gt;
pre {
  white-space: pre-wrap;
}
.remark-code {
  background: #f8f8f8;
}
.remark-inline-code {
  background: "white";
}
.remark-code {
  font-size: 22px;
}
.huge .remark-code { /*Change made here*/
  font-size: 200% !important;
}
.tiny .remark-code { /*Change made here*/
  font-size: 60% !important;
}
.smaller .remark-code { /*Change made here*/
  font-size: 90% !important;
}
.smaller2 .remark-code { /*Change made here*/
  font-size: 80% !important;
}
.smaller3 .remark-code { /*Change made here*/
  font-size: 70% !important;
}
&lt;/style&gt;

&lt;style&gt;
div.blue { background-color:#e8f2f6; border-radius: 5px; padding: 20px;}
&lt;/style&gt;

# Learning objectives (lectures 11 and 12)

&lt;big&gt;

* Understand *why* and *when* between and within sample normalization are needed

* Apply common between and within sample normalization approaches to RNA-seq counts

* Understand why the *count nature* of RNA-seq data requires modification to the Differential Expression approaches applied to microarray data (e.g. `limma`)

* Apply models such as **limma-trend, limma-voom, `DESeq2` and `edgeR`** for inference of Differential Expression 

---



## How do we handle these M-V relationships in our analysis?

.pull-left[
&lt;big&gt;
Options we discussed last time:

* Use a non-parametric test 
  
* Make adjustments and model as usual
  
* Use a model specific for count data
]

.pull-right[
&lt;img src="lect11-RNAseq2_files/figure-html/unnamed-chunk-3-1.png" width="468" style="display: block; margin: auto;" /&gt;
]

---

# One option: Voom


### Mean-variance modelling at the observational level

* Falls under the category *"Make adjustments and model as usual"*

* Specifically, adjustment to regular `lm` to take into account the M-V relationship

* **Key idea 1**: heteroskedasticity leads to higher variance observations getting more weight in minimization of error than they should

--

* **Key idea 2**: modeling the mean-variance relationship is more important than getting the probability distribution exactly right (i.e. don't bother with other distributions like Poisson, Binomial, etc)

* Proposed in ["voom: precision weights unlock linear model analysis tools for RNA-seq read counts" by Law et al. (2014)](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29)

---

# Voom implementation

&lt;big&gt;

* Input:
  1. **raw counts** (required to estimate M-V relationship), but modeling is done on log-transformed CPM values `\(((log_2(CPM + 0.5)\)` to be precise) 
  
  2. design matrix
  
* Output: precision weights and moderated `\(t\)`-statistics

* Implemented in [`limma` package](https://bioconductor.org/packages/release/bioc/html/limma.html): `voom()` function

---


# Voom steps


1. Fit linear model to `\(log_2(CPM_{ig} + 0.5)\)` values (samples `\(i\)`) for each gene `\(g\)`

2. Extract the fitted quarter-root error variance estimates `\(s^{1/2}_g = \sqrt{sd(\hat{\varepsilon_{ig}})}\)`

3. Fit a smoothed line `\(\hat{f}\)` to the trend between mean log counts and `\(s^{1/2}_g\)` using [lowess](https://rafalab.github.io/dsbook/smoothing.html#local-weighted-regression-loess) (locally weighted regression)

--

4. Use the fitted lowess curve to estimate **precision weights**: `\(w_{ig} = \frac{1}{\hat{f}(\hat{c}_{ig})^4}\)` where `\(\hat{c}_{ig}\)` are the `\(log_2\)` *fitted* counts (estimated from model in step 1)

5. Fit linear model to `\(log_2(CPM_{ig} + 0.5)\)` values using **precision weights** `\(w_{ig}\)`

6. Compute moderated `\(t\)`-statistics as before (using `eBayes` from `limma`)

---

# Voom illustration

![](https://media.springernature.com/full/springer-static/image/art%3A10.1186%2Fgb-2014-15-2-r29/MediaObjects/13059_2013_Article_3227_Fig2_HTML.jpg)

Figure 2, [Law et. al, 2014](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4053721/#B34)

--

`$$w_{ig} = \frac{1}{\hat{f}(\hat{c}_{ig})^4} = \frac{1}{(\sqrt{s_{ig}})^4} = \frac{1}{s_{ig}^2}$$`

---

# lowess

.pull-left[

* locally weighted regression fits a smooth curve to approximate the relationship between independent and dependent variables

* Each smoothed value is given by a weighted linear least squares regression over the **span** (a neighborhood of the independent variable)

* Smoothing span is adjustable

* Generalization to locally weighted polynomial regression and inclusion of multiple independent variables: `loess`

GIF source (not rendered in PDF): ["Introduction to Data Science" by Irizarry](https://rafalab.github.io/dsbook/smoothing.html#local-weighted-regression-loess)
]

.pull-right[
[![](https://rafalab.github.io/dsbook/ml/img/loess-animation.gif)](https://rafalab.github.io/dsbook/ml/img/loess-animation.gif)

]

---

# Why quarter-root variance?

* The **coefficient of variation** `\((CV = \frac{\sigma}{\mu})\)` for RNA-seq counts is roughly `\(\sqrt{\frac{1}{\lambda} + \phi}\)`
  
  * `\(\lambda:\)` expected size of count; arises from technical variability associated with sequencing and gradually decreases with increasing count size 

  * `\(\phi:\)` measure of biological variation (*overdispersion*); roughly constant

--

* Standard deviation of `\(log_2(CPM)\)` is approximately equal to CV of the counts (by Taylor's theorem) 

`$$sd(log_2(CPM)) \approx \sqrt{\frac{1}{\lambda} + \phi}$$`
---

## Why quarter-root variance?

&lt;div class="blue"&gt; CV of RNA-seq counts should be a decreasing function of count size for small to moderate counts, and asymptote to a value that depends on biological variability&lt;/div&gt;

&lt;img src="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4053721/bin/gb-2014-15-2-r29-1.jpg" width="450" style="display: block; margin: auto;" /&gt;

[Law et al. 2014](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4053721/): Panels (a)-(e) represent datasets with increasing expected biological variability

--

Square root of standard deviation used as distribution is more symmetric (i.e. less skewed) 

---

class: middle

### How can we incorporate these precision weights in the regression fit?

---

# Weighted least squares (WLS) regression

* OLS: `\(\boldsymbol{\hat{\beta}}_g = \boldsymbol{(X^TX)^{-1}X^T}\boldsymbol{y}_g\)`

* WLS: `\(\boldsymbol{\hat{\beta}}_g = \boldsymbol{(X^T}\boldsymbol{W}_g\boldsymbol{X)^{-1}X^T}\boldsymbol{W}_g\boldsymbol{y}_g\)`, where `\(\boldsymbol{W}_g\)` is a diagonal matrix of weights for gene `\(g\)`

* **Intuition**: in minimizing the sum of squared errors, we put less weight on data points that are less precise: 

`$$\boldsymbol{\hat{\beta}}_{g} = argmin_{\beta_{g1}, ..., \beta_{gp}} \Big( \sum_{i=1}^n w_{ig}(x_{i1} \beta_{g1} + ... + x_{ip} \beta_{gp} - y_{ig})^2 \Big)$$`
--

* Optimal weights for this purpose: inverse variance

* Remedies heteroskedasticity 

* Note: parameter estimates `\(\hat{\boldsymbol{\beta}}_g\)` assume weights (variances) are known

---

# limma-voom


* **limma-voom** is the application of `limma` to `\(log_2(CPM + 0.5)\)` values, with inverse variance observational weights estimated from the M-V trend

* This alleviates the problem of heteroskedasticity and (hopefully) improves estimates of residual standard error

* Gene-specific variance estimates are 'shrunken' to borrow information across all genes:

`$$\tilde{s}^2_g = \frac{d_0s_0^2 + ds^2_g}{d_0 + d}$$` 

--

* Note that `\(s^2_g\)` estimates are affected by voom weights
  
  * recall that `\(s^2_g\)` is the sum of squared residuals `\(\frac{1}{n-p}\hat{\boldsymbol{\epsilon}_g}^T\hat{\boldsymbol{\epsilon}_g}\)`
  * under WLS `\(\boldsymbol{\hat{\epsilon}}_g = \boldsymbol{y}_g - \boldsymbol{X\hat{\beta}}_g = \boldsymbol{y}_g - \boldsymbol{X(X^T}\boldsymbol{W}_g\boldsymbol{X)^{-1}X^T}\boldsymbol{W}_g\boldsymbol{y}_g\)`
  
---

# limma-voom, continued


* Moderated `\(t\)` statistics are then calculated using the shrunken gene-specific variance estimates: `\(\tilde{t}_{g} = \frac{\hat{\beta}_{ig}}{\tilde{s}_g \sqrt{v_{ii}}}\)`

  * recall that under OLS, `\(v_{ii}\)` is the `\(i^{th}\)` diagonal element of `\(\boldsymbol{(X^TX)}^{-1}\)`
  * under WLS, `\(v_{ii}\)` is the `\(i^{th}\)` diagonal element of `\((\boldsymbol{X^T}\boldsymbol{W}_g\boldsymbol{X})^{-1}\)`

* Degrees of freedom for moderated `\(t\)` statistic: `\(n-p+d_0\)` 

* If `\(d_0\)` is large compared to `\(n-p\)`, moderated statistics have a bigger effect compared to using regular `\(t\)` statistics
  * i.e. in general, shrinkage matters more for small sample sizes

---

# Differential expression analysis on Chd8 data

* Recall: Our **additive** model for each gene to test for Group (Chd8 mutant vs WT) effect, and adjust for:

  * Sex (M vs F)
  
  * DPC (days post conception, 5 levels)

`$$Y_{i} = \theta + \tau_{Mut} x_{i,Mut} + \tau_{F} x_{i,F} + \tau_{D14.5} x_{i,D14.5} + \tau_{D17.5} x_{i,D17.5} + \tau_{D21} x_{i,D21} + \tau_{D77} x_{i,D77} + \epsilon_i$$`
&lt;small&gt;
`$$x_{i,Mut} = \bigg\{\begin{array}{l} 
1\text{ if sample } i \text{ is Mutant} \\
0 \text{ otherwise}\\
\end{array}, \hspace{1em} x_{i,F} = \bigg\{\begin{array}{l} 
1\text{ if sample } i \text{ is Female} \\
0 \text{ otherwise}\\
\end{array}, \hspace{1em}x_{i,D\#} = \bigg\{\begin{array}{l} 
1\text{ if sample } i \text{ is DPC#} \\
0 \text{ otherwise}\\
\end{array}$$`

where `\(D\# \in \{D14.5, D17.5, D21, D77\}\)`
&lt;big&gt;

* Our model has `\(n-p=44-7=37\)` degrees of freedom

* We will focus on the null hypothesis of the **main effect** of Group `\(H_0: \tau_{Mut}=0\)` 


---

# limma-voom in action

.smaller2[

```r
vw &lt;- voom(assays(sumexp)$counts, 
           design = model.matrix(~ Sex + Group + DPC, data = colData(sumexp)), 
           plot = TRUE, span = 0.5)  
```

&lt;img src="lect11-RNAseq2_files/figure-html/unnamed-chunk-5-1.png" width="576" style="display: block; margin: auto;" /&gt;

```r
lvfit &lt;- lmFit(vw, model.matrix(~ Sex + Group + DPC, data = colData(sumexp)))
lvfit &lt;- eBayes(lvfit)
```
]

---

# limma-voom vs limma

&lt;img src="lect11-RNAseq2_files/figure-html/unnamed-chunk-6-1.png" width="720" style="display: block; margin: auto;" /&gt;

---

# Another option: limma-trend

#### Limma-trend uses the M-V relationship at the **gene** level, whereas voom uses **observational** level trends ([Law et. al, 2014](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4053721/#B34))


* Gene-wise variances are shrunken toward a **global M-V trend**, instead of toward a constant pooled variance:

`$$\tilde{s}^2_g = \frac{d_0s_{0g}^2 + ds^2_g}{d_0 + d}$$` 
--

* Notice the `\(g\)` subscript on `\(s_{0g}^2\)`! The prior variance is different for each gene (unlike in regular limma)

* Based on the M-V trend, `\(s_{0g}^2\)` is (typically) higher for lowly expressed genes

---

# limma-trend vs voom? 

![](https://media.springernature.com/full/springer-static/image/art%3A10.1186%2Fgb-2014-15-2-r29/MediaObjects/13059_2013_Article_3227_Fig2_HTML.jpg)

Figure 2, [Law et. al, 2014](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4053721/#B34)

---

# limma-trend in action

.pull-left[
.smaller2[

```r
mm &lt;- model.matrix(~ Sex + Group + DPC, 
                   data = colData(sumexp))
ltfit &lt;- lmFit(cpm(assays(sumexp)$counts, 
                   log = TRUE), 
               design =  mm)
ltfit &lt;- eBayes(ltfit, trend = TRUE)

# limma-trend s^2_{0g}
str(ltfit$s2.prior)
```

```
##  Named num [1:12021] 0.0287 0.051 0.0274 0.023 0.0268 ...
##  - attr(*, "names")= chr [1:12021] "0610007P14Rik" "0610009B22Rik" "0610009O20Rik" "0610010F05Rik" ...
```

```r
# regular limma s^2_0
str(lfit$s2.prior)
```

```
##  num 0.0334
```
]]

.pull-right[
&lt;img src="lect11-RNAseq2_files/figure-html/unnamed-chunk-8-1.png" width="576" style="display: block; margin: auto;" /&gt;
]

---

# Nuances for limma-trend and limma-voom

&lt;big&gt;

* If M-V relationship is flat, limma-voom and limma-trend have practically no effect

  * for limma-voom, weights will be all equal
  
  * for limma-trend, `\(s_{0g}^2\)` will be constant across genes

* Even if M-V isn't flat, impact is most prominent in lowly expressed genes

---

## limma-voom 'false positives'?

One of the top DE genes by Group according to voom (but not other methods):

&lt;img src="lect11-RNAseq2_files/figure-html/unnamed-chunk-9-1.png" width="1008" style="display: block; margin: auto;" /&gt;

--

* Why does this happen? 
  * Voom weighting causes very low expression values to have little effect on model fit
  * Weights for this gene are about 30-40x higher for DPC 77 observations
  
* Whether this is a false positive is a matter of opinion, but lesson is: *always look at the data*

---

# Another option: directly model counts

.left-column[
&lt;img src="https://raw.githubusercontent.com/Bioconductor/BiocStickers/master/DESeq2/DESeq2.png" width="100" style="display: block; margin: auto;" /&gt;&lt;img src="https://raw.githubusercontent.com/Bioconductor/BiocStickers/master/edgeR/edgeR.png" width="100" style="display: block; margin: auto;" /&gt;
]

.right-column[
* Methods: [`edgeR`](https://bioconductor.org/packages/release/bioc/html/edgeR.html), [`DESeq2`](https://bioconductor.org/packages/release/bioc/html/DESeq2.html)

* Both assume counts have underlying *Negative Binomial distribution* and fit **generalized linear models**

  * **Generalized linear models** (GLM) are a generalization of OLS that allows for response variables that have error distribution models other than a normal distribution
  
* Still fit models gene-by-gene as we've discussed so far

* Many similarities with limma: empirical Bayes-based moderation of parameters and addressing the M-V trend
]

---

# Why Negative Binomial distribution?

* Negative Binomial is also known as a **Poisson-Gamma** mixture

  * i.e. A Poisson with a rate parameter that is Gamma-distributed (instead of fixed)
  
  * The Gamma distribution on means captures the biological variance (overdispersion) that can't be accommodated by Poisson alone

* "Overdispersed Poisson" (variance `\(&gt;\)` mean)

* **Key problem**: estimating dispersion from small datasets is tricky

&lt;img src="img/overdisp.png" width="400" style="display: block; margin: auto;" /&gt;

---

# Negative Binomial GLM

* Gene-specific variance under NB: `\(\sigma_g^2 = \mu_g + \mu_g^2\phi_g\)` 

  * `\(\phi_g\)` is the **dispersion** for gene `\(g\)`

  * if `\(\phi_g=0\)`, get Poisson!

* We can perform inference about `\(\mu_g\)` using GLM (e.g. using likelihood ratio tests)

--

* To do so, we need to treat `\(\phi_g\)` as known (*so first need to estimate it*)

--

### Estimation of dispersion is the main issue addressed by methods like `edgeR` and `DEseq2`

---

## Disperson estimation

* One option is to assume `\(\phi_g\)` is a set parametric function of the mean `\(\mu_g\)` (e.g. quadratic)

* More flexible approach is to use empirical Bayes techniques:

  * Dispersion is gene-specific but moderated toward the observed trend with the mean


&lt;img src="img/dispersion.png" width="700" style="display: block; margin: auto;" /&gt;


&lt;small&gt;
source: [Statistical Analysis of Next Generation Sequencing Data, Chapter 3 by Chen, Lun, and Smyth (2014)](https://link.springer.com/book/10.1007/978-3-319-07212-8)


---

# `DESeq2` vs `edgeR`

* These methods are very similar overall 

* Major differences between the methods lie in how they filter low-count genes, estimate prior degrees of freedom, deal with outliers in dispersion estimation, and moderate dispersion of genes with high within-group variance or low counts
  * Also slight differences in specific types of hypothesis tests (quasi-likelihood in edgeR and Wald test in DESeq2)

* Many of these choices can be altered by changing default parameter settings in both methods (see user manuals)

---

# DESeq2 vs edgeR

.pull-left[
**edgeR**
.tiny[

```r
dge &lt;- DGEList(assays(sumexp)$counts)
dge &lt;- calcNormFactors(dge)
dge &lt;- estimateDisp(dge, 
                    design = model.matrix(~ Sex + Group + DPC, 
                                      data = colData(sumexp)), 
                    robust = TRUE)

edgeR_fit &lt;- glmQLFit(dge, 
                      design = model.matrix(~ Sex + Group + DPC, 
                                      data = colData(sumexp)))
```
]
]
.pull-right[
**DESeq2**
.tiny[

```r
dds &lt;- DESeqDataSet(sumexp, 
                    design = model.matrix(~ Sex + Group + DPC, 
                                      data = colData(sumexp)))
dds &lt;- estimateSizeFactors(dds)
dds &lt;- DESeq(dds)
```

```
## using supplied model matrix
```

```
## using pre-existing size factors
```

```
## estimating dispersions
```

```
## gene-wise dispersion estimates
```

```
## mean-dispersion relationship
```

```
## final dispersion estimates
```

```
## fitting model and testing
```
]
]

---

# How to choose a method?

&lt;img src="img/comparisons.png" width="1200" style="display: block; margin: auto;" /&gt;

---

## Example comparison 1: [Love et al. (2018)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6178912.2/)

&lt;img src="https://f1000researchdata.s3.amazonaws.com/manuscripts/17982/ea0127cc-e217-43f8-ab4c-f0f93e48eb46_figure14.gif" width="500" style="display: block; margin: auto;" /&gt;


---

## Example comparison 2 (for single-cell RNA-seq)

&lt;img src="https://media.springernature.com/lw685/springer-static/image/art%3A10.1038%2Fnmeth.4612/MediaObjects/41592_2018_Article_BFnmeth4612_Fig5_HTML.jpg?as=webp" width="350" style="display: block; margin: auto;" /&gt;

[Soneson &amp; Robinson (2018)](https://www.nature.com/articles/nmeth.4612)

---

# How to choose a method?


* No established gold standards

  * Simulations somewhat unsatisfying (depend on specific settings)
  
  * In real data, the truth is unknown
  
--

&lt;div class = "blue"&gt;
&lt;b&gt;The most popular and widely used methods tend to give similar results&lt;/b&gt;
&lt;div&gt;

* **edgeR** and **DESeq2** are very similar in design
  
  * might be expected to work better for small sample sizes or low read depth
    
* **limma-trend** or **limma-voom** also sound choices

  * work equally well when library sizes don't vary much
  
  * might not do as well when sample size or depth is very low
  
---

# Comparing methods on the Chd8 dataset

#### tl;dr version: there isn't a big difference

Possible reasons why: 

* methods have been converging in approach 

* modeling count data directly with GLMs is more important for smaller samples sizes, lower read depth

Check out the comparisons in detail in the [companion notes](https://github.com/STAT540-UBC/STAT540-UBC.github.io/blob/master/examples/rnaseqdiffex-examples/examples-RNAseq.md)

---

### Comparisons of p-values for Chd8 mutation effect

&lt;img src="https://raw.githubusercontent.com/STAT540-UBC/resources/main/rnaseqdiffex-examples/examples-RNAseq_files/figure-gfm/comparemethods-1.png" width="500" style="display: block; margin: auto;" /&gt;

---

### Correlation of p-value ranks for the effect of Chd8 mutation

&lt;img src="https://raw.githubusercontent.com/STAT540-UBC/resources/main/rnaseqdiffex-examples/examples-RNAseq_files/figure-gfm/unnamed-chunk-28-1.png" width="600" style="display: block; margin: auto;" /&gt;

---

### Overlap of genes with FDR &lt; 0.05 for the effect of Chd8 mutation

&lt;img src="https://raw.githubusercontent.com/STAT540-UBC/resources/main/rnaseqdiffex-examples/examples-RNAseq_files/figure-gfm/unnamed-chunk-29-1.png" width="900" style="display: block; margin: auto;" /&gt;

---

### Heatmap of top 30 genes by limma-voom applied to TMM

&lt;img src="https://raw.githubusercontent.com/STAT540-UBC/resources/main/rnaseqdiffex-examples/examples-RNAseq_files/figure-gfm/hm-2.png" width="900" style="display: block; margin: auto;" /&gt;

---

### Heatmap of top 30 genes by limma-trend, adjusted for DPC effect

&lt;img src="https://raw.githubusercontent.com/STAT540-UBC/resources/main/rnaseqdiffex-examples/examples-RNAseq_files/figure-gfm/adjhm-1.png" width="900" style="display: block; margin: auto;" /&gt;





    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false,
"ratio": "16:9"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
