<!DOCTYPE html>
<html lang="en"><head>
<script src="lect09-multipleTesting_files/libs/clipboard/clipboard.min.js"></script>
<script src="lect09-multipleTesting_files/libs/quarto-html/tabby.min.js"></script>
<script src="lect09-multipleTesting_files/libs/quarto-html/popper.min.js"></script>
<script src="lect09-multipleTesting_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="lect09-multipleTesting_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="lect09-multipleTesting_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="lect09-multipleTesting_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="lect09-multipleTesting_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.3.450">

  <meta name="author" content="Keegan Korthauer">
  <meta name="dcterms.date" content="2025-02-06">
  <title>More limma &amp; multiple testing</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="lect09-multipleTesting_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="lect09-multipleTesting_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="lect09-multipleTesting_files/libs/revealjs/dist/theme/quarto.css">
  <link href="lect09-multipleTesting_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="lect09-multipleTesting_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="lect09-multipleTesting_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="lect09-multipleTesting_files/libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="lect09-multipleTesting_files/libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="lect09-multipleTesting_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" data-background-color="#197aa0" data-background-image="https://github.com/STAT540-UBC/stat540-ubc.github.io/raw/main/images/stat540-logo-s.png" data-background-opacity="0.9" data-background-position="95% 90%" data-background-size="12%" class="quarto-title-block center">
  <h1 class="title">More <code>limma</code> &amp; multiple testing</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Keegan Korthauer 
</div>
</div>
</div>

  <p class="date">February 6, 2025</p>
</section>
<section id="learning-objectives" class="slide level2">
<h2>Learning objectives</h2>
<ol type="1">
<li><p>Use <code>limma</code> to perform <strong>genome-wide differential expression testing</strong> on microarray data</p></li>
<li><p>Understand the key advantages between <code>limma</code> and <code>lm</code> (standard linear regression)</p></li>
<li><p>Explain why <strong>multiple testing</strong> increases the number of errors we make by chance</p></li>
<li><p>Be able to adjust for multiple comparisons by controlling the <strong>False Discovery Rate</strong></p>
<ul>
<li>e.g.&nbsp;using Benjamini-Hochberg or Storey’s q-value</li>
</ul></li>
</ol>
</section>
<section id="recall-the-hybrid-estimator-in-limma" class="slide level2">
<h2>Recall: The hybrid estimator in <code>limma</code></h2>
<p><span class="math display">\[\tilde{s}^2_g = \frac{d_0}{d_0 + d}s_0^2 + \frac{d}{d_0 + d}s^2_g\]</span></p>
<p>Recall that <span class="math inline">\((s_0, d_0)\)</span> are the <em>prior</em> parameters for <span class="math inline">\(\sigma^2_g\)</span> (random variable):</p>
<p><span class="math display">\[\frac{1}{\sigma^2_g} \sim \frac{1}{d_0s_0^2} \chi^2_{d_0}\]</span></p>
<div class="fragment">
<p>The prior parameters incorporate information from all genes which allows us to <strong>shrink/nudge the gene-specific variances toward a common consensus</strong></p>
<ul>
<li><p>they are estimated from the data - the formulas for <span class="math inline">\((s_0, d_0)\)</span> and their derivations are beyond the scope of the course, but <code>limma</code> takes care of the details for us</p></li>
<li><p>note that <span class="math inline">\((s_0, d_0)\)</span> <strong>do not depend on</strong> <span class="math inline">\(g\)</span></p></li>
</ul>
</div>
</section>
<section id="illustration" class="slide level2">
<h2>Illustration</h2>

<img data-src="https://media.springernature.com/lw685/springer-static/image/art%3A10.1186%2Fs12859-019-3248-9/MediaObjects/12859_2019_3248_Fig5_HTML.png?as=webp" class="r-stretch quarto-figure-center"><p>Figure 5 from <a href="https://doi.org/10.1186/s12859-019-3248-9">https://doi.org/10.1186/s12859-019-3248-9</a></p>
</section>
<section id="moderated-t-statistic" class="slide level2">
<h2>Moderated <em>t</em> statistic</h2>
<ul>
<li>plug in this posterior mean estimate to obtain a ‘moderated’ <em>t</em> statistic:</li>
</ul>
<p><span class="math display">\[\large \tilde{t}_{gj} = \frac{\hat{\alpha}_{gj}}{\tilde{s}_g \sqrt{v_{jj}}}\]</span></p>
<ul>
<li><p>Under limma’s assumptions, we know the null distribution for <span class="math inline">\(\tilde{t}_{gj}\)</span> under <span class="math inline">\(H_0\)</span>: <span class="math display">\[\tilde{t}_{gj} \sim t_{d_0+d}\]</span></p></li>
<li><p>parameters from the prior <span class="math inline">\(d_0\)</span> and <span class="math inline">\(s_0^2\)</span> are estimated from the data</p></li>
</ul>
<div class="fragment">
<div class="columns">
<div class="column" style="width:80%;">
<ul>
<li>This is how limma is a <strong>hybrid</strong> of frequentist (<em>t</em>-statistic) and Bayesian (hierarchical model) approaches (i.e.&nbsp;empirical Bayes)</li>
</ul>
</div><div class="column" style="width:20%;">
<p><img data-src="img/hybrid.jpg"></p>
</div>
</div>
</div>
</section>
<section id="side-by-side-comparison-of-key-quantities-and-results" class="slide level2">
<h2>Side-by-side comparison of key quantities and results</h2>
<table style="width:100%;">
<colgroup>
<col style="width: 42%">
<col style="width: 28%">
<col style="width: 28%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th style="text-align: center;">OLS</th>
<th style="text-align: center;">limma</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Estimated gene-wise residual variance<sup>1</sup><sup>2</sup>:</td>
<td style="text-align: center;"><span class="math inline">\({s}_g^2 = \frac{1}{d} \hat{\boldsymbol\varepsilon}^T \hat{\boldsymbol\varepsilon}\)</span></td>
<td style="text-align: center;"><span class="math inline">\(\tilde{s}^2_g = \frac{d_0s_0^2 + ds^2_g}{d_0 + d}\)</span></td>
</tr>
<tr class="even">
<td><em>t</em>-statistic for <span class="math inline">\(H_0: \alpha_{gj} = 0\)</span>:</td>
<td style="text-align: center;"><span class="math inline">\({t}_{gj} = \frac{\hat{\alpha}_{gj}}{s_g \sqrt{v_{jj}}}\)</span></td>
<td style="text-align: center;"><span class="math inline">\(\tilde{t}_{gj} = \frac{\hat{\alpha}_{gj}}{\tilde{s}_g \sqrt{v_{jj}}}\)</span></td>
</tr>
<tr class="odd">
<td>Distribution of <em>t</em>-statistic under <span class="math inline">\(H_0\)</span>:</td>
<td style="text-align: center;"><span class="math inline">\({t}_{gj} \sim t_{d}\)</span></td>
<td style="text-align: center;"><span class="math inline">\(\tilde{t}_{gj} \sim t_{d_0+d}\)</span></td>
</tr>
</tbody>
</table>
<aside><ol class="aside-footnotes"><li id="fn1"><p>limma can handle more complicated models where this is not the same for each gene, so this is actually a vector of 45K copies of the number 35</p></li><li id="fn2"><p>Here the degrees of freedom of our linear model are <span class="math inline">\(d = n - p - 1\)</span>, where <span class="math inline">\(n\)</span> is the sample size and <span class="math inline">\(p\)</span> is the number of covariates</p></li></ol></aside></section>
<section id="moderated-limma-vs-traditional-lm-tests" class="slide level2">
<h2>Moderated (<code>limma</code>) vs traditional (<code>lm</code>) tests</h2>
<ul>
<li><strong>moderated variances</strong> will be <em>“shrunk” toward the typical gene-wise variance</em>, relative to to raw sample residual variances</li>
</ul>
<div class="fragment">
<ul>
<li><strong>degrees of freedom</strong> for null distribution <em>increase</em> relative to default <span class="math inline">\((d \text{ vs } d_0 + d)\)</span>
<ul>
<li><span class="math inline">\(\rightarrow\)</span> makes it closer to a standard normal</li>
<li><span class="math inline">\(\rightarrow\)</span> makes tail probabilities (p-values) smaller</li>
<li><span class="math inline">\(\rightarrow\)</span> easier to reject the null</li>
</ul></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>overall, when all is well, limma will deliver statistical results that are <strong>more stable</strong> and <strong>more powerful</strong></li>
</ul>
</div>
</section>
<section id="overview-limma-workflow" class="slide level2">
<h2>Overview: <code>limma</code> workflow</h2>

<img data-src="img/limma_flow.png" style="width:68.0%" class="r-stretch quarto-figure-center"></section>
<section id="functions-that-make-your-life-easier" class="slide level2">
<h2>Functions that make your life easier</h2>
<table>
<colgroup>
<col style="width: 42%">
<col style="width: 57%">
</colgroup>
<thead>
<tr class="header">
<th>Function</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>model.matrix</code></td>
<td>Takes in your data frame and makes a design matrix</td>
</tr>
<tr class="even">
<td><code>limma::lmFit</code></td>
<td>Fits the linear model to each gene separately – replace gene with “feature” depending on your data (‘industrial scale’ <code>lm</code>)</td>
</tr>
<tr class="odd">
<td><code>limma::eBayes</code></td>
<td>Use output of linear regression to compute moderated <em>t</em> statistics</td>
</tr>
<tr class="even">
<td><code>limma::topTable</code></td>
<td>Query your results; sort your p-values; sort genes; Adjust for multiple comparisons</td>
</tr>
<tr class="odd">
<td><code>limma::decideTests</code></td>
<td>Identify which genes are significantly differentially expressed for each contrast</td>
</tr>
<tr class="even">
<td><code>limma::makeContrasts</code></td>
<td>Create the contrast matrix <span class="math inline">\(C\)</span> that you desire</td>
</tr>
<tr class="odd">
<td><code>limma::contrast.fit</code></td>
<td>Apply a contrast to your estimates</td>
</tr>
</tbody>
</table>
</section>
<section id="getting-help" class="slide level2">
<h2>Getting help</h2>

<img data-src="img/limma_help.png" class="r-stretch quarto-figure-center"><ul>
<li><a href="https://bioconductor.org/packages/release/bioc/html/limma.html" data-preview-link="true">Bioconductor homepage for limma</a></li>
<li>Bring up help pages for specific functions in RStudio, e.g.&nbsp;<code>?limma::topTable</code></li>
</ul>
</section>
<section id="limma-step-one-lmfit" class="slide level2">
<h2><code>limma</code> step one: <code>lmFit</code></h2>
<div class="columns">
<div class="column" style="width:70%;">
<p><span class="math display">\[\mathbf{Y}_g = \mathbf{X} \boldsymbol\alpha_g + \boldsymbol\varepsilon_g,\]</span></p>
<ul>
<li><p>Within each gene observations are iid / constant variance</p></li>
<li><p><code>lmFit()</code> carries out multiple linear regression on each gene</p></li>
<li><p>Usage: <code>lmFit(object, design)</code></p>
<ul>
<li><p><code>object</code> is a data.frame or matrix with features in rows and samples in columns (<em>G</em> genes by <em>N</em> samples), or other Bioconductor object such as <a href="https://bioconductor.org/packages/devel/workflows/vignettes/maEndToEnd/inst/doc/MA-Workflow.html#42_Bioconductor_ExpressionSets"><code>ExpressionSet</code></a></p></li>
<li><p><code>design</code> is a design matrix (output of <code>model.matrix(y ~ x)</code>; <em>N</em> samples by <span class="math inline">\(p+1\)</span> parameters)</p></li>
</ul></li>
</ul>
</div><div class="column" style="width:30%;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/exprset.png"></p>
</figure>
</div>
<p>Bioconductor’s ExpressionSet class; <a href="http://dx.doi.org/10.12688/f1000research.8967.1">image source</a></p>
</div>
</div>
</section>
<section id="lets-run-limma-for-the-interactive-model-with-age-continuous-and-genotype-factor" class="slide level2">
<h2>Let’s run <code>limma</code> for the interactive model with age (continuous) and genotype (factor)</h2>
<p><span class="math display">\[\Large y_{ig} = \theta + \tau_{KO}x_{ig,KO} + \tau_{Age}x_{ig,Age} + \tau_{KO:Age}x_{ig,KO}x_{ig,Age}\]</span></p>
<ul>
<li><p><span class="math inline">\(i\)</span> indexes mouse, <span class="math inline">\(g\)</span> indexes genes</p></li>
<li><p><span class="math inline">\(x_{ig,KO}\)</span> is the indicator variable for the NrlKO group</p></li>
<li><p><span class="math inline">\(x_{ig,Age}\)</span> is the continuous age variable</p></li>
</ul>
</section>
<section id="interactive-model-with-age-and-genotype" class="slide level2">
<h2>Interactive model with age and genotype</h2>
<p>Example gene (but we want to fit this model on all genes):</p>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="co"># read in our dataset</span></span>
<span id="cb1-2"><a href="#cb1-2"></a>eset <span class="ot">&lt;-</span> <span class="fu">getGEO</span>(<span class="st">"GSE4051"</span>, <span class="at">getGPL =</span> <span class="cn">FALSE</span>)[[<span class="dv">1</span>]]</span>
<span id="cb1-3"><a href="#cb1-3"></a></span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="co"># recode time points</span></span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="fu">pData</span>(eset) <span class="ot">&lt;-</span> <span class="fu">pData</span>(eset) <span class="sc">%&gt;%</span></span>
<span id="cb1-6"><a href="#cb1-6"></a>  <span class="fu">mutate</span>(<span class="at">sample_id =</span> geo_accession) <span class="sc">%&gt;%</span></span>
<span id="cb1-7"><a href="#cb1-7"></a>  <span class="fu">mutate</span>(<span class="at">dev_stage =</span>  <span class="fu">case_when</span>(</span>
<span id="cb1-8"><a href="#cb1-8"></a>    <span class="fu">grepl</span>(<span class="st">"E16"</span>, title) <span class="sc">~</span> <span class="st">"E16"</span>,</span>
<span id="cb1-9"><a href="#cb1-9"></a>    <span class="fu">grepl</span>(<span class="st">"P2"</span>, title) <span class="sc">~</span> <span class="st">"P2"</span>,</span>
<span id="cb1-10"><a href="#cb1-10"></a>    <span class="fu">grepl</span>(<span class="st">"P6"</span>, title) <span class="sc">~</span> <span class="st">"P6"</span>,</span>
<span id="cb1-11"><a href="#cb1-11"></a>    <span class="fu">grepl</span>(<span class="st">"P10"</span>, title) <span class="sc">~</span> <span class="st">"P10"</span>,</span>
<span id="cb1-12"><a href="#cb1-12"></a>    <span class="fu">grepl</span>(<span class="st">"4 weeks"</span>, title) <span class="sc">~</span> <span class="st">"P28"</span></span>
<span id="cb1-13"><a href="#cb1-13"></a>  )) <span class="sc">%&gt;%</span></span>
<span id="cb1-14"><a href="#cb1-14"></a>  <span class="fu">mutate</span>(<span class="at">genotype =</span> <span class="fu">case_when</span>(</span>
<span id="cb1-15"><a href="#cb1-15"></a>    <span class="fu">grepl</span>(<span class="st">"Nrl-ko"</span>, title) <span class="sc">~</span> <span class="st">"NrlKO"</span>,</span>
<span id="cb1-16"><a href="#cb1-16"></a>    <span class="fu">grepl</span>(<span class="st">"wt"</span>, title) <span class="sc">~</span> <span class="st">"WT"</span></span>
<span id="cb1-17"><a href="#cb1-17"></a>  ))</span>
<span id="cb1-18"><a href="#cb1-18"></a></span>
<span id="cb1-19"><a href="#cb1-19"></a><span class="co"># reorder factor levels, add continous age variable</span></span>
<span id="cb1-20"><a href="#cb1-20"></a><span class="fu">pData</span>(eset) <span class="ot">&lt;-</span> <span class="fu">pData</span>(eset) <span class="sc">%&gt;%</span></span>
<span id="cb1-21"><a href="#cb1-21"></a>  <span class="fu">mutate</span>(<span class="at">dev_stage =</span> <span class="fu">fct_relevel</span>(dev_stage, <span class="st">"E16"</span>, <span class="st">"P2"</span>, <span class="st">"P6"</span>, <span class="st">"P10"</span>, <span class="st">"P28"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb1-22"><a href="#cb1-22"></a>  <span class="fu">mutate</span>(<span class="at">genotype =</span> <span class="fu">as.factor</span>(genotype)) <span class="sc">%&gt;%</span></span>
<span id="cb1-23"><a href="#cb1-23"></a>  <span class="fu">mutate</span>(<span class="at">genotype =</span> <span class="fu">fct_relevel</span>(genotype, <span class="st">"WT"</span>, <span class="st">"NrlKO"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb1-24"><a href="#cb1-24"></a>  <span class="fu">mutate</span>(<span class="at">age =</span> <span class="fu">ifelse</span>(dev_stage <span class="sc">==</span> <span class="st">"E16"</span>, <span class="sc">-</span><span class="dv">4</span>,</span>
<span id="cb1-25"><a href="#cb1-25"></a>                            <span class="fu">ifelse</span>(dev_stage <span class="sc">==</span> <span class="st">"P2"</span>, <span class="dv">2</span>, </span>
<span id="cb1-26"><a href="#cb1-26"></a>                                   <span class="fu">ifelse</span>(dev_stage <span class="sc">==</span> <span class="st">"P6"</span>, <span class="dv">6</span>, </span>
<span id="cb1-27"><a href="#cb1-27"></a>                                          <span class="fu">ifelse</span>(dev_stage <span class="sc">==</span> <span class="st">"P10"</span>, <span class="dv">10</span>, <span class="dv">28</span>)))))</span>
<span id="cb1-28"><a href="#cb1-28"></a></span>
<span id="cb1-29"><a href="#cb1-29"></a><span class="co"># function to return tidy data that merges expression matrix and metadata</span></span>
<span id="cb1-30"><a href="#cb1-30"></a>toLongerMeta <span class="ot">&lt;-</span> <span class="cf">function</span>(expset) {</span>
<span id="cb1-31"><a href="#cb1-31"></a>    <span class="fu">stopifnot</span>(<span class="fu">class</span>(expset) <span class="sc">==</span> <span class="st">"ExpressionSet"</span>)</span>
<span id="cb1-32"><a href="#cb1-32"></a>    </span>
<span id="cb1-33"><a href="#cb1-33"></a>    expressionMatrix <span class="ot">&lt;-</span> lonExpressionressionMatrix <span class="ot">&lt;-</span> <span class="fu">exprs</span>(expset) <span class="sc">%&gt;%</span> </span>
<span id="cb1-34"><a href="#cb1-34"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb1-35"><a href="#cb1-35"></a>    <span class="fu">rownames_to_column</span>(<span class="st">"gene"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb1-36"><a href="#cb1-36"></a>    <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">!</span>gene, </span>
<span id="cb1-37"><a href="#cb1-37"></a>                 <span class="at">values_to =</span> <span class="st">"expression"</span>,</span>
<span id="cb1-38"><a href="#cb1-38"></a>                 <span class="at">names_to =</span> <span class="st">"sample_id"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb1-39"><a href="#cb1-39"></a>    <span class="fu">left_join</span>(<span class="fu">pData</span>(expset) <span class="sc">%&gt;%</span> <span class="fu">select</span>(sample_id, dev_stage, age, genotype),</span>
<span id="cb1-40"><a href="#cb1-40"></a>            <span class="at">by =</span> <span class="st">"sample_id"</span>)</span>
<span id="cb1-41"><a href="#cb1-41"></a>  <span class="fu">return</span>(expressionMatrix)</span>
<span id="cb1-42"><a href="#cb1-42"></a>}</span>
<span id="cb1-43"><a href="#cb1-43"></a></span>
<span id="cb1-44"><a href="#cb1-44"></a><span class="co"># pull out gene of interest</span></span>
<span id="cb1-45"><a href="#cb1-45"></a>Tmem_dat <span class="ot">&lt;-</span> <span class="fu">toLongerMeta</span>(eset) <span class="sc">%&gt;%</span> </span>
<span id="cb1-46"><a href="#cb1-46"></a>  <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"1441811_x_at"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb1-47"><a href="#cb1-47"></a>  <span class="fu">mutate</span>(<span class="at">gene =</span> <span class="st">"Tmem176a"</span>) </span>
<span id="cb1-48"><a href="#cb1-48"></a></span>
<span id="cb1-49"><a href="#cb1-49"></a>Tmem_plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(Tmem_dat, <span class="fu">aes</span>(<span class="at">x =</span> age, <span class="at">y =</span> expression, <span class="at">colour =</span> genotype)) <span class="sc">+</span> </span>
<span id="cb1-50"><a href="#cb1-50"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb1-51"><a href="#cb1-51"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>, <span class="at">se =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb1-52"><a href="#cb1-52"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Tmem176a"</span>) <span class="sc">+</span></span>
<span id="cb1-53"><a href="#cb1-53"></a>  <span class="fu">ylim</span>(<span class="dv">5</span>, <span class="dv">11</span>) <span class="sc">+</span></span>
<span id="cb1-54"><a href="#cb1-54"></a>  <span class="fu">xlab</span>(<span class="st">"age (days)"</span>) </span>
<span id="cb1-55"><a href="#cb1-55"></a>Tmem_plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="lect09-multipleTesting_files/figure-revealjs/unnamed-chunk-2-1.png" width="960" class="r-stretch quarto-figure-center"></section>



<section id="computation-is-fast" class="slide level2">
<h2>Computation is fast</h2>
<ul>
<li>Bioconductor way:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="fu">system.time</span>(gFit <span class="ot">&lt;-</span> <span class="fu">lmFit</span>(eset, </span>
<span id="cb14-2"><a href="#cb14-2"></a>                          <span class="fu">model.matrix</span>(<span class="sc">~</span> genotype<span class="sc">*</span>age, <span class="at">data =</span> <span class="fu">pData</span>(eset))))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   user  system elapsed 
  0.038   0.003   0.042 </code></pre>
</div>
</div>
<ul>
<li>Equivalently, using the ‘separate’ way:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a>gFit <span class="ot">&lt;-</span> <span class="fu">lmFit</span>(myDat, </span>
<span id="cb16-2"><a href="#cb16-2"></a>              <span class="fu">model.matrix</span>(<span class="sc">~</span> genotype<span class="sc">*</span>age, <span class="at">data =</span> myMeta))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Using <code>lmFit</code> to fit an interactive model on 45K probesets takes a fraction of a second</li>
</ul>
<div class="fragment">
<ul>
<li>The time-intensive parts of an analysis: data wrangling, selecting a model, choosing how to parameterize it, and interpreting it</li>
</ul>
</div>
</section>
<section id="output-of-step-1-lmfit" class="slide level2">
<h2>Output of Step 1 (<code>lmFit</code>)</h2>
<div class="columns">
<div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a><span class="fu">summary</span>(gFit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                 Length Class  Mode     
coefficients     180404 -none- numeric  
rank                  1 -none- numeric  
assign                4 -none- numeric  
qr                    5 qr     list     
df.residual       45101 -none- numeric  
sigma             45101 -none- numeric  
cov.coefficients     16 -none- numeric  
stdev.unscaled   180404 -none- numeric  
pivot                 4 -none- numeric  
Amean             45101 -none- numeric  
method                1 -none- character
design              156 -none- numeric  </code></pre>
</div>
</div>
</div><div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a><span class="fu">nrow</span>(eset)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Features 
   45101 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a><span class="fu">nrow</span>(eset)<span class="sc">*</span><span class="dv">4</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Features 
  180404 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a><span class="fu">ncol</span>(eset)<span class="sc">*</span><span class="dv">4</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Samples 
    156 </code></pre>
</div>
</div>
<div class="fragment">
<ul>
<li><p>OK… but where are the shrunken variance estimates?? How do I pull out p-values??</p></li>
<li><p>Actually, we haven’t carried out the empirical Bayesian computation yet – still need to run <code>eBayes()</code>!</p></li>
</ul>
</div>
</div>
</div>
</section>
<section id="step-2-moderated-t-tests-using-ebayes" class="slide level2">
<h2>Step 2: Moderated <em>t</em>-tests using <code>eBayes</code></h2>
<div class="columns">
<div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a><span class="fu">summary</span>(gFit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                 Length Class  Mode     
coefficients     180404 -none- numeric  
rank                  1 -none- numeric  
assign                4 -none- numeric  
qr                    5 qr     list     
df.residual       45101 -none- numeric  
sigma             45101 -none- numeric  
cov.coefficients     16 -none- numeric  
stdev.unscaled   180404 -none- numeric  
pivot                 4 -none- numeric  
Amean             45101 -none- numeric  
method                1 -none- character
design              156 -none- numeric  </code></pre>
</div>
</div>
</div><div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a><span class="fu">summary</span>(ebFit <span class="ot">&lt;-</span> <span class="fu">eBayes</span>(gFit))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                 Length Class  Mode     
coefficients     180404 -none- numeric  
rank                  1 -none- numeric  
assign                4 -none- numeric  
qr                    5 qr     list     
df.residual       45101 -none- numeric  
sigma             45101 -none- numeric  
cov.coefficients     16 -none- numeric  
stdev.unscaled   180404 -none- numeric  
pivot                 4 -none- numeric  
Amean             45101 -none- numeric  
method                1 -none- character
design              156 -none- numeric  
df.prior              1 -none- numeric  
s2.prior              1 -none- numeric  
var.prior             4 -none- numeric  
proportion            1 -none- numeric  
s2.post           45101 -none- numeric  
t                180404 -none- numeric  
df.total          45101 -none- numeric  
p.value          180404 -none- numeric  
lods             180404 -none- numeric  
F                 45101 -none- numeric  
F.p.value         45101 -none- numeric  </code></pre>
</div>
</div>
</div>
</div>
</section>
<section id="components-of-the-empirical-bayes-estimators" class="slide level2">
<h2>Components of the empirical Bayes estimators</h2>
<table>
<colgroup>
<col style="width: 13%">
<col style="width: 38%">
<col style="width: 23%">
<col style="width: 11%">
<col style="width: 13%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">math</th>
<th style="text-align: left;">description</th>
<th style="text-align: left;"><code>limma::eBayes</code></th>
<th style="text-align: left;">dimension</th>
<th style="text-align: center;">in <code>lm</code>?</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;"><span class="math inline">\(s_g^2\)</span></td>
<td style="text-align: left;">gene-specific residual variance</td>
<td style="text-align: left;"><code>sigma^2</code></td>
<td style="text-align: left;">45K</td>
<td style="text-align: center;"><span class="math inline">\(\checkmark\)</span></td>
</tr>
<tr class="even">
<td style="text-align: center;"><span class="math inline">\(d\)</span></td>
<td style="text-align: left;">residual degrees of freedom <span class="math inline">\((n-p-1)\)</span></td>
<td style="text-align: left;"><code>df.residual</code></td>
<td style="text-align: left;">Scalar<sup>1</sup></td>
<td style="text-align: center;"><span class="math inline">\(\checkmark\)</span></td>
</tr>
<tr class="odd">
<td style="text-align: center;"><span class="math inline">\(s_0^2\)</span></td>
<td style="text-align: left;">mean of inverse <span class="math inline">\(\chi^2\)</span> prior for <span class="math inline">\(s_g^2\)</span></td>
<td style="text-align: left;"><code>s2.prior</code></td>
<td style="text-align: left;">Scalar</td>
<td style="text-align: center;"></td>
</tr>
<tr class="even">
<td style="text-align: center;"><span class="math inline">\(d_0\)</span></td>
<td style="text-align: left;">degrees of freedom for the prior</td>
<td style="text-align: left;"><code>df.prior</code></td>
<td style="text-align: left;">Scalar</td>
<td style="text-align: center;"></td>
</tr>
<tr class="odd">
<td style="text-align: center;"><span class="math inline">\(\tilde{s}_g^2\)</span></td>
<td style="text-align: left;">posterior mean of <span class="math inline">\(s_g^2\)</span> (moderated gene-specific residual variance)</td>
<td style="text-align: left;"><code>s2.post</code></td>
<td style="text-align: left;">45K</td>
<td style="text-align: center;"></td>
</tr>
<tr class="even">
<td style="text-align: center;"><span class="math inline">\(\tilde{t}_g\)</span></td>
<td style="text-align: left;">moderated t statistics</td>
<td style="text-align: left;"><code>t</code></td>
<td style="text-align: left;">45K<span class="math inline">\(\times\)</span>p</td>
<td style="text-align: center;"></td>
</tr>
</tbody>
</table>
<aside><ol class="aside-footnotes"><li id="fn3"><p>limma can handle more complicated models where this is not the same for each gene, so this is actually a vector of 45K copies of the number 35</p></li></ol></aside></section>
<section id="step-3-toptable-to-extract-test-output" class="slide level2">
<h2>Step 3: <code>topTable()</code> to extract test output</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a><span class="fu">topTable</span>(fit, <span class="at">coef =</span> <span class="cn">NULL</span>, <span class="at">number =</span> <span class="dv">10</span>, <span class="at">genelist =</span> fit<span class="sc">$</span>genes, <span class="at">adjust.method =</span> <span class="st">"BH"</span>,</span>
<span id="cb29-2"><a href="#cb29-2"></a>         <span class="at">sort.by =</span> <span class="st">"B"</span>, <span class="at">resort.by =</span> <span class="cn">NULL</span>, <span class="at">p.value =</span> <span class="dv">1</span>, <span class="at">lfc =</span> <span class="dv">0</span>, <span class="at">confint =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li><p>Refer to the help page (<code>?topTable</code>) for full details of each argument</p></li>
<li><p>Summary of key <code>topTable</code> arguments:</p>
<ul>
<li><code>coef</code> is the argument where you specify the coefficient(s) you want to test for equality with zero (default is NULL; must be specified)</li>
<li><code>number</code> lets you control size of hit list (default is 10)</li>
<li><code>p.value</code> lets you specify a minimum adjusted p-value cutoff (default is 1)</li>
<li><code>lfc</code> lets you specify a minimum observed effect size - log2 fold change (default is 0)</li>
<li><code>sort.by</code> and <code>resort.by</code> give control over the ordering (default is by “B”: log-odds that the gene is differentially expressed)</li>
<li><code>adjust.method</code> specifies how/if to adjust p-values for multiple testing (default is BH)</li>
</ul></li>
</ul>
</section>
<section id="toptable-in-action-genotypenrlko" class="slide level2">
<h2><code>topTable</code> in action: <code>genotypeNrlKO</code></h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a><span class="fu">topTable</span>(ebFit, <span class="at">coef =</span> <span class="st">"genotypeNrlKO"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                  logFC  AveExpr          t      P.Value    adj.P.Val        B
1450946_at   -4.7134891 8.733103 -15.583683 5.390931e-18 2.431364e-13 28.40938
1442070_at   -0.9669400 8.268122  -9.787407 6.896220e-12 1.555132e-07 16.46086
1443457_at   -1.1530057 6.049428  -9.091412 4.947029e-11 6.029428e-07 14.68843
1422679_s_at -2.1187394 9.495435  -9.002625 6.388954e-11 6.029428e-07 14.45703
1431708_a_at -2.1610625 8.591450  -8.967950 7.062007e-11 6.029428e-07 14.36634
1433050_at   -1.5083887 8.468891  -8.923932 8.021234e-11 6.029428e-07 14.25096
1426288_at   -4.2100657 9.323796  -8.772209 1.246491e-10 8.031144e-07 13.85106
1418108_at    0.8662091 8.260647   8.537984 2.475294e-10 1.395478e-06 13.22713
1450770_at    1.3103576 7.357033   8.128847 8.332369e-10 3.747976e-06 12.11870
1457802_at   -0.8362030 7.778189  -8.117112 8.629984e-10 3.747976e-06 12.08657</code></pre>
</div>
</div>
<ul>
<li><p><code>topTable(ebFit, coef = 2)</code> is equivalent here, but <em>much less informative</em>!!</p></li>
<li><p>What is the null hypothesis here?</p></li>
</ul>
<aside class="notes">
<p><span class="math inline">\(H_0: \tau_{KO}=0\)</span>; this finds genes where the KO differs from WT <em>when age is zero</em></p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="plot-the-top-6-probes-for-genotypenrlko" class="slide level2">
<h2>Plot the top 6 probes for <code>genotypeNrlKO</code></h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1"></a><span class="co"># grab the gene names of the top 6 genes from topTable</span></span>
<span id="cb32-2"><a href="#cb32-2"></a>keep <span class="ot">&lt;-</span> <span class="fu">topTable</span>(ebFit, <span class="at">coef =</span> <span class="st">"genotypeNrlKO"</span>, <span class="at">number =</span> <span class="dv">6</span>) <span class="sc">%&gt;%</span></span>
<span id="cb32-3"><a href="#cb32-3"></a>  <span class="fu">rownames</span>()</span>
<span id="cb32-4"><a href="#cb32-4"></a></span>
<span id="cb32-5"><a href="#cb32-5"></a><span class="co"># extract the data for all 6 genes in tidy format</span></span>
<span id="cb32-6"><a href="#cb32-6"></a>topSixGenotype <span class="ot">&lt;-</span> <span class="fu">toLongerMeta</span>(eset) <span class="sc">%&gt;%</span> </span>
<span id="cb32-7"><a href="#cb32-7"></a>  <span class="fu">filter</span>(gene <span class="sc">%in%</span> keep) <span class="sc">%&gt;%</span></span>
<span id="cb32-8"><a href="#cb32-8"></a>  <span class="fu">mutate</span>(<span class="at">gene =</span> <span class="fu">factor</span>(gene, <span class="at">levels =</span> keep))</span>
<span id="cb32-9"><a href="#cb32-9"></a></span>
<span id="cb32-10"><a href="#cb32-10"></a><span class="fu">ggplot</span>(topSixGenotype, </span>
<span id="cb32-11"><a href="#cb32-11"></a>       <span class="fu">aes</span>(<span class="at">x =</span> age, <span class="at">y =</span> expression, <span class="at">color =</span> genotype)) <span class="sc">+</span> </span>
<span id="cb32-12"><a href="#cb32-12"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb32-13"><a href="#cb32-13"></a>  <span class="fu">xlab</span>(<span class="st">"age (days)"</span>) <span class="sc">+</span></span>
<span id="cb32-14"><a href="#cb32-14"></a>  <span class="fu">facet_wrap</span>( <span class="sc">~</span> gene) <span class="sc">+</span> </span>
<span id="cb32-15"><a href="#cb32-15"></a>  <span class="fu">stat_smooth</span>(<span class="at">method=</span><span class="st">"lm"</span>, <span class="at">se=</span><span class="cn">FALSE</span>, <span class="at">cex=</span><span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="lect09-multipleTesting_files/figure-revealjs/unnamed-chunk-17-1.png" width="1344" class="r-stretch quarto-figure-center"></section>
<section id="toptable-in-action-age" class="slide level2">
<h2><code>topTable</code> in action: <code>age</code></h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1"></a><span class="fu">topTable</span>(ebFit, <span class="at">coef =</span> <span class="st">"age"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                  logFC  AveExpr        t      P.Value    adj.P.Val        B
1451635_at   0.19294061 7.791237 20.78935 3.204717e-22 1.445359e-17 40.60768
1425222_x_at 0.17604111 6.514475 19.72633 1.962954e-21 4.426561e-17 38.79502
1433699_at   0.14875031 6.939973 17.80467 6.442530e-20 7.957605e-16 35.29412
1452243_at   0.15889179 6.840975 17.75658 7.057586e-20 7.957605e-16 35.20253
1435436_at   0.13067119 8.274676 17.61808 9.187120e-20 8.286966e-16 34.93763
1458418_at   0.21277027 6.270187 16.43686 9.326171e-19 7.006477e-15 32.60718
1424977_at   0.07814988 6.505464 16.36079 1.087456e-18 7.006477e-15 32.45260
1431174_at   0.16158579 7.418000 16.01901 2.183055e-18 1.230725e-14 31.75106
1421818_at   0.11250919 7.982278 15.66666 4.531133e-18 2.270651e-14 31.01562
1419069_at   0.09325234 8.222096 15.54775 5.813459e-18 2.621928e-14 30.76457</code></pre>
</div>
</div>
<ul>
<li><p><code>topTable(ebFit, coef = 3)</code> is equivalent here, but <em>much less informative</em>!!</p></li>
<li><p>What is the null hypothesis here?</p></li>
</ul>
<aside class="notes">
<p><span class="math inline">\(H_0: \tau_{Age}=0\)</span>; this finds genes where age significantly affects gene expression <em>for WT</em></p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="plot-the-top-6-probes-for-age" class="slide level2">
<h2>Plot the top 6 probes for <code>age</code></h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1"></a><span class="co"># grab the gene names of the top 6 genes from topTable</span></span>
<span id="cb35-2"><a href="#cb35-2"></a>keep <span class="ot">&lt;-</span> <span class="fu">topTable</span>(ebFit, <span class="at">coef =</span> <span class="st">"age"</span>, <span class="at">number =</span> <span class="dv">6</span>) <span class="sc">%&gt;%</span></span>
<span id="cb35-3"><a href="#cb35-3"></a>  <span class="fu">rownames</span>()</span>
<span id="cb35-4"><a href="#cb35-4"></a></span>
<span id="cb35-5"><a href="#cb35-5"></a><span class="co"># extract the data for all 6 genes in tidy format</span></span>
<span id="cb35-6"><a href="#cb35-6"></a>topSixAge <span class="ot">&lt;-</span> <span class="fu">toLongerMeta</span>(eset) <span class="sc">%&gt;%</span> </span>
<span id="cb35-7"><a href="#cb35-7"></a>  <span class="fu">filter</span>(gene <span class="sc">%in%</span> keep) <span class="sc">%&gt;%</span></span>
<span id="cb35-8"><a href="#cb35-8"></a>  <span class="fu">mutate</span>(<span class="at">gene =</span> <span class="fu">factor</span>(gene, <span class="at">levels =</span> keep))</span>
<span id="cb35-9"><a href="#cb35-9"></a></span>
<span id="cb35-10"><a href="#cb35-10"></a><span class="fu">ggplot</span>(topSixAge, </span>
<span id="cb35-11"><a href="#cb35-11"></a>       <span class="fu">aes</span>( <span class="at">x =</span> age, <span class="at">y =</span> expression, <span class="at">color =</span> genotype)) <span class="sc">+</span> </span>
<span id="cb35-12"><a href="#cb35-12"></a>             <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb35-13"><a href="#cb35-13"></a>             <span class="fu">xlab</span>(<span class="st">"age (days)"</span>) <span class="sc">+</span></span>
<span id="cb35-14"><a href="#cb35-14"></a>             <span class="fu">facet_wrap</span>( <span class="sc">~</span> gene) <span class="sc">+</span> </span>
<span id="cb35-15"><a href="#cb35-15"></a>             <span class="fu">stat_smooth</span>(<span class="at">method=</span><span class="st">"lm"</span>, <span class="at">se=</span><span class="cn">FALSE</span>, <span class="at">cex=</span><span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="lect09-multipleTesting_files/figure-revealjs/unnamed-chunk-19-1.png" width="1344" class="r-stretch quarto-figure-center"></section>
<section id="toptable-in-action-genotypenrlkoage" class="slide level2">
<h2><code>topTable</code> in action: <code>genotypeNrlKO:age</code></h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1"></a><span class="fu">topTable</span>(ebFit, <span class="at">coef =</span> <span class="st">"genotypeNrlKO:age"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                  logFC  AveExpr         t      P.Value    adj.P.Val        B
1451635_at   -0.1871800 7.791237 -14.00463 1.684302e-16 7.596369e-12 27.52010
1425222_x_at -0.1700908 6.514475 -13.23451 9.961393e-16 2.246344e-11 25.74877
1441809_at    0.1336356 6.649440  12.55702 5.037221e-15 7.572790e-11 24.13110
1452243_at   -0.1531326 6.840975 -11.88284 2.670693e-14 3.011273e-10 22.46414
1431174_at   -0.1655894 7.418000 -11.39882 9.161231e-14 8.263614e-10 21.23123
1458418_at   -0.2037314 6.270187 -10.92850 3.122990e-13 2.347500e-09 20.00384
1435436_at   -0.1154500 8.274676 -10.80855 4.289363e-13 2.763637e-09 19.68613
1446715_at   -0.1409774 5.613122 -10.57909 7.912260e-13 4.325169e-09 19.07307
1442190_at   -0.1323038 5.896077 -10.54672 8.630965e-13 4.325169e-09 18.98601
1416306_at    0.1702059 6.560050  10.41285 1.238250e-12 5.584630e-09 18.62456</code></pre>
</div>
</div>
<ul>
<li><p><code>topTable(ebFit, coef = 4)</code> is equivalent here, but <em>much less informative</em>!!</p></li>
<li><p>What is the null hypothesis here?</p></li>
</ul>
<aside class="notes">
<p><span class="math inline">\(H_0: \tau_{KO:Age}=0\)</span>; this finds genes where the effect of age is significantly different in each genotype</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="plot-the-top-6-probes-for-genotypenrlkoage" class="slide level2">
<h2>Plot the top 6 probes for <code>genotypeNrlKO:age</code></h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1"></a><span class="co"># grab the gene names of the top 6 genes from topTable</span></span>
<span id="cb38-2"><a href="#cb38-2"></a>keep <span class="ot">&lt;-</span> <span class="fu">topTable</span>(ebFit, <span class="at">coef =</span> <span class="st">"genotypeNrlKO:age"</span>, <span class="at">number =</span> <span class="dv">6</span>) <span class="sc">%&gt;%</span></span>
<span id="cb38-3"><a href="#cb38-3"></a>  <span class="fu">rownames</span>()</span>
<span id="cb38-4"><a href="#cb38-4"></a></span>
<span id="cb38-5"><a href="#cb38-5"></a><span class="co"># extract the data for all 6 genes in tidy format</span></span>
<span id="cb38-6"><a href="#cb38-6"></a>topSixItx <span class="ot">&lt;-</span> <span class="fu">toLongerMeta</span>(eset) <span class="sc">%&gt;%</span> </span>
<span id="cb38-7"><a href="#cb38-7"></a>  <span class="fu">filter</span>(gene <span class="sc">%in%</span> keep) <span class="sc">%&gt;%</span></span>
<span id="cb38-8"><a href="#cb38-8"></a>  <span class="fu">mutate</span>(<span class="at">gene =</span> <span class="fu">factor</span>(gene, <span class="at">levels =</span> keep))</span>
<span id="cb38-9"><a href="#cb38-9"></a></span>
<span id="cb38-10"><a href="#cb38-10"></a><span class="fu">ggplot</span>(topSixItx, </span>
<span id="cb38-11"><a href="#cb38-11"></a>       <span class="fu">aes</span>( <span class="at">x =</span> age, <span class="at">y =</span> expression, <span class="at">color =</span> genotype)) <span class="sc">+</span> </span>
<span id="cb38-12"><a href="#cb38-12"></a>             <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb38-13"><a href="#cb38-13"></a>             <span class="fu">xlab</span>(<span class="st">"age (days)"</span>) <span class="sc">+</span></span>
<span id="cb38-14"><a href="#cb38-14"></a>             <span class="fu">facet_wrap</span>( <span class="sc">~</span> gene) <span class="sc">+</span> </span>
<span id="cb38-15"><a href="#cb38-15"></a>             <span class="fu">stat_smooth</span>(<span class="at">method=</span><span class="st">"lm"</span>, <span class="at">se=</span><span class="cn">FALSE</span>, <span class="at">cex=</span><span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="lect09-multipleTesting_files/figure-revealjs/unnamed-chunk-21-1.png" width="1344" class="r-stretch quarto-figure-center"></section>
<section id="toptable-in-action-any-effect-of-genotype" class="slide level2">
<h2><code>topTable</code> in action: any effect of genotype</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1"></a><span class="fu">topTable</span>(ebFit, <span class="at">coef =</span> <span class="fu">c</span>(<span class="st">"genotypeNrlKO"</span>, <span class="st">"genotypeNrlKO:age"</span>)) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             genotypeNrlKO genotypeNrlKO.age  AveExpr        F      P.Value
1450946_at      -4.7134891       -0.05411298 8.733103 246.5762 2.403638e-22
1451635_at       0.4868705       -0.18718005 7.791237 129.8112 1.307185e-17
1451617_at      -0.9885684       -0.22329233 7.419244 128.9935 1.450111e-17
1425222_x_at     0.3698846       -0.17009077 6.514475 120.0041 4.719613e-17
1441809_at      -0.1576064        0.13363563 6.649440 117.5571 6.594308e-17
1416306_at       0.4278776        0.17020589 6.560050 113.2125 1.212332e-16
1439143_at       0.6570269        0.06170447 5.856507 111.2102 1.616023e-16
1449526_a_at     0.9935801        0.04232958 7.372946 110.5598 1.775851e-16
1452243_at       0.1639373       -0.15313265 6.840975 106.2784 3.344140e-16
1451763_at      -0.6819899       -0.14010359 8.160065 102.7010 5.771358e-16
                adj.P.Val
1450946_at   1.084065e-17
1451635_at   2.180049e-13
1451617_at   2.180049e-13
1425222_x_at 5.321482e-13
1441809_at   5.948198e-13
1416306_at   9.112895e-13
1439143_at   1.001158e-12
1449526_a_at 1.001158e-12
1452243_at   1.675823e-12
1451763_at   2.442992e-12</code></pre>
</div>
</div>
<ul>
<li><p><code>topTable(ebFit, coef = c(2,4))</code> is equivalent here, but <em>much less informative</em>!!</p></li>
<li><p>What is the null hypothesis here?</p></li>
</ul>
<aside class="notes">
<p><span class="math inline">\(H_0: \tau_{KO} = \tau_{KO:Age}=0\)</span>; this finds genes where any (additive and/or interaction) effect of genotype is significant</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="plot-the-top-6-probes-for-any-effect-of-genotype" class="slide level2">
<h2>Plot the top 6 probes for any effect of genotype</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1"></a><span class="co"># grab the gene names of the top 6 genes from topTable</span></span>
<span id="cb41-2"><a href="#cb41-2"></a>keep <span class="ot">&lt;-</span> <span class="fu">topTable</span>(ebFit, <span class="at">coef =</span> <span class="fu">c</span>(<span class="st">"genotypeNrlKO"</span>, <span class="st">"genotypeNrlKO:age"</span>), <span class="at">number =</span> <span class="dv">6</span>) <span class="sc">%&gt;%</span></span>
<span id="cb41-3"><a href="#cb41-3"></a>  <span class="fu">rownames</span>()</span>
<span id="cb41-4"><a href="#cb41-4"></a></span>
<span id="cb41-5"><a href="#cb41-5"></a><span class="co"># extract the data for all 6 genes in tidy format</span></span>
<span id="cb41-6"><a href="#cb41-6"></a>topSixGenotypeMarginal <span class="ot">&lt;-</span> <span class="fu">toLongerMeta</span>(eset) <span class="sc">%&gt;%</span> </span>
<span id="cb41-7"><a href="#cb41-7"></a>  <span class="fu">filter</span>(gene <span class="sc">%in%</span> keep) <span class="sc">%&gt;%</span></span>
<span id="cb41-8"><a href="#cb41-8"></a>  <span class="fu">mutate</span>(<span class="at">gene =</span> <span class="fu">factor</span>(gene, <span class="at">levels =</span> keep))</span>
<span id="cb41-9"><a href="#cb41-9"></a></span>
<span id="cb41-10"><a href="#cb41-10"></a><span class="fu">ggplot</span>(topSixGenotypeMarginal, </span>
<span id="cb41-11"><a href="#cb41-11"></a>       <span class="fu">aes</span>( <span class="at">x =</span> age, <span class="at">y =</span> expression, <span class="at">color =</span> genotype)) <span class="sc">+</span> </span>
<span id="cb41-12"><a href="#cb41-12"></a>             <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb41-13"><a href="#cb41-13"></a>             <span class="fu">xlab</span>(<span class="st">"age (days)"</span>) <span class="sc">+</span></span>
<span id="cb41-14"><a href="#cb41-14"></a>             <span class="fu">facet_wrap</span>( <span class="sc">~</span> gene) <span class="sc">+</span> </span>
<span id="cb41-15"><a href="#cb41-15"></a>             <span class="fu">stat_smooth</span>(<span class="at">method=</span><span class="st">"lm"</span>, <span class="at">se=</span><span class="cn">FALSE</span>, <span class="at">cex=</span><span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="lect09-multipleTesting_files/figure-revealjs/unnamed-chunk-23-1.png" width="1344" class="r-stretch quarto-figure-center"></section>
<section id="comparison-of-s_g2-and-tildes_g2-shrinkage" class="slide level2">
<h2>Comparison of <span class="math inline">\(s_g^2\)</span> and <span class="math inline">\(\tilde{s}_g^2\)</span> (shrinkage!)</h2>
<div class="columns">
<div class="column" style="width:35%;">
<p>Fill in the blank (increases or decreases):</p>
<ol type="1">
<li><p>For <strong>large</strong> variances, limma ___________ the gene-specific variance estimates.</p></li>
<li><p>For <strong>small</strong> variances, limma ___________ the gene-specific variance estimates.</p></li>
</ol>
</div><div class="column" style="width:65%;">
<div class="panel-tabset">
<ul id="tabset-2" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-2-1">original scale</a></li><li><a href="#tabset-2-2">log-scale</a></li></ul>
<div class="tab-content">
<div id="tabset-2-1">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">limma =</span> ebFit<span class="sc">$</span>s2.post, </span>
<span id="cb42-2"><a href="#cb42-2"></a>                 <span class="at">lm =</span> gFit<span class="sc">$</span>sigma<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb42-3"><a href="#cb42-3"></a>shrinkplot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">y =</span> limma, <span class="at">x =</span> lm)) <span class="sc">+</span></span>
<span id="cb42-4"><a href="#cb42-4"></a>  <span class="fu">geom_point</span>(<span class="at">alpha=</span><span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb42-5"><a href="#cb42-5"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept=</span><span class="dv">0</span>, <span class="at">slope=</span><span class="dv">1</span>, <span class="at">color=</span><span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb42-6"><a href="#cb42-6"></a>  <span class="fu">ggtitle</span>(<span class="st">"Ests of gene-specific variance"</span>)</span>
<span id="cb42-7"><a href="#cb42-7"></a>shrinkplot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img data-src="lect09-multipleTesting_files/figure-revealjs/unnamed-chunk-24-1.png" width="672"></p>
</div>
</div>
</div>
<div id="tabset-2-2">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1"></a>shrinkplot <span class="sc">+</span></span>
<span id="cb43-2"><a href="#cb43-2"></a>  <span class="fu">scale_x_log10</span>() <span class="sc">+</span></span>
<span id="cb43-3"><a href="#cb43-3"></a>  <span class="fu">scale_y_log10</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img data-src="lect09-multipleTesting_files/figure-revealjs/unnamed-chunk-25-1.png" width="672"></p>
</div>
</div>
<p>This plot is on the log-scale to ‘zoom in’ on the low range</p>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="comparison-of-interaction-coefficient-p-values" class="slide level2">
<h2>Comparison of interaction coefficient p-values</h2>

<img data-src="lect09-multipleTesting_files/figure-revealjs/unnamed-chunk-26-1.png" width="768" class="r-stretch quarto-figure-center"><ul>
<li>17479 genes where limma p-value is <em>larger</em> than lm</li>
<li>27622 genes where limma p-value is <em>smaller</em> than lm</li>
</ul>
</section>
<section id="multiple-testing" class="slide level2">
<h2>Multiple testing</h2>
<p>What do we do with our lists of thousands of p-values?</p>
</section>
<section id="recall-the-two-types-of-errors" class="slide level2">
<h2>Recall the two types of errors</h2>

<img data-src="img/errors.png" class="r-stretch quarto-figure-center"></section>
<section id="error-rates" class="slide level2">
<h2>Error rates</h2>

<img data-src="img/hypError.png" class="r-stretch quarto-figure-center"><p><span class="math display">\[\alpha = P(\text{Type I Error}), \text{   } \beta = P(\text{Type II Error}), \text{   Power} = 1- \beta\]</span></p>
</section>
<section id="type-i-error-rate-for-m-tests" class="slide level2">
<h2>Type I Error rate for <span class="math inline">\(m\)</span> tests</h2>
<div class="columns">
<div class="column">
<ul>
<li><p><span class="math inline">\(P(\text{incorrect decision} | H_0) = \alpha\)</span></p>
<ul>
<li>let <span class="math inline">\(\alpha=0.05\)</span></li>
</ul></li>
<li><p><span class="math inline">\(P(\text{correct decision} | H_0) = 1-\alpha = 0.95\)</span></p></li>
</ul>
<div class="fragment">
<ul>
<li><p><span class="math inline">\(P(\text{correct decision on }m\text{ tests} | H_0) =\)</span> <span class="math display">\[(1-\alpha)^m = 0.95^m\]</span></p></li>
<li><p><span class="math inline">\(P(\text{at least one error on }m\text{ tests} | H_0) =\)</span> <span class="math display">\[1 - (1-\alpha)^m =\]</span> <span class="math display">\[1-0.95^m = \alpha_{FWER}\]</span></p></li>
<li><p>FWER stands for “Family-wise error rate”</p></li>
</ul>
</div>
</div><div class="column">
<div class="fragment">
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">m =</span> <span class="fu">seq</span>(<span class="dv">1</span>,<span class="dv">200</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb44-2"><a href="#cb44-2"></a>  <span class="fu">mutate</span>(<span class="at">y =</span> <span class="dv">1</span><span class="fl">-0.95</span><span class="sc">^</span>m)</span>
<span id="cb44-3"><a href="#cb44-3"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x =</span> m, <span class="at">y =</span> y)) <span class="sc">+</span></span>
<span id="cb44-4"><a href="#cb44-4"></a>  <span class="fu">geom_line</span>(<span class="at">size =</span> <span class="fl">1.5</span>, <span class="at">alpha =</span> <span class="fl">0.7</span>, <span class="at">colour =</span> <span class="st">"dodgerblue"</span>) <span class="sc">+</span></span>
<span id="cb44-5"><a href="#cb44-5"></a>  <span class="fu">xlab</span>(<span class="st">"m (number of tests)"</span>) <span class="sc">+</span></span>
<span id="cb44-6"><a href="#cb44-6"></a>  <span class="fu">ylab</span>(<span class="fu">expression</span>(<span class="fu">paste</span>(<span class="dv">1</span><span class="fl">-0.95</span><span class="sc">^</span>m, <span class="st">" (FWER)"</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="lect09-multipleTesting_files/figure-revealjs/unnamed-chunk-27-1.png" width="576"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="multiple-comparisons-in-genomics" class="slide level2">
<h2>Multiple comparisons in genomics</h2>
<div class="r-stack">
<p><img data-src="img/mc1.png" height="800"></p>
<div class="fragment fade-in-then-out">
<p><img data-src="img/mc2.png" height="800"></p>
</div>
<div class="fragment fade-in-then-out">
<p><img data-src="img/mc3.png" height="800"></p>
</div>
<div class="fragment fade-in-then-out">
<p><img data-src="img/mc4.png" height="800"></p>
</div>
<div class="fragment fade-in">
<p><img data-src="img/mc5.png" height="800"></p>
</div>
</div>
</section>
<section id="controlling-fwer-with-the-bonferroni-correction" class="slide level2">
<h2>Controlling FWER with the Bonferroni correction</h2>
<ul>
<li><strong>FWER</strong> is the probability of making at least one error when testing <span class="math inline">\(m\)</span> hypotheses</li>
</ul>
<div class="fragment">
<ul>
<li><p><strong>Control the FWER</strong>: limit the probability of making at least one incorrect decision</p></li>
<li><p>One example: the <strong>Bonferroni</strong> correction for <span class="math inline">\(\alpha=0.05\)</span>:</p>
<p><span class="math display">\[\text{We want: } P(\text{at least one error on }m \text{ tests}) \lt \alpha\]</span></p>
<p><span class="math display">\[\text{We know: } P(\text{at least one error on }m \text{ tests}) \lt \sum_{i=1}^m P(\text{error on test }i)\]</span></p>
<p><span class="math display">\[\text{We let: } P(\text{error on test }i) = \alpha_{Bon} \Rightarrow \sum_{i=1}^m P(\text{error on test }i) = m  \alpha_{Bon} \]</span></p>
<p><span class="math display">\[\Rightarrow \alpha_{Bon} = \frac{\alpha}{m} = \frac{0.05}{m} \text{ (Reject if p-value } p_i \lt \frac{\alpha}{m} \text{)}\]</span></p></li>
</ul>
</div>
</section>

<section id="can-we-do-better" class="slide level2">
<h2>Can we do better?</h2>
<ul>
<li><p>Bonferroni correction is very <strong>conservative</strong> (i.e.&nbsp;controls the FWER even lower than <span class="math inline">\(\alpha\)</span> in many settings)</p></li>
<li><p>Several other options are better</p></li>
<li><p>For example, the Holm procedure: multiplier for p-value correction is not the same for all genes; more powerful</p></li>
</ul>
<p><span class="math display">\[p_{Holm(1)} = mp_{(1)}\]</span></p>
<p><span class="math display">\[p_{Holm(2)} = (m-1)p_{(2)}\]</span> <span class="math display">\[p_{Holm(3)} = (m-2)p_{(3)}\]</span> <span class="math display">\[\vdots\]</span> <span class="math display">\[\Rightarrow FWER \le \alpha\]</span> —</p>
</section>
<section id="how-practical-is-the-fwer-in-high-throughput-biology" class="slide level2">
<h2>How practical is the FWER in high-throughput biology?</h2>
<ul>
<li>Why do we care so much about making <em>one single error</em>??</li>
</ul>
<center>
<iframe src="https://giphy.com/embed/8mndEBLsg9Whg2Sduv" width="300" height="300" frameborder="0" class="giphy-embed" allowfullscreen=""></iframe><p><a href="https://giphy.com/gifs/idk-preity-zinta-confuse-8mndEBLsg9Whg2Sduv"></a></p>
</center>
<div class="fragment">
<ul>
<li><p>One easy way to ensure no Type I errors: reject no hypotheses! 😃</p>
<ul>
<li>However, then our power is zero… 😭</li>
</ul></li>
</ul>
</div>
<div class="fragment">
<div class="callout callout-important callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Important</strong></p>
</div>
<div class="callout-content">
<p>Being overly strict about Type I error leads to greater Type II error (loss of power)</p>
</div>
</div>
</div>
</div>
</section>
<section id="radical-idea" class="slide level2">
<h2>Radical idea</h2>
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>FWER is not that relevant in high-throughput studies</strong></p>
</div>
<div class="callout-content">
<p>It’s OK to make multiple errors, as long as you also have a <em>comparatively</em> large number of true positives!</p>
</div>
</div>
</div>
<center>
<iframe src="https://giphy.com/embed/LqajRC2pU0Je8" width="480" height="289" frameborder="0" class="giphy-embed" allowfullscreen=""></iframe><p><a href="https://giphy.com/gifs/mar-08-LqajRC2pU0Je8"></a></p>
<center>
</center></center></section>
<section id="enter-the-false-discovery-rate-fdr" class="slide level2">
<h2>Enter: the False Discovery Rate (FDR)</h2>
<div class="columns">
<div class="column" style="width:65%;">
<p><img data-src="img/bh.png"></p>
</div><div class="column" style="width:35%;">
<p><a href="https://doi.org/10.1111/j.2517-6161.1995.tb02031.x">Benjamini Y, Hochberg Y. “<strong>Controlling the False Discovery Rate: a Practical and Powerful Approach to Multiple Testing.</strong>” <em>Journal of the Royal Statistical Society: Series B (Methodological)</em>. 1995 Jan;57(1): 289-300.</a></p>
<p>Over 110,000 citations to date!!</p>
</div>
</div>
</section>
<section id="false-discovery-rate" class="slide level2">
<h2>False Discovery Rate</h2>

<img data-src="img/fdrtable.png" class="r-stretch quarto-figure-center"><p>FDR is designed to control the expected proportion of false positives (V) among all hypotheses where the null has been rejected (R)</p>
<div class="fragment">
<p><span class="math display">\[FDR = E \Big[ \frac{V}{R} \Big]\]</span></p>
</div>
</section>
<section id="fdr-vs-fpr-vs-fwer" class="slide level2">
<h2>FDR vs FPR vs FWER</h2>
<ul>
<li><strong>False Discovery Rate (FDR)</strong> is the rate that significant features <span class="math inline">\((R)\)</span> are truly null</li>
</ul>
<p><span class="math display">\[FDR = E \Big[ \frac{V}{R} \Big]\]</span></p>
<ul>
<li><strong>False Positive Rate (FPR)</strong> is the rate that truly null features <span class="math inline">\((m_0)\)</span> are called significant</li>
</ul>
<p><span class="math display">\[FPR =  E\Big[\frac{V}{m_0}\Big]\]</span></p>
<ul>
<li><strong>Family-Wise Error Rate (FWER)</strong> is the probability that the number of truly null features rejected <span class="math inline">\((V)\)</span> is at least 1</li>
</ul>
<p><span class="math display">\[\text{FWER } = P(V\ge1) \]</span></p>
</section>
<section id="benjamini-hochberg-fdr-bh-procedure" class="slide level2">
<h2>Benjamini-Hochberg FDR (BH procedure)</h2>
<ul>
<li><p>Proposed the idea of controlling FDR instead of FWER</p></li>
<li><p>Proposed a procedure for doing so</p>
<ul>
<li>note that we know <span class="math inline">\(R\)</span>, but we don’t know <span class="math inline">\(V\)</span></li>
</ul></li>
<li><p>Procedure: control FDR at level <span class="math inline">\(q\)</span></p>
<ol type="1">
<li><p>order the raw p-values <span class="math inline">\(p_{(1)} \le p_{(2)} \le ...\le p_{(m)}\)</span></p></li>
<li><p>find test with highest rank <span class="math inline">\(j\)</span> such that <span class="math inline">\(p_{(j)} &lt; \frac{jq}{m}\)</span></p></li>
<li><p>declare all smaller ranks up to <span class="math inline">\(j\)</span> significant</p></li>
</ol></li>
</ul>
</section>
<section id="controlling-fdr-at-level-q-0.05" class="slide level2">
<h2>Controlling FDR at level <span class="math inline">\(q\)</span> = 0.05</h2>
<div class="r-stack">
<div class="fragment fade-in-then-out">
<table>
<thead>
<tr class="header">
<th style="text-align: center;">Rank <span class="math inline">\((j)\)</span></th>
<th style="text-align: center;">P-value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">1</td>
<td style="text-align: center;">0.0008</td>
</tr>
<tr class="even">
<td style="text-align: center;">2</td>
<td style="text-align: center;">0.009</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3</td>
<td style="text-align: center;">0.127</td>
</tr>
<tr class="even">
<td style="text-align: center;">4</td>
<td style="text-align: center;">0.205</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5</td>
<td style="text-align: center;">0.396</td>
</tr>
<tr class="even">
<td style="text-align: center;">6</td>
<td style="text-align: center;">0.450</td>
</tr>
<tr class="odd">
<td style="text-align: center;">7</td>
<td style="text-align: center;">0.641</td>
</tr>
<tr class="even">
<td style="text-align: center;">8</td>
<td style="text-align: center;">0.781</td>
</tr>
<tr class="odd">
<td style="text-align: center;">9</td>
<td style="text-align: center;">0.900</td>
</tr>
<tr class="even">
<td style="text-align: center;">10</td>
<td style="text-align: center;">0.993</td>
</tr>
</tbody>
</table>
<p><span class="math inline">\(\text{}\)</span></p>
</div>
<div class="fragment fade-in-then-out">
<table>
<thead>
<tr class="header">
<th style="text-align: center;">Rank <span class="math inline">\((j)\)</span></th>
<th style="text-align: center;">P-value</th>
<th style="text-align: center;">BH-threshold: jq/m</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">1</td>
<td style="text-align: center;">0.0008</td>
<td style="text-align: center;">0.005</td>
</tr>
<tr class="even">
<td style="text-align: center;">2</td>
<td style="text-align: center;">0.009</td>
<td style="text-align: center;">0.010</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3</td>
<td style="text-align: center;">0.127</td>
<td style="text-align: center;">0.015</td>
</tr>
<tr class="even">
<td style="text-align: center;">4</td>
<td style="text-align: center;">0.205</td>
<td style="text-align: center;">0.020</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5</td>
<td style="text-align: center;">0.396</td>
<td style="text-align: center;">0.025</td>
</tr>
<tr class="even">
<td style="text-align: center;">6</td>
<td style="text-align: center;">0.450</td>
<td style="text-align: center;">0.030</td>
</tr>
<tr class="odd">
<td style="text-align: center;">7</td>
<td style="text-align: center;">0.641</td>
<td style="text-align: center;">0.035</td>
</tr>
<tr class="even">
<td style="text-align: center;">8</td>
<td style="text-align: center;">0.781</td>
<td style="text-align: center;">0.040</td>
</tr>
<tr class="odd">
<td style="text-align: center;">9</td>
<td style="text-align: center;">0.900</td>
<td style="text-align: center;">0.045</td>
</tr>
<tr class="even">
<td style="text-align: center;">10</td>
<td style="text-align: center;">0.993</td>
<td style="text-align: center;">0.050</td>
</tr>
</tbody>
</table>
<p><span class="math inline">\(\text{}\)</span></p>
</div>
<div class="fragment fade-in-then-out">
<table>
<colgroup>
<col style="width: 13%">
<col style="width: 18%">
<col style="width: 31%">
<col style="width: 36%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">Rank <span class="math inline">\((j)\)</span></th>
<th style="text-align: center;">P-value</th>
<th style="text-align: center;">BH-threshold: jq/m</th>
<th style="text-align: center;">BH-adjusted p-value: <span class="math inline">\(p_{(i)}^{BH}= \text{min}(\text{min}_{j\ge i}(mp_{(j)}/j), 1)\)</span></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">1</td>
<td style="text-align: center;">0.0008</td>
<td style="text-align: center;">0.005</td>
<td style="text-align: center;">0.0080</td>
</tr>
<tr class="even">
<td style="text-align: center;">2</td>
<td style="text-align: center;">0.009</td>
<td style="text-align: center;">0.010</td>
<td style="text-align: center;">0.0450</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3</td>
<td style="text-align: center;">0.127</td>
<td style="text-align: center;">0.015</td>
<td style="text-align: center;">0.4233</td>
</tr>
<tr class="even">
<td style="text-align: center;">4</td>
<td style="text-align: center;">0.205</td>
<td style="text-align: center;">0.020</td>
<td style="text-align: center;">0.5125</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5</td>
<td style="text-align: center;">0.396</td>
<td style="text-align: center;">0.025</td>
<td style="text-align: center;">0.7500</td>
</tr>
<tr class="even">
<td style="text-align: center;">6</td>
<td style="text-align: center;">0.450</td>
<td style="text-align: center;">0.030</td>
<td style="text-align: center;">0.7500</td>
</tr>
<tr class="odd">
<td style="text-align: center;">7</td>
<td style="text-align: center;">0.641</td>
<td style="text-align: center;">0.035</td>
<td style="text-align: center;">0.9157</td>
</tr>
<tr class="even">
<td style="text-align: center;">8</td>
<td style="text-align: center;">0.781</td>
<td style="text-align: center;">0.040</td>
<td style="text-align: center;">0.9763</td>
</tr>
<tr class="odd">
<td style="text-align: center;">9</td>
<td style="text-align: center;">0.900</td>
<td style="text-align: center;">0.045</td>
<td style="text-align: center;">0.9930</td>
</tr>
<tr class="even">
<td style="text-align: center;">10</td>
<td style="text-align: center;">0.993</td>
<td style="text-align: center;">0.050</td>
<td style="text-align: center;">0.9930</td>
</tr>
</tbody>
</table>
<p><span class="math inline">\(\text{}\)</span></p>
</div>
<aside class="notes">
<p>This formula looks more complicated than it really is. It says:</p>
<ol type="1">
<li>First, order all p-values from small to large. Then multiply each 𝑝 -value by the total number of tests 𝑚 and divide by its rank order. Second, make sure that the resulting sequence is non-decreasing: if it ever starts decreasing, make the preceding 𝑝 -value equal to the subsequent (repeatedly, until the whole sequence becomes non-decreasing). If any 𝑝 -value ends up larger than 1, make it equal to 1.</li>
</ol>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
<div class="fragment fade-in-then-out">
<table>
<thead>
<tr class="header">
<th style="text-align: center;">Rank <span class="math inline">\((j)\)</span></th>
<th style="text-align: center;">P-value</th>
<th style="text-align: center;">BH-threshold: jq/m</th>
<th style="text-align: center;">Reject <span class="math inline">\(H_0\)</span>?</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">1</td>
<td style="text-align: center;">0.0008</td>
<td style="text-align: center;">0.005</td>
<td style="text-align: center;"><span class="math inline">\(\checkmark\)</span></td>
</tr>
<tr class="even">
<td style="text-align: center;">2</td>
<td style="text-align: center;">0.009</td>
<td style="text-align: center;">0.010</td>
<td style="text-align: center;"><span class="math inline">\(\checkmark\)</span></td>
</tr>
<tr class="odd">
<td style="text-align: center;">3</td>
<td style="text-align: center;">0.127</td>
<td style="text-align: center;">0.015</td>
<td style="text-align: center;"></td>
</tr>
<tr class="even">
<td style="text-align: center;">4</td>
<td style="text-align: center;">0.205</td>
<td style="text-align: center;">0.020</td>
<td style="text-align: center;"></td>
</tr>
<tr class="odd">
<td style="text-align: center;">5</td>
<td style="text-align: center;">0.396</td>
<td style="text-align: center;">0.025</td>
<td style="text-align: center;"></td>
</tr>
<tr class="even">
<td style="text-align: center;">6</td>
<td style="text-align: center;">0.450</td>
<td style="text-align: center;">0.030</td>
<td style="text-align: center;"></td>
</tr>
<tr class="odd">
<td style="text-align: center;">7</td>
<td style="text-align: center;">0.641</td>
<td style="text-align: center;">0.035</td>
<td style="text-align: center;"></td>
</tr>
<tr class="even">
<td style="text-align: center;">8</td>
<td style="text-align: center;">0.781</td>
<td style="text-align: center;">0.040</td>
<td style="text-align: center;"></td>
</tr>
<tr class="odd">
<td style="text-align: center;">9</td>
<td style="text-align: center;">0.900</td>
<td style="text-align: center;">0.045</td>
<td style="text-align: center;"></td>
</tr>
<tr class="even">
<td style="text-align: center;">10</td>
<td style="text-align: center;">0.993</td>
<td style="text-align: center;">0.050</td>
<td style="text-align: center;"><span class="math inline">\(\text{ }\)</span></td>
</tr>
</tbody>
</table>
<p><span class="math inline">\(\text{}\)</span></p>
</div>
<div class="fragment fade-in">
<table>
<thead>
<tr class="header">
<th style="text-align: center;">Rank <span class="math inline">\((j)\)</span></th>
<th>P-val</th>
<th style="text-align: center;">BH-thr: jq/m</th>
<th style="text-align: center;">Reject <span class="math inline">\(H_0\)</span>?</th>
<th style="text-align: center;"><span class="math inline">\(FWER &lt; 0.05\)</span>?</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">1</td>
<td>0.0008</td>
<td style="text-align: center;">0.005</td>
<td style="text-align: center;"><span class="math inline">\(\checkmark\)</span></td>
<td style="text-align: center;"><span class="math inline">\(\checkmark\)</span></td>
</tr>
<tr class="even">
<td style="text-align: center;">2</td>
<td>0.009</td>
<td style="text-align: center;">0.010</td>
<td style="text-align: center;"><span class="math inline">\(\checkmark\)</span></td>
<td style="text-align: center;"></td>
</tr>
<tr class="odd">
<td style="text-align: center;">3</td>
<td>0.127</td>
<td style="text-align: center;">0.015</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
</tr>
<tr class="even">
<td style="text-align: center;">4</td>
<td>0.205</td>
<td style="text-align: center;">0.020</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
</tr>
<tr class="odd">
<td style="text-align: center;">5</td>
<td>0.396</td>
<td style="text-align: center;">0.025</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
</tr>
<tr class="even">
<td style="text-align: center;">6</td>
<td>0.450</td>
<td style="text-align: center;">0.030</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
</tr>
<tr class="odd">
<td style="text-align: center;">7</td>
<td>0.641</td>
<td style="text-align: center;">0.035</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
</tr>
<tr class="even">
<td style="text-align: center;">8</td>
<td>0.781</td>
<td style="text-align: center;">0.040</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
</tr>
<tr class="odd">
<td style="text-align: center;">9</td>
<td>0.900</td>
<td style="text-align: center;">0.045</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
</tr>
<tr class="even">
<td style="text-align: center;">10</td>
<td>0.993</td>
<td style="text-align: center;">0.050</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
</tr>
</tbody>
</table>
<p>Where <span class="math inline">\(\alpha_{Bon}=0.05/10=0.005\)</span></p>
</div>
</div>
</section>
<section id="bh-fdr-values-given-in-limma-by-default" class="slide level2">
<h2>BH FDR values given in <code>limma</code> by default</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1"></a><span class="fu">topTable</span>(ebFit, <span class="at">coef =</span> <span class="st">"genotypeNrlKO"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                  logFC  AveExpr          t      P.Value    adj.P.Val        B
1450946_at   -4.7134891 8.733103 -15.583683 5.390931e-18 2.431364e-13 28.40938
1442070_at   -0.9669400 8.268122  -9.787407 6.896220e-12 1.555132e-07 16.46086
1443457_at   -1.1530057 6.049428  -9.091412 4.947029e-11 6.029428e-07 14.68843
1422679_s_at -2.1187394 9.495435  -9.002625 6.388954e-11 6.029428e-07 14.45703
1431708_a_at -2.1610625 8.591450  -8.967950 7.062007e-11 6.029428e-07 14.36634
1433050_at   -1.5083887 8.468891  -8.923932 8.021234e-11 6.029428e-07 14.25096
1426288_at   -4.2100657 9.323796  -8.772209 1.246491e-10 8.031144e-07 13.85106
1418108_at    0.8662091 8.260647   8.537984 2.475294e-10 1.395478e-06 13.22713
1450770_at    1.3103576 7.357033   8.128847 8.332369e-10 3.747976e-06 12.11870
1457802_at   -0.8362030 7.778189  -8.117112 8.629984e-10 3.747976e-06 12.08657</code></pre>
</div>
</div>
<div class="fragment">
<p>Or, obtain them yourself for any vector of p-values <code>p</code> with <code>p.adjust(p, method="BH")</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1"></a>pvals <span class="ot">&lt;-</span> <span class="fu">topTable</span>(ebFit, <span class="at">coef =</span> <span class="st">"genotypeNrlKO"</span>, <span class="at">number =</span> <span class="cn">Inf</span>)<span class="sc">$</span>P.Value</span>
<span id="cb47-2"><a href="#cb47-2"></a><span class="fu">p.adjust</span>(pvals, <span class="at">method =</span> <span class="st">"BH"</span>) <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2.431364e-13 1.555132e-07 6.029428e-07 6.029428e-07 6.029428e-07
[6] 6.029428e-07</code></pre>
</div>
</div>
</div>
</section>
<section id="other-ways-to-control-fdr" class="slide level2">
<h2>Other ways to control FDR</h2>
<ul>
<li><p>BH is just one (the first) method to control FDR</p></li>
<li><p>Since the publication of the BH method, other methods have been proposed</p></li>
<li><p>One of the most popular is <a href="https://doi.org/10.1073/pnas.1530509100">Storey’s q-value</a></p></li>
</ul>

<img data-src="img/qval.png" class="r-stretch quarto-figure-center"><ul>
<li><code>qvalue</code> package implementation: provides adjusted p-values</li>
</ul>
</section>
<section id="storeys-q-value-vs-bh-conceptual" class="slide level2">
<h2>Storey’s q-value vs BH (Conceptual)</h2>
<ul>
<li><p>Just like BH, is focused on the proportion of discoveries that are false positives</p></li>
<li><p><em>Conceptual</em> difference between BH and Storey’s q-value is:</p>
<ul>
<li><p>BH <strong>controls</strong> the FDR</p></li>
<li><p>q-values give an unbiased <strong>estimate</strong> of the FDR (will control the FDR <em>on average</em>)</p></li>
</ul></li>
</ul>
</section>
<section id="storeys-q-value-vs-bh-mathematical" class="slide level2">
<h2>Storey’s q-value vs BH (Mathematical)</h2>
<ul>
<li><p>Mathematically, the difference between the two is in how <span class="math inline">\(m_0\)</span> is estimated</p>
<ul>
<li><p>Or equivalently, how <span class="math inline">\(\pi_0=\frac{m_0}{m}\)</span> is estimated (since <span class="math inline">\(m\)</span> is known)</p></li>
<li><p><span class="math inline">\(\pi_0\)</span> represents the overall proportion of tests that are truly null</p></li>
<li><p>BH conservatively assumes that <span class="math inline">\(\pi_0=1\)</span>; q-value estimates it from distribution of p-values</p></li>
</ul></li>
<li><p>q-value:</p></li>
</ul>
<p><span class="math display">\[\hat{q}(p_i) = \min_{i\le j} \Big( \frac{\hat{\pi}_0m}{j}p_{(j)}\Big)\]</span></p>
<ul>
<li>q-value and BH-adjusted p-values are equivalent when <span class="math inline">\(\pi_0=1\)</span></li>
</ul>
<p><span class="math display">\[\hat{p}_{BH}(p_i) = \min_{i\le j} \Big(\frac{m}{j}p_{(j)}\Big)\]</span></p>
</section>
<section id="bh-vs-q-value-in-toy-example" class="slide level2">
<h2>BH vs q-value in toy example</h2>
<table>
<colgroup>
<col style="width: 22%">
<col style="width: 16%">
<col style="width: 30%">
<col style="width: 30%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">Rank <span class="math inline">\((j)\)</span></th>
<th>P-value</th>
<th style="text-align: center;"><span class="math inline">\(\hat{p}_{BH}(p_i)\)</span></th>
<th style="text-align: center;"><span class="math inline">\(\hat{q}(p_i)\)</span><sup>1</sup></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">1</td>
<td>0.0008</td>
<td style="text-align: center;">0.045</td>
<td style="text-align: center;">0.045</td>
</tr>
<tr class="even">
<td style="text-align: center;">2</td>
<td>0.009</td>
<td style="text-align: center;">0.045</td>
<td style="text-align: center;">0.045</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3</td>
<td>0.127</td>
<td style="text-align: center;">0.423</td>
<td style="text-align: center;">0.423</td>
</tr>
<tr class="even">
<td style="text-align: center;">4</td>
<td>0.205</td>
<td style="text-align: center;">0.513</td>
<td style="text-align: center;">0.513</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5</td>
<td>0.396</td>
<td style="text-align: center;">0.750</td>
<td style="text-align: center;">0.750</td>
</tr>
<tr class="even">
<td style="text-align: center;">6</td>
<td>0.450</td>
<td style="text-align: center;">0.750</td>
<td style="text-align: center;">0.750</td>
</tr>
<tr class="odd">
<td style="text-align: center;">7</td>
<td>0.641</td>
<td style="text-align: center;">0.916</td>
<td style="text-align: center;">0.916</td>
</tr>
<tr class="even">
<td style="text-align: center;">8</td>
<td>0.781</td>
<td style="text-align: center;">0.976</td>
<td style="text-align: center;">0.976</td>
</tr>
<tr class="odd">
<td style="text-align: center;">9</td>
<td>0.900</td>
<td style="text-align: center;">0.993</td>
<td style="text-align: center;">0.993</td>
</tr>
<tr class="even">
<td style="text-align: center;">10</td>
<td>0.993</td>
<td style="text-align: center;">0.993</td>
<td style="text-align: center;">0.993</td>
</tr>
</tbody>
</table>
<aside><ol class="aside-footnotes"><li id="fn4"><p><span class="math inline">\(\hat{\pi}_0=1\)</span> in this case</p></li></ol></aside></section>
<section id="bh-vs-q-value-in-effect-of-genotypenrlkoage" class="slide level2">
<h2>BH vs q-value in effect of <code>genotypeNrlKO:age</code></h2>
<div class="panel-tabset">
<ul id="tabset-3" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-3-1">Original scale</a></li><li><a href="#tabset-3-2">log-scale</a></li></ul>
<div class="tab-content">
<div id="tabset-3-1">
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1"></a><span class="co"># requires qvalue package</span></span>
<span id="cb49-2"><a href="#cb49-2"></a>pvals <span class="ot">&lt;-</span> <span class="fu">topTable</span>(ebFit, <span class="at">coef =</span> <span class="st">"genotypeNrlKO:age"</span>, <span class="at">number =</span> <span class="cn">Inf</span>)<span class="sc">$</span>P.Value</span>
<span id="cb49-3"><a href="#cb49-3"></a></span>
<span id="cb49-4"><a href="#cb49-4"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">fdr.bh =</span> <span class="fu">p.adjust</span>(pvals, <span class="at">method =</span> <span class="st">"BH"</span>),</span>
<span id="cb49-5"><a href="#cb49-5"></a>                 <span class="at">fdr.q =</span> qvalue<span class="sc">::</span><span class="fu">qvalue</span>(pvals)<span class="sc">$</span>qvalues)</span>
<span id="cb49-6"><a href="#cb49-6"></a></span>
<span id="cb49-7"><a href="#cb49-7"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">y =</span> fdr.q, <span class="at">x =</span> fdr.bh)) <span class="sc">+</span></span>
<span id="cb49-8"><a href="#cb49-8"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb49-9"><a href="#cb49-9"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept =</span> <span class="dv">0</span>, <span class="at">slope =</span> <span class="dv">1</span>, <span class="at">color=</span><span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb49-10"><a href="#cb49-10"></a>  <span class="fu">ggtitle</span>(<span class="st">"BH vs q-value for genotypeNrlKO:age"</span>) <span class="sc">+</span></span>
<span id="cb49-11"><a href="#cb49-11"></a>  <span class="fu">xlim</span>(<span class="dv">0</span>,<span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb49-12"><a href="#cb49-12"></a>  <span class="fu">ylim</span>(<span class="dv">0</span>,<span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="lect09-multipleTesting_files/figure-revealjs/unnamed-chunk-30-1.png" width="768"></p>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-3-2">
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1"></a>df <span class="ot">&lt;-</span> <span class="fu">mutate</span>(df, </span>
<span id="cb50-2"><a href="#cb50-2"></a>             <span class="at">l_fdr.bh =</span> <span class="sc">-</span><span class="fu">log10</span>(fdr.bh),</span>
<span id="cb50-3"><a href="#cb50-3"></a>             <span class="at">l_fdr.q =</span> <span class="sc">-</span><span class="fu">log10</span>(fdr.q))</span>
<span id="cb50-4"><a href="#cb50-4"></a></span>
<span id="cb50-5"><a href="#cb50-5"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">y =</span> l_fdr.q, <span class="at">x =</span> l_fdr.bh)) <span class="sc">+</span></span>
<span id="cb50-6"><a href="#cb50-6"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb50-7"><a href="#cb50-7"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept =</span> <span class="dv">0</span>, <span class="at">slope =</span> <span class="dv">1</span>, <span class="at">color=</span><span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb50-8"><a href="#cb50-8"></a>  <span class="fu">ggtitle</span>(<span class="st">"BH vs q-value for genotypeNrlKO:age"</span>) <span class="sc">+</span></span>
<span id="cb50-9"><a href="#cb50-9"></a>  <span class="fu">xlab</span>(<span class="st">"-log10(BH adjusted p-value)"</span>) <span class="sc">+</span></span>
<span id="cb50-10"><a href="#cb50-10"></a>  <span class="fu">ylab</span>(<span class="st">"-log10(q-value)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="lect09-multipleTesting_files/figure-revealjs/unnamed-chunk-31-1.png" width="768"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="fdr-control-is-an-active-area-of-research" class="slide level2">
<h2>FDR control is an active area of research</h2>
<div class="columns">
<div class="column">
<p><img data-src="img/practicalguide.png"></p>
<p><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1716-1">Korthauer*, Kimes* et al.&nbsp;2019</a></p>
</div><div class="column">
<p><img data-src="img/practicalguidefig.png"></p>
</div>
</div>
</section>
<section id="compounding-issues-of-multiple-comparisons" class="slide level2">
<h2>Compounding issues of multiple comparisons</h2>
<ul>
<li><p>What if you’re not only testing 20K genes, but also multiple tests per gene (e.g.&nbsp;multiple contrasts, such as several two-group comparisons)?</p></li>
<li><p>Classical procedures for adjustment:</p>
<ul>
<li>Tukey multiple comparison procedure</li>
<li>Scheffe multiple comparison procedure</li>
<li>Bonferroni or Holm FWER correction</li>
</ul></li>
<li><p>In our setting, we can also apply BH to all p-values globally</p>
<ul>
<li><p><code>limma::decideTests(pvals, method="global")</code> for a matrix of p-values or <code>eBayes</code> output (e.g.&nbsp;rows = genes, columns = contrasts)</p></li>
<li><p>p-values are combined, adjusted globally, then separated back out and sorted</p></li>
</ul></li>
</ul>
</section>
<section id="assumptions-about-p-values" class="slide level2">
<h2>Assumptions about p-values</h2>
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Note</strong></p>
</div>
<div class="callout-content">
<p>Implicit assumption for all multiple testing correction methods: p-value distribution is “well-behaved”</p>
</div>
</div>
</div>
<div class="fragment">
<p>What does this mean?</p>
<p>Primarily, that the distribution of p-values <em>under the null</em> is <strong>uniform</strong> (i.e.&nbsp;flat)</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="lect09-multipleTesting_files/figure-revealjs/unnamed-chunk-32-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
</div>
</section>
<section id="p-value-distributions" class="slide level2">
<h2>p-value distributions</h2>
<p>Spike of small p-values indicates non-null tests.</p>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1"></a><span class="co"># plot histogram of p-values for genotypeNrlKO</span></span>
<span id="cb51-2"><a href="#cb51-2"></a><span class="fu">hist</span>(pvals)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="lect09-multipleTesting_files/figure-revealjs/unnamed-chunk-33-1.png" width="960" class="r-stretch quarto-figure-center"><p>Great primer on how things can go wrong: <a href="http://varianceexplained.org/statistics/interpreting-pvalue-histogram/" class="uri">http://varianceexplained.org/statistics/interpreting-pvalue-histogram/</a></p>
</section>
<section id="what-if-p-values-are-poorly-behaved" class="slide level2">
<h2>What if p-values are <em>poorly behaved</em>?</h2>
<ul>
<li><p>FDR estimates can be invalid (assumptions are violated)</p>
<ul>
<li>Some possible causes: test assumptions violated, model misspecification, uncontrolled variation/batch effects, selection bias/filtering</li>
</ul></li>
<li><p>Possible solution: nonparametric test</p>
<ul>
<li>Limitation: lack of flexibility to adjust for multiple covariates</li>
</ul></li>
</ul>
<div class="fragment">
<ul>
<li><p>Another possible solution: estimate sampling distribution or p-values “empirically” using resampling techniques</p>
<ul>
<li><p><strong>Permutation</strong>: construct a simulated version of your dataset that satisfies the null hypothesis and compute statistic (e.g.&nbsp;shuffle group labels for a two-group comparison); repeat many times and use permutation statistics as your sampling distribution rather than a t, Normal, F, <span class="math inline">\(\chi^2\)</span>, etc</p></li>
<li><p>Limitation: computationally intensive for genomics; not optimal for small sample sizes</p></li>
</ul></li>
</ul>
</div>
</section>
<section id="next-up-regression-with-count-measures-rna-seq" class="slide level2">
<h2>Next up: regression with count measures (RNA-seq)!</h2>
<ul>
<li><p>So far, all of our modeling assumed our outcome <span class="math inline">\(Y_i\)</span> (gene expression measurements for a particular gene) were <strong>continuous</strong> and with <strong>constant variance</strong> (iid).</p></li>
<li><p>From RNA-seq, we obtain <strong>discrete</strong>, skewed distributions (with <strong>non-constant variance</strong>) of counts that represent gene expression levels across genes. These violate modeling assumptions for standard linear regression / limma.</p></li>
<li><p>Next time, we’ll explore:</p>
<ul>
<li><p>Data transformations and extensions to limma framework to apply deal with these violations</p></li>
<li><p>Generalized linear models (negative binomial regression) to directly model counts</p></li>
</ul></li>
</ul>

<img src="https://github.com/STAT540-UBC/stat540-ubc.github.io/raw/main/images/stat540-logo-s.png" class="slide-logo r-stretch"><div class="footer footer-default">

</div>
</section>

    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="lect09-multipleTesting_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="lect09-multipleTesting_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="lect09-multipleTesting_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="lect09-multipleTesting_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="lect09-multipleTesting_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="lect09-multipleTesting_files/libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="lect09-multipleTesting_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="lect09-multipleTesting_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="lect09-multipleTesting_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="lect09-multipleTesting_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="lect09-multipleTesting_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1600,

        height: 900,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>