<!DOCTYPE html>
<html lang="en"><head>
<script src="lect05-twogroup_files/libs/clipboard/clipboard.min.js"></script>
<script src="lect05-twogroup_files/libs/quarto-html/tabby.min.js"></script>
<script src="lect05-twogroup_files/libs/quarto-html/popper.min.js"></script>
<script src="lect05-twogroup_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="lect05-twogroup_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="lect05-twogroup_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="lect05-twogroup_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="lect05-twogroup_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.3.450">

  <meta name="author" content="Keegan Korthauer">
  <meta name="dcterms.date" content="2025-01-21">
  <title>Two group comparisons</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="lect05-twogroup_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="lect05-twogroup_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="lect05-twogroup_files/libs/revealjs/dist/theme/quarto.css">
  <link href="lect05-twogroup_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="lect05-twogroup_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="lect05-twogroup_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="lect05-twogroup_files/libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="lect05-twogroup_files/libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="lect05-twogroup_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" data-background-color="#197aa0" data-background-image="https://github.com/STAT540-UBC/stat540-ubc.github.io/raw/main/images/stat540-logo-s.png" data-background-opacity="0.9" data-background-position="95% 90%" data-background-size="12%" class="quarto-title-block center">
  <h1 class="title">Two group comparisons</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Keegan Korthauer 
</div>
</div>
</div>

  <p class="date">January 21, 2025</p>
</section>
<section id="reminders" class="slide level2">
<h2>Reminders</h2>
<ul>
<li><p>Project groups posted to Canvas last week</p></li>
<li><p>Intro Assignment due <strong>Thursday Jan 23</strong></p></li>
<li><p><a href="https://stat540-ubc.github.io/group_project_rubrics.html#initial-proposal-5-pts">Initial Project Proposal</a> due <strong>Monday Jan 27</strong> (Feedback in-class Jan 30)</p></li>
</ul>
</section>
<section id="todays-learning-objectives" class="slide level2">
<h2>Today’s learning objectives</h2>
<ul>
<li><p>Understand <strong>how</strong> and <strong>when</strong> to carry out a t-test for comparing two population means</p></li>
<li><p>Identify when alternative approaches (e.g.&nbsp;nonparametric) are more appropriate</p></li>
<li><p>Avoid common pitfalls in interpretation of hypothesis tests and p-values</p></li>
</ul>
</section>
<section id="central-dogma-of-statistics" class="slide level2">
<h2>Central dogma of statistics</h2>

<img data-src="img/statInference.png" class="r-stretch quarto-figure-center"><p>We want to understand a <strong>population</strong> (e.g.&nbsp;all individuals with a certain disease) but we can only study a <strong>random sample</strong> from it</p>
<p><a href="https://www.gs.washington.edu/academics/courses/akey/56008/lecture/lecture1.pdf">Image source: Josh Akey’s Lecture notes</a></p>
</section>
<section id="hypothesis-testing-in-genomics" class="slide level2">
<h2>Hypothesis Testing in Genomics</h2>
<div class="columns">
<div class="column">
<ul>
<li><p>Retina presents a model system for investigating <strong>regulatory networks</strong> underlying neuronal differentiation</p></li>
<li><p><strong>Nrl</strong> transcription factor is known to be important for Rod development</p></li>
</ul>
</div><div class="column">
<p><img data-src="img/paperRef.PNG" data-fig-align="center"> <img data-src="img/cellTypes.png" data-fig-align="center"> <a href="https://doi.org/10.1073/pnas.0508214103">Akimoto et al.&nbsp;(2006)</a></p>
</div>
</div>
<div class="fragment">
<p><strong>What happens if you delete <em>Nrl</em>?</strong></p>
</div>
</section>
<section id="why-a-hypothesis-test" class="slide level2">
<h2>Why a Hypothesis Test?</h2>
<p>From the <a href="https://doi.org/10.1073/pnas.0508214103">Akimoto et al.&nbsp;(2006) paper</a>:</p>
<blockquote>
<p>“we hypothesized that Nrl is the ideal transcription factor to gain insights into gene expression changes …”</p>
</blockquote>
<div class="fragment">
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Biological question</strong></p>
</div>
<div class="callout-content">
<p>Is the expression level of gene <em>A</em> affected by knockout of the <em>Nrl</em> gene?</p>
</div>
</div>
</div>
</div>
<div class="fragment">
<p>We can use <strong>statistical inference</strong> to answer this biological question!</p>
</div>
</section>
<section id="statistical-inference" class="slide level2">
<h2>Statistical inference</h2>
<div class="columns">
<div class="column">
<ul>
<li>Let’s observe and study a <strong>random sample</strong> to make conclusions about a population: measure gene expression on a random sample of mice</li>
<li><strong>Experimental design</strong>:
<ul>
<li>5 developmental stages (E16, P2, P6, P10, 4Weeks)</li>
<li>2 genotypes: Wild type (WT), Nrl Knockout (NrlKO)</li>
<li>3-4 replicates for each combination</li>
</ul></li>
</ul>
</div><div class="column">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/NrlKOmouse.png"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="reading-in-exploring-the-data" class="slide level2">
<h2>Reading in / exploring the data</h2>
<ul>
<li><p>Data obtained from the <a href="https://www.ncbi.nlm.nih.gov/geo/">Gene Expression Omnibus (GEO)</a> repository (accession <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE4051">GSE4051</a>)</p></li>
<li><p>Load directly into R session using <a href="https://www.bioconductor.org/packages/release/bioc/html/GEOquery.html"><code>GEOquery</code> package</a> - see code below (which also refactors some of the metadata for convenience)</p></li>
<li><p>Practice with this in Seminars 4 and 5 (Review lecture 3 for general principles)</p></li>
</ul>
<div class="cell columns column-output-location">
<div class="column">
<div class="sourceCode cell-code" id="cb1" data-code-line-numbers="2,8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="co"># load libraries</span></span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="fu">library</span>(GEOquery)</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="fu">library</span>(gridExtra)</span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="fu">theme_set</span>(<span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">20</span>))</span>
<span id="cb1-6"><a href="#cb1-6"></a></span>
<span id="cb1-7"><a href="#cb1-7"></a><span class="co"># download and read in dataset</span></span>
<span id="cb1-8"><a href="#cb1-8"></a>eset <span class="ot">&lt;-</span> <span class="fu">getGEO</span>(<span class="st">"GSE4051"</span>, <span class="at">getGPL =</span> <span class="cn">FALSE</span>)[[<span class="dv">1</span>]]</span>
<span id="cb1-9"><a href="#cb1-9"></a></span>
<span id="cb1-10"><a href="#cb1-10"></a><span class="co"># recode time points</span></span>
<span id="cb1-11"><a href="#cb1-11"></a><span class="fu">pData</span>(eset) <span class="ot">&lt;-</span> <span class="fu">pData</span>(eset) <span class="sc">%&gt;%</span></span>
<span id="cb1-12"><a href="#cb1-12"></a>  <span class="fu">mutate</span>(<span class="at">sample_id =</span> geo_accession) <span class="sc">%&gt;%</span></span>
<span id="cb1-13"><a href="#cb1-13"></a>  <span class="fu">mutate</span>(<span class="at">dev_stage =</span>  <span class="fu">case_when</span>(</span>
<span id="cb1-14"><a href="#cb1-14"></a>    <span class="fu">grepl</span>(<span class="st">"E16"</span>, title) <span class="sc">~</span> <span class="st">"E16"</span>,</span>
<span id="cb1-15"><a href="#cb1-15"></a>    <span class="fu">grepl</span>(<span class="st">"P2"</span>, title) <span class="sc">~</span> <span class="st">"P2"</span>,</span>
<span id="cb1-16"><a href="#cb1-16"></a>    <span class="fu">grepl</span>(<span class="st">"P6"</span>, title) <span class="sc">~</span> <span class="st">"P6"</span>,</span>
<span id="cb1-17"><a href="#cb1-17"></a>    <span class="fu">grepl</span>(<span class="st">"P10"</span>, title) <span class="sc">~</span> <span class="st">"P10"</span>,</span>
<span id="cb1-18"><a href="#cb1-18"></a>    <span class="fu">grepl</span>(<span class="st">"4 weeks"</span>, title) <span class="sc">~</span> <span class="st">"4_weeks"</span></span>
<span id="cb1-19"><a href="#cb1-19"></a>  )) <span class="sc">%&gt;%</span></span>
<span id="cb1-20"><a href="#cb1-20"></a>  <span class="fu">mutate</span>(<span class="at">genotype =</span> <span class="fu">case_when</span>(</span>
<span id="cb1-21"><a href="#cb1-21"></a>    <span class="fu">grepl</span>(<span class="st">"Nrl-ko"</span>, title) <span class="sc">~</span> <span class="st">"NrlKO"</span>,</span>
<span id="cb1-22"><a href="#cb1-22"></a>    <span class="fu">grepl</span>(<span class="st">"wt"</span>, title) <span class="sc">~</span> <span class="st">"WT"</span></span>
<span id="cb1-23"><a href="#cb1-23"></a>  ))</span>
<span id="cb1-24"><a href="#cb1-24"></a></span>
<span id="cb1-25"><a href="#cb1-25"></a><span class="fu">pData</span>(eset) <span class="ot">&lt;-</span> <span class="fu">pData</span>(eset) <span class="sc">%&gt;%</span></span>
<span id="cb1-26"><a href="#cb1-26"></a>  <span class="fu">mutate</span>(<span class="at">dev_stage =</span> <span class="fu">fct_relevel</span>(dev_stage, <span class="st">"E16"</span>, <span class="st">"P2"</span>, <span class="st">"P6"</span>, <span class="st">"P10"</span>, <span class="st">"4_weeks"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb1-27"><a href="#cb1-27"></a>  <span class="fu">mutate</span>(<span class="at">genotype =</span> <span class="fu">as.factor</span>(genotype))</span>
<span id="cb1-28"><a href="#cb1-28"></a>eset</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column">
<div class="cell-output cell-output-stdout">
<pre><code>ExpressionSet (storageMode: lockedEnvironment)
assayData: 45101 features, 39 samples 
  element names: exprs 
protocolData: none
phenoData
  sampleNames: GSM92610 GSM92611 ... GSM92648 (39 total)
  varLabels: title geo_accession ... genotype (39 total)
  varMetadata: labelDescription
featureData: none
experimentData: use 'experimentData(object)'
  pubMedIds: 16505381 
Annotation: GPL1261 </code></pre>
</div>
</div>
</div>
</section>
<section id="two-example-genes-irs4-and-nrl" class="slide level2">
<h2>Two example genes: <em>Irs4</em> and <em>Nrl</em></h2>
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Biological questions</strong></p>
</div>
<div class="callout-content">
<ol type="1">
<li><p>Is the expression level of gene <em>Irs4</em> truly different in NrlKO compared to WT?</p></li>
<li><p>Is the expression level of gene <em>Nrl</em> truly different in NrlKO compared to WT?</p></li>
</ol>
</div>
</div>
</div>
<div class="fragment">
<p>We can’t answer these questions in general; we can <em>only</em> study these genes in collected data (<strong>gene expression values from a random sample of mice</strong>)</p>
</div>
</section>
<section id="extract-the-two-genes-of-interest" class="slide level2">
<h2>Extract the two genes of interest</h2>
<div class="cell columns column-output-location">
<div class="column">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="co"># function to convert to tidy format</span></span>
<span id="cb3-2"><a href="#cb3-2"></a>toLongerMeta <span class="ot">&lt;-</span> <span class="cf">function</span>(expset) {</span>
<span id="cb3-3"><a href="#cb3-3"></a>    <span class="fu">stopifnot</span>(<span class="fu">class</span>(expset) <span class="sc">==</span> <span class="st">"ExpressionSet"</span>)</span>
<span id="cb3-4"><a href="#cb3-4"></a>    </span>
<span id="cb3-5"><a href="#cb3-5"></a>    expressionMatrix <span class="ot">&lt;-</span> <span class="fu">exprs</span>(expset) <span class="sc">%&gt;%</span> </span>
<span id="cb3-6"><a href="#cb3-6"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb3-7"><a href="#cb3-7"></a>    <span class="fu">rownames_to_column</span>(<span class="st">"gene"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb3-8"><a href="#cb3-8"></a>    <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">!</span>gene, </span>
<span id="cb3-9"><a href="#cb3-9"></a>                 <span class="at">values_to =</span> <span class="st">"Expression"</span>,</span>
<span id="cb3-10"><a href="#cb3-10"></a>                 <span class="at">names_to =</span> <span class="st">"sample_id"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb3-11"><a href="#cb3-11"></a>    <span class="fu">left_join</span>(<span class="fu">pData</span>(expset) <span class="sc">%&gt;%</span> <span class="fu">select</span>(sample_id, dev_stage, genotype),</span>
<span id="cb3-12"><a href="#cb3-12"></a>            <span class="at">by =</span> <span class="st">"sample_id"</span>)</span>
<span id="cb3-13"><a href="#cb3-13"></a>  <span class="fu">return</span>(expressionMatrix)</span>
<span id="cb3-14"><a href="#cb3-14"></a>}</span>
<span id="cb3-15"><a href="#cb3-15"></a></span>
<span id="cb3-16"><a href="#cb3-16"></a><span class="co"># convert to tidy format and extract two genes</span></span>
<span id="cb3-17"><a href="#cb3-17"></a>twoGenes <span class="ot">&lt;-</span> <span class="fu">toLongerMeta</span>(eset) <span class="sc">%&gt;%</span> </span>
<span id="cb3-18"><a href="#cb3-18"></a>  <span class="fu">filter</span>(gene <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"1422248_at"</span>, <span class="st">"1450946_at"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb3-19"><a href="#cb3-19"></a>  <span class="fu">mutate</span>(<span class="at">gene =</span> <span class="fu">ifelse</span>(gene <span class="sc">==</span> <span class="st">"1422248_at"</span>, <span class="st">"Irs4"</span>, <span class="st">"Nrl"</span>)) </span>
<span id="cb3-20"><a href="#cb3-20"></a>twoGenes</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 78 × 5
   gene  sample_id Expression dev_stage genotype
   &lt;chr&gt; &lt;chr&gt;          &lt;dbl&gt; &lt;fct&gt;     &lt;fct&gt;   
 1 Irs4  GSM92610        7.71 4_weeks   NrlKO   
 2 Irs4  GSM92611        7.77 4_weeks   NrlKO   
 3 Irs4  GSM92612        7.73 4_weeks   NrlKO   
 4 Irs4  GSM92613        7.57 4_weeks   NrlKO   
 5 Irs4  GSM92614        7.95 E16       NrlKO   
 6 Irs4  GSM92615        7.52 E16       NrlKO   
 7 Irs4  GSM92616        8.08 E16       NrlKO   
 8 Irs4  GSM92617        7.71 P10       NrlKO   
 9 Irs4  GSM92618        7.87 P10       NrlKO   
10 Irs4  GSM92619        7.75 P10       NrlKO   
# ℹ 68 more rows</code></pre>
</div>
</div>
</div>
<p>What do you notice?</p>
</section>
<section id="visualizing-irs4-and-nrl-genes-in-our-sample" class="slide level2">
<h2>Visualizing <em>Irs4</em> and <em>Nrl</em> genes in our sample</h2>
<div class="panel-tabset">
<ul id="tabset-1" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-1-1">Code</a></li><li><a href="#tabset-1-2">Output</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1">
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>irsLim <span class="ot">&lt;-</span> <span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"Irs4"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb5-2"><a href="#cb5-2"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">y =</span> Expression, <span class="at">x =</span> genotype, <span class="at">colour =</span> genotype)) <span class="sc">+</span> </span>
<span id="cb5-3"><a href="#cb5-3"></a>  <span class="fu">geom_jitter</span>(<span class="at">size =</span> <span class="dv">2</span>, <span class="at">alpha =</span> <span class="fl">0.8</span>, <span class="at">width =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb5-4"><a href="#cb5-4"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Irs4 gene"</span>) <span class="sc">+</span></span>
<span id="cb5-5"><a href="#cb5-5"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb5-6"><a href="#cb5-6"></a>             </span>
<span id="cb5-7"><a href="#cb5-7"></a>nrlLim <span class="ot">&lt;-</span> <span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"Nrl"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb5-8"><a href="#cb5-8"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">y =</span> Expression, <span class="at">x =</span> genotype, <span class="at">colour =</span> genotype)) <span class="sc">+</span> </span>
<span id="cb5-9"><a href="#cb5-9"></a>  <span class="fu">geom_jitter</span>(<span class="at">size =</span> <span class="dv">2</span>, <span class="at">alpha =</span> <span class="fl">0.8</span>, <span class="at">width =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb5-10"><a href="#cb5-10"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Nrl gene"</span>) <span class="sc">+</span></span>
<span id="cb5-11"><a href="#cb5-11"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>) </span>
<span id="cb5-12"><a href="#cb5-12"></a></span>
<span id="cb5-13"><a href="#cb5-13"></a><span class="fu">grid.arrange</span>(irsLim <span class="sc">+</span> <span class="fu">ylim</span>(<span class="dv">5</span>, <span class="dv">13</span>), nrlLim <span class="sc">+</span> <span class="fu">ylim</span>(<span class="dv">5</span>, <span class="dv">13</span>), <span class="at">ncol =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-1-2">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="lect05-twogroup_files/figure-revealjs/unnamed-chunk-4-1.png" width="960"></p>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="formulating-our-hypotheses" class="slide level2">
<h2>Formulating our hypotheses</h2>
<ul>
<li><p><strong>Experimental design:</strong> (ignoring developmental time for now)</p>
<ul>
<li>2 conditions: WT <em>vs</em> NrlKO</li>
<li>observe the expression of many genes in a random sample of ~20 mice from each condition</li>
</ul></li>
<li><p><strong>Biological hypothesis:</strong> for <em>some</em> genes, the expression levels are different between conditions</p></li>
<li><p><strong>Statistical hypotheses:</strong> (for each gene <span class="math inline">\(g=1,...,G\)</span>)</p>
<ul>
<li>H<sub>0</sub> (null hypothesis): the expression level of gene <span class="math inline">\(g\)</span> is the <strong><em>same</em></strong> in both conditions</li>
<li>H<sub>A</sub> (alternative hypothesis): the expression level of gene <span class="math inline">\(g\)</span> is <strong><em>different</em></strong> between conditions</li>
</ul></li>
</ul>
<div class="fragment">
<div class="callout callout-none no-icon callout-style-simple">
<div class="callout-body">
<div class="callout-content">
<p>How might we test H<sub>0</sub>?</p>
</div>
</div>
</div>
</div>
</section>
<section id="notation1" class="slide level2">
<h2>Notation<sup>1</sup></h2>
<p>Population parameters (unknown/unobservable):</p>
<p><span class="math inline">\(\mu_Y = E[Y]\)</span> : the (population) expected expression of gene <span class="math inline">\(g\)</span> in WT mice</p>
<p><span class="math inline">\(\mu_Z = E[Z]\)</span> : the (population) expected expression of gene <span class="math inline">\(g\)</span> in NrlKO mice</p>
<aside><ol class="aside-footnotes"><li id="fn1"><p>ignoring subscript for gene <span class="math inline">\(g\)</span> for now</p></li></ol></aside></section>
<section id="notation1-1" class="slide level2">
<h2>Notation<sup>1</sup></h2>
<p>Random variables and statistics (we can estimate from data):</p>
<ul>
<li><p><span class="math inline">\(Y_i\)</span> : expression of gene <span class="math inline">\(g\)</span> in the WT sample <span class="math inline">\(i\)</span></p></li>
<li><p><span class="math inline">\(Z_i\)</span>: expression of gene <span class="math inline">\(g\)</span> in NrlKO sample <span class="math inline">\(i\)</span></p></li>
<li><p><span class="math inline">\(Y_1, Y_2,..., Y_{n_Y}\)</span> : a <strong>random sample</strong> of size <span class="math inline">\(n_Y\)</span> WT mice</p></li>
<li><p><span class="math inline">\(Z_1, Z_2,..., Z_{n_Z}\)</span> : a <strong>random sample</strong> of size <span class="math inline">\(n_Z\)</span> NrlKO mice</p></li>
<li><p><span class="math inline">\(\bar{Y}=\frac{\sum_{i=1}^{n_Y}Y_i}{n_Y}\)</span>: sample mean of gene <span class="math inline">\(g\)</span> expression from WT mice</p></li>
<li><p><span class="math inline">\(\bar{Z}=\frac{\sum_{i=1}^{n_Z}Z_i}{n_Z}\)</span>: sample mean of gene <span class="math inline">\(g\)</span> expression from NrlKO mice</p></li>
</ul>
<aside><ol class="aside-footnotes"><li id="fn2"><p>ignoring subscript for gene <span class="math inline">\(g\)</span> for now</p></li></ol></aside></section>
<section id="is-there-enough-evidence-to-reject-h0" class="slide level2">
<h2>Is there enough evidence to reject H<sub>0</sub>?</h2>
<h3 id="h_0-mu_y-mu_z"><span class="math inline">\(H_0: \mu_Y = \mu_Z\)</span></h3>

<img data-src="lect05-twogroup_files/figure-revealjs/unnamed-chunk-5-1.png" width="960" class="r-stretch"><p><strong>Statistical Inference</strong>: random samples are used to learn about the population</p>
</section>
<section id="what-we-observe-sample-averages-bary-vs-barz" class="slide level2">
<h2>What we observe: sample averages: <span class="math inline">\(\bar{Y}\)</span> vs <span class="math inline">\(\bar{Z}\)</span></h2>
<div class="columns">
<div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="co"># calculate mean of each gene and genotype</span></span>
<span id="cb6-2"><a href="#cb6-2"></a>meanExp <span class="ot">&lt;-</span> twoGenes <span class="sc">%&gt;%</span></span>
<span id="cb6-3"><a href="#cb6-3"></a>  <span class="fu">group_by</span>(gene, genotype) <span class="sc">%&gt;%</span></span>
<span id="cb6-4"><a href="#cb6-4"></a>  <span class="fu">summarize</span>(<span class="at">meanExpr =</span> <span class="fu">mean</span>(Expression)) <span class="sc">%&gt;%</span></span>
<span id="cb6-5"><a href="#cb6-5"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> genotype, <span class="at">values_from =</span> meanExpr) <span class="sc">%&gt;%</span> </span>
<span id="cb6-6"><a href="#cb6-6"></a>  <span class="fu">mutate</span>(<span class="at">diffExp =</span> NrlKO <span class="sc">-</span> WT)</span>
<span id="cb6-7"><a href="#cb6-7"></a>meanExp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 4
# Groups:   gene [2]
  gene  NrlKO    WT diffExp
  &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;
1 Irs4   7.74  7.77 -0.0261
2 Nrl    6.09 11.2  -5.15  </code></pre>
</div>
</div>
</div><div class="column">
<p>This code uses <a href="https://raw.githubusercontent.com/rstudio/cheatsheets/main/tidyr.pdf">tidy data wrangling functions</a> to calculate:</p>
<ul>
<li>the mean expression of each gene per genotype group</li>
<li>the difference in mean expression of each gene in Nrl KO vs WT groups</li>
</ul>
</div>
</div>
</section>
<section id="is-the-difference-between-bary-and-barz-enough-to-reject-h0" class="slide level2">
<h2>Is the difference between <span class="math inline">\(\bar{Y}\)</span> and <span class="math inline">\(\bar{Z}\)</span> enough to reject H<sub>0</sub>?</h2>

<img data-src="lect05-twogroup_files/figure-revealjs/unnamed-chunk-7-1.png" width="960" class="r-stretch"><div class="fragment">
<ul>
<li><p>The sample means, <span class="math inline">\(\bar{Y}\)</span> vs <span class="math inline">\(\bar{Z}\)</span>, by themselves are not enough to make conclusions about the population</p></li>
<li><p>What is a “large” difference? “Large” relative to what?</p></li>
</ul>
</div>
</section>
<section id="consider-this-artificial-version-of-nrl" class="slide level2">
<h2>Consider this artificial version of <em>Nrl</em></h2>

<img data-src="lect05-twogroup_files/figure-revealjs/unnamed-chunk-8-1.png" width="960" class="r-stretch"><div class="fragment">
<p>What can we use to interpret the size of the mean difference? <span class="math inline">\(\frac{\bar{Y}-\bar{Z}}{??}\)</span></p>
</div>
</section>
<section id="large-difference-relative-to-what" class="slide level2">
<h2>“Large” difference relative to what?</h2>
<div class="columns">
<div class="column">
<p>“Large” relative to the <strong>observed variation:</strong></p>
</div><div class="column">
<p><span class="math display">\[\frac{\bar{Y}-\bar{Z}}{\sqrt{Var(\bar{Y}-\bar{Z})}}\]</span></p>
</div>
</div>

<img data-src="lect05-twogroup_files/figure-revealjs/unnamed-chunk-9-1.png" width="960" class="r-stretch"></section>
<section id="quantifying-observed-variation" class="slide level2">
<h2>Quantifying observed variation</h2>
<ul>
<li><p>Recall that if <span class="math inline">\(Var(Y_i)=\sigma_Y^2\)</span>, then <span class="math inline">\(Var(\bar{Y})=\frac{\sigma_Y^2}{n_Y}\)</span></p></li>
<li><p>Assume that the random variables within each group are <em>independent and identically distributed</em> (iid), and that the groups are independent. More specifically, that</p>
<ol type="1">
<li><span class="math inline">\(Y_1, Y_2,..., Y_{n_Y}\)</span> are iid,</li>
<li><span class="math inline">\(Z_1, Z_2,..., Z_{n_Z}\)</span> are iid, and</li>
<li><span class="math inline">\(Y, Z\)</span> are independent.</li>
</ol></li>
<li><p>Then, it follows that <span class="math inline">\(Var(\bar{Z}-\bar{Y})=\frac{\sigma_Z^2}{n_Z}+\frac{\sigma_Y^2}{n_Y}\)</span></p></li>
</ul>
<div class="fragment">
<ul>
<li>If we also assume equal population variances: <span class="math inline">\(\sigma_Z^2=\sigma_Y^2=\sigma^2\)</span>, then <span class="math display">\[Var(\bar{Z}-\bar{Y})=\frac{\sigma_Z^2}{n_Z}+\frac{\sigma_Y^2}{n_Y}=\sigma^2\left[\frac{1}{n_Z}+\frac{1}{n_Y}\right]\]</span></li>
</ul>
</div>
</section>
<section id="reflect" class="slide level2">
<h2>Reflect</h2>
<div class="callout callout-caution callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Stop!</strong></p>
</div>
<div class="callout-content">
<p>But how can we calculate population variance <span class="math inline">\(\sigma\)</span> if it is <strong>unknown</strong>?</p>
</div>
</div>
</div>
</section>
<section id="using-the-______________-combined-somehow" class="slide level2">
<h2>…using the ______________ (combined, somehow)!</h2>
<div class="columns">
<div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="co"># calculate sample variance for each gene and genotype</span></span>
<span id="cb8-2"><a href="#cb8-2"></a>twoGenes <span class="sc">%&gt;%</span></span>
<span id="cb8-3"><a href="#cb8-3"></a>  <span class="fu">group_by</span>(gene, genotype) <span class="sc">%&gt;%</span></span>
<span id="cb8-4"><a href="#cb8-4"></a>  <span class="fu">summarize</span>(<span class="at">groupVar =</span> <span class="fu">var</span>(Expression))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 3
# Groups:   gene [2]
  gene  genotype groupVar
  &lt;chr&gt; &lt;fct&gt;       &lt;dbl&gt;
1 Irs4  NrlKO      0.0233
2 Irs4  WT         0.0240
3 Nrl   NrlKO      0.594 
4 Nrl   WT         1.22  </code></pre>
</div>
</div>
<p>For example, for Nrl in WT: <span class="math display">\[\hat{\sigma}_Y^2 = S_Y^2=\frac{1}{n_Y}\sum_{i=1}^{n_Y}(Y_i-\bar{Y})^2=1.22\]</span></p>
</div><div class="column">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="lect05-twogroup_files/figure-revealjs/unnamed-chunk-11-1.png" width="432"></p>
</div>
</div>
</div>
</div>
</section>
<section id="combining-sample-variances" class="slide level2">
<h2>Combining sample variances</h2>
<p>Plug these estimates into chosen formula for the variance of difference of sample means:</p>
<ul>
<li>Assuming <strong>equal</strong> variance of Y’s and Z’s</li>
</ul>
<p><span class="math display">\[\hat{Var}(\bar{Z_n}-\bar{Y_n})=\hat{\sigma}_{\text{pooled}}^2\left[\frac{1}{n_Y}+\frac{1}{n_Z}\right]\]</span> <span class="math display">\[\hat{\sigma}_{\text{pooled}}^2=S_Y^2\frac{n_Y-1}{n_Y+n_Z-2}+S_Z^2\frac{n_z-1}{n_Y+n_Z-2}\]</span></p>
<ul>
<li>Assuming <strong>unequal</strong> variance of Y’s and Z’s (Welch’s t-test)</li>
</ul>
<p><span class="math display">\[\hat{Var}(\bar{Z_n}-\bar{Y_n})=\hat{\sigma}_{\bar{Z}_n-\bar{Y}_n}^2=\frac{S_Y^2}{n_Y}+\frac{S_Z^2}{n_Z}\]</span></p>
<p>Recall: the ‘hat’ (^) is used to distinguish an ‘estimate’ from a ‘parameter’</p>
</section>
<section id="test-statistic" class="slide level2">
<h2>Test Statistic</h2>
<p>‘Manual’ calculation of <span class="math inline">\(T=\frac{\bar{Z}_n-\bar{Y}_n}{\sqrt{\hat{Var}(\bar{Z_n}-\bar{Y_n})}}\)</span> (for illustration):</p>
<div class="panel-tabset">
<ul id="tabset-2" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-2-1">Pooled variances</a></li><li><a href="#tabset-2-2">t-statistics</a></li></ul>
<div class="tab-content">
<div id="tabset-2-1">
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="do">## compute sample variance of each gene/genotype</span></span>
<span id="cb10-2"><a href="#cb10-2"></a>theVars <span class="ot">&lt;-</span> twoGenes <span class="sc">%&gt;%</span></span>
<span id="cb10-3"><a href="#cb10-3"></a>  <span class="fu">group_by</span>(gene, genotype) <span class="sc">%&gt;%</span></span>
<span id="cb10-4"><a href="#cb10-4"></a>  <span class="fu">summarize</span>(<span class="at">groupVar =</span> <span class="fu">var</span>(Expression))</span>
<span id="cb10-5"><a href="#cb10-5"></a></span>
<span id="cb10-6"><a href="#cb10-6"></a><span class="do">## compute sample size in each group</span></span>
<span id="cb10-7"><a href="#cb10-7"></a>nY <span class="ot">&lt;-</span> <span class="fu">with</span>(twoGenes, <span class="fu">sum</span>(genotype <span class="sc">==</span> <span class="st">"WT"</span> <span class="sc">&amp;</span> gene <span class="sc">==</span> <span class="st">"Nrl"</span>))</span>
<span id="cb10-8"><a href="#cb10-8"></a>nZ <span class="ot">&lt;-</span> <span class="fu">with</span>(twoGenes, <span class="fu">sum</span>(genotype <span class="sc">==</span> <span class="st">"NrlKO"</span> <span class="sc">&amp;</span> gene <span class="sc">==</span> <span class="st">"Nrl"</span>))</span>
<span id="cb10-9"><a href="#cb10-9"></a></span>
<span id="cb10-10"><a href="#cb10-10"></a><span class="do">## assuming unequal true variance</span></span>
<span id="cb10-11"><a href="#cb10-11"></a>s2DiffWelch <span class="ot">&lt;-</span> theVars <span class="sc">%&gt;%</span> </span>
<span id="cb10-12"><a href="#cb10-12"></a>    <span class="fu">mutate</span>(<span class="at">s2Welch =</span> groupVar <span class="sc">/</span> <span class="fu">ifelse</span>(genotype <span class="sc">==</span> <span class="st">"WT"</span>, nY, nZ)) <span class="sc">%&gt;%</span></span>
<span id="cb10-13"><a href="#cb10-13"></a>    <span class="fu">group_by</span>(gene) <span class="sc">%&gt;%</span></span>
<span id="cb10-14"><a href="#cb10-14"></a>    <span class="fu">summarize</span>(<span class="at">s2Welch =</span> <span class="fu">sum</span>(s2Welch))</span>
<span id="cb10-15"><a href="#cb10-15"></a>meanExp<span class="sc">$</span>s2DiffWelch <span class="ot">&lt;-</span> s2DiffWelch<span class="sc">$</span>s2Welch</span>
<span id="cb10-16"><a href="#cb10-16"></a></span>
<span id="cb10-17"><a href="#cb10-17"></a><span class="do">## assuming equal true variance</span></span>
<span id="cb10-18"><a href="#cb10-18"></a>s2Pooled <span class="ot">&lt;-</span> theVars <span class="sc">%&gt;%</span> </span>
<span id="cb10-19"><a href="#cb10-19"></a>    <span class="fu">mutate</span>(<span class="at">s2Pool =</span> groupVar <span class="sc">*</span> <span class="fu">ifelse</span>(genotype <span class="sc">==</span> <span class="st">"WT"</span>, </span>
<span id="cb10-20"><a href="#cb10-20"></a>                                       (nY <span class="sc">-</span> <span class="dv">1</span>) <span class="sc">/</span> (nY <span class="sc">+</span> nZ <span class="sc">-</span> <span class="dv">2</span>),</span>
<span id="cb10-21"><a href="#cb10-21"></a>                                       (nZ <span class="sc">-</span> <span class="dv">1</span>) <span class="sc">/</span> (nY <span class="sc">+</span> nZ <span class="sc">-</span> <span class="dv">2</span>))) <span class="sc">%&gt;%</span></span>
<span id="cb10-22"><a href="#cb10-22"></a>  <span class="fu">group_by</span>(gene) <span class="sc">%&gt;%</span></span>
<span id="cb10-23"><a href="#cb10-23"></a>  <span class="fu">summarize</span>(<span class="at">s2Pool =</span> <span class="fu">sum</span>(s2Pool))</span>
<span id="cb10-24"><a href="#cb10-24"></a>meanExp<span class="sc">$</span>s2Diff <span class="ot">&lt;-</span> s2Pooled<span class="sc">$</span>s2Pool <span class="sc">*</span> (<span class="dv">1</span><span class="sc">/</span>nY <span class="sc">+</span> <span class="dv">1</span><span class="sc">/</span>nZ)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-2-2">
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>meanExp <span class="sc">%&gt;%</span> </span>
<span id="cb11-2"><a href="#cb11-2"></a>    <span class="fu">mutate</span>(<span class="at">t =</span> diffExp <span class="sc">/</span> <span class="fu">sqrt</span>(s2Diff)) <span class="sc">%&gt;%</span></span>
<span id="cb11-3"><a href="#cb11-3"></a>    <span class="fu">mutate</span>(<span class="at">tWelch =</span> diffExp <span class="sc">/</span> <span class="fu">sqrt</span>(s2DiffWelch))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 8
# Groups:   gene [2]
  gene  NrlKO    WT diffExp s2DiffWelch  s2Diff       t  tWelch
  &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;       &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;
1 Irs4   7.74  7.77 -0.0261     0.00243 0.00243  -0.529  -0.529
2 Nrl    6.09 11.2  -5.15       0.0925  0.0942  -16.8   -17.0  </code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Can we now say whether the observed differences are ‘big’?</p>
<div class="fragment">
<p>The difference is about half a standard deviation for <em>Irs4</em> and ~17 standard deviations for <em>Nrl</em></p>
</div>
</section>
<section id="what-to-do-with-this-statistic" class="slide level2">
<h2>What to do with this statistic?</h2>
<ul>
<li><p>The test statistic <span class="math inline">\(T\)</span> is a <strong>random variable</strong> because it’s based on our <strong>random sample</strong></p></li>
<li><p>We need a measure of its <strong>uncertainty</strong> to determine how extreme our observed <span class="math inline">\(T\)</span> is:</p>
<ul>
<li>If we were to repeat the experiment many times, what’s the probability of observing a value of <span class="math inline">\(T\)</span> <strong>as extreme</strong> as the one we observed?</li>
</ul></li>
</ul>
<div class="fragment">
<ul>
<li><p>We need a probability distribution!</p></li>
<li><p>However, this is unknown to us so we need to <strong>make more assumptions</strong></p></li>
</ul>
</div>
</section>
<section id="null-distribution-assumptions" class="slide level2">
<h2>Null distribution assumptions</h2>
<ul>
<li><p>If we know how our statistic behaves when the <em>null hypothesis is true</em>, then we can evaluate how extreme our observed data is</p>
<ul>
<li>The <strong>null distribution</strong> is the probability distribution of <span class="math inline">\(T\)</span> under H<sub>0</sub></li>
</ul></li>
</ul>
<div class="fragment">
<ul>
<li><p>Let’s assume that <span class="math inline">\(Y_i\)</span> and <span class="math inline">\(Z_i\)</span> follow (unknown) probability distributions called <span class="math inline">\(F\)</span> and <span class="math inline">\(G\)</span>:<br>
<span class="math display">\[(Y_i \sim F, \text{ and } Z_i \sim G)\]</span></p></li>
<li><p>Depending on the assumptions we make about <span class="math inline">\(F\)</span> and <span class="math inline">\(G\)</span>, theory tells us specific <strong>null distributions</strong> for our test statistic</p></li>
</ul>
</div>
</section>
<section id="section" class="slide level2">
<h2></h2>
<h3 id="willing-to-assume-that-f-and-g-are-normal-distributions">Willing to assume that F and G are normal distributions?</h3>
<div class="columns">
<div class="column">
<p><strong>2-sample <em>t</em>-test</strong> (equal variances): <span class="math display">\[T\sim t_{n_Y+n_Z-2}\]</span></p>
</div><div class="column">
<p><strong>Welch’s 2-sample <em>t</em>-test</strong> (unequal variances): <span class="math display">\[T\sim t_{&lt;something\,ugly&gt;}\]</span></p>
</div>
</div>
<div class="fragment">
<!-- -->
<h3 id="unwilling-to-assume-that-f-and-g-are-normal-distributions">Unwilling to assume that F and G are normal distributions?</h3>
<p>But you feel that n<sub>Y</sub> and n<sub>Z</sub> are large enough?</p>
<p>Then the t-distributions above (or even a normal distribution) are decent approximations</p>
</div>
<div class="fragment">
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Review</strong></p>
</div>
<div class="callout-content">
<p>Why could we assume the sampling distribution of <span class="math inline">\(T\)</span> is normally distributed when we have a large sample size?</p>
</div>
</div>
</div>
</div>
</section>
<section id="students-t-distribution" class="slide level2">
<h2>Student’s <em>t</em>-distribution</h2>
<p>Summary: <span class="math inline">\(T=\frac{\bar{Z}_n-\bar{Y}_n}{\sqrt{\hat{Var}(\bar{Z_n}-\bar{Y_n})}}\)</span> is a <strong>random variable</strong>, and under certain assumptions, we can prove that <span class="math inline">\(T\)</span> follows a <em>t</em>-distribution</p>

<img data-src="img/tdist.png" class="r-stretch quarto-figure-center"><p>Recall that the <em>t</em>-distribution has one parameter: df = degrees of freedom</p>
</section>
<section id="hypothesis-testing-step-1" class="slide level2">
<h2>Hypothesis testing: Step 1</h2>
<h3 id="formulate-your-hypothesis-as-a-statistical-hypothesis">1. Formulate your hypothesis as a statistical hypothesis</h3>
<p>In our example:</p>
<p><span class="math display">\[H_0: \mu_Y = \mu_Z \, \text{ vs} \,\,\, H_A: \mu_Y \neq \mu_Z\]</span></p>
</section>
<section id="hypothesis-testing-step-2" class="slide level2">
<h2>Hypothesis testing: Step 2</h2>
<h3 id="a.-choose-a-test-statistic">2a. Choose a test statistic</h3>
<p>In our example: 2-sample <em>t</em>-test, with equal variance</p>
<h3 id="b.-compute-the-observed-value-for-the-test-statistic">2b. Compute the observed value for the test statistic</h3>
<p>For our two example genes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>twoGenes <span class="sc">%&gt;%</span> </span>
<span id="cb13-2"><a href="#cb13-2"></a>  <span class="fu">group_by</span>(gene) <span class="sc">%&gt;%</span></span>
<span id="cb13-3"><a href="#cb13-3"></a>  <span class="fu">summarize</span>(<span class="at">t =</span> <span class="fu">t.test</span>(Expression <span class="sc">~</span> genotype, <span class="at">var.equal=</span><span class="cn">TRUE</span>)<span class="sc">$</span>statistic)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 2
  gene        t
  &lt;chr&gt;   &lt;dbl&gt;
1 Irs4   -0.529
2 Nrl   -16.8  </code></pre>
</div>
</div>
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Tip</strong></p>
</div>
<div class="callout-content">
<p>This code uses a shortcut to computing the t-statistic using the <code>t.test</code> function</p>
</div>
</div>
</div>
</section>
<section id="hypothesis-testing-step-3" class="slide level2">
<h2>Hypothesis testing: Step 3</h2>
<h3 id="compute-the-p-value">3. Compute the p-value</h3>
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Definition</strong></p>
</div>
<div class="callout-content">
<p><strong>p-value</strong>: Probability of observing a test statistic at least as extreme as that observed, under the <em>null sampling distribution</em></p>
</div>
</div>
</div>
<p>For our two example genes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>twoGenes <span class="sc">%&gt;%</span> </span>
<span id="cb15-2"><a href="#cb15-2"></a>  <span class="fu">group_by</span>(gene) <span class="sc">%&gt;%</span></span>
<span id="cb15-3"><a href="#cb15-3"></a>  <span class="fu">summarize</span>(<span class="at">pvalue =</span> <span class="fu">t.test</span>(Expression <span class="sc">~</span> genotype, <span class="at">var.equal=</span><span class="cn">TRUE</span>)<span class="sc">$</span>p.value)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 2
  gene    pvalue
  &lt;chr&gt;    &lt;dbl&gt;
1 Irs4  6.00e- 1
2 Nrl   6.73e-19</code></pre>
</div>
</div>
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Tip</strong></p>
</div>
<div class="callout-content">
<p>The <code>t.test</code> function also computes the p-value for us</p>
</div>
</div>
</div>
</section>
<section id="in-other-words-assuming-that-h0-is-true" class="slide level2">
<h2>In other words, assuming that H<sub>0</sub> is true:</h2>
<p>For <em>Irs4</em>, the probability of seeing a test statistic as extreme as that observed <span class="math inline">\((t = -0.53)\)</span> is pretty high <span class="math inline">\((p = 0.6)\)</span>.</p>
<p>But for <em>Nrl</em>, the probability of seeing a test statistic as extreme as that observed <span class="math inline">\((t = -16.8)\)</span> is extremely low <span class="math inline">\((p=6.76 \times 10^{-19})\)</span></p>

<img data-src="lect05-twogroup_files/figure-revealjs/unnamed-chunk-17-1.png" width="960" class="r-stretch"></section>
<section id="hypothesis-testing-step-4" class="slide level2">
<h2>Hypothesis Testing: Step 4</h2>
<h3 id="make-a-decision-about-significance-of-results">4. Make a decision about significance of results</h3>
<ul>
<li><p>The decision should be based on a pre-specified significance level (<span class="math inline">\(\alpha\)</span>)</p></li>
<li><p><span class="math inline">\(\alpha\)</span> is often set at 0.05. However, this value is arbitrary and may depend on the study.</p></li>
</ul>
<div class="columns">
<div class="column">
<p><strong><em>Irs4</em></strong></p>
<p>Using <span class="math inline">\(\alpha=0.05\)</span>, since the p-value for the Irs4 test is greater than 0.05, we conclude that there is <strong>not enough evidence</strong> in the data to claim that Irs4 has differential expression in WT compared to NrlKO models.</p>
<p>We do not reject H<sub>0</sub>!</p>
</div><div class="column">
<p><strong><em>Nrl</em></strong></p>
<p>Using <span class="math inline">\(\alpha=0.05\)</span>, since the p-value for the Nrl test is much less than 0.05, we conclude that there is <strong>significant</strong> evidence in the data to claim that <em>Nrl</em> has differential expression in WT compared to NrlKO models.</p>
<p>We reject H<sub>0</sub>!</p>
</div>
</div>
</section>
<section id="t.test-function-in-r" class="slide level2">
<h2><code>t.test</code> function in R</h2>
<div class="columns">
<div class="column">
<p>Assuming equal variances</p>
<div class="cell" data-linewidth="50">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>twoGenes <span class="sc">%&gt;%</span> </span>
<span id="cb17-2"><a href="#cb17-2"></a>  <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"Nrl"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb17-3"><a href="#cb17-3"></a>  <span class="fu">t.test</span>(Expression <span class="sc">~</span> genotype, </span>
<span id="cb17-4"><a href="#cb17-4"></a>         <span class="at">var.equal=</span><span class="cn">TRUE</span>, <span class="at">data =</span> .)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Two Sample t-test

data:  Expression by genotype
t = -16.798, df = 37, p-value &lt; 2.2e-16
alternative hypothesis: true difference in means between group NrlKO and group WT is not equal to 0
95 percent confidence interval:
 -5.776672 -4.533071
sample estimates:
mean in group NrlKO    mean in group WT 
           6.089579           11.244451 </code></pre>
</div>
</div>
</div><div class="column">
<p>Not assuming equal variances</p>
<div class="cell" data-linewidth="50">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>twoGenes <span class="sc">%&gt;%</span> </span>
<span id="cb19-2"><a href="#cb19-2"></a>  <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"Nrl"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb19-3"><a href="#cb19-3"></a>  <span class="fu">t.test</span>(Expression <span class="sc">~</span> genotype, </span>
<span id="cb19-4"><a href="#cb19-4"></a>         <span class="at">var.equal=</span><span class="cn">FALSE</span>, <span class="at">data =</span> .)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Welch Two Sample t-test

data:  Expression by genotype
t = -16.951, df = 34.01, p-value &lt; 2.2e-16
alternative hypothesis: true difference in means between group NrlKO and group WT is not equal to 0
95 percent confidence interval:
 -5.772864 -4.536879
sample estimates:
mean in group NrlKO    mean in group WT 
           6.089579           11.244451 </code></pre>
</div>
</div>
</div>
</div>
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Tip</strong></p>
</div>
<div class="callout-content">
<p>Check out <code>?t.test</code> for more options, including how to specify one-sided tests</p>
</div>
</div>
</div>
</section>
<section id="interpreting-p-values" class="slide level2">
<h2>Interpreting p-values</h2>
<p>Which of the following are true? (select all that apply)</p>
<ol type="a">
<li>If the effect size is very small, but the sample size is large enough, it is possible to have a statistically significant p-value</li>
<li>A study may show a relatively large magnitude of association (effect size), but a statistically insignificant p-value if the sample size is small</li>
<li>A very small p-value indicates there is a very small chance the finding is a false positive</li>
</ol>
</section>
<section id="common-p-value-pitfalls" class="slide level2">
<h2>Common p-value pitfalls</h2>
<div class="callout callout-caution callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Caution</strong></p>
</div>
<div class="callout-content">
<p>Valid inference using p-values depends on accurate assumptions about null sampling distribution</p>
</div>
</div>
</div>
<div class="callout callout-caution callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Caution</strong></p>
</div>
<div class="callout-content">
<p>A p-value is <strong>NOT</strong>:</p>
<ul>
<li><p>The probability that the null hypothesis is true</p></li>
<li><p>The probability that the finding is a “fluke”</p></li>
<li><p>A measure of the size or importance of observed effects</p></li>
</ul>
</div>
</div>
</div>
<iframe data-external="1" src="https://www.youtube.com/embed/ax0tDcFkPic" title="" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen=""></iframe>
</section>
<section id="preview-genome-wide-testing-of-differential-expression" class="slide level2">
<h2>Preview: “Genome-wide” testing of differential expression</h2>
<ul>
<li><p>In genomics, we often perform thousands of statistical tests (e.g., a <em>t</em>-test per gene)</p></li>
<li><p>The distribution of p-values across all tests provides good diagnostics/insights</p></li>
<li><p>Is it mostly uniform (flat)? If not, is the departure from uniform expected based on biological knowledge?</p></li>
<li><p>We will revisit these topics in greater detail in later lectures</p></li>
</ul>
</section>
<section id="different-kinds-of-t-tests" class="slide level2">
<h2>Different kinds of <em>t</em>-tests:</h2>
<ul>
<li><p>One sample <em>or</em> <strong>two samples</strong></p></li>
<li><p>One-sided <em>or</em> <strong>two sided</strong></p></li>
<li><p>Paired <em>or</em> <strong>unpaired</strong></p></li>
<li><p>Equal variance <em>or</em> unequal variance</p></li>
</ul>
</section>
<section id="types-of-errors-in-hypothesis-testing" class="slide level2">
<h2>Types of Errors in Hypothesis Testing</h2>

<img data-src="img/hypError.png" class="r-stretch quarto-figure-center"><p><span class="math display">\[ \alpha = P(\text{Type I Error}), \text{   } \beta = P(\text{Type II Error}), \text{   Power} = 1- \beta\]</span></p>
</section>
<section id="h0-innocent-until-proven-guilty" class="slide level2">
<h2>H<sub>0</sub>: “<em>Innocent until proven guilty</em>”</h2>
<ul>
<li><p>The default state is <span class="math inline">\(H_0 \rightarrow\)</span> we only reject if we have enough evidence</p></li>
<li><p>If <span class="math inline">\(H_0\)</span>: Innocent and <span class="math inline">\(H_A\)</span>: Guilty, then</p>
<ul>
<li><p>Type I Error <span class="math inline">\((\alpha)\)</span>: Wrongfully convict innocent (<em>False Positive</em>)</p></li>
<li><p>Type II Error <span class="math inline">\((\beta)\)</span>: Fail to convict criminal (<em>False Negative</em>)</p></li>
</ul></li>
</ul>
</section>
<section id="section-1" class="slide level2">
<h2></h2>
<h3 id="willing-to-assume-that-f-and-g-are-normal-distributions-1">Willing to assume that F and G are normal distributions?</h3>
<div class="columns">
<div class="column">
<p><strong>2-sample <em>t</em>-test</strong> (equal variances): <span class="math display">\[T\sim t_{n_Y+n_Z-2}\]</span></p>
</div><div class="column">
<p><strong>Welch’s 2-sample <em>t</em>-test</strong> (unequal variances): <span class="math display">\[T\sim t_{&lt;something\,ugly&gt;}\]</span></p>
</div>
</div>
<h3 id="unwilling-to-assume-that-f-and-g-are-normal-distributions-1">Unwilling to assume that F and G are normal distributions?</h3>
<p>But you feel that n<sub>Y</sub> and n<sub>Z</sub> are large enough?</p>
<p>Then the t-distributions above (or even a normal distribution) are decent approximations</p>
<div class="callout callout-caution callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Stop!</strong></p>
</div>
<div class="callout-content">
<p>What if we aren’t comfortable assuming the underlying data generating process is normal <strong>AND</strong> we aren’t sure our sample is large enough to invoke the CLT?</p>
</div>
</div>
</div>
</section>
<section id="what-are-alternatives-to-the-t-test" class="slide level2">
<h2>What are alternatives to the <em>t</em>-test?</h2>
<ul>
<li><p>First, one could use the t test statistic but use a <strong>permutation approach</strong> to compute its p-value; we’ll revisit this topic later</p></li>
<li><p><em>Non-parametric</em> tests are an alternative:</p>
<ul>
<li><p><strong>Wilcoxon rank sum test</strong> (Mann Whitney) uses ranks to test differences in population means</p></li>
<li><p><strong>Kolmogorov-Smirnov test</strong> uses the empirical CDF to test differences in population cumulative distributions</p></li>
</ul></li>
</ul>
</section>
<section id="wilcoxon-rank-sum-test" class="slide level2">
<h2>Wilcoxon rank sum test</h2>
<ul>
<li><p>Rank all data, <strong>ignoring the grouping</strong> variable</p></li>
<li><p><strong>Test statistic</strong> = sum of the ranks for one group (optionally, subtract the minimum possible which is <span class="math inline">\(\frac{n_Y(n_Y+1)}{2}\)</span>)</p>
<ul>
<li>Alternative but equivalent formulation based on the number of <span class="math inline">\(y_i, z_i\)</span> pairs for which <span class="math inline">\(y_i \geq z_i\)</span></li>
</ul></li>
<li><p>The null distribution of such statistics can be worked out or approximated</p></li>
</ul>
</section>
<section id="wilcox.test-function-in-r" class="slide level2">
<h2><code>wilcox.test</code> function in R</h2>
<div class="panel-tabset">
<ul id="tabset-3" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-3-1"><em>Irs4</em></a></li><li><a href="#tabset-3-2"><em>Nrl</em></a></li></ul>
<div class="tab-content">
<div id="tabset-3-1">
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a><span class="fu">wilcox.test</span>(Expression <span class="sc">~</span> genotype, </span>
<span id="cb21-2"><a href="#cb21-2"></a>            <span class="at">data =</span> twoGenes <span class="sc">%&gt;%</span> <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"Irs4"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Wilcoxon rank sum exact test

data:  Expression by genotype
W = 160, p-value = 0.4115
alternative hypothesis: true location shift is not equal to 0</code></pre>
</div>
</div>
</div>
<div id="tabset-3-2">
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a><span class="fu">wilcox.test</span>(Expression <span class="sc">~</span> genotype, </span>
<span id="cb23-2"><a href="#cb23-2"></a>            <span class="at">data =</span> twoGenes <span class="sc">%&gt;%</span> <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"Nrl"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Wilcoxon rank sum exact test

data:  Expression by genotype
W = 1, p-value = 5.804e-11
alternative hypothesis: true location shift is not equal to 0</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="kolmogorov-smirnov-test-two-sample" class="slide level2">
<h2>Kolmogorov-Smirnov test (two sample)</h2>
<ul>
<li><p><strong>Null hypothesis</strong>: F = G, i.e.&nbsp;the distributions are the same</p></li>
<li><p>Estimate each CDF with the empirical CDF (ECDF)</p></li>
</ul>
<p><span class="math display">\[\hat{F}(x)=\frac{1}{n}\sum_{i=1}^n{I[x_i\leq{x}]}\]</span></p>
<ul>
<li><strong>Test statistic</strong> is the maximum of the absolute difference between the ECDFs<sup>1</sup></li>
</ul>
<p><span class="math display">\[max|\hat{F}(x)-\hat{G}(x)|\]</span></p>
<ul>
<li>Null distribution does not depend on F, G (!)</li>
</ul>
<aside><ol class="aside-footnotes"><li id="fn3"><p>I’m suppressing detail here</p></li></ol></aside></section>
<section id="kolmogorov-smirnov-test-two-sample-1" class="slide level2">
<h2>Kolmogorov-Smirnov test (two sample)</h2>

<img data-src="lect05-twogroup_files/figure-revealjs/unnamed-chunk-22-1.png" width="960" class="r-stretch"></section>
<section id="ks.test-function-in-r" class="slide level2">
<h2><code>ks.test</code> function in R</h2>
<div class="panel-tabset">
<ul id="tabset-4" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-4-1"><em>Irs4</em></a></li><li><a href="#tabset-4-2"><em>Nrl</em></a></li></ul>
<div class="tab-content">
<div id="tabset-4-1">
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a>Irs4gene <span class="ot">&lt;-</span> twoGenes <span class="sc">%&gt;%</span> </span>
<span id="cb25-2"><a href="#cb25-2"></a>  <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"Irs4"</span>)</span>
<span id="cb25-3"><a href="#cb25-3"></a><span class="fu">ks.test</span>(Irs4gene<span class="sc">$</span>Expression[Irs4gene<span class="sc">$</span>genotype <span class="sc">==</span> <span class="st">"WT"</span>],</span>
<span id="cb25-4"><a href="#cb25-4"></a>        Irs4gene<span class="sc">$</span>Expression[Irs4gene<span class="sc">$</span>genotype <span class="sc">==</span> <span class="st">"NrlKO"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Exact two-sample Kolmogorov-Smirnov test

data:  Irs4gene$Expression[Irs4gene$genotype == "WT"] and Irs4gene$Expression[Irs4gene$genotype == "NrlKO"]
D = 0.28421, p-value = 0.3278
alternative hypothesis: two-sided</code></pre>
</div>
</div>
</div>
<div id="tabset-4-2">
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a>Nrlgene <span class="ot">&lt;-</span> twoGenes <span class="sc">%&gt;%</span> </span>
<span id="cb27-2"><a href="#cb27-2"></a>  <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"Nrl"</span>)</span>
<span id="cb27-3"><a href="#cb27-3"></a><span class="fu">ks.test</span>(Nrlgene<span class="sc">$</span>Expression[Nrlgene<span class="sc">$</span>genotype <span class="sc">==</span> <span class="st">"WT"</span>],</span>
<span id="cb27-4"><a href="#cb27-4"></a>        Nrlgene<span class="sc">$</span>Expression[Nrlgene<span class="sc">$</span>genotype <span class="sc">==</span> <span class="st">"NrlKO"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Exact two-sample Kolmogorov-Smirnov test

data:  Nrlgene$Expression[Nrlgene$genotype == "WT"] and Nrlgene$Expression[Nrlgene$genotype == "NrlKO"]
D = 0.95, p-value = 5.804e-10
alternative hypothesis: two-sided</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="discussion" class="slide level2">
<h2>Discussion</h2>
<ol type="1">
<li><p>What test(s) might be appropriate if your sample size is just barely large enough to invoke CLT, but you also have suspected outliers?</p></li>
<li><p>If more than one test is appropriate (e.g.&nbsp;<em>t</em>-test, Wilcoxon, and KS), which should we report?</p></li>
<li><p>What is generally more important for results interpretation: the effect size or the p-value?</p></li>
<li><p>What should you do if methods that are equally appropriate and defensible give very different answers?</p></li>
</ol>

<img src="https://github.com/STAT540-UBC/stat540-ubc.github.io/raw/main/images/stat540-logo-s.png" class="slide-logo r-stretch"><div class="footer footer-default">

</div>
</section>

    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="lect05-twogroup_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="lect05-twogroup_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="lect05-twogroup_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="lect05-twogroup_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="lect05-twogroup_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="lect05-twogroup_files/libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="lect05-twogroup_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="lect05-twogroup_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="lect05-twogroup_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="lect05-twogroup_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="lect05-twogroup_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1600,

        height: 900,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>