<!DOCTYPE html>
<html lang="en"><head>
<script src="lect06-anova_files/libs/clipboard/clipboard.min.js"></script>
<script src="lect06-anova_files/libs/quarto-html/tabby.min.js"></script>
<script src="lect06-anova_files/libs/quarto-html/popper.min.js"></script>
<script src="lect06-anova_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="lect06-anova_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="lect06-anova_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="lect06-anova_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="lect06-anova_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.3.450">

  <meta name="author" content="Keegan Korthauer">
  <meta name="dcterms.date" content="2025-01-23">
  <title>Linear models and ANOVA</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="lect06-anova_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="lect06-anova_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="lect06-anova_files/libs/revealjs/dist/theme/quarto.css">
  <link href="lect06-anova_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="lect06-anova_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="lect06-anova_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="lect06-anova_files/libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="lect06-anova_files/libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="lect06-anova_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" data-background-color="#197aa0" data-background-image="https://github.com/STAT540-UBC/stat540-ubc.github.io/raw/main/images/stat540-logo-s.png" data-background-opacity="0.9" data-background-position="95% 90%" data-background-size="12%" class="quarto-title-block center">
  <h1 class="title">Linear models and ANOVA</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Keegan Korthauer 
</div>
</div>
</div>

  <p class="date">January 23, 2025</p>
</section>
<section id="recap-are-these-genes-different-in-nrlko-vs-wt" class="slide level2">
<h2>Recap: Are these genes different in NrlKO vs WT?</h2>
<p>H<sub>0</sub>: the expression level of gene <span class="math inline">\(g\)</span> is the same in both genotypes</p>
<p>Is there <strong>enough</strong> evidence in the data to reject H<sub>0</sub>?</p>

<img data-src="lect06-anova_files/figure-revealjs/unnamed-chunk-2-1.png" width="1056" class="r-stretch quarto-figure-center"></section>
<section id="learn-about-a-population-from-a-random-sample" class="slide level2">
<h2>Learn about a population from a random sample</h2>
<div class="columns">
<div class="column">
<p><strong>Population</strong> (Unknown) <span class="math display">\[Y \sim F, \,\, Z \sim G\]</span> <span class="math display">\[E[Y] = \mu_Y, \,\, E[Z] = \mu_Z\]</span> <span class="math display">\[Var[Y] = {\sigma}_Y^2,\,\, Var[Z] = {\sigma}_Z^2\]</span> <span class="math display">\[H_0: \mu_Y = \mu_Z\]</span></p>
<p><span class="math display">\[H_A: \mu_Y \neq \mu_Z\]</span></p>
<div class="fragment">
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Review:</strong></p>
</div>
<div class="callout-content">
<p>Which quantities on this slide are <strong>statistics</strong> and which are <strong>parameters</strong>?</p>
</div>
</div>
</div>
</div>
</div><div class="column">
<p><strong>Sample</strong> (Observed, with randomness)</p>
<p><span class="math display">\[(Y_1, Y_2, ..., Y_{n_Y}) \text{ and } (Z_1, Z_2, ..., Z_{n_Z})\]</span> <span class="math display">\[\hat{\mu}_Y = \bar{Y} = \frac{\sum_{i=1}^{n_Y} Y_i}{n_Y}\]</span></p>
<p><span class="math display">\[\hat{\sigma}_Y^2 = S_Y^2=\frac{1}{n_Y-1}\sum_{i=1}^{n_Y}(Y_i-\bar{Y})^2\]</span></p>
<p>(with similar quantities for <span class="math inline">\(Z:\)</span> <span class="math inline">\(\bar{Z}\)</span> and <span class="math inline">\(S^2_Z)\)</span></p>
<p><span class="math display">\[T = \frac{\bar{Y}-\bar{Z}}{\sqrt{\hat{Var}(\bar{Y}-\bar{Z}))}}\]</span></p>
</div>
</div>
</section>
<section id="summary-hypothesis-testing" class="slide level2">
<h2>Summary: Hypothesis testing</h2>
<ol type="1">
<li><p>Formulate scientific hypothesis as a <strong>statistical hypothesis</strong> <span class="math inline">\((H_0 \text{ vs } H_A)\)</span></p></li>
<li><p>Choose a <strong>test statistic</strong> to test <span class="math inline">\(H_0\)</span> and compute its <strong>observed value</strong>. For example:</p></li>
</ol>
<ul>
<li>2-sample <em>t</em>-test</li>
<li>Welch’s <em>t</em>-test (unequal variance)</li>
<li>Wilcoxon rank-sum test</li>
<li>Kolmogorov-Smirnov test</li>
</ul>
<ol start="3" type="1">
<li>Compute the probability of seeing a test statistic as extreme as that observed, under the <strong>null sampling distribution</strong> (p-value)</li>
<li>Make a decision about the <strong>significance</strong> of the results, based on a pre-specified significance level ( <span class="math inline">\(\alpha\)</span> )</li>
</ol>
</section>
<section id="we-can-run-these-tests-in-r" class="slide level2">
<h2>We can run these tests in R</h2>
<p>Example: use the <code>t.test</code> function to test H<sub>0</sub> using a 2-sample <em>t</em>-test with equal variance:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"Irs4"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb1-2"><a href="#cb1-2"></a>  <span class="fu">t.test</span>(expression <span class="sc">~</span> genotype, <span class="at">data =</span> ., <span class="at">var.equal =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Two Sample t-test

data:  expression by genotype
t = 0.52854, df = 37, p-value = 0.6003
alternative hypothesis: true difference in means between group WT and group NrlKO is not equal to 0
95 percent confidence interval:
 -0.07384018  0.12595821
sample estimates:
   mean in group WT mean in group NrlKO 
           7.765671            7.739612 </code></pre>
</div>
</div>
</section>
<section id="discussion---choice-and-interpretation-of-hypothesis-tests" class="slide level2">
<h2>Discussion - choice and interpretation of hypothesis tests</h2>
<ol type="1">
<li><p>If more than one test is appropriate (e.g.&nbsp;<em>t</em>-test, Wilcoxon, and KS), which should I choose?</p></li>
<li><p>What should I do if several tests (e.g.&nbsp;<em>t</em>-test, Wilcoxon, KS) give very different answers?</p></li>
<li><p>What is generally more important for results interpretation: the effect size or the p-value?</p></li>
</ol>
<div class="fragment">
<div class="callout callout-important callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Key question</strong></p>
</div>
<div class="callout-content">
<p>Why should I bother with the assumptions of the t-test, which aren’t necessary for the Wilcoxon or KS?</p>
</div>
</div>
</div>
<aside class="notes">
<p>What test should I use?</p>
<ul>
<li><p>How to decide which test to carry out (e.g.&nbsp;t-test, Wilcoxon, KS)?</p></li>
<li><p>Are assumptions met for each one?</p>
<ul>
<li><p>If so, parametric tests (e.g.&nbsp;t-test) tend to have slightly higher power (ability to reject H<sub>0</sub> when H<sub>0</sub> is false)</p></li>
<li><p>But if assumptions are potentially violated, non-parametric tests (Wilcoxon, KS) are a safer choice (albeit conservative)</p></li>
</ul></li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</div>
</section>
<section id="todays-learning-objectives" class="slide level2">
<h2>Today’s Learning Objectives</h2>
<ol type="1">
<li><p>Compare means of different groups (2 or more) using a <strong>linear regression model</strong></p></li>
<li><p>Write a linear model using matrix notation and understand which matrix is built by R</p></li>
<li><p>Use ‘indicator’ variables to represent the levels of a qualitative explanatory variable</p></li>
<li><p>Distinguish between <strong>single</strong> and <strong>joint</strong> hypothesis tests (e.g.&nbsp;<span class="math inline">\(t\)</span>-tests vs <span class="math inline">\(F\)</span>-tests)</p></li>
</ol>
</section>
<section id="ways-to-test-h0-mu_1-mu_2" class="slide level2">
<h2>3 ways to test H<sub>0</sub>: <span class="math inline">\(\mu_1 = \mu_2\)</span></h2>
<div class="panel-tabset">
<ul id="tabset-1" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-1-1">t-test</a></li><li><a href="#tabset-1-2">ANOVA</a></li><li><a href="#tabset-1-3">linear regresion</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1">
<p><strong>2-sample t-test (with equal variance)</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3" data-code-line-numbers="2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"Irs4"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb3-2"><a href="#cb3-2"></a>  <span class="fu">t.test</span>(expression <span class="sc">~</span> genotype, <span class="at">data =</span> ., <span class="at">var.equal =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Two Sample t-test

data:  expression by genotype
t = 0.52854, df = 37, p-value = 0.6003
alternative hypothesis: true difference in means between group WT and group NrlKO is not equal to 0
95 percent confidence interval:
 -0.07384018  0.12595821
sample estimates:
   mean in group WT mean in group NrlKO 
           7.765671            7.739612 </code></pre>
</div>
</div>
</div>
<div id="tabset-1-2">
<p><strong>(One-way) Analysis of Variance (ANOVA)</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5" data-code-line-numbers="2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"Irs4"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb5-2"><a href="#cb5-2"></a>  <span class="fu">aov</span>(expression <span class="sc">~</span> genotype, <span class="at">data =</span> .) <span class="sc">%&gt;%</span></span>
<span id="cb5-3"><a href="#cb5-3"></a>  <span class="fu">summary</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            Df Sum Sq  Mean Sq F value Pr(&gt;F)
genotype     1 0.0066 0.006617   0.279    0.6
Residuals   37 0.8764 0.023685               </code></pre>
</div>
</div>
</div>
<div id="tabset-1-3">
<p><strong>Linear Regression</strong><sup>1</sup></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7" data-code-line-numbers="2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"Irs4"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb7-2"><a href="#cb7-2"></a>  <span class="fu">lm</span>(expression <span class="sc">~</span> genotype, <span class="at">data =</span> .) <span class="sc">%&gt;%</span></span>
<span id="cb7-3"><a href="#cb7-3"></a>  <span class="fu">summary</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = expression ~ genotype, data = .)

Residuals:
     Min       1Q   Median       3Q      Max 
-0.36553 -0.09067  0.01724  0.09396  0.36496 

Coefficients:
              Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)    7.76567    0.03441 225.659   &lt;2e-16 ***
genotypeNrlKO -0.02606    0.04930  -0.529      0.6    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.1539 on 37 degrees of freedom
Multiple R-squared:  0.007494,  Adjusted R-squared:  -0.01933 
F-statistic: 0.2794 on 1 and 37 DF,  p-value: 0.6003</code></pre>
</div>
</div>
</div>
</div>
</div>
<aside><ol class="aside-footnotes"><li id="fn1"><p>Note differences in sign between t-test &amp; linear regression: pay attention to which group is ‘reference’</p></li></ol></aside></section>
<section id="these-are-not-coincidences" class="slide level2">
<h2>These are not coincidences!</h2>
<div class="panel-tabset">
<ul id="tabset-2" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-2-1">t-test</a></li><li><a href="#tabset-2-2">ANOVA</a></li><li><a href="#tabset-2-3">linear regresion</a></li></ul>
<div class="tab-content">
<div id="tabset-2-1">
<p><strong>2-sample t-test (with equal variance)</strong></p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>$`t statistic`
        t 
0.5285386 

$`p-value`
[1] 0.6002819

$`mean difference`
[1] 0.02605902

$`(t statistic)^2`
       t 
0.279353 </code></pre>
</div>
</div>
</div>
<div id="tabset-2-2">
<p><strong>(One-way) Analysis of Variance (ANOVA)</strong><sup>1</sup></p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>$`F statistic`
[1] 0.279353

$`p-value`
[1] 0.6002819</code></pre>
</div>
</div>
</div>
<div id="tabset-2-3">
<p><strong>Linear Regression</strong></p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>$`t statistic`
[1] -0.5285386

$`p-value`
[1] 0.6002819

$`coefficient estimate`
[1] -0.02605902</code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="fragment">
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Key Question</strong></p>
</div>
<div class="callout-content">
<p>Why are these giving us the same results?</p>
</div>
</div>
</div>
</div>
<aside><ol class="aside-footnotes"><li id="fn2"><p>Note that the t-statistic squared is equal to the ANOVA F statistic</p></li></ol></aside></section>
<section id="t-test-vs-linear-regression-wheres-the-line" class="slide level2">
<h2><em>t</em>-test vs <em>line</em>ar regression: where’s the <em>line</em><sup>1</sup>?</h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="lect06-anova_files/figure-revealjs/unnamed-chunk-11-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
<div class="fragment">
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Key Question</strong></p>
</div>
<div class="callout-content">
<p>Why can we run a t-test with a <strong>linear</strong> regression model?</p>
</div>
</div>
</div>
</div>
<aside><ol class="aside-footnotes"><li id="fn3"><p>Note that the <span class="math inline">\(x\)</span>-axis in these plots is not numerical, thus a line in this space does not have any mathematical meaning.</p></li></ol></aside></section>
<section id="from-t-test-to-linear-regression" class="slide level2">
<h2>From <em>t</em>-test to linear regression</h2>
<p>Let’s change the notation to give a common framework to all methods</p>
<span class="math display">\[Y \sim G; \; E[Y] = \mu_Y\]</span>
<center>
<strong>↓</strong>
</center>
<p><span class="math display">\[Y = \mu_Y + \varepsilon_Y; \; \varepsilon_Y \sim G; \; E[\varepsilon_Y] = 0\]</span></p>
<div class="fragment">
<h3 id="why-is-this-equivalent">Why is this equivalent?</h3>
<p><span class="math display">\[E[Y] = E[\mu_Y + \varepsilon_Y] = \mu_Y + E[\varepsilon_Y] = \mu_Y\]</span> We are just rewriting <span class="math inline">\(Y\)</span> here</p>
</div>
</section>
<section id="from-t-test-to-linear-regression-1" class="slide level2">
<h2>From <em>t</em>-test to linear regression</h2>
<p>Let’s change the notation to give a common framework to all methods</p>
<span class="math display">\[Y \sim G; \; E[Y] = \mu_Y\]</span>
<center>
<strong>↓</strong>
</center>
<p><span class="math display">\[Y = \mu_Y + \varepsilon_Y; \; \varepsilon_Y \sim G; \; E[\varepsilon_Y] = 0\]</span></p>
<h3 id="we-can-use-indices-to-accommodate-multiple-groups-i.e.">We can use indices to accommodate multiple groups, i.e.,</h3>
<p><span class="math display">\[Y_{ij} = \mu_j + \varepsilon_{ij};\; \; \varepsilon_{ij} \sim G_j; \; \;E[\varepsilon_{ij}] = 0;\]</span></p>
<p>where <span class="math inline">\(j = \textrm{\{WT, NrlKO}\}\)</span> (or <span class="math inline">\(j=\textrm{\{1, 2}\}\)</span> ) identifies the groups; and <span class="math inline">\(i=1, \ldots, n_j\)</span> identifies the observations within each group</p>
<div class="fragment">
<p>For example: <span class="math inline">\(Y_{11}\)</span> is the first observation in group 1 or WT</p>
</div>
</section>
<section id="this-is-called-the-cell-means-model" class="slide level2">
<h2>This is called the <strong>cell-means model</strong></h2>
<p>Using data from the model</p>
<p><span class="math display">\[Y_{ij} = \mu_j + \varepsilon_{ij};\; \; \varepsilon_{ij} \sim G; \; \;E[\varepsilon_{ij}] = 0;\]</span></p>
<p>where <span class="math inline">\(j\)</span> indexes groups (e.g.&nbsp;WT vs NrlKO) and <span class="math inline">\(i\)</span> indexes samples within group, the goal is to test <span class="math inline">\(H_0 : \mu_1 = \mu_2\)</span></p>
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Note</strong></p>
</div>
<div class="callout-content">
<p>In the <strong>cell-means</strong> model parameterization, we have a parameter <span class="math inline">\(E[Y_{ij}] = \mu_j\)</span> that represents the population mean of each group (in our example: genotype)</p>
</div>
</div>
</div>
<div class="callout callout-important callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Important</strong></p>
</div>
<div class="callout-content">
<p>We assume a common distribution <span class="math inline">\(G\)</span> for all groups (equal variance assumption)</p>
</div>
</div>
</div>
<div class="fragment">
<p>Why the name? ‘Cell’ here refers to a cell of a table - e.g.&nbsp;make a table of means by group, and <span class="math inline">\(\mu_j\)</span> represents the population value for each cell <span class="math inline">\(j\)</span> in the table</p>
</div>
</section>
<section id="recall-sample-mean-estimator-of-population-mean" class="slide level2">
<h2>Recall: sample mean estimator of population mean</h2>
<ul>
<li><p>For each group <span class="math inline">\(j\)</span>, the <strong>population</strong> mean is given by <span class="math inline">\(E[Y_{ij}] = \mu_j\)</span></p></li>
<li><p>A natural <em>estimator</em> of the population mean <span class="math inline">\(\mu_j\)</span> is the <strong>sample</strong> mean <span class="math inline">\(\hat{\mu}_j = \bar{Y_j} = \frac{\sum_{i=1}^{n_j}Y_{ij}}{n_j}\)</span></p></li>
<li><p>Recall that the <code>t.test</code> function calculates these for us in R</p></li>
</ul>
</section>
<section id="but-why-does-the-lm-function-report-different-estimates" class="slide level2">
<h2>But why does the <code>lm</code> function report different estimates?</h2>
<div class="panel-tabset">
<ul id="tabset-3" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-3-1"><code>t.test</code></a></li><li><a href="#tabset-3-2"><code>lm</code></a></li></ul>
<div class="tab-content">
<div id="tabset-3-1">
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="co"># t.test</span></span>
<span id="cb12-2"><a href="#cb12-2"></a><span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"Irs4"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb12-3"><a href="#cb12-3"></a>  <span class="fu">t.test</span>(expression <span class="sc">~</span> genotype, <span class="at">data =</span> ., <span class="at">var.equal =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Two Sample t-test

data:  expression by genotype
t = 0.52854, df = 37, p-value = 0.6003
alternative hypothesis: true difference in means between group WT and group NrlKO is not equal to 0
95 percent confidence interval:
 -0.07384018  0.12595821
sample estimates:
   mean in group WT mean in group NrlKO 
           7.765671            7.739612 </code></pre>
</div>
</div>
</div>
<div id="tabset-3-2">
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="co"># lm</span></span>
<span id="cb14-2"><a href="#cb14-2"></a><span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"Irs4"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb14-3"><a href="#cb14-3"></a>  <span class="fu">lm</span>(expression <span class="sc">~</span> genotype, <span class="at">data =</span> .) <span class="sc">%&gt;%</span> </span>
<span id="cb14-4"><a href="#cb14-4"></a>  <span class="fu">summary</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = expression ~ genotype, data = .)

Residuals:
     Min       1Q   Median       3Q      Max 
-0.36553 -0.09067  0.01724  0.09396  0.36496 

Coefficients:
              Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)    7.76567    0.03441 225.659   &lt;2e-16 ***
genotypeNrlKO -0.02606    0.04930  -0.529      0.6    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.1539 on 37 degrees of freedom
Multiple R-squared:  0.007494,  Adjusted R-squared:  -0.01933 
F-statistic: 0.2794 on 1 and 37 DF,  p-value: 0.6003</code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="fragment">
<ul>
<li><code>(Intercept)</code> estimate from <code>lm</code> is the <strong>sample mean</strong> of WT group</li>
<li><code>genotypeNrlKO</code> estimate from <code>lm</code> is <strong>not</strong> the sample mean of the NrlKO group… what is it?</li>
</ul>
</div>
</section>
<section id="parameterization-how-to-write-the-model" class="slide level2">
<h2>Parameterization: how to write the model?</h2>
<ul>
<li><p>By default, the <code>lm</code> function does not use the cell-means parameterization</p></li>
<li><p>Usually, the goal is to <em>compare</em> the means, not to study each in isolation</p></li>
</ul>
<p>Let’s let <span class="math inline">\(\theta = \mu_1\)</span> and rewrite <span class="math inline">\(\mu_j = \theta + \tau_j\)</span>, and plug into <strong>cell-means</strong> <span class="math inline">\((\mu_j)\)</span> model: <span class="math display">\[Y_{ij} = \mu_j + \varepsilon_{ij};\; \; \varepsilon_{ij} \sim G; \; \;E[\varepsilon_{ij}] = 0;\]</span></p>
<center>
↓
</center>
<p>This gives us the <strong>reference-treatment effect</strong> <span class="math inline">\((\theta,\tau_j)\)</span> model: <span class="math display">\[Y_{ij} = \theta+\tau_j + \varepsilon_{ij};\; \; \tau_1=0, \; \; \varepsilon_{ij} \sim G; \; \;E[\varepsilon_{ij}] = 0;\]</span></p>
</section>
<section id="reference-treatment-effect-parameterization" class="slide level2">
<h2>Reference-treatment effect parameterization</h2>
<p><strong>Reference-treatment effect</strong> <span class="math inline">\((\theta,\tau_j)\)</span> model: <span class="math display">\[Y_{ij} = \theta+\tau_j + \varepsilon_{ij};\; \; \tau_1=0, \; \; \varepsilon_{ij} \sim G; \; \;E[\varepsilon_{ij}] = 0;\]</span></p>
<ul>
<li><p>Note that for each group, the population mean is given by <span class="math inline">\(E[Y_{ij}] = \theta+\tau_j=\mu_j,\)</span> and <span class="math inline">\(\tau_2=\mu_2-\mu_1=E[Y_{i2}] -E[Y_{i1}]\)</span> <em>compares</em> the means</p></li>
<li><p><span class="math inline">\(\tau_1\)</span> must be set to zero, since group 1 is the <em>reference</em> group</p></li>
</ul>
<div class="fragment">
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Note</strong></p>
</div>
<div class="callout-content">
<p>In the <strong>reference-treatment effect</strong> model parameterization, we have the following parameters:</p>
<ul>
<li><span class="math inline">\(\theta\)</span> represents the population mean of the reference group (in our example: WT)</li>
<li><span class="math inline">\(\tau_j\)</span> represents the difference in the population mean of group <span class="math inline">\(j\)</span> compared to the reference (in our example: NrlKO - WT)</li>
</ul>
</div>
</div>
</div>
</div>
</section>
<section id="relationship-between-parameterizations" class="slide level2">
<h2>Relationship between parameterizations</h2>

<img data-src="img/param_2.png" class="r-stretch quarto-figure-center"></section>
<section id="lm-output" class="slide level2">
<h2><code>lm</code> output</h2>
<ul>
<li>the sample mean of the WT group (<strong>reference</strong>): <span class="math inline">\(\hat\theta\)</span></li>
<li>the difference in sample mean of NrlKO and WT groups (<strong>treatment effect</strong>): <span class="math inline">\(\hat\tau_2\)</span></li>
</ul>
<div class="panel-tabset">
<ul id="tabset-4" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-4-1">Irs4</a></li><li><a href="#tabset-4-2">Nrl</a></li></ul>
<div class="tab-content">
<div id="tabset-4-1">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a><span class="fu">library</span>(broom)</span>
<span id="cb16-2"><a href="#cb16-2"></a><span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"Irs4"</span>)  <span class="sc">%&gt;%</span></span>
<span id="cb16-3"><a href="#cb16-3"></a>  <span class="fu">group_by</span>(genotype) <span class="sc">%&gt;%</span></span>
<span id="cb16-4"><a href="#cb16-4"></a>  <span class="fu">summarize</span>(<span class="at">meanExpr =</span> <span class="fu">mean</span>(expression)) <span class="sc">%&gt;%</span></span>
<span id="cb16-5"><a href="#cb16-5"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> genotype, <span class="at">values_from =</span> meanExpr) <span class="sc">%&gt;%</span> </span>
<span id="cb16-6"><a href="#cb16-6"></a>  <span class="fu">mutate</span>(<span class="at">diffExp =</span> NrlKO <span class="sc">-</span> WT)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 3
     WT NrlKO diffExp
  &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;
1  7.77  7.74 -0.0261</code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"Irs4"</span>)  <span class="sc">%&gt;%</span></span>
<span id="cb18-2"><a href="#cb18-2"></a>  <span class="fu">lm</span>(expression <span class="sc">~</span> genotype, <span class="at">data =</span> .) <span class="sc">%&gt;%</span> <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 5
  term          estimate std.error statistic  p.value
  &lt;chr&gt;            &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)     7.77      0.0344   226.    1.10e-59
2 genotypeNrlKO  -0.0261    0.0493    -0.529 6.00e- 1</code></pre>
</div>
</div>
</div>
<div id="tabset-4-2">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a><span class="fu">library</span>(broom)</span>
<span id="cb20-2"><a href="#cb20-2"></a><span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"Nrl"</span>)  <span class="sc">%&gt;%</span></span>
<span id="cb20-3"><a href="#cb20-3"></a>  <span class="fu">group_by</span>(genotype) <span class="sc">%&gt;%</span></span>
<span id="cb20-4"><a href="#cb20-4"></a>  <span class="fu">summarize</span>(<span class="at">meanExpr =</span> <span class="fu">mean</span>(expression)) <span class="sc">%&gt;%</span></span>
<span id="cb20-5"><a href="#cb20-5"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> genotype, <span class="at">values_from =</span> meanExpr) <span class="sc">%&gt;%</span> </span>
<span id="cb20-6"><a href="#cb20-6"></a>  <span class="fu">mutate</span>(<span class="at">diffExp =</span> NrlKO <span class="sc">-</span> WT)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 3
     WT NrlKO diffExp
  &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;
1  11.2  6.09   -5.15</code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a><span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"Nrl"</span>)  <span class="sc">%&gt;%</span></span>
<span id="cb22-2"><a href="#cb22-2"></a>  <span class="fu">lm</span>(expression <span class="sc">~</span> genotype, <span class="at">data =</span> .) <span class="sc">%&gt;%</span> <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 5
  term          estimate std.error statistic  p.value
  &lt;chr&gt;            &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)      11.2      0.214      52.5 2.39e-36
2 genotypeNrlKO    -5.15     0.307     -16.8 6.73e-19</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="we-still-havent-answered-our-question-wheres-the-line" class="slide level2">
<h2>We still haven’t answered our question… where’s the line??</h2>
<p><span class="math display">\[Y_{ij} = \theta+\tau_j + \varepsilon_{ij};\; \; \tau_1=0, \; \; \varepsilon_{ij} \sim G; \; \;E[\varepsilon_{ij}] = 0;\]</span></p>

<img data-src="lect06-anova_files/figure-revealjs/unnamed-chunk-16-1.png" width="960" class="r-stretch quarto-figure-center"></section>
<section id="indicator-variables" class="slide level2">
<h2>Indicator variables</h2>
<p>Let’s re-write our model using <strong>indicator</strong> (aka ‘dummy’) variables:</p>
<span class="math display">\[Y_{ij} = \theta+\tau_j + \varepsilon_{ij}\;\; \text{where} \; \; \tau_1=0; \; \; \varepsilon_{ij} \sim G; \; \;E[\varepsilon_{ij}] = 0;\]</span>
<center>
<strong>↓</strong>
</center>
<p><span class="math display">\[Y_{ij} = \theta+\tau x_{ij} + \varepsilon_{ij} \;\; \text{where} \; \; x_{ij}=\bigg\{\begin{array}{l}
1\text{ if } j=2\\
0 \text{ otherwise}\\
\end{array}\]</span></p>
<div class="fragment">
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Note</strong></p>
</div>
<div class="callout-content">
<p>Note that <span class="math inline">\(Y_{i1} = \theta + \varepsilon_{i1}\)</span>, because <span class="math inline">\(x_{i1}=0\)</span> and <span class="math inline">\(Y_{i2} = \theta + \tau+ \varepsilon_{i2}\)</span>, because <span class="math inline">\(x_{i2}=1\)</span> (for all <span class="math inline">\(i\)</span>)</p>
</div>
</div>
</div>
<p>The second form is written as a <em>linear</em> ( <span class="math inline">\(y=a + bx +\varepsilon\)</span> ) regression model, with an (<strong>indicator</strong>) explanatory variable <span class="math inline">\(x_{ij}\)</span></p>
</div>
</section>
<section id="t-test-with-a-linear-model" class="slide level2">
<h2>t-test with a linear model</h2>
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Note</strong></p>
</div>
<div class="callout-content">
<p>Using indicator variables to model our categorical variable <code>genotype</code>, we can perform a <strong>2-sample t-test</strong> with a linear model</p>
</div>
</div>
</div>
<p><span class="math display">\[Y_{ij} = \theta+\tau x_{ij} + \varepsilon_{ij}\;\text{where}\; \; x_{ij}=\bigg\{\begin{array}{l}
1 \text{ if } j=2\\
0 \text{ if } j=1\\
\end{array}\]</span></p>
<ul>
<li><p>The standalone t-test is carried out on <span class="math inline">\(H_0: \mu_1 = \mu_2\)</span></p></li>
<li><p>The t-test in the linear model is carried out on <span class="math inline">\(H_0: \tau = 0\)</span>, where <span class="math inline">\(\tau\)</span> is the difference in population means (here NrlKO - WT)</p></li>
<li><p>Recall that <span class="math inline">\(\tau = \mu_2 - \mu_1\)</span> - this is why these are equivalent tests!</p></li>
</ul>
</section>
<section id="beyond-2-group-comparisons" class="slide level2">
<h2>Beyond 2-group comparisons</h2>

<img data-src="img/more_2_groups.png" class="r-stretch quarto-figure-center"></section>
<section id="section" class="slide level2">
<h2></h2>
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Note</strong></p>
</div>
<div class="callout-content">
<p>Indicator variables can be used to model one <em>or more</em> categorical variables, each with 2 <em>or more</em> levels!</p>
</div>
</div>
</div>
<p><strong>2-sample <em>t</em>-test</strong> using a linear model</p>
<p><span class="math display">\[Y_{ij} = \theta+\tau x_{ij} + \varepsilon_{ij}\;\; \text{where} \; \; x_{ij}=\bigg\{\begin{array}{l}
1 \text{ if } j=2\\
0 \text{ if } j=1\\
\end{array}\]</span></p>
<p><strong>1-way ANOVA with many levels</strong><sup>1</sup> using a linear model - e.g for 3 groups: <span class="math display">\[Y_{ij} = \theta+\tau_2 x_{ij2} + \tau_3 x_{ij3} +\varepsilon_{ij}\;\; \text{where} \; x_{ij2}=\bigg\{\begin{array}{l}
1\text{ if } j=2\\
0 \text{ otherwise}\\
\end{array}\; \text{ and } \; x_{ij3}=\bigg\{\begin{array}{l}
1\text{ if } j=3\\
0 \text{ otherwise}\\
\end{array}\]</span></p>
<div class="fragment">
<div class="callout callout-important callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Important</strong></p>
</div>
<div class="callout-content">
<p>This equivalence is why R can estimate all of them with <code>lm()</code></p>
</div>
</div>
</div>
</div>
<aside><ol class="aside-footnotes"><li id="fn4"><p>in general; yet <em>another</em> parameterization can be used to present ANOVA</p></li></ol></aside></section>
<section id="connections" class="slide level2">
<h2>Connections</h2>
<div class="callout callout-important callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Important</strong></p>
</div>
<div class="callout-content">
<ul>
<li><p>The <strong>t-test</strong> is a special case of <strong>ANOVA</strong>, but with ANOVA you can compare <strong>more than two groups</strong> and <strong>more than one factor</strong>.</p></li>
<li><p><strong>ANOVA</strong> is a special case of <strong>linear regression</strong>, but with linear regression you can include <strong>quantitative variables</strong> in the model.</p></li>
<li><p><strong>Linear regression</strong> provides a unifying framework to model the association between a response and <strong>many quantitative and qualitative variables</strong>.</p></li>
<li><p><strong>In R</strong> all three can be computed using the <code>lm()</code> function.</p></li>
</ul>
</div>
</div>
</div>
</section>
<section id="linear-models-using-matrix-notation" class="slide level2">
<h2>Linear models using matrix notation</h2>

<img data-src="img/linear_form.png" class="r-stretch quarto-figure-center"><p>It will become handy to write our model using matrix notation</p>
</section>
<section id="design-matrix" class="slide level2">
<h2>Design matrix</h2>
<p>Let’s form a <strong><a href="http://genomicsclass.github.io/book/pages/expressing_design_formula.html">design matrix</a></strong> <span class="math inline">\((X)\)</span> for a 3-group comparison</p>
<div class="columns">
<div class="column">
<p><span class="math display">\[Y_{ij} = \theta+\tau_2  x_{ij2} + \tau_3 x_{ij3} +\varepsilon_{ij}\]</span></p>
<p><img data-src="img/model_matrix_I.png"></p>
</div><div class="column">
<p>First column in <span class="math inline">\(X\)</span> for reference treatment parameterization is all 1s</p>
<p>Second &amp; third columns contain <span class="math inline">\(x_{ij2}\)</span> and <span class="math inline">\(x_{ij3}\)</span>:</p>
<ul>
<li><p><span class="math inline">\(x_{i12}=x_{i13}=0\)</span> for the reference group</p></li>
<li><p><span class="math inline">\(x_{i22}=1\)</span> for the 2nd group</p></li>
<li><p><span class="math inline">\(x_{i33}=1\)</span> for the 3rd group</p></li>
</ul>
</div>
</div>
</section>
<section id="section-1" class="slide level2">
<h2></h2>

<img data-src="img/model_matrix_II.png" class="r-stretch"><p><font color="red"> <span class="math inline">\(Y_{i1}= 1 \times \theta + 0 \times \tau_2 + 0 \times \tau_3 + \varepsilon_{i1} =\theta + \varepsilon_{i1}\)</span></font></p>
<p><font color="blue"> <span class="math inline">\(Y_{i2}= 1 \times \theta + 1 \times \tau_2 + 0 \times \tau_3 + \varepsilon_{i2}=\theta + \tau_2+\varepsilon_{i2}\)</span></font></p>
<p><font color="green"> <span class="math inline">\(Y_{i3}= 1 \times \theta + 0 \times \tau_2 + 1 \times \tau_3 + \varepsilon_{i3}=\theta + \tau_3+\varepsilon_{i3}\)</span></font></p>
<p><font color="black"> <span class="math display">\[\; Y_{ij} = \theta +\tau_2  x_{ij2} + \tau_3  x_{ij3} + \varepsilon_{ij}\]</span></font></p>
</section>
<section id="section-2" class="slide level2">
<h2></h2>

<img data-src="img/rf_tx_matrix.png" class="r-stretch quarto-figure-center"><p>The model is still written with a reference-treatment parameterization (difference of means)</p>
<p><span class="math inline">\(E[Y_{i1}]=\theta\)</span></p>
<p><span class="math inline">\(E[Y_{i2}]=\theta+\tau_2 \; \rightarrow \tau_2=E[Y_{i2}]-E[Y_{i1}]=\mu_2-\mu_1\)</span></p>
<p><span class="math inline">\(E[Y_{i3}]=\theta+\tau_3 \; \rightarrow \tau_3=E[Y_{i3}]-E[Y_{i1}]=\mu_3-\mu_1\)</span></p>
</section>
<section id="linear-regression-can-include-quantitative-qualitative-covariates" class="slide level2">
<h2>Linear<sup>1</sup> regression can include <em>quantitative</em> &amp; <em>qualitative</em> covariates</h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/LM_vbles.png"></p>
</figure>
</div>
<aside><ol class="aside-footnotes"><li id="fn5"><p>Here we mean <strong>linear</strong> in the parameters <span class="math inline">\(\boldsymbol{\alpha}\)</span>; <span class="math inline">\(X\)</span> can contain <span class="math inline">\(x^2\)</span>, <span class="math inline">\(log(x)\)</span>, etc</p></li></ol></aside></section>
<section id="how-it-works-in-practice-using-lm-in-r" class="slide level2">
<h2>How it works in practice using <code>lm()</code> in R</h2>
<span class="math display">\[Y = X\alpha + \varepsilon\]</span>
<center>
<strong>↓</strong>
</center>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a><span class="fu">lm</span>(y <span class="sc">~</span> x, <span class="at">data =</span> yourData)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="columns">
<div class="column">
<p><strong><code>y ~ x</code>:</strong> formula</p>
<p><strong><code>y</code>:</strong> numeric</p>
<p><strong><code>x</code>:</strong> numeric and/or factor</p>
</div><div class="column">
<p><strong><code>yourData</code>:</strong> <code>data.frame</code> (or <code>tibble</code>) in which <code>x</code> and <code>y</code> are to be found</p>
</div>
</div>
<p>By default, R uses the reference-treatment parameterization<sup>1</sup></p>
<aside><ol class="aside-footnotes"><li id="fn6"><p>but <a href="http://genomicsclass.github.io/book/pages/expressing_design_formula.html">you can change that</a>!</p></li></ol></aside></section>
<section id="factor-class-in-r" class="slide level2">
<h2><code>factor</code> class in R</h2>
<p>Mathematically, the design matrix <span class="math inline">\(X\)</span> in <span class="math inline">\(Y=X\alpha+\varepsilon\)</span> needs to be a numeric matrix</p>
<div class="callout callout-important callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Important</strong></p>
</div>
<div class="callout-content">
<ul>
<li><p>If your data contains categorical variables (e.g., <code>genotype</code>), you need to set them as <strong>factors</strong></p>
<ul>
<li><p>especially important if your categorical variables are encoded numerically!!</p></li>
<li><p><code>lm</code> will automatically treat character variables as factors</p></li>
</ul></li>
<li><p><code>lm</code> automatically creates the appropriate indicator variables (numeric) for factors!</p></li>
</ul>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a><span class="fu">str</span>(twoGenes<span class="sc">$</span>genotype)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> Factor w/ 2 levels "WT","NrlKO": 2 2 2 2 2 2 2 2 2 2 ...</code></pre>
</div>
</div>
</section>
<section id="under-the-hood-lm-creates-a-numeric-x" class="slide level2">
<h2>Under the hood, <code>lm</code> creates a numeric <span class="math inline">\(X\)</span></h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a><span class="co"># create design matrix</span></span>
<span id="cb27-2"><a href="#cb27-2"></a>mm <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>( <span class="sc">~</span> genotype, <span class="at">data =</span> twoGenes) </span>
<span id="cb27-3"><a href="#cb27-3"></a></span>
<span id="cb27-4"><a href="#cb27-4"></a><span class="co"># show first 3 and last 3 rows of model.matrix</span></span>
<span id="cb27-5"><a href="#cb27-5"></a><span class="fu">head</span>(mm, <span class="dv">3</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  (Intercept) genotypeNrlKO
1           1             1
2           1             1
3           1             1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a><span class="fu">tail</span>(mm, <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   (Intercept) genotypeNrlKO
76           1             0
77           1             0
78           1             0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a><span class="co"># show first 3 and last 3 values of genotype</span></span>
<span id="cb31-2"><a href="#cb31-2"></a>twoGenes <span class="sc">%&gt;%</span> </span>
<span id="cb31-3"><a href="#cb31-3"></a>  <span class="fu">slice</span>(<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, (<span class="fu">n</span>()<span class="sc">-</span><span class="dv">3</span>)<span class="sc">:</span><span class="fu">n</span>())) <span class="sc">%&gt;%</span></span>
<span id="cb31-4"><a href="#cb31-4"></a>  <span class="fu">pull</span>(genotype)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] NrlKO NrlKO NrlKO WT    WT    WT    WT   
Levels: WT NrlKO</code></pre>
</div>
</div>
</section>
<section id="beyond-2-group-comparisons-in-our-case-study" class="slide level2">
<h2>Beyond 2-group comparisons in our case study</h2>
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Biological question</strong></p>
</div>
<div class="callout-content">
<p>Is the expression of gene X the same at all developmental stages?</p>
</div>
</div>
</div>
<div class="fragment">
<p><span class="math display">\[H_0 : \mu_{E16} = \mu_{P2} = \mu_{P6} = \mu_{P10} = \mu_{P28}\]</span></p>
</div>
<div class="fragment">
<p>Let’s look at another two genes for some variety</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="lect06-anova_files/figure-revealjs/unnamed-chunk-21-1.png" width="1200"></p>
</figure>
</div>
</div>
</div>
</div>
</section>
<section id="the-sample-means-hatmu_e16-hatmu_p2-hatmu_p6-hatmu_p10-hatmu_p28" class="slide level2">
<h2>The sample means: <span class="math inline">\(\hat\mu_{E16}, \; \hat\mu_{P2}, \; \hat\mu_{P6}, \; \hat\mu_{P10}, \; \hat\mu_{P28}\)</span></h2>
<div class="columns">
<div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1"></a>twoGenes <span class="sc">%&gt;%</span> </span>
<span id="cb33-2"><a href="#cb33-2"></a>  <span class="fu">group_by</span>(gene, dev_stage) <span class="sc">%&gt;%</span></span>
<span id="cb33-3"><a href="#cb33-3"></a>  <span class="fu">summarize</span>(<span class="at">meanExpr =</span> <span class="fu">mean</span>(expression)) <span class="sc">%&gt;%</span></span>
<span id="cb33-4"><a href="#cb33-4"></a>  <span class="fu">pivot_wider</span>(<span class="at">values_from =</span> meanExpr, </span>
<span id="cb33-5"><a href="#cb33-5"></a>              <span class="at">names_from =</span> gene)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 3
  dev_stage BB114814 Cdc14a
  &lt;fct&gt;        &lt;dbl&gt;  &lt;dbl&gt;
1 E16           5.54   7.54
2 P2            5.84   7.58
3 P6            5.78   7.55
4 P10           6.38   7.57
5 P28           9.17   7.56</code></pre>
</div>
</div>
</div><div class="column">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="lect06-anova_files/figure-revealjs/unnamed-chunk-23-1.png" width="576"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="bb114814-gene-with-notable-time-effect" class="slide level2">
<h2>BB114814 gene with notable time effect</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1"></a>twoGenes <span class="sc">%&gt;%</span> <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"BB114814"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb35-2"><a href="#cb35-2"></a>  <span class="fu">group_by</span>(dev_stage) <span class="sc">%&gt;%</span></span>
<span id="cb35-3"><a href="#cb35-3"></a>  <span class="fu">summarize</span>(<span class="at">cellMeans =</span> <span class="fu">mean</span>(expression)) <span class="sc">%&gt;%</span></span>
<span id="cb35-4"><a href="#cb35-4"></a>  <span class="fu">mutate</span>(<span class="at">timeEffect =</span> cellMeans <span class="sc">-</span> cellMeans[<span class="dv">1</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 3
  dev_stage cellMeans timeEffect
  &lt;fct&gt;         &lt;dbl&gt;      &lt;dbl&gt;
1 E16            5.54      0    
2 P2             5.84      0.304
3 P6             5.78      0.243
4 P10            6.38      0.834
5 P28            9.17      3.63 </code></pre>
</div>
</div>
<p>“Effect” here is relative to reference/baseline (E16)</p>
</section>
<section id="bb114814-gene-with-notable-time-effect-1" class="slide level2">
<h2>BB114814 gene with notable time effect</h2>
<div class="columns">
<div class="column">
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 3
  dev_stage cellMeans timeEffect
  &lt;fct&gt;         &lt;dbl&gt;      &lt;dbl&gt;
1 E16            5.54      0    
2 P2             5.84      0.304
3 P6             5.78      0.243
4 P10            6.38      0.834
5 P28            9.17      3.63 </code></pre>
</div>
</div>
</div><div class="column">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="lect06-anova_files/figure-revealjs/unnamed-chunk-26-1.png" width="768"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
<div class="fragment">
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Check your understanding</strong></p>
</div>
<div class="callout-content">
<p>Can you guess the size of the <span class="math inline">\(X\)</span> matrix needed to test for any time differences compared to E16? How many indicator variables do we need?</p>
</div>
</div>
</div>
</div>
</section>
<section id="gene-bb114814-with-notable-time-effect" class="slide level2">
<h2>Gene BB114814 with notable time effect</h2>
<p>We need <code>__</code> indicator variables to estimate and test <code>__</code> time differences (between <code>__</code> time points), and our design matrix will have <code>__</code> columns:</p>
<aside class="notes">
<p>We need 4 indicator variables to estimate and test 4 time differences (between 5 time points): - <span class="math inline">\(x_{P2}\)</span>: P2 vs E16 - <span class="math inline">\(x_{P6}\)</span>: P6 vs E16 - <span class="math inline">\(x_{P10}\)</span>: P10 vs E16 - <span class="math inline">\(x_{P28}\)</span>: P28 vs E16</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
<div class="fragment">
<p>Mathematically:</p>
<p><span class="math display">\[Y_{ij}=\theta+\tau_{P2} x_{ijP2}+\tau_{P6} x_{ijP6}+\tau_{P10} x_{ijP10}+\tau_{P28} x_{ijP28}+\varepsilon_{ij}\]</span></p>
<p><strong>Notation</strong>: <span class="math inline">\(x_{ijk}\)</span>:</p>
<ul>
<li><span class="math inline">\(i\)</span> indexes for the observation/sample within group</li>
<li><span class="math inline">\(j\)</span> indexes the group (here: level of <code>dev_stage</code>)</li>
<li><span class="math inline">\(k\)</span> is the name of the indicator variable</li>
</ul>
</div>
</section>
<section id="under-the-hood-lm-creates-a-numeric-x-1" class="slide level2">
<h2>Under the hood, <code>lm</code> creates a numeric <span class="math inline">\(X\)</span></h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1"></a><span class="fu">str</span>(twoGenes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>tibble [78 × 5] (S3: tbl_df/tbl/data.frame)
 $ gene      : chr [1:78] "BB114814" "BB114814" "BB114814" "BB114814" ...
 $ sample_id : chr [1:78] "GSM92610" "GSM92611" "GSM92612" "GSM92613" ...
 $ expression: num [1:78] 8.59 8.39 9.34 9.49 5.39 ...
 $ dev_stage : Factor w/ 5 levels "E16","P2","P6",..: 5 5 5 5 1 1 1 4 4 4 ...
 $ genotype  : Factor w/ 2 levels "WT","NrlKO": 2 2 2 2 2 2 2 2 2 2 ...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1"></a><span class="fu">model.matrix</span>( <span class="sc">~</span> dev_stage, <span class="at">data =</span> twoGenes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   (Intercept) dev_stageP2 dev_stageP6 dev_stageP10 dev_stageP28
1            1           0           0            0            1
2            1           0           0            0            1
3            1           0           0            0            1
4            1           0           0            0            1
5            1           0           0            0            0
6            1           0           0            0            0
7            1           0           0            0            0
8            1           0           0            1            0
9            1           0           0            1            0
10           1           0           0            1            0
11           1           0           0            1            0
12           1           1           0            0            0
13           1           1           0            0            0
14           1           1           0            0            0
15           1           1           0            0            0
16           1           0           1            0            0
17           1           0           1            0            0
18           1           0           1            0            0
19           1           0           1            0            0
20           1           0           0            0            1
21           1           0           0            0            1
22           1           0           0            0            1
23           1           0           0            0            1
24           1           0           0            0            0
25           1           0           0            0            0
26           1           0           0            0            0
27           1           0           0            0            0
28           1           0           0            1            0
29           1           0           0            1            0
30           1           0           0            1            0
31           1           0           0            1            0
32           1           1           0            0            0
33           1           1           0            0            0
34           1           1           0            0            0
35           1           1           0            0            0
36           1           0           1            0            0
37           1           0           1            0            0
38           1           0           1            0            0
39           1           0           1            0            0
40           1           0           0            0            1
41           1           0           0            0            1
42           1           0           0            0            1
43           1           0           0            0            1
44           1           0           0            0            0
45           1           0           0            0            0
46           1           0           0            0            0
47           1           0           0            1            0
48           1           0           0            1            0
49           1           0           0            1            0
50           1           0           0            1            0
51           1           1           0            0            0
52           1           1           0            0            0
53           1           1           0            0            0
54           1           1           0            0            0
55           1           0           1            0            0
56           1           0           1            0            0
57           1           0           1            0            0
58           1           0           1            0            0
59           1           0           0            0            1
60           1           0           0            0            1
61           1           0           0            0            1
62           1           0           0            0            1
63           1           0           0            0            0
64           1           0           0            0            0
65           1           0           0            0            0
66           1           0           0            0            0
67           1           0           0            1            0
68           1           0           0            1            0
69           1           0           0            1            0
70           1           0           0            1            0
71           1           1           0            0            0
72           1           1           0            0            0
73           1           1           0            0            0
74           1           1           0            0            0
75           1           0           1            0            0
76           1           0           1            0            0
77           1           0           1            0            0
78           1           0           1            0            0
attr(,"assign")
[1] 0 1 1 1 1
attr(,"contrasts")
attr(,"contrasts")$dev_stage
[1] "contr.treatment"</code></pre>
</div>
</div>
</section>
<section id="hypothesis-tests-in-lm-output" class="slide level2">
<h2>Hypothesis tests in <code>lm</code> output</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 3
  dev_stage cellMeans timeEffect
  &lt;fct&gt;         &lt;dbl&gt;      &lt;dbl&gt;
1 E16            5.54      0    
2 P2             5.84      0.304
3 P6             5.78      0.243
4 P10            6.38      0.834
5 P28            9.17      3.63 </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1"></a>twoGenes <span class="sc">%&gt;%</span> <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"BB114814"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb43-2"><a href="#cb43-2"></a>  <span class="fu">lm</span>(expression <span class="sc">~</span> dev_stage, <span class="at">data =</span> .) <span class="sc">%&gt;%</span> <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 5
  term         estimate std.error statistic  p.value
  &lt;chr&gt;           &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)     5.54      0.102     54.2  1.31e-34
2 dev_stageP2     0.304     0.140      2.17 3.69e- 2
3 dev_stageP6     0.243     0.140      1.74 9.11e- 2
4 dev_stageP10    0.834     0.140      5.96 9.62e- 7
5 dev_stageP28    3.63      0.140     26.0  5.30e-24</code></pre>
</div>
</div>
<div class="columns">
<div class="column">
<p><span class="math inline">\(H_0: \theta=0\)</span> or <span class="math inline">\(H_0: \mu_{E16}=0\)</span></p>
<p><strong>Estimate</strong>: <span class="math inline">\(\hat\theta=\hat\mu_{E16}=\bar{Y}_{\cdot E16}\)</span></p>
</div><div class="column">
<blockquote>
<p>we are not usually interested in testing this hypothesis: baseline mean = 0</p>
</blockquote>
</div>
</div>
</section>
<section id="hypothesis-tests-in-lm-output-1" class="slide level2">
<h2>Hypothesis tests in <code>lm</code> output</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 3
  dev_stage cellMeans timeEffect
  &lt;fct&gt;         &lt;dbl&gt;      &lt;dbl&gt;
1 E16            5.54      0    
2 P2             5.84      0.304
3 P6             5.78      0.243
4 P10            6.38      0.834
5 P28            9.17      3.63 </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1"></a>twoGenes <span class="sc">%&gt;%</span> <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"BB114814"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb46-2"><a href="#cb46-2"></a>  <span class="fu">lm</span>(expression <span class="sc">~</span> dev_stage, <span class="at">data =</span> .) <span class="sc">%&gt;%</span> <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 5
  term         estimate std.error statistic  p.value
  &lt;chr&gt;           &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)     5.54      0.102     54.2  1.31e-34
2 dev_stageP2     0.304     0.140      2.17 3.69e- 2
3 dev_stageP6     0.243     0.140      1.74 9.11e- 2
4 dev_stageP10    0.834     0.140      5.96 9.62e- 7
5 dev_stageP28    3.63      0.140     26.0  5.30e-24</code></pre>
</div>
</div>
<div class="columns">
<div class="column">
<p><span class="math inline">\(H_0: \tau_{P2}=0\)</span> or <span class="math inline">\(H_0: \mu_{P2}=\mu_{E16}\)</span></p>
<p><strong>Estimate</strong>: <span class="math inline">\(\hat{\tau}_{P2}=\hat{\mu}_{P2}-\hat{\mu}_{E16}=\bar{Y}_{\cdot P2}-\bar{Y}_{\cdot E16}\)</span></p>
</div><div class="column">
<blockquote>
<p>we <em>are</em> usually interested in testing this hypothesis: change from E16 to 2 days old = 0</p>
</blockquote>
</div>
</div>
</section>
<section id="hypothesis-tests-in-lm-output-2" class="slide level2">
<h2>Hypothesis tests in <code>lm</code> output</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 3
  dev_stage cellMeans timeEffect
  &lt;fct&gt;         &lt;dbl&gt;      &lt;dbl&gt;
1 E16            5.54      0    
2 P2             5.84      0.304
3 P6             5.78      0.243
4 P10            6.38      0.834
5 P28            9.17      3.63 </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1"></a>twoGenes <span class="sc">%&gt;%</span> <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"BB114814"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb49-2"><a href="#cb49-2"></a>  <span class="fu">lm</span>(expression <span class="sc">~</span> dev_stage, <span class="at">data =</span> .) <span class="sc">%&gt;%</span> <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 5
  term         estimate std.error statistic  p.value
  &lt;chr&gt;           &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)     5.54      0.102     54.2  1.31e-34
2 dev_stageP2     0.304     0.140      2.17 3.69e- 2
3 dev_stageP6     0.243     0.140      1.74 9.11e- 2
4 dev_stageP10    0.834     0.140      5.96 9.62e- 7
5 dev_stageP28    3.63      0.140     26.0  5.30e-24</code></pre>
</div>
</div>
<div class="columns">
<div class="column">
<p><span class="math inline">\(H_0: \tau_{P28}=0\)</span> or <span class="math inline">\(H_0: \mu_{P28}=\mu_{E16}\)</span></p>
<p><strong>Estimate</strong>: <span class="math inline">\(\hat\tau_{P28}=\hat\mu_{P28}-\hat\mu_{E16}=\bar{Y}_{\cdot P28}-\bar{Y}_{\cdot E16}\)</span></p>
</div><div class="column">
<blockquote>
<p>we <em>are</em> usually interested in testing this hypothesis: change from E16 to 4 weeks old = 0</p>
</blockquote>
</div>
</div>
</section>
<section id="notice-the-standard-error-estimates" class="slide level2">
<h2>Notice the standard error estimates</h2>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1"></a>twoGenes <span class="sc">%&gt;%</span> <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"BB114814"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb51-2"><a href="#cb51-2"></a>  <span class="fu">lm</span>(expression <span class="sc">~</span> dev_stage, <span class="at">data =</span> .) <span class="sc">%&gt;%</span></span>
<span id="cb51-3"><a href="#cb51-3"></a>  <span class="fu">summary</span>() <span class="sc">%&gt;%</span> .<span class="sc">$</span>coef</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>              Estimate Std. Error   t value     Pr(&gt;|t|)
(Intercept)  5.5409162  0.1021560 54.239748 1.314828e-34
dev_stageP2  0.3037855  0.1398829  2.171713 3.694652e-02
dev_stageP6  0.2432795  0.1398829  1.739166 9.105366e-02
dev_stageP10 0.8341163  0.1398829  5.962962 9.620151e-07
dev_stageP28 3.6323772  0.1398829 25.967276 5.303201e-24</code></pre>
</div>
</div>
<div class="fragment">
<p>All data points are used to estimate the variance of the error term for the indicator variables</p>
</div>
</section>
<section id="two-types-of-null-hypotheses-single-vs-joint" class="slide level2">
<h2>Two types of null hypotheses: single vs joint</h2>
<p><span class="math display">\[Y = X \alpha + \varepsilon\]</span> <span class="math display">\[\alpha = (\theta, \tau_{P2}, \tau_{P6}, \tau_{P10}, \tau_{P28})\]</span></p>
<div class="columns">
<div class="column">
<p><span class="math inline">\(H_0: \tau_j = 0\)</span> vs <span class="math inline">\(H_1: \tau_j \neq 0\)</span></p>
<p><strong>for each <em>j</em> individually</strong></p>
<p>For example: Is gene <em>A</em> differentially expressed 2 days after birth (compared to embryonic day 16)?</p>
<p><span class="math display">\[H_0: \tau_{P2}=0\]</span></p>
<div class="fragment">
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Note</strong></p>
</div>
<div class="callout-content">
<p>This single hypothesis can be tested with a <strong>t-test</strong></p>
</div>
</div>
</div>
</div>
</div><div class="column">
<p><span class="math inline">\(H_0: \tau_j = 0\)</span> vs <span class="math inline">\(H_1: \tau_j \neq 0\)</span></p>
<p><strong>for all <em>j</em> at the same time</strong></p>
<p>For example: Is gene <em>A</em> significantly affected by time? In other words, is gene <em>A</em> differentially expressed at <em>any</em> time point?</p>
<p><span class="math display">\[H_0: \tau_{P2}=\tau_{P6}=\tau_{P10}=\tau_{P28}=0\]</span></p>
<div class="fragment">
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Key Question</strong></p>
</div>
<div class="callout-content">
<p>How do we test this joint hypothesis?</p>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="f-test-and-overall-significance-of-one-or-more-coefficients" class="slide level2">
<h2><em>F</em>-test and overall significance of one or more coefficients</h2>
<ul>
<li><p>the <em>t</em>-test in linear regression allows us to test single hypotheses: <span class="math display">\[H_0 : \tau_j = 0\]</span> <span class="math display">\[H_A : \tau_j \neq 0\]</span></p></li>
<li><p>but we often like to test multiple hypotheses <em>simultaneously</em>: <span class="math display">\[H_0 : \tau_{P2} = \tau_{P6} = \tau_{P10} = \tau_{P28}=0\textrm{ [AND statement]}\]</span> <span class="math display">\[H_A : \tau_j \neq 0 \textrm{ for some j [OR statement]}\]</span></p></li>
<li><p>the <strong><em>F</em>-test</strong> allows us to test such compound tests</p>
<ul>
<li>more on this type of test next week</li>
</ul></li>
</ul>
</section>
<section id="single-and-joint-tests-in-lm-output" class="slide level2">
<h2>Single and joint tests in <code>lm</code> output</h2>
<p>Can you locate the results of each type of test in the <code>lm</code> output?</p>
<p><span class="math inline">\(H_0: \tau_j = 0\)</span> vs <span class="math inline">\(H_1: \tau_j \neq 0\)</span> for each <span class="math inline">\(j\)</span> <strong>individually</strong></p>
<p><span class="math inline">\(H_0: \tau_j = 0\)</span> vs <span class="math inline">\(H_1: \tau_j \neq 0\)</span> for all <span class="math inline">\(j\)</span> <strong>together</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1"></a>twoGenes <span class="sc">%&gt;%</span> <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"BB114814"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb53-2"><a href="#cb53-2"></a>  <span class="fu">lm</span>(expression <span class="sc">~</span> dev_stage, <span class="at">data =</span> .) <span class="sc">%&gt;%</span></span>
<span id="cb53-3"><a href="#cb53-3"></a>  <span class="fu">summary</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = expression ~ dev_stage, data = .)

Residuals:
     Min       1Q   Median       3Q      Max 
-0.78553 -0.13324 -0.04796  0.17038  0.51846 

Coefficients:
             Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)    5.5409     0.1022  54.240  &lt; 2e-16 ***
dev_stageP2    0.3038     0.1399   2.172   0.0369 *  
dev_stageP6    0.2433     0.1399   1.739   0.0911 .  
dev_stageP10   0.8341     0.1399   5.963 9.62e-07 ***
dev_stageP28   3.6324     0.1399  25.967  &lt; 2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.2703 on 34 degrees of freedom
Multiple R-squared:  0.9662,    Adjusted R-squared:  0.9623 
F-statistic: 243.3 on 4 and 34 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
</section>
<section id="to-conclude---key-ideas-from-today" class="slide level2">
<h2>To conclude - key ideas from today</h2>
<ol type="1">
<li><p>We can compare group means (2 or more) using a linear model</p></li>
<li><p>We can use different parameterizations (<strong>cell means</strong> and <strong>reference-treatment effect</strong>) to write statistical models</p></li>
<li><p>We can write a <strong>linear model</strong> using matrix notation: <span class="math inline">\(Y = X \alpha + \varepsilon\)</span></p></li>
<li><p>Linear models can include <strong>quantitative &amp; qualitative covariates</strong></p></li>
<li><p>We use different tests to distinguish between <strong>single</strong> and <strong>joint</strong> hypotheses (e.g.&nbsp;<span class="math inline">\(t\)</span>-tests vs <span class="math inline">\(F\)</span>-tests)</p></li>
</ol>

<img src="https://github.com/STAT540-UBC/stat540-ubc.github.io/raw/main/images/stat540-logo-s.png" class="slide-logo r-stretch"><div class="footer footer-default">

</div>
</section>

    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="lect06-anova_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="lect06-anova_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="lect06-anova_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="lect06-anova_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="lect06-anova_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="lect06-anova_files/libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="lect06-anova_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="lect06-anova_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="lect06-anova_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="lect06-anova_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="lect06-anova_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1600,

        height: 900,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>