<!DOCTYPE html>
<html lang="en"><head>
<script src="lect07-linearModels_files/libs/clipboard/clipboard.min.js"></script>
<script src="lect07-linearModels_files/libs/quarto-html/tabby.min.js"></script>
<script src="lect07-linearModels_files/libs/quarto-html/popper.min.js"></script>
<script src="lect07-linearModels_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="lect07-linearModels_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="lect07-linearModels_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="lect07-linearModels_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="lect07-linearModels_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.3.450">

  <meta name="author" content="Keegan Korthauer">
  <meta name="dcterms.date" content="2025-01-28">
  <title>Linear models with multiple factors</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="lect07-linearModels_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="lect07-linearModels_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="lect07-linearModels_files/libs/revealjs/dist/theme/quarto.css">
  <link href="lect07-linearModels_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="lect07-linearModels_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="lect07-linearModels_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="lect07-linearModels_files/libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="lect07-linearModels_files/libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="lect07-linearModels_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" data-background-color="#197aa0" data-background-image="https://github.com/STAT540-UBC/stat540-ubc.github.io/raw/main/images/stat540-logo-s.png" data-background-opacity="0.9" data-background-position="95% 90%" data-background-size="12%" class="quarto-title-block center">
  <h1 class="title">Linear models with multiple factors</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Keegan Korthauer 
</div>
</div>
</div>

  <p class="date">January 28, 2025</p>
</section>
<section id="last-class" class="slide level2">
<h2>Last class…</h2>
<ol type="1">
<li>How to compare means of different groups (2 or more) using a linear regression model
<ul>
<li>indicator variables to model the levels of a qualitative explanatory variable</li>
</ul></li>
<li>Write a linear model using matrix notation
<ul>
<li>understand which matrix is built by R</li>
</ul></li>
<li>Distinguish between <strong>single</strong> and <strong>joint</strong> hypotheses
<ul>
<li><span class="math inline">\(t\)</span>-tests vs <span class="math inline">\(F\)</span>-tests</li>
</ul></li>
</ol>
</section>
<section id="comparing-more-than-two-groups" class="slide level2">
<h2>Comparing more than two groups</h2>
<ul>
<li><p><strong>Biological question</strong>: do gene expression levels differ by developmental stage?</p></li>
<li><p><strong>Statistical question</strong>: are gene expression generated by a single common distribution across all developmental stages? Or do the distributions differ by timepoint?</p></li>
</ul>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a>bb1plot <span class="ot">&lt;-</span> twoGenes <span class="sc">%&gt;%</span> <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"BB114814"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb1-2"><a href="#cb1-2"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> dev_stage, <span class="at">y =</span> expression)) <span class="sc">+</span> </span>
<span id="cb1-3"><a href="#cb1-3"></a>             <span class="fu">geom_jitter</span>(<span class="at">width =</span> <span class="fl">0.2</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb1-4"><a href="#cb1-4"></a>             <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"BB114814"</span>) <span class="sc">+</span></span>
<span id="cb1-5"><a href="#cb1-5"></a>             <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>) <span class="sc">+</span></span>
<span id="cb1-6"><a href="#cb1-6"></a>             <span class="fu">ylim</span>(<span class="dv">5</span>, <span class="dv">10</span>) <span class="sc">+</span></span>
<span id="cb1-7"><a href="#cb1-7"></a>             <span class="fu">xlab</span>(<span class="st">""</span>) <span class="sc">+</span></span>
<span id="cb1-8"><a href="#cb1-8"></a>             <span class="fu">stat_summary</span>(<span class="fu">aes</span>(<span class="at">group=</span><span class="dv">1</span>), <span class="at">fun=</span>mean, <span class="at">geom=</span><span class="st">"line"</span>, <span class="at">colour=</span><span class="st">"red"</span>)</span>
<span id="cb1-9"><a href="#cb1-9"></a></span>
<span id="cb1-10"><a href="#cb1-10"></a>cdcplot <span class="ot">&lt;-</span> twoGenes <span class="sc">%&gt;%</span> <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"Cdc14a"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb1-11"><a href="#cb1-11"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> dev_stage, <span class="at">y =</span> expression)) <span class="sc">+</span> </span>
<span id="cb1-12"><a href="#cb1-12"></a>             <span class="fu">geom_jitter</span>(<span class="at">width =</span> <span class="fl">0.2</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb1-13"><a href="#cb1-13"></a>             <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Cdc14a"</span>) <span class="sc">+</span></span>
<span id="cb1-14"><a href="#cb1-14"></a>             <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>) <span class="sc">+</span></span>
<span id="cb1-15"><a href="#cb1-15"></a>             <span class="fu">ylim</span>(<span class="dv">5</span>, <span class="dv">10</span>) <span class="sc">+</span></span>
<span id="cb1-16"><a href="#cb1-16"></a>             <span class="fu">xlab</span>(<span class="st">""</span>) <span class="sc">+</span></span>
<span id="cb1-17"><a href="#cb1-17"></a>             <span class="fu">stat_summary</span>(<span class="fu">aes</span>(<span class="at">group=</span><span class="dv">1</span>), <span class="at">fun=</span>mean, <span class="at">geom=</span><span class="st">"line"</span>, <span class="at">colour=</span><span class="st">"red"</span>)</span>
<span id="cb1-18"><a href="#cb1-18"></a><span class="fu">grid.arrange</span>(bb1plot, cdcplot, <span class="at">nrow =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="lect07-linearModels_files/figure-revealjs/unnamed-chunk-3-1.png" width="1200" class="r-stretch quarto-figure-center"></section>
<section id="quick-review-from-t-test-to-linear-regression" class="slide level2">
<h2><strong>Quick review</strong>: from <span class="math inline">\(t\)</span>-test to linear regression</h2>
<center>
<p><strong>2-sample t-test</strong></p>
<p><span class="math display">\[
Y \sim F; \ E[Y]= \mu_Y ; \ Z \sim G; \ E[Z]= \mu_Z
\]</span> <span class="math display">\[ H_0: \mu_Y = \mu_Z\]</span></p>
<h4 id="section"><strong>↓</strong></h4>
<h4 id="how-why"><strong>How? Why?</strong></h4>
<h4 id="section-1"><strong>↓</strong></h4>
<p><strong>Linear regression</strong></p>
<p><span class="math display">\[ Y = X\alpha + \epsilon; \text{ }\hspace{1em} H_0: \alpha_j = 0\]</span></p>
</center></section>
<section id="how-using-matrix-notation" class="slide level2">
<h2>How: Using matrix notation</h2>
<div class="columns">
<div class="column">
<p>2 group comparison: <br></p>
<p><span class="math inline">\(Y_{ij}=\theta+ \tau_2 x_{ij2} + \varepsilon_{ij}\)</span> <strong>→</strong> <span class="math inline">\(\bf{Y} = \bf{X\alpha+\epsilon}\)</span> <br></p>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/matrix_not.png"></p>
</figure>
</div>
</div><div class="column">
<ul>
<li><p><span class="math inline">\(x_{ij2}\)</span> is the second column of <span class="math inline">\(X\)</span> (design matrix)</p></li>
<li><p>Tip: examine design matrix in R with <code>model.matrix()</code></p></li>
</ul>
<p><br></p>
<p><span class="math display">\[\color{red}{Y_{11} = 1*\theta + 0*\tau_2 + \epsilon_{11} =\theta + \epsilon_{11}}\]</span></p>
<p><span class="math display">\[\color{blue}{Y_{12} = 1*\theta + 1*\tau_2 + \epsilon_{12} = \theta + \tau_2 + \epsilon_{12}}\]</span></p>
</div>
</div>
<div class="fragment">
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Recall</strong></p>
</div>
<div class="callout-content">
<p>For comparisons involving more than 2 groups (ANOVA), we add indicator variables (columns of <span class="math inline">\(X\)</span>)</p>
</div>
</div>
</div>
</div>
</section>
<section id="why-flexible-framework" class="slide level2">
<h2>Why: Flexible framework</h2>
<p><span class="math inline">\(\bf{Y} = \bf{X\alpha+\epsilon}\)</span> gives us a very flexible framework</p>

<img data-src="img/LM_vbles.png" class="r-stretch quarto-figure-center"><p>These (and many more) can be accommodated by the design matrix (X)!</p>
</section>
<section id="parameterizations" class="slide level2">
<h2>Parameterizations</h2>
<p>Different ways of writing the <span class="math inline">\(\bf{X\alpha}=\)</span> <strong>[design matrix][parameter vector]</strong> pair correspond to different <strong>parameterizations</strong> of the model</p>
<h4 id="bfy-bfxalpha-varepsilon"><span class="math display">\[\bf{Y} = \bf{X\alpha} + \varepsilon\]</span></h4>
<p>Understanding these concepts makes it easier:</p>
<ul>
<li>to interpret and compare fitted models</li>
<li>to fit models such that comparisons you care most about are directly addressed in the output</li>
</ul>
</section>
<section id="we-can-tell-r-to-use-the-cell-means-parameterization" class="slide level2">
<h2>We can tell R to use the cell-means parameterization</h2>
<p>Write the formula as <code>Y ~ 0 + x</code> in the <code>lm</code> call to remove the intercept <span class="math inline">\((\theta)\)</span> parameter and fit cell means parameters instead</p>
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"BB114814"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb2-2"><a href="#cb2-2"></a>  <span class="fu">lm</span>(expression <span class="sc">~</span> <span class="dv">0</span> <span class="sc">+</span> dev_stage, <span class="at">data =</span> .) <span class="sc">%&gt;%</span></span>
<span id="cb2-3"><a href="#cb2-3"></a>  <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 5
  term         estimate std.error statistic  p.value
  &lt;chr&gt;           &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 dev_stageE16     5.54    0.102       54.2 1.31e-34
2 dev_stageP2      5.84    0.0956      61.2 2.30e-36
3 dev_stageP6      5.78    0.0956      60.5 3.27e-36
4 dev_stageP10     6.38    0.0956      66.7 1.23e-37
5 dev_stageP28     9.17    0.0956      96.0 5.56e-43</code></pre>
</div>
</div>
<p>What null hypotheses does the <em>t</em>-test column now represent?</p>
</div>
</section>
<section id="converting-between-parameterizations" class="slide level2">
<h2>Converting between parameterizations</h2>

<img data-src="img/paramconvert.png" class="r-stretch quarto-figure-center"><div class="columns">
<div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"BB114814"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb4-2"><a href="#cb4-2"></a>  <span class="fu">lm</span>(expression <span class="sc">~</span> <span class="dv">0</span> <span class="sc">+</span> dev_stage, <span class="at">data =</span> .) <span class="sc">%&gt;%</span></span>
<span id="cb4-3"><a href="#cb4-3"></a>  <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 5
  term         estimate std.error statistic  p.value
  &lt;chr&gt;           &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 dev_stageE16     5.54    0.102       54.2 1.31e-34
2 dev_stageP2      5.84    0.0956      61.2 2.30e-36
3 dev_stageP6      5.78    0.0956      60.5 3.27e-36
4 dev_stageP10     6.38    0.0956      66.7 1.23e-37
5 dev_stageP28     9.17    0.0956      96.0 5.56e-43</code></pre>
</div>
</div>
</div><div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"BB114814"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb6-2"><a href="#cb6-2"></a>  <span class="fu">lm</span>(expression <span class="sc">~</span> dev_stage, <span class="at">data =</span> .) <span class="sc">%&gt;%</span></span>
<span id="cb6-3"><a href="#cb6-3"></a>  <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 5
  term         estimate std.error statistic  p.value
  &lt;chr&gt;           &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)     5.54      0.102     54.2  1.31e-34
2 dev_stageP2     0.304     0.140      2.17 3.69e- 2
3 dev_stageP6     0.243     0.140      1.74 9.11e- 2
4 dev_stageP10    0.834     0.140      5.96 9.62e- 7
5 dev_stageP28    3.63      0.140     26.0  5.30e-24</code></pre>
</div>
</div>
</div>
</div>
</section>
<section id="learning-objectives-for-today" class="slide level2">
<h2>Learning objectives for today</h2>
<ol type="1">
<li>Model more than one factor with multiple levels
<ul>
<li>build models with multiple categorical variables and their interaction</li>
</ul></li>
<li>Distinguish between <strong>simple</strong> and <strong>main</strong> effects
<ul>
<li><code>lm</code> vs <code>anova</code> tests</li>
</ul></li>
<li>Test main effects using <strong>nested</strong> models
<ul>
<li><span class="math inline">\(t\)</span>-tests vs <span class="math inline">\(F\)</span>-tests</li>
</ul></li>
</ol>
</section>
<section id="what-if-you-have-2-categorical-variables" class="slide level2">
<h2>What if you have <em>2</em> categorical variables?</h2>
<p>For example: <code>genotype</code> and <code>dev_stage</code> (for simplicity, let’s consider only E16 and P28)</p>
<ul>
<li><p>ANOVA is usually used to study models with one or more categorical variables (factors)</p></li>
<li><p>Can we combine 2 levels in each of 2 factors into 4 groups (treat as one-way ANOVA)?</p></li>
</ul>

<img data-src="lect07-linearModels_files/figure-revealjs/unnamed-chunk-8-1.png" width="528" class="r-stretch"></section>
<section id="what-if-you-have-2-categorical-variables-1" class="slide level2">
<h2>What if you have <em>2</em> categorical variables?</h2>
<p>For example: <code>genotype</code> and <code>dev_stage</code> (for simplicity, let’s consider only E16 and P28)</p>
<ul>
<li><p>ANOVA is usually used to study models with one or more categorical variables (factors)</p></li>
<li><p>Can we combine 2 levels in each of 2 factors into 4 groups (treat as one-way ANOVA)?</p>
<ul>
<li>no way to separate effects of each factor, or their interaction</li>
</ul></li>
</ul>

<img data-src="lect07-linearModels_files/figure-revealjs/unnamed-chunk-9-1.png" width="1248" class="r-stretch"></section>
<section id="two-way-anova-or-a-linear-model-with-interaction" class="slide level2">
<h2>Two-way ANOVA (or a linear model with interaction)</h2>
<div class="columns">
<div class="column">
<p>Which group means are we comparing in a model with 2 factors?</p>
<p><span class="math display">\[\mu_1=E[Y_{(WT,E16)}]\]</span> <span class="math display">\[\mu_2=E[Y_{(NrlKO,E16)}]\]</span> <span class="math display">\[\mu_3=E[Y_{(WT,P28)}]\]</span> <span class="math display">\[\mu_4=E[Y_{(NrlKO,P28)}]\]</span></p>
</div><div class="column">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="lect07-linearModels_files/figure-revealjs/unnamed-chunk-10-1.png" width="768"></p>
</div>
</div>
</div>
</div>
</section>
<section id="reference-treatment-effect-parameterization" class="slide level2">
<h2>Reference-treatment effect parameterization</h2>
<ul>
<li><p>By default, <code>lm</code> assumes a <strong>reference-treatment effect</strong> parameterization</p></li>
<li><p>Mathematically, we just need <em>more</em> indicator variables, see <a href="https://github.com/STAT540-UBC/resources/blob/main/linearModels/lect07_companion_linear_models.md">companion notes</a> for more details</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a>twoFactFit <span class="ot">&lt;-</span> <span class="fu">lm</span>(expression <span class="sc">~</span> genotype <span class="sc">*</span> dev_stage, oneGene)</span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="fu">tidy</span>(twoFactFit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 5
  term                       estimate std.error statistic  p.value
  &lt;chr&gt;                         &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)                   9.91      0.157     62.9  2.02e-15
2 genotypeNrlKO                -0.984     0.240     -4.09 1.78e- 3
3 dev_stageP28                  1.64      0.223      7.35 1.44e- 5
4 genotypeNrlKO:dev_stageP28   -2.04      0.328     -6.23 6.47e- 5</code></pre>
</div>
</div>
</section>
<section id="cell-means-and-treatment-effects-in-the-two-way-model" class="slide level2">
<h2>Cell-means and treatment effects in the two-way model</h2>
<p>Why do we need more indicator variables?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="fu">table</span>(oneGene<span class="sc">$</span>dev_stage, oneGene<span class="sc">$</span>genotype)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     
      WT NrlKO
  E16  4     3
  P28  4     4</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>(means<span class="fl">.2</span>Fact <span class="ot">&lt;-</span> oneGene <span class="sc">%&gt;%</span></span>
<span id="cb12-2"><a href="#cb12-2"></a>   <span class="fu">group_by</span>(dev_stage, genotype) <span class="sc">%&gt;%</span> </span>
<span id="cb12-3"><a href="#cb12-3"></a>   <span class="fu">summarize</span>(<span class="at">cellMeans =</span> <span class="fu">mean</span>(expression)) <span class="sc">%&gt;%</span> </span>
<span id="cb12-4"><a href="#cb12-4"></a>   <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span></span>
<span id="cb12-5"><a href="#cb12-5"></a>   <span class="fu">mutate</span>(<span class="at">txEffects =</span> cellMeans <span class="sc">-</span> cellMeans[<span class="dv">1</span>],</span>
<span id="cb12-6"><a href="#cb12-6"></a>          <span class="at">lmEst =</span> <span class="fu">tidy</span>(twoFactFit)<span class="sc">$</span>estimate))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 5
  dev_stage genotype cellMeans txEffects  lmEst
  &lt;fct&gt;     &lt;fct&gt;        &lt;dbl&gt;     &lt;dbl&gt;  &lt;dbl&gt;
1 E16       WT            9.91     0      9.91 
2 E16       NrlKO         8.92    -0.984 -0.984
3 P28       WT           11.5      1.64   1.64 
4 P28       NrlKO         8.52    -1.39  -2.04 </code></pre>
</div>
</div>
</section>
<section id="what-is-the-reference-group-here" class="slide level2">
<h2>What is the reference group here?</h2>
<p>Reference group: <strong>WT &amp; E16</strong></p>
<p>As before, comparisons are relative to a reference but in this case there is a reference level <em>in each factor</em>: <strong>WT and E16</strong></p>

<img data-src="lect07-linearModels_files/figure-revealjs/unnamed-chunk-14-1.png" width="720" class="r-stretch quarto-figure-center"></section>
<section id="the-reference-wt-e16" class="slide level2">
<h2>The reference: WT &amp; E16</h2>
<p><strong>Mean of reference group</strong>: <span class="math inline">\(\theta=E[Y_{WT,E16}]\)</span></p>
<p><strong><code>lm</code> estimate</strong>: <span class="math inline">\(\hat{\theta}\)</span> is the sample mean of the group</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="fu">tidy</span>(twoFactFit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 5
  term                       estimate std.error statistic  p.value
  &lt;chr&gt;                         &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)                   9.91      0.157     62.9  2.02e-15
2 genotypeNrlKO                -0.984     0.240     -4.09 1.78e- 3
3 dev_stageP28                  1.64      0.223      7.35 1.44e- 5
4 genotypeNrlKO:dev_stageP28   -2.04      0.328     -6.23 6.47e- 5</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a>means<span class="fl">.2</span>Fact</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 5
  dev_stage genotype cellMeans txEffects  lmEst
  &lt;fct&gt;     &lt;fct&gt;        &lt;dbl&gt;     &lt;dbl&gt;  &lt;dbl&gt;
1 E16       WT            9.91     0      9.91 
2 E16       NrlKO         8.92    -0.984 -0.984
3 P28       WT           11.5      1.64   1.64 
4 P28       NrlKO         8.52    -1.39  -2.04 </code></pre>
</div>
</div>
<p>In general, one is not interested in: <span class="math inline">\(H_0: \theta=0\)</span></p>
</section>
<section id="simple-genotype-effect-wt-vs-nrlko-at-e16" class="slide level2">
<h2>Simple genotype effect: WT <em>vs</em> NrlKO <font color="blue"> at E16 </font></h2>
<p>And now the “treatment effects”…</p>

<img data-src="lect07-linearModels_files/figure-revealjs/unnamed-chunk-17-1.png" width="768" class="r-stretch quarto-figure-center"><div class="fragment">
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Important: Simple/Conditional vs Main/Marginal effects</strong></p>
</div>
<div class="callout-content">
<p>“Treatment effect” parameters represent <strong>conditional effects</strong>: effects at a given level of the other factor (e.g.&nbsp;effect of genotype at E16). These are also called <strong>simple effects</strong>. They do <strong><em>not</em></strong> represent marginal effects.</p>
<p>A <strong>marginal effect</strong>, on the other hand, is the overall effect of a factor, averaged over all levels of the other factor (e.g.&nbsp;the overall effect of genotype, averaged over all levels of developmental time). These are also called <strong>main effects</strong>.</p>
</div>
</div>
</div>
</div>
</section>
<section id="simple-genotype-effect-wt-vs-nrlko-at-e16-1" class="slide level2">
<h2>Simple <strong>genotype</strong> effect: WT <em>vs</em> NrlKO <font color="blue"> at E16 </font></h2>
<p><strong>Effect of genotype at E16</strong>: <span class="math inline">\(\tau_{KO}=E[Y_{NrlKO,E16}]-E[Y_{WT,E16}]\)</span></p>
<p><strong><code>lm</code> estimate</strong>: <span class="math inline">\(\hat{\tau}_{KO}\)</span> is the <em>difference</em> of sample respective means (check below)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="fu">tidy</span>(twoFactFit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 5
  term                       estimate std.error statistic  p.value
  &lt;chr&gt;                         &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)                   9.91      0.157     62.9  2.02e-15
2 genotypeNrlKO                -0.984     0.240     -4.09 1.78e- 3
3 dev_stageP28                  1.64      0.223      7.35 1.44e- 5
4 genotypeNrlKO:dev_stageP28   -2.04      0.328     -6.23 6.47e- 5</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>means<span class="fl">.2</span>Fact</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 5
  dev_stage genotype cellMeans txEffects  lmEst
  &lt;fct&gt;     &lt;fct&gt;        &lt;dbl&gt;     &lt;dbl&gt;  &lt;dbl&gt;
1 E16       WT            9.91     0      9.91 
2 E16       NrlKO         8.92    -0.984 -0.984
3 P28       WT           11.5      1.64   1.64 
4 P28       NrlKO         8.52    -1.39  -2.04 </code></pre>
</div>
</div>
<p><strong>But</strong>, do you want to test the <em>conditional</em> effect at E16: <span class="math inline">\(H_0: \tau_{KO}=0\)</span>??</p>
</section>
<section id="simple-developmental-effect-e16-vs-p28-in-wt" class="slide level2">
<h2>Simple <strong>developmental</strong> effect: E16 <em>vs</em> P28 <font color="blue"> in WT </font></h2>
<p>Similarly, for the other factor: <span class="math inline">\(\tau_{P28}\)</span> is the effect of developmental time (P28 vs E16) <strong>in WT</strong></p>
<p>If <span class="math inline">\(\tau_{P28}=0\)</span>, what would the mean be in the WT group at P28?</p>

<img data-src="lect07-linearModels_files/figure-revealjs/unnamed-chunk-20-1.png" width="768" class="r-stretch quarto-figure-center"></section>
<section id="simple-developmental-effect-e16-vs-p28-in-wt-1" class="slide level2">
<h2>Simple <strong>developmental</strong> effect: E16 <em>vs</em> P28 <font color="blue"> in WT </font></h2>
<p>Similarly, for the other factor: <span class="math inline">\(\tau_{P28}\)</span> is the effect of developmental time (P28 vs E16) <strong>in WT</strong></p>
<p>If <span class="math inline">\(\tau_{P28}=0\)</span>, what would the mean be in the WT group at P28?</p>

<img data-src="lect07-linearModels_files/figure-revealjs/unnamed-chunk-21-1.png" width="768" class="r-stretch quarto-figure-center"></section>
<section id="simple-developmental-effect-e16-vs-p28-in-wt-2" class="slide level2">
<h2>Simple developmental effect: E16 <em>vs</em> P28 <font color="blue"> in WT </font></h2>
<p><strong>Effect of development in WT</strong>: <span class="math inline">\(\tau_{P28}=E[Y_{WT,P28}]-E[Y_{WT,E16}]\)</span></p>
<p><strong><code>lm</code> estimate</strong>: <span class="math inline">\(\hat{\tau}_{P28}\)</span> is the <em>difference</em> of respective sample means (check below)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a><span class="fu">tidy</span>(twoFactFit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 5
  term                       estimate std.error statistic  p.value
  &lt;chr&gt;                         &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)                   9.91      0.157     62.9  2.02e-15
2 genotypeNrlKO                -0.984     0.240     -4.09 1.78e- 3
3 dev_stageP28                  1.64      0.223      7.35 1.44e- 5
4 genotypeNrlKO:dev_stageP28   -2.04      0.328     -6.23 6.47e- 5</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a>means<span class="fl">.2</span>Fact</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 5
  dev_stage genotype cellMeans txEffects  lmEst
  &lt;fct&gt;     &lt;fct&gt;        &lt;dbl&gt;     &lt;dbl&gt;  &lt;dbl&gt;
1 E16       WT            9.91     0      9.91 
2 E16       NrlKO         8.92    -0.984 -0.984
3 P28       WT           11.5      1.64   1.64 
4 P28       NrlKO         8.52    -1.39  -2.04 </code></pre>
</div>
</div>
<p><strong>Again</strong>, do you want to test the <em>conditional</em> effect in WT: <span class="math inline">\(H_0: \tau_{P28}=0\)</span>??</p>
</section>
<section id="interaction-effect" class="slide level2">
<h2>Interaction effect</h2>
<p>Is the effect of genotype the same at different developmental stages?</p>
<p>Equivalently: Is the development effect the same for both genotypes?</p>

<img data-src="lect07-linearModels_files/figure-revealjs/unnamed-chunk-24-1.png" width="768" class="r-stretch quarto-figure-center"></section>
<section id="interaction-effect-1" class="slide level2">
<h2>Interaction effect</h2>
<p>Is the effect of genotype the same at different developmental stages?</p>
<p>Equivalently: Is the development effect the same for both genotypes?</p>

<img data-src="lect07-linearModels_files/figure-revealjs/unnamed-chunk-25-1.png" width="768" class="r-stretch quarto-figure-center"></section>
<section id="interaction-effect-2" class="slide level2">
<h2>Interaction effect</h2>
<p>Is the effect of genotype the same at different developmental stages?</p>
<p>Equivalently: Is the development effect the same for both genotypes?</p>

<img data-src="lect07-linearModels_files/figure-revealjs/unnamed-chunk-26-1.png" width="768" class="r-stretch quarto-figure-center"></section>
<section id="interaction-effect-3" class="slide level2">
<h2>Interaction effect</h2>
<p>The genotype effect at E16 is <span class="math inline">\(\tau_{KO}\)</span>. However, <span class="math inline">\(\tau_{KO}\)</span> does not seem to be the effect at P28.</p>
<p>The difference is the interaction effect! If there’s no interaction effect, <span class="math inline">\(\tau_{KO:P28}=0\)</span></p>

<img data-src="lect07-linearModels_files/figure-revealjs/unnamed-chunk-27-1.png" width="768" class="r-stretch quarto-figure-center"><p><a href="#/how-do-we-test-for-a-main-effect">Back</a></p>
</section>
<section id="interaction-effect-4" class="slide level2">
<h2>Interaction effect</h2>
<p>Difference of differences:</p>
<p><span class="math display">\[\tau_{KO:P28}= (E[Y_{NrlKO,P28}]-E[Y_{WT,P28}]) - (E[Y_{NrlKO,E16}]-E[Y_{WT,E16}])\]</span></p>
<p>In <code>lm</code> output:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a><span class="fu">tidy</span>(twoFactFit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 5
  term                       estimate std.error statistic  p.value
  &lt;chr&gt;                         &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)                   9.91      0.157     62.9  2.02e-15
2 genotypeNrlKO                -0.984     0.240     -4.09 1.78e- 3
3 dev_stageP28                  1.64      0.223      7.35 1.44e- 5
4 genotypeNrlKO:dev_stageP28   -2.04      0.328     -6.23 6.47e- 5</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a>means<span class="fl">.2</span>Fact</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 5
  dev_stage genotype cellMeans txEffects  lmEst
  &lt;fct&gt;     &lt;fct&gt;        &lt;dbl&gt;     &lt;dbl&gt;  &lt;dbl&gt;
1 E16       WT            9.91     0      9.91 
2 E16       NrlKO         8.92    -0.984 -0.984
3 P28       WT           11.5      1.64   1.64 
4 P28       NrlKO         8.52    -1.39  -2.04 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a>(means<span class="fl">.2</span>Fact<span class="sc">$</span>cellMeans[<span class="dv">4</span>] <span class="sc">-</span> means<span class="fl">.2</span>Fact<span class="sc">$</span>cellMeans[<span class="dv">3</span>]) <span class="sc">-</span> (means<span class="fl">.2</span>Fact<span class="sc">$</span>cellMeans[<span class="dv">2</span>] <span class="sc">-</span> means<span class="fl">.2</span>Fact<span class="sc">$</span>cellMeans[<span class="dv">1</span>]) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] -2.040372</code></pre>
</div>
</div>
</section>
<section id="summary-of-model-parameters-with-interaction" class="slide level2">
<h2>Summary of model parameters: with interaction</h2>
<table>
<colgroup>
<col style="width: 30%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th>model parameter</th>
<th><code>lm</code> estimate</th>
<th>stats</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><span class="math inline">\(\theta\)</span></td>
<td><code>(Intercept)</code></td>
<td><span class="math inline">\(E[Y_{WT,E16}]\)</span></td>
</tr>
<tr class="even">
<td><span class="math inline">\(\tau_{KO}\)</span></td>
<td><code>genotypeNrlKO</code></td>
<td><span class="math inline">\(E[Y_{NrlKO,E16}] - E[Y_{WT,E16}]\)</span></td>
</tr>
<tr class="odd">
<td><span class="math inline">\(\tau_{P28}\)</span></td>
<td><code>dev_stageP28</code></td>
<td><span class="math inline">\(E[Y_{WT,P28}] - E[Y_{WT,E16}]\)</span></td>
</tr>
<tr class="even">
<td><span class="math inline">\(\tau_{KO:P28}\)</span></td>
<td><code>genotypeNrlKO:dev_stageP28</code></td>
<td><span class="math inline">\(E[Y_{NrlKO,P28}] - E[Y_{WT,P28}] - \tau_{KO}\)</span></td>
</tr>
</tbody>
</table>
<p>It is important to remember that <code>lm</code> reports <strong>simple, not main</strong> effects!</p>
<p>Why? Because of the parameterization used!  (see <a href="https://github.com/STAT540-UBC/resources/blob/main/linearModels/lect07_companion_linear_models.md">companion notes</a>)</p>
<p>It can also be shown that <span class="math inline">\(\tau_{KO:P28}=E[Y_{NrlKO,P28}]-\tau_{P28}-\tau_{KO}-\theta\)</span> (see previous slide and companion notes)</p>
</section>
<section id="lets-examine-these-parameters-closer" class="slide level2">
<h2>Let’s examine these parameters closer</h2>
<p>For our model, <code>lm</code> tests 4 hypotheses:</p>
<div class="columns">
<div class="column">
<blockquote>
<h3 id="h_0-theta0"><span class="math inline">\(H_0: \theta=0\)</span></h3>
<h3 id="h_0-tau_ko0"><span class="math inline">\(H_0: \tau_{KO}=0\)</span></h3>
<h3 id="h_0-tau_p280"><span class="math inline">\(H_0: \tau_{P28}=0\)</span></h3>
<h3 id="h_0-tau_kop280"><span class="math inline">\(H_0: \tau_{KO:P28}=0\)</span></h3>
</blockquote>
</div><div class="column">
<p><img data-src="img/2way_param.png"></p>
</div>
</div>
<p>We may not be interested in these hypotheses, e.g., <span class="math inline">\(\tau_{KO}\)</span> and <span class="math inline">\(\tau_{P28}\)</span> are <em>conditional</em> effects <em>at</em> a given level of a factor (<em>simple effects</em>)</p>
</section>
<section id="ex-1-nothing-statistically-significant-very-flat-genes" class="slide level2">
<h2>Ex 1: nothing statistically significant, very flat genes</h2>
<div class="panel-tabset">
<ul id="tabset-1" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-1-1">Plots</a></li><li><a href="#tabset-1-2"><code>lm</code> output</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="lect07-linearModels_files/figure-revealjs/unnamed-chunk-31-1.png" width="1344"></p>
</div>
</div>
</div>
<div id="tabset-1-2">
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1"></a><span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"1442080_at"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb32-2"><a href="#cb32-2"></a>  <span class="fu">lm</span>(expression <span class="sc">~</span> genotype <span class="sc">*</span> dev_stage, <span class="at">data =</span> .) <span class="sc">%&gt;%</span></span>
<span id="cb32-3"><a href="#cb32-3"></a>  <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 5
  term                       estimate std.error statistic  p.value
  &lt;chr&gt;                         &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)                  6.86      0.0703    97.7   1.61e-17
2 genotypeNrlKO               -0.0378    0.107     -0.352 7.32e- 1
3 dev_stageP28                 0.0795    0.0994     0.800 4.41e- 1
4 genotypeNrlKO:dev_stageP28   0.149     0.146      1.02  3.29e- 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1"></a><span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"1448243_at"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb34-2"><a href="#cb34-2"></a>  <span class="fu">lm</span>(expression <span class="sc">~</span> genotype <span class="sc">*</span> dev_stage, <span class="at">data =</span> .) <span class="sc">%&gt;%</span></span>
<span id="cb34-3"><a href="#cb34-3"></a>  <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 5
  term                       estimate std.error statistic  p.value
  &lt;chr&gt;                         &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)                   8.52      0.256    33.3   2.16e-12
2 genotypeNrlKO                -0.434     0.391    -1.11  2.91e- 1
3 dev_stageP28                 -0.253     0.362    -0.700 4.99e- 1
4 genotypeNrlKO:dev_stageP28    0.551     0.533     1.03  3.24e- 1</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="ex-2-statistically-significant-interaction-non-parallel" class="slide level2">
<h2>Ex 2: statistically significant interaction (non-parallel)</h2>
<div class="panel-tabset">
<ul id="tabset-2" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-2-1">Plots</a></li><li><a href="#tabset-2-2"><code>lm</code> output</a></li></ul>
<div class="tab-content">
<div id="tabset-2-1">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="lect07-linearModels_files/figure-revealjs/unnamed-chunk-34-1.png" width="1344"></p>
</div>
</div>
</div>
<div id="tabset-2-2">
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1"></a><span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"1434709_at"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb36-2"><a href="#cb36-2"></a>  <span class="fu">lm</span>(expression <span class="sc">~</span> genotype <span class="sc">*</span> dev_stage, <span class="at">data =</span> .) <span class="sc">%&gt;%</span></span>
<span id="cb36-3"><a href="#cb36-3"></a>  <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 5
  term                       estimate std.error statistic  p.value
  &lt;chr&gt;                         &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)                    8.59     0.221     38.8  4.08e-13
2 genotypeNrlKO                 -1.59     0.338     -4.71 6.43e- 4
3 dev_stageP28                  -2.22     0.313     -7.09 2.02e- 5
4 genotypeNrlKO:dev_stageP28     4.97     0.461     10.8  3.48e- 7</code></pre>
</div>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1"></a><span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"1458220_at"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb38-2"><a href="#cb38-2"></a>  <span class="fu">lm</span>(expression <span class="sc">~</span> genotype <span class="sc">*</span> dev_stage, <span class="at">data =</span> .) <span class="sc">%&gt;%</span></span>
<span id="cb38-3"><a href="#cb38-3"></a>  <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 5
  term                       estimate std.error statistic  p.value
  &lt;chr&gt;                         &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)                    9.47     0.185     51.2  1.94e-14
2 genotypeNrlKO                 -2.33     0.283     -8.23 4.97e- 6
3 dev_stageP28                  -1.37     0.262     -5.24 2.75e- 4
4 genotypeNrlKO:dev_stageP28     1.51     0.385      3.92 2.40e- 3</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="disagreement-in-simple-effects-with-interaction" class="slide level2">
<h2>Disagreement in simple effects with interaction</h2>
<ul>
<li><p>Note that a significant interaction means the <strong>simple</strong> effects may not agree</p></li>
<li><p>For the gene 1434709_at on the previous slide, compare the effect of genotype at E16 and P28:</p></li>
</ul>
<table>
<thead>
<tr class="header">
<th>Effect</th>
<th><code>lm</code> output</th>
<th style="text-align: center;">Estimate</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Genotype at E16</td>
<td><code>genotypeNrlKO</code></td>
<td style="text-align: center;"></td>
</tr>
<tr class="even">
<td>Genotype at P28</td>
<td></td>
<td style="text-align: center;"></td>
</tr>
</tbody>
</table>
<div class="fragment">
<ul>
<li><p><strong>Main</strong> effects (overall): does genotype have an effect on gene expression?</p>
<ul>
<li>We can’t (yet) answer this question! It depends (on the level of <code>dev_stage</code>)! (more later)</li>
</ul></li>
</ul>
</div>
</section>
<section id="ex-3-balanced-only-genotype-at-e16-is-significant" class="slide level2">
<h2>Ex 3: <em>BALANCED</em> &amp; only genotype at E16 is significant</h2>
<p>For simplicity here, we’ll add a fake observation in the NrlKO &amp; E16 group (close to its mean) so that we have a <em>balanced</em> design</p>
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Note</strong></p>
</div>
<div class="callout-content">
<p>In <em>unbalanced</em> designs the <em>main</em> effects are a <em>weighted</em> average of the simple effects, and the weights are not easy to interpret (beyond the scope of this course but worth noting the issue!)</p>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1"></a><span class="co"># recall our unbalanced design</span></span>
<span id="cb40-2"><a href="#cb40-2"></a><span class="fu">table</span>(<span class="fu">pData</span>(eset)<span class="sc">$</span>genotype, <span class="fu">pData</span>(eset)<span class="sc">$</span>dev_stage)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       
        E16 P2 P6 P10 P28
  WT      4  4  4   4   4
  NrlKO   3  4  4   4   4</code></pre>
</div>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1"></a><span class="co"># Duplicate sample GSM92615 (E16 NrlKO) and add noise expression</span></span>
<span id="cb42-2"><a href="#cb42-2"></a>twoGenes <span class="ot">&lt;-</span> <span class="fu">filter</span>(twoGenes, sample_id <span class="sc">==</span> <span class="st">"GSM92615"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb42-3"><a href="#cb42-3"></a>  <span class="fu">mutate</span>(<span class="at">expression =</span> expression <span class="sc">+</span> <span class="fu">rnorm</span>(<span class="fu">n</span>(), <span class="dv">0</span>, <span class="fl">0.05</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb42-4"><a href="#cb42-4"></a>  <span class="fu">rbind</span>(twoGenes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="ex-3-balanced-only-genotype-at-e16-is-significant-1" class="slide level2">
<h2>Ex 3: <em>BALANCED</em> &amp; only genotype at E16 is significant</h2>
<div class="panel-tabset">
<ul id="tabset-3" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-3-1">Plots</a></li><li><a href="#tabset-3-2"><code>lm</code> output</a></li></ul>
<div class="tab-content">
<div id="tabset-3-1">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="lect07-linearModels_files/figure-revealjs/unnamed-chunk-38-1.png" width="1344"></p>
</div>
</div>
</div>
<div id="tabset-3-2">
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1"></a><span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"1447753_at"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb43-2"><a href="#cb43-2"></a>  <span class="fu">lm</span>(expression <span class="sc">~</span> genotype <span class="sc">*</span> dev_stage, <span class="at">data =</span> .) <span class="sc">%&gt;%</span></span>
<span id="cb43-3"><a href="#cb43-3"></a>  <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 5
  term                       estimate std.error statistic  p.value
  &lt;chr&gt;                         &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)                  6.48      0.0454    143.   9.31e-21
2 genotypeNrlKO                0.300     0.0641      4.68 5.35e- 4
3 dev_stageP28                -0.0768    0.0641     -1.20 2.54e- 1
4 genotypeNrlKO:dev_stageP28  -0.113     0.0907     -1.24 2.38e- 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1"></a><span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"1431651_at"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb45-2"><a href="#cb45-2"></a>  <span class="fu">lm</span>(expression <span class="sc">~</span> genotype <span class="sc">*</span> dev_stage, <span class="at">data =</span> .) <span class="sc">%&gt;%</span></span>
<span id="cb45-3"><a href="#cb45-3"></a>  <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 5
  term                       estimate std.error statistic  p.value
  &lt;chr&gt;                         &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)                 6.50       0.0710   91.6    1.92e-18
2 genotypeNrlKO               0.220      0.100     2.19   4.88e- 2
3 dev_stageP28               -0.00250    0.100    -0.0249 9.81e- 1
4 genotypeNrlKO:dev_stageP28 -0.105      0.142    -0.736  4.76e- 1</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="ex-3-balanced-only-genotype-at-e16-is-significant-2" class="slide level2">
<h2>Ex 3: <em>BALANCED</em> &amp; only genotype at E16 is significant</h2>
<p>For both of these genes:</p>
<ul>
<li><p>The interaction effect is not significant (almost parallel pattern)</p></li>
<li><p>The effect of developmental stage is not significant for WT (almost flat pattern)</p></li>
<li><p>There is a significant genotype effect at E16</p></li>
<li><p>There may be a genotype effect <em>regardless</em> of the developmental stage (<strong>main</strong> effect). However, that hypothesis is <strong>not</strong> tested here!!</p></li>
</ul>
<div class="fragment">
<ul>
<li>How do we test a <strong>main</strong> effect??</li>
</ul>
</div>
</section>
<section id="how-do-we-test-for-a-main-effect" class="slide level2">
<h2>How do we test for a <strong>main</strong> effect?</h2>
<ul>
<li><p>The main effect measures the <em>overall</em> association between the response and a factor - it is the (weighted) average of an effect over the levels of the other factor</p></li>
<li><p><code>anova()</code> can be used to test the main effects</p></li>
<li><p>The following is a way to write the null hypothesis that there is no main effect of genotype:</p></li>
</ul>
<p><span class="math display">\[H_0: \frac{(E[Y_{KO,E16}]-E[Y_{WT,E16}])+(E[Y_{KO,P28}]-E[Y_{WT,P28}])}{2}=0\]</span></p>
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Note</strong></p>
</div>
<div class="callout-content">
<p>For unbalanced experiments <span class="math inline">\(H_0: w_1 \text{effect}_{E16}+ w_2 \text{effect}_{P28}=0\)</span>, where <span class="math inline">\(w_1\)</span> and <span class="math inline">\(w_2\)</span> are sample size weights</p>
</div>
</div>
</div>
<div class="fragment">
<p>How would we write this in terms of the Interaction model parameters? (Refer back to <a href="#/interaction-effect-3">this slide</a>)</p>
</div>
</section>
<section id="main-effects-using-anova-f-tests" class="slide level2">
<h2>Main effects using <code>anova</code> F tests</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1"></a><span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"1447753_at"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb47-2"><a href="#cb47-2"></a>  <span class="fu">lm</span>(expression <span class="sc">~</span> genotype <span class="sc">*</span> dev_stage, <span class="at">data =</span> .) <span class="sc">%&gt;%</span></span>
<span id="cb47-3"><a href="#cb47-3"></a>  <span class="fu">anova</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb47-4"><a href="#cb47-4"></a>  <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 6
  term                  df  sumsq  meansq statistic   p.value
  &lt;chr&gt;              &lt;int&gt;  &lt;dbl&gt;   &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
1 genotype               1 0.237  0.237       28.8   0.000168
2 dev_stage              1 0.0709 0.0709       8.62  0.0125  
3 genotype:dev_stage     1 0.0127 0.0127       1.54  0.238   
4 Residuals             12 0.0987 0.00823     NA    NA       </code></pre>
</div>
</div>
<p>As we suspected, there is a <strong>significant genotype effect</strong> for this probe (1447753_at), i.e., its mean expression changes in NrlKO group (compared to WT), on average over developmental stages.</p>
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Technical note:</strong></p>
</div>
<div class="callout-content">
<p><code>anova()</code> uses <em>type I sums of squares</em> (sequential; conditional on previous terms), thus order matters in unbalanced designs! See this <a href="https://mcfromnz.wordpress.com/2011/03/02/anova-type-iiiiii-ss-explained/">primer</a> on types of sums of squares for an intuitive explanation.</p>
</div>
</div>
</div>
</section>
<section id="main-interaction-effects-important-notes" class="slide level2">
<h2>Main &amp; interaction effects: important notes</h2>
<ul>
<li><p>A <strong>significant interaction effect</strong> means that the effect of one factor depends on the levels of another</p>
<ul>
<li>e.g., the effect of genotype depends on developmental stage</li>
</ul></li>
<li><p><strong>Main effects</strong>: are the (weighted) average of an effect over the levels of the other factor</p></li>
<li><p>A <strong>non-significant main effect</strong> means that, on average, there’s no evidence of a factor’s effect</p>
<ul>
<li>e.g., no evidence of a genotype effect, on average over both developmental stages</li>
</ul></li>
</ul>
<div class="callout callout-caution callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Caution</strong></p>
</div>
<div class="callout-content">
<p>If the interaction is significant, it is possible that one or both simple effects are significant but the average effect (i.e., the main effect) is not. This is because the effect of a factor <em>depends on</em> the level of the other one. Looking at main effects alone may mask interesting results!</p>
</div>
</div>
</div>
</section>
<section id="additive-models" class="slide level2">
<h2>Additive models</h2>
<ul>
<li><p>In some applications, we need to/want to test the interaction term</p></li>
<li><p>However, additive models are simpler and smaller</p></li>
<li><p>If there are no statistical or biological grounds to include the interaction term, additive models are preferred</p></li>
<li><p>Additive effects: <span class="math inline">\(E[Y_{NrlKO,P28}]-E[Y_{WT,E16}]=\tau_{KO}+\tau_{P28}\)</span></p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1"></a><span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"1447753_at"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb49-2"><a href="#cb49-2"></a>  <span class="fu">lm</span>(expression <span class="sc">~</span> genotype <span class="sc">+</span> dev_stage, <span class="at">data =</span> .) <span class="sc">%&gt;%</span></span>
<span id="cb49-3"><a href="#cb49-3"></a>  <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 5
  term          estimate std.error statistic  p.value
  &lt;chr&gt;            &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)      6.51     0.0401    162.   6.96e-23
2 genotypeNrlKO    0.244    0.0463      5.26 1.54e- 4
3 dev_stageP28    -0.133    0.0463     -2.88 1.30e- 2</code></pre>
</div>
</div>
</section>
<section id="exercise---main-effect-agreement-in-balanced-models" class="slide level2">
<h2>Exercise - Main effect agreement in Balanced models</h2>
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Check for yourself!</strong></p>
</div>
<div class="callout-content">
<p>Note the agreement between coefficients in this (balanced) additive model and estimates of the main effects using the output of the interaction model!</p>
</div>
</div>
</div>
<div class="panel-tabset">
<ul id="tabset-4" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-4-1">Additive model</a></li><li><a href="#tabset-4-2">Interaction model</a></li><li><a href="#tabset-4-3">Main effect estimate from Interaction model</a></li></ul>
<div class="tab-content">
<div id="tabset-4-1">
<div class="cell" data-highlight.output="5">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1"></a><span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"1447753_at"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb51-2"><a href="#cb51-2"></a>  <span class="fu">lm</span>(expression <span class="sc">~</span> genotype <span class="sc">+</span> dev_stage, <span class="at">data =</span> .) <span class="sc">%&gt;%</span></span>
<span id="cb51-3"><a href="#cb51-3"></a>  <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 5
  term          estimate std.error statistic  p.value
  &lt;chr&gt;            &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)      6.51     0.0401    162.   6.96e-23
2 genotypeNrlKO    0.244    0.0463      5.26 1.54e- 4
3 dev_stageP28    -0.133    0.0463     -2.88 1.30e- 2</code></pre>
</div>
</div>
</div>
<div id="tabset-4-2">
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1"></a><span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"1447753_at"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb53-2"><a href="#cb53-2"></a>  <span class="fu">lm</span>(expression <span class="sc">~</span> genotype<span class="sc">*</span>dev_stage, <span class="at">data =</span> .) <span class="sc">%&gt;%</span></span>
<span id="cb53-3"><a href="#cb53-3"></a>  <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 5
  term                       estimate std.error statistic  p.value
  &lt;chr&gt;                         &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)                  6.48      0.0454    143.   9.31e-21
2 genotypeNrlKO                0.300     0.0641      4.68 5.35e- 4
3 dev_stageP28                -0.0768    0.0641     -1.20 2.54e- 1
4 genotypeNrlKO:dev_stageP28  -0.113     0.0907     -1.24 2.38e- 1</code></pre>
</div>
</div>
</div>
<div id="tabset-4-3">
<ul>
<li><p>Main effect estimate of genotype = <span class="math display">\[\hat{\tau}_{KO} + \frac{1}{2}\hat{\tau}_{KO:P28}\]</span></p></li>
<li><p>Main effect estimate of dev_stage = <span class="math display">\[\hat{\tau}_{P28} + \frac{1}{2}\hat{\tau}_{KO:P28}\]</span></p></li>
</ul>
</div>
</div>
</div>
</section>
<section id="additive-models-and-balanced-designs" class="slide level2">
<h2><strong>Additive models</strong> and balanced designs</h2>
<ul>
<li><p>In an additive model, the <code>lm()</code> parameters for balanced designs are <strong>average effects</strong>, over the levels of the other factor - same as in <code>anova()</code>!</p>
<ul>
<li>Note the agreement between <code>lm</code> and <code>anova</code>; this is gone in unbalanced designs since weights are computed differently!</li>
</ul></li>
<li><p>The intercept parameter is now <span class="math inline">\(\bar{Y} - \bar{x}_{ij,KO}\hat{\tau}_{KO} - \bar{x}_{ij,P28}\hat{\tau}_{P28}\)</span></p></li>
</ul>
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Note</strong></p>
</div>
<div class="callout-content">
<p><em>Type III sum of squares</em> (partial; conditional on all other terms in the model) are required for agreement in unbalanced designs (use <code>car::Anova()</code> to obtain) - beyond our scope</p>
</div>
</div>
</div>
</section>
<section id="parameters-in-additive-models-represent-main-effects" class="slide level2">
<h2>Parameters in additive models represent main effects</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1"></a>(fit <span class="ot">&lt;-</span> <span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"1447753_at"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb55-2"><a href="#cb55-2"></a>  <span class="fu">lm</span>(expression <span class="sc">~</span> genotype <span class="sc">+</span> dev_stage, <span class="at">data =</span> .)) <span class="sc">%&gt;%</span></span>
<span id="cb55-3"><a href="#cb55-3"></a>  <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 5
  term          estimate std.error statistic  p.value
  &lt;chr&gt;            &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)      6.51     0.0401    162.   6.96e-23
2 genotypeNrlKO    0.244    0.0463      5.26 1.54e- 4
3 dev_stageP28    -0.133    0.0463     -2.88 1.30e- 2</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1"></a><span class="fu">tidy</span>(fit)<span class="sc">$</span>statistic[<span class="dv">2</span>]<span class="sc">^</span><span class="dv">2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 27.69005</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1"></a>fit <span class="sc">%&gt;%</span> <span class="fu">anova</span>() <span class="sc">%&gt;%</span> <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 6
  term         df  sumsq  meansq statistic   p.value
  &lt;chr&gt;     &lt;int&gt;  &lt;dbl&gt;   &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
1 genotype      1 0.237  0.237       27.7   0.000154
2 dev_stage     1 0.0709 0.0709       8.27  0.0130  
3 Residuals    13 0.111  0.00857     NA    NA       </code></pre>
</div>
</div>
</section>
<section id="additive-vs-interaction-models" class="slide level2">
<h2>Additive vs interaction models</h2>

<img data-src="lect07-linearModels_files/figure-revealjs/unnamed-chunk-47-1.png" width="1152" class="r-stretch quarto-figure-center"></section>
<section id="interactions-with-multi-level-factors-more-than-2-groups" class="slide level2">
<h2>Interactions with multi-level factors (more than 2 groups)</h2>
<p>Back to our old friend the BB114814 gene</p>

<img data-src="lect07-linearModels_files/figure-revealjs/unnamed-chunk-48-1.png" width="960" class="r-stretch quarto-figure-center"></section>
<section id="interactions-with-multi-level-factors-more-than-2-groups-1" class="slide level2">
<h2>Interactions with multi-level factors (more than 2 groups)</h2>
<p>We can generalize the regression model to factors with more levels (e.g., E16, P2, P10 and P28): we just add more indicator variables (and parameters)!</p>
<p><strong>With interaction</strong></p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1"></a>bb1gene <span class="ot">&lt;-</span> <span class="fu">toLongerMeta</span>(eset) <span class="sc">%&gt;%</span> </span>
<span id="cb61-2"><a href="#cb61-2"></a>  <span class="fu">filter</span>(gene <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"1440645_at"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb61-3"><a href="#cb61-3"></a>  <span class="fu">mutate</span>(<span class="at">gene =</span> <span class="st">"BB114814"</span>)</span>
<span id="cb61-4"><a href="#cb61-4"></a></span>
<span id="cb61-5"><a href="#cb61-5"></a>(itxFit <span class="ot">&lt;-</span> <span class="fu">lm</span>(expression <span class="sc">~</span> genotype<span class="sc">*</span>dev_stage, <span class="at">data =</span> bb1gene)) <span class="sc">%&gt;%</span></span>
<span id="cb61-6"><a href="#cb61-6"></a>  <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 5
   term                       estimate std.error statistic  p.value
   &lt;chr&gt;                         &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
 1 (Intercept)                  5.43       0.124    43.8   4.76e-28
 2 genotypeNrlKO                0.252      0.189     1.33  1.95e- 1
 3 dev_stageP2                  0.399      0.175     2.27  3.05e- 2
 4 dev_stageP6                  0.195      0.175     1.11  2.75e- 1
 5 dev_stageP10                 0.920      0.175     5.24  1.29e- 5
 6 dev_stageP28                 3.96       0.175    22.6   5.97e-20
 7 genotypeNrlKO:dev_stageP2   -0.226      0.258    -0.877 3.88e- 1
 8 genotypeNrlKO:dev_stageP6    0.0599     0.258     0.232 8.18e- 1
 9 genotypeNrlKO:dev_stageP10  -0.208      0.258    -0.804 4.28e- 1
10 genotypeNrlKO:dev_stageP28  -0.694      0.258    -2.69  1.18e- 2</code></pre>
</div>
</div>
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Note</strong></p>
</div>
<div class="callout-content">
<p>All the <code>dev_stage</code> parameters are still <strong>simple</strong> effects, but we now have more: one for each level compared to the reference</p>
</div>
</div>
</div>
</section>
<section id="factors-with-multiple-levels-cont." class="slide level2">
<h2>Factors with multiple levels (cont.)</h2>
<p><strong>Without interaction: additive</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1"></a>(addFit <span class="ot">&lt;-</span> <span class="fu">lm</span>(expression <span class="sc">~</span> genotype <span class="sc">+</span> dev_stage, <span class="at">data =</span> bb1gene)) <span class="sc">%&gt;%</span></span>
<span id="cb63-2"><a href="#cb63-2"></a>  <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 5
  term          estimate std.error statistic  p.value
  &lt;chr&gt;            &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)     5.53      0.110     50.2   9.62e-33
2 genotypeNrlKO   0.0317    0.0878     0.361 7.21e- 1
3 dev_stageP2     0.302     0.142      2.13  4.11e- 2
4 dev_stageP6     0.241     0.142      1.70  9.87e- 2
5 dev_stageP10    0.832     0.142      5.86  1.44e- 6
6 dev_stageP28    3.63      0.142     25.6   2.43e-23</code></pre>
</div>
</div>
<p>Parameters are now <strong>main</strong> effects (on average over the levels of the other factor), but we have more!</p>
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Question</strong></p>
</div>
<div class="callout-content">
<p>Does developmental stage have a significant effect on this gene’s expression?</p>
</div>
</div>
</div>
<div class="fragment">
<p>We haven’t tested that!!</p>
</div>
</section>
<section id="recall-f-test-and-overall-significance" class="slide level2">
<h2>Recall: <em>F</em>-test and overall significance</h2>
<ul>
<li>the <em>t</em>-test in linear regression allows us to test single hypotheses; these are given in the summary of <code>lm</code> <span class="math display">\[H_0 : \tau_i = 0\]</span> <span class="math display">\[H_A : \tau_j \neq 0\]</span></li>
<li>but we often like to test multiple hypotheses <em>simultaneously</em>: <span class="math display">\[H_0 : \tau_{P2} = \tau_{P6} = \tau_{P10} = \tau_{P28}=0\textrm{ [AND statement]}\]</span> <span class="math display">\[H_A : \tau_j \neq 0 \textrm{ for at least one j [OR statement]}\]</span> the <em>F</em>-test allows us to test such compound tests</li>
</ul>
</section>
<section id="compound-tests-in-2-factor-interaction-models" class="slide level2">
<h2>Compound tests in 2-factor Interaction models</h2>
<ul>
<li><code>lm</code> tests each factor level separately (OR) - simple effects in reference category of other factor</li>
<li><code>anova</code> tests each factor level jointly (AND) - main effect, controlling for the previous ones</li>
</ul>
<div class="cell" data-code-folding="true">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1"></a>itxFit <span class="sc">%&gt;%</span> <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 5
   term                       estimate std.error statistic  p.value
   &lt;chr&gt;                         &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
 1 (Intercept)                  5.43       0.124    43.8   4.76e-28
 2 genotypeNrlKO                0.252      0.189     1.33  1.95e- 1
 3 dev_stageP2                  0.399      0.175     2.27  3.05e- 2
 4 dev_stageP6                  0.195      0.175     1.11  2.75e- 1
 5 dev_stageP10                 0.920      0.175     5.24  1.29e- 5
 6 dev_stageP28                 3.96       0.175    22.6   5.97e-20
 7 genotypeNrlKO:dev_stageP2   -0.226      0.258    -0.877 3.88e- 1
 8 genotypeNrlKO:dev_stageP6    0.0599     0.258     0.232 8.18e- 1
 9 genotypeNrlKO:dev_stageP10  -0.208      0.258    -0.804 4.28e- 1
10 genotypeNrlKO:dev_stageP28  -0.694      0.258    -2.69  1.18e- 2</code></pre>
</div>
</div>
<div class="cell" data-code-folding="true">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1"></a><span class="fu">anova</span>(itxFit) <span class="sc">%&gt;%</span> <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 6
  term                  df   sumsq  meansq statistic   p.value
  &lt;chr&gt;              &lt;int&gt;   &lt;dbl&gt;   &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
1 genotype               1  0.0693  0.0693      1.13  2.97e- 1
2 dev_stage              4 71.0    17.8       288.    6.72e-23
3 genotype:dev_stage     4  0.689   0.172       2.80  4.43e- 2
4 Residuals             29  1.78    0.0616     NA    NA       </code></pre>
</div>
</div>
</section>
<section id="compound-tests-in-2-factor-additive-models" class="slide level2">
<h2>Compound tests in 2-factor Additive models</h2>
<ul>
<li><code>lm</code> tests each factor level separately (OR) - main effects averaged over levels of other factor</li>
<li><code>anova</code> tests each factor level jointly (AND) - main effect, controlling for the previous ones</li>
</ul>
<div class="cell" data-code-folding="true">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1"></a>addFit <span class="sc">%&gt;%</span> <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 5
  term          estimate std.error statistic  p.value
  &lt;chr&gt;            &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)     5.53      0.110     50.2   9.62e-33
2 genotypeNrlKO   0.0317    0.0878     0.361 7.21e- 1
3 dev_stageP2     0.302     0.142      2.13  4.11e- 2
4 dev_stageP6     0.241     0.142      1.70  9.87e- 2
5 dev_stageP10    0.832     0.142      5.86  1.44e- 6
6 dev_stageP28    3.63      0.142     25.6   2.43e-23</code></pre>
</div>
</div>
<div class="cell" data-code-folding="true">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1"></a><span class="fu">anova</span>(addFit) <span class="sc">%&gt;%</span> <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 6
  term         df   sumsq  meansq statistic   p.value
  &lt;chr&gt;     &lt;int&gt;   &lt;dbl&gt;   &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
1 genotype      1  0.0693  0.0693     0.925  3.43e- 1
2 dev_stage     4 71.0    17.8      237.     8.45e-24
3 Residuals    33  2.47    0.0750    NA     NA       </code></pre>
</div>
</div>
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Note</strong></p>
</div>
<div class="callout-content">
<p>The <span class="math inline">\(t\)</span>-test in <code>lm</code> and the <span class="math inline">\(F\)</span>-test (1 df) in <code>anova</code> for genotype are not equivalent here due to unbalancedness (order matters)</p>
</div>
</div>
</div>
</section>
<section id="these-examples-are-just-special-cases-of-nested-models" class="slide level2">
<h2>These examples are just special cases of <em>nested models</em></h2>
<p>For example: does development have a significant effect on gene expression?</p>
<p>Compare the models with and without <code>dev_stage</code>!</p>
<p><strong>Model 1</strong>: <code>expression ~ genotype</code></p>
<p><strong>Model 2</strong>: <code>expression ~ genotype + dev_stage</code></p>
<p>Mathematically:</p>
<p><strong>Model 1</strong>: <span class="math inline">\(Y_{ijk}=\theta + \tau_{KO}x_{KO,ijk} + \varepsilon\)</span></p>
<p><strong>Model 2</strong>: <span class="math inline">\(Y_{ijk}=\theta + \tau_{KO}x_{KO,ijk} + \tau_{P2}x_{P2,ijk}+\tau_{P6}x_{P6,ijk}+\tau_{P10}x_{P10,ijk}+\tau_{P28}x_{P28,ijk}+ \varepsilon\)</span></p>
<p><span class="math display">\[H_0: \tau_{P2}=\tau_{P6}=\tau_{P10}=\tau_{P28}=0\]</span></p>
<p>The <span class="math inline">\(x_{**,ijk}\)</span> are indicator variables (see <a href="https://github.com/STAT540-UBC/resources/blob/main/linearModels/lect07_companion_linear_models.md">companion notes</a>)</p>
</section>
<section id="more-general-f-test-to-compare-nested-models" class="slide level2">
<h2>More general: F-test to compare nested models</h2>
<p><span class="math inline">\(H_0: \alpha_{k+1} = ... = \alpha_{k+p}\)</span></p>
<p><span class="math display">\[F =  \frac{(SS_{reduced} - SS_{full})/(p)}{SS_{full} / (n-p-k-1)} \sim \mathbf{F}_{p, \,n-p-k-1}\]</span> This <span class="math inline">\(F\)</span>-statistic compares the following two models:</p>
<ul>
<li>Reduced (k + 1 parameters): <span class="math display">\[y_i = \alpha_0 + \alpha_1 x_{i1} + ... +  \alpha_k x_{ik} + \epsilon_i\]</span></li>
<li>Full (p + k + 1 parameters): <span class="math display">\[y_i = \alpha_0 + \alpha_1 x_{i1} + ... +  \alpha_k x_{ik} + ... +  \alpha_p x_{ip} + \epsilon_i\]</span></li>
</ul>
<div class="fragment">
<p>A <em>significant</em> F-statistic here means that the full model explains significantly more variation in the outcome variable than the reduced model</p>
</div>
</section>
<section id="nested-models-in-r" class="slide level2">
<h2>Nested models in R</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1"></a>addReduced <span class="ot">&lt;-</span> <span class="fu">lm</span>(expression <span class="sc">~</span> genotype, <span class="at">data =</span> bb1gene)</span>
<span id="cb73-2"><a href="#cb73-2"></a>addFull <span class="ot">&lt;-</span> <span class="fu">lm</span>(expression <span class="sc">~</span> genotype <span class="sc">+</span> dev_stage, <span class="at">data =</span> bb1gene)</span>
<span id="cb73-3"><a href="#cb73-3"></a><span class="fu">anova</span>(addReduced,addFull)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Variance Table

Model 1: expression ~ genotype
Model 2: expression ~ genotype + dev_stage
  Res.Df    RSS Df Sum of Sq      F    Pr(&gt;F)    
1     37 73.497                                  
2     33  2.474  4    71.023 236.84 &lt; 2.2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1"></a><span class="fu">anova</span>(addFull) <span class="sc">%&gt;%</span> <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 6
  term         df   sumsq  meansq statistic   p.value
  &lt;chr&gt;     &lt;int&gt;   &lt;dbl&gt;   &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
1 genotype      1  0.0693  0.0693     0.925  3.43e- 1
2 dev_stage     4 71.0    17.8      237.     8.45e-24
3 Residuals    33  2.47    0.0750    NA     NA       </code></pre>
</div>
</div>
</section>
<section id="another-special-case-overall-goodness-of-fit" class="slide level2">
<h2>Another special case: overall goodness of fit!</h2>
<p>Compare the full <em>vs</em> the intercept-only models (compound test)! <span class="math display">\[H_0: \tau_{KO}=\tau_{P2}=\tau_{P6}=\tau_{P10}=\tau_{P28}=0 \, \ (5 \text{ df})\]</span></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1"></a>addReduced <span class="ot">&lt;-</span> <span class="fu">lm</span>(expression <span class="sc">~</span> <span class="dv">1</span>, <span class="at">data =</span> bb1gene)</span>
<span id="cb77-2"><a href="#cb77-2"></a><span class="fu">anova</span>(addReduced,addFull)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Variance Table

Model 1: expression ~ 1
Model 2: expression ~ genotype + dev_stage
  Res.Df    RSS Df Sum of Sq      F    Pr(&gt;F)    
1     38 73.566                                  
2     33  2.474  5    71.092 189.66 &lt; 2.2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
</section>
<section id="goodness-of-fit-also-given-in-output-of-lm" class="slide level2">
<h2>Goodness of fit also given in output of <code>lm</code></h2>
<div class="cell" data-highlight.output="21" data-output.lines="[2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22]">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1"></a><span class="fu">summary</span>(addFull)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = expression ~ genotype + dev_stage, data = bb1gene)

Residuals:
     Min       1Q   Median       3Q      Max 
-0.80137 -0.12454 -0.03212  0.17038  0.50036 

Coefficients:
              Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)    5.52734    0.11012  50.192  &lt; 2e-16 ***
genotypeNrlKO  0.03167    0.08785   0.361   0.7207    
dev_stageP2    0.30152    0.14185   2.126   0.0411 *  
dev_stageP6    0.24102    0.14185   1.699   0.0987 .  
dev_stageP10   0.83185    0.14185   5.864 1.44e-06 ***
dev_stageP28   3.63011    0.14185  25.592  &lt; 2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.2738 on 33 degrees of freedom
Multiple R-squared:  0.9664,    Adjusted R-squared:  0.9613 
F-statistic: 189.7 on 5 and 33 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
</section>
<section id="summary-so-far" class="slide level2">
<h2>Summary so far</h2>
<ul>
<li><p><strong><em>t</em>-tests</strong> can be used to test the equality of <strong>2</strong> population means</p></li>
<li><p><strong>ANOVA</strong> can be used to test the equality of <strong>more than 2</strong> population means simultaneously (main effects)</p></li>
<li><p><strong>Linear regression</strong> provides a general framework for modelling the relationship between a response and different type of explanatory variables</p>
<ul>
<li><p><em>t</em>-tests are used to test the significance of <strong>simple effects</strong> (<em>individual</em> coefficients)</p></li>
<li><p><em>F</em>-tests are used to test the significance of <strong>main effects</strong> (<em>simultaneously</em> multiple coefficients)</p></li>
<li><p><em>F</em>-tests are used to compare nested models (<strong>overall</strong> effects or <strong>goodness of fit</strong>)</p></li>
</ul></li>
<li><p>Next up: continuous explanatory variables! Multiple genes!</p></li>
</ul>

<img src="https://github.com/STAT540-UBC/stat540-ubc.github.io/raw/main/images/stat540-logo-s.png" class="slide-logo r-stretch"><div class="footer footer-default">

</div>
</section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="lect07-linearModels_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="lect07-linearModels_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="lect07-linearModels_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="lect07-linearModels_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="lect07-linearModels_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="lect07-linearModels_files/libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="lect07-linearModels_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="lect07-linearModels_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="lect07-linearModels_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="lect07-linearModels_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="lect07-linearModels_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1600,

        height: 900,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>