---
title: "Gene set analysis"
author: "Yongjin Park"
classoption: "aspectratio=169"
---

```{r setup, include=FALSE}
library(tidyverse)
library(data.table)
library(patchwork)
source("Util.R")
source("Setup.R")
fig.dir <- "Fig/"
setup.env(fig.dir, .echo=TRUE)
dir.create("Data", showWarnings=FALSE)
```

## What is Gene Set Analysis?

:::::: {.columns}
::: {.column width=.45}

### (Discrete) Gene Set Analysis

**Input:**

1. A dictionary of gene sets that map genes to sets (gene-to-set mapping)
	
2. A list of **top** genes identified in our own study (after FDR control)

:::
::: {.column width=.45}

### (Rank-based) Gene Set Enrichment

**Input:**

1. A dictionary of gene sets that map genes to sets
	
2. A **full** list of gene-level **scores** (e.g., p-values)

:::
::::::

**Output:**

A table of scores for all the gene sets in the dictionary.

## What have we learned from our differential expression analysis?

```{r results="asis", out.width=".7\\linewidth", echo = FALSE}
knitr::include_graphics("./Vis/geneset_1.pdf")
```

\hfill $\to$ **coloured:** some pathway of interest

## Let's drop the edges in biological networks

Prior Knowledge of biological pathways define a set of genes (coloured)

```{r results="asis", out.width=".7\\linewidth", echo = FALSE}
knitr::include_graphics("./Vis/geneset_2.pdf")
```

## Does this pathway overlap with our DEG list?

```{r results="asis", out.width=".7\\linewidth", echo = FALSE}
knitr::include_graphics("./Vis/geneset_3.pdf")
```

## Say that we found $q$ genes are overlapping with this pathway

```{r results="asis", out.width=".7\\linewidth", echo = FALSE}
knitr::include_graphics("./Vis/geneset_4.pdf")
```

## Gene Set Analysis testing over-representation of DEGs

:::::: {.columns}
::: {.column width=.65}

```{r results="asis", out.width=".95\\linewidth", echo = FALSE}
knitr::include_graphics("./Vis/geneset_5.pdf")
```

:::
::: {.column width=.35}

### What are the numbers to count?

* $N$: \# genes in this universe

* $m$: \# genes in this set

* $n$: \# genes *not* in this gene set

* $k$: \# DEGs in our analysis

* $q$: \# DEGs (of $k$) overlapping with the set of $m$ genes

:::
::::::

## Is this overlap of $q=6$ of $k=8$ genes significant?

:::::: {.columns}
::: {.column width=.65}

```{r results="asis", out.width=".95\\linewidth", echo = FALSE}
knitr::include_graphics("./Vis/geneset_5.pdf")
```

:::
::: {.column width=.35}

* $N = 24$ # a total of 24 genes

* $m = 9$ genes in this set

* $n = N - m = 15$

* $k = 8$ DEGs

* $q=6$ out of $k=8$ overlap

:::
::::::

## Is this overlap of $q=6$ of $k=8$ genes significant?

:::::: {.columns}
::: {.column width=.65}

```{r results="asis", out.width=".95\\linewidth", echo = FALSE}
knitr::include_graphics("./Vis/geneset_5.pdf")
```

:::
::: {.column width=.35}

Questions:

* Is it meaningful enough to report?

* Is it surprising enough that we recapitulated 6/9 (~`r round(6/9*100)` %)?

* What is the null distribution?

* What is the generative/simulation scheme?

:::
::::::

## How do we find $q$ out of $k$ DEGs overlapping with a gene set of $m$ genes?

:::::: {.columns}
::: {.column width=.55}

```{r results="asis", out.width=".95\\linewidth", echo = FALSE}
knitr::include_graphics("./Vis/geneset_5.pdf")
```

:::
::: {.column width=.45}

### Under the null of hypergeometric distribution

\Large

1. Sample $k$ DEGs out of $N$ genes

2. Of these $k$ genes, $q$ overlap with a gene set consisting of $m$ genes

3. The rest $k-q$ genes overlap with genes outside of the gene set $N-m$

:::
::::::

## *Binomial coefficient*: Let's see if we can estimate the null distribution by counting

\Large

* How many all possible ways to select $k=8$ out of $N=24$ genes, ignoring the order of $k$ selected genes and $N-k$ *not* selected genes?

* We can think of this as three steps: (1) enumerating $N$ genes, (2) partition them into the first $k$ genes and the rest, (3) ignore the order within each partition.

$${24 \choose 8} = \frac{
\onslide<1->{\color{teal}
	\textsf{\{all possible ways to enumerate }24 \textsf{ genes}\}
}
	}{
\onslide<2->{\color{magenta}
		\{ \textsf{enumerating }8 \textsf{ genes} \}
}
\onslide<3->{\color{blue}
		\{ \textsf{enumerating }16 \textsf{ genes} \}
}
}
= \frac{
\onslide<1->{\color{teal}
24!
}
}{
\onslide<2->{\color{magenta}
8! 
}
\onslide<3->{\color{blue}
16!
}
}$$

\vfill

\normalsize

Okay, using binomial coefficient, let's count the numbers.

## Hypergeometric distribution

\Large
\begin{itemize}
\item<1-> What is the probability of uniformly selecting $k$ DEGs out of $N$ genes? ${N \choose k}^{-1}$
\item<2-> How many possible ways of finding $q$ DEGs overlapping with $m$ genes in the gene set? ${m \choose q}$
\item<3-> How many possible ways of finding the rest $k-q$ DEGs overlapping with ($N-m=n$) genes in the gene set? ${N-m \choose k-q}$
\end{itemize}

\normalsize

## Hypergeometric distribution

\begin{eqnarray*}
  P_{0}(q|N,m,k) &=&
                     \sum_{\begin{array}{l}
                     {\color{blue}                             
	\textsf{\# ways to select }q\textsf{ out of }m } \\
                     {\color{teal}
        \textsf{\# ways to select }(k-q)\textsf{ out of }N-m }
        \end{array}}
{\color{magenta}
\begin{array}{l}
  \textsf{the probability} \\
  \textsf{of choosing a set size }k \\
  \textsf{out of total }N
\end{array}} \\
&=&
\onslide<2->{
\underbrace{\color{blue} {m \choose q}}_{
\begin{array}{l}
\textsf{\# ways to choose } \\
q\, \textsf{ overlap out of }m
\end{array} }
}
\onslide<3->{
    \times 
\underbrace{\color{teal} {N-m \choose k - q}}_{
\begin{array}{l}
\textsf{\# ways to choose } \\
(k-q) \textsf{ out of }N-m
\end{array} }
}
\onslide<4->{
\times
{\color{magenta} {N \choose k}^{-1} }
}
\end{eqnarray*}


## What is the probability of $k$ overlapping DEGs?

:::::: {.columns}
::: {.column width=.45}

### Hypergenometric PMF

\Large

$$p(x|N,m,k) = \frac{{m \choose x} {N - m \choose k-x}}{{N \choose k}}$$

:::
::: {.column width=.45}

### Hypergeometric CDF

\Large

$$p(q|N,m,k) = \sum_{x = 0}^{q} \frac{{m \choose x} {N - m \choose k-x}}{{N \choose k}}$$

:::
::::::

\normalsize

## Hypergeometric test for testing significant overlap

\Large

$$H_{0}:\, x \le q \quad \textsf{vs.} \quad H_{1}:\, x > q$$

\normalsize

We may observe overlap $q$ genes by random sampling of $k$ genes **without** replacement.

Therefore, we can calculate the p-value:

\only<1>{
What is the definition of p-value?
}

\only<2->{

\Large

$$P(x > q|n, m, k) = 1 - \sum_{x = 0}^{q} \frac{{m \choose x } {n \choose k-x}}{{n+m \choose k}}$$

\normalsize

}

```{r size="large"}
phyper(q=6, m=9, n=15, k=8, lower.tail=FALSE)
```


## How significant is $q$ overlap in our discovery?

:::::: {.columns}
::: {.column width=.33}

$m=9$, $n=15$, and $k=8$,

```{r fig.width=1.8, fig.height=2.7, echo = FALSE, onslide.plot="1-"}
.m <- 9
.n <- 15
.k <- 8
.dt <- data.table(pp = dhyper(0:8, m=.m, n=.n, k=.k),
                  cc = phyper(0:8, m=.m, n=.n, k=.k),
                  x = 0:8,
                  .null = "0_null")

plot.hyper.cutoff <- function(.dt) {
    
    .y.axis <- scale_y_continuous(breaks=c(0, .3, .6, .9))
    .x.axis <- scale_x_continuous(breaks=0:8)

    p1 <- .gg.plot(.dt, aes(x, pp, fill=.null)) +
        geom_bar(stat="identity") +
        .y.axis + .x.axis + xlab("") +
        scale_fill_manual(values=c("gray","red"), guide="none") +
        ylab("probability")

    p2 <- .gg.plot(.dt, aes(x,cc, fill=.null)) +
        geom_bar(stat="identity") +
        .y.axis + .x.axis + xlab("") +
        scale_fill_manual(values=c("gray","red"), guide="none") +
        ylab("cumulative\nprobability")

    .dt.sig <- .dt %>% filter(.null != "0_null")

    p3 <-
        .gg.plot(.dt, aes(x, 1-cc, fill=.null)) +
        geom_line() +
        geom_point(pch=21) +
        scale_y_sqrt(breaks=c(1,.5, 0.05, 1e-2, 0)) + .x.axis +
        scale_fill_manual(values=c("gray","red"), guide="none") +
        ylab("1 - cumulative\nprobability")

    if(nrow(.dt.sig) > 0){
        p3 <- p3 + geom_text(aes(y=.1 + 1 - cc, label = num.sci(1 - cc)),
                             data = .dt.sig,
                             vjust = 0, hjust = 0,
                             size = 2, angle=90)
    }

    wrap_plots(p1,p2,p3,ncol=1, heights=c(1,1,2))
}

plot.hyper.cutoff(.dt)
```

:::
::: {.column width=.33}

\onslide<2->{If $q=3$:}

```{r fig.width=1.8, fig.height=2.7, echo = FALSE, onslide.plot="2-"}
.dt %>% mutate(.null = if_else(x > 3, "1_alt", "0_null")) %>% 
plot.hyper.cutoff()
```

:::
::: {.column width=.33}

\onslide<3->{If $q=6$:}

```{r fig.width=1.8, fig.height=2.7, echo = FALSE, onslide.plot="3-"}
.dt %>% mutate(.null = if_else(x > 6, "1_alt", "0_null")) %>% 
plot.hyper.cutoff()
```

:::
::::::

## Summary of the gene set analysis by hypergeometric test

:::::: {.columns}
::: {.column width=.5}

**When does it work?**

- A routine to construct a set of differentially expressed genes by
  handling multiple hypothesis testing

- Gene sets are of similar sizes and *nearly* disjoint/independent from one another

- Genes are *nearly* independent (there is no overwhelmingly favourite genes)

:::
::: {.column width=.5}

**When does it not work?**

- **Don't** have a good way to make a set:

  - Our discovery data may lack statistical power, i.e., no (or a few)
    significant genes left after multiple hypothesis correction

- There is a hidden factor that can affect two steps: (1) gene set
  selection (annotations/knowledge) and (2) differential expression
  calling

:::
::::::


## Rank-based Gene Set Enrichment Analysis method (Subramanian _et al._ 2005)

\Large

- A collection of gene-sets: $\mathcal{C}_{1}, \ldots, \mathcal{C}_{K}$

- A vector of gene-level scores ($G$ genes): $z_{1}, \ldots, z_{G}$

- Each $z_{g}$ could come from differential expression analysis

\normalsize

## GSEA algorithm

\Large

- For each $k$

    - Compute a set-level score $S_{k}(\mathbf{z},\mathcal{C}_{k})$

    - E.g., Kolmogorov-Smirnov statistic comparing
	
$$\{z_{g}:\,g \in \mathcal{C}_{k}\} \, \textsf{vs.} \, \{z_{g}:\,g \notin \mathcal{C}_{k}\}$$

\normalsize


## Gene Set Enrichment Analysis method (Subramanian _et al._ 2005)

- Construct null distribution of $S_{1}, \ldots, S_{K}$ by sample label (case-control) or gene-to-set membership permutation

- Using null distribution by permutation, estimate p-values and false discovery rates

- If we knew null distribution, we would not need expensive permutations.

:::::: {.columns}
::: {.column width=.45}

**Good**:

- No cutoff/assumptions needed to estimate null distribution

- Aggregate scores across many genes! (boost the power)

:::
::: {.column width=.45}

**Bad**:

- What is an appropriate statistic?

- What should be permuted? For how long?

:::
::::::


## Simulation (Efron and Tibshirani 2007)

1. Generate basal gene expression 

$$X_{i,g} \sim \mathcal{N}\!\left(0,1\right)$$

2. Sample case vs. control membership (the rows of $X$) uniformly at random

3. Sample membership gene to gene set uniformly at random 

4. For the first gene set, select a certain fraction of genes to perturb

5. For the selected genes $g^{*}$, add some $\Delta$ value to $X_{i,g^*}$ if the sample $i$ belongs to the control group


## Let's simulate some gene set data (Efron and Tibshirani 2007)

:::::: {.columns}
::: {.column width=.55}

```{r}
simulate.data <-
    function(G = 1000,        # genes
             K = 150,         # gene sets
             n.samp = 100,    # sample size
             delta = .4,      # perturbation
             p.perturb = 1) { # Pr of petrub
  case.control <- sample(0:1, n.samp, TRUE)
  S <- sample(K, G, TRUE)       # gene sets
  X <- .rnorm(n.samp, G)        # All the other genes
  ## Perturbation of the first gene set
  .genes.1 <- which(S == 1)
  n1 <- length(.genes.1)
  n.perturb <- max(floor(n1 * p.perturb),1)
  .genes.1 <- sample(.genes.1, n.perturb)
  .case <- case.control == 1
  X[.case, .genes.1] <- X[.case, .genes.1] + delta
  require(Matrix)
  .membership <- sparseMatrix(j=1:G, i=S, x=rep(1,G))
  list(X=X, S=.membership, Y=case.control)
}
```

:::
::: {.column width=.35}

```{r}
set.seed(1)
dat <- simulate.data()
```

Let's run t-test for each gene:
```{r}
run.t.test <- function(X, Y){
  .case <- Y == 1
  .ctrl <- Y == 0
  .fun <- function(x){
    t.test(x[.case],
           x[.ctrl])$statistic
  }
  apply(X, 2, .fun)
}
```

:::
::::::


## The goal is to come up with a representative score for all the genes within each set


```{r echo=FALSE, fig.width=6, fig.height=2}
zz <-
    data.frame(z = run.t.test(dat$X, dat$Y)) %>%
    mutate(Var2 = 1:n()) %>%
    left_join(reshape2::melt(as.matrix(dat$S), value.name="s"))

.show <- zz %>% filter(Var1 <= 10) %>%
    mutate(s = if_else(s == 0, "outside", "inside"))

.gg.plot(.show, aes(s, `z`, fill = s)) +
    facet_grid(. ~ Var1) +
    geom_violin(scale="width", size=.1) +
    geom_boxplot(width=.2, fill="white", outlier.stroke=0, outlier.size=.2) +
    xlab("first 10 gene sets") +
    ylab("t-test statistics") +
    theme(axis.text.x = element_blank()) +
    scale_fill_brewer("genes", palette="Paired")
```

## What will be a proper gene set score?

\Large

Can we simply aggregate gene-level z-scores (or t-statistics) within each set?

::: {.block}

### Irizarry _et al._ (2009), using Stouffer Z-score

$$S_{k} = \sum_{g \in \mathcal{C}_{k}} z_{g} /  \sqrt{|\mathcal{C}_{k}|} \sim \mathcal{N}\!\left(0, 1\right)$$
if $Z_{g} \sim \mathcal{N}\!\left(0,1\right),\,\forall g$ 

:::

## Aggregating z-scores within a set to have one number for the set

:::::: {.columns}
::: {.column width=.4}

```{r}
geneset.score <- function(X, S, Y) {
    z.genes <- run.t.test(X, Y)
    n.sets <- apply(S, 1, sum)
    z.sets <- (S %*% z.genes /
               sqrt(n.sets))
}
```

:::
::: {.column width=.45}


```{r}
z.sets <- geneset.score(dat$X, dat$S, dat$Y)                        
```

```{r include=FALSE}
.qq.norm <- function(zz, ylab = "z-scores"){
    .qq <- qqnorm(as.numeric(zz), plot.it=FALSE)
    .dt <- setDT(.qq)
    .dt.sig <- .dt[abs(y) > 2 & abs(y) > 1.25*abs(x)]
    .gg.plot(.dt, aes(x,y)) +
        geom_point(stroke=0, size=.7) +
        geom_point(data=.dt.sig, size=2, pch=21, colour="red") +
        geom_abline(slope=1, colour=2) +
        xlab("z ~ N(0,1)") +
        ylab(ylab)
}
```


```{r fig.width=2, fig.height=1.25, echo=FALSE, only.plot="2"}
.qq.norm(z.sets, "set-level\n z-scores")
```

```{r include=FALSE}
.pv.plot <- function(zz) {
    .dt <- data.table(z = as.numeric(zz)) %>%
        mutate(i = 1:n()) %>%
        mutate(pv = 2 * pnorm(abs(z), lower.tail=FALSE)) %>%
        mutate(l10.pv = -log10(pv))
    .dt.sig <- .dt[pv < 1e-2]
    .gg.plot(.dt, aes(as.factor(i),l10.pv)) +
        geom_point(stroke=0, size=.7) +
        geom_point(data=.dt.sig, size=2, pch=21, colour="red") +
        theme(axis.text.x = element_blank()) +
        theme(axis.ticks.x = element_blank()) +
        geom_hline(yintercept = -log10(0.01), colour="red", lty=2) +
        xlab("gene sets") +
        scale_y_continuous("p-val",
                           labels=function(x) num.sci(10^(-x)))
}
```

```{r fig.width=2, fig.height=1.25, echo=FALSE, only.plot="3"}
.pv.plot(z.sets)
```

:::
::::::

## Constructing the null distribution by gene permutation

What if we don't know the distribution of set-wise scores?

:::::: {.columns}
::: {.column width=.55}

```{r size="large"}
S.perm <- t(apply(dat$S, 1, sample))
z.perm <- geneset.score(dat$X, S.perm, dat$Y)
```

```{r fig.width=3.2, fig.height=2, echo=FALSE, onslide.plot="2"}
.qq.norm(z.perm, "permuted\nz-scores") | .pv.plot(z.perm)
```

:::
::: {.column width=.45}

How it goes:

* Repeatedly permuting gene set membership matrix while
  preserving the number of genes within each set

* Compute set-level z-scores (or a similar kind) and construct null
  distribution

* Calculate p-values by counting the frequency of observed $S_{k}^{*} > S_{k}^{\textsf{perm}}$

:::
::::::

## Constructing the null distribution by sample permutation

:::::: {.columns}
::: {.column width=.55}


```{r}
Y.perm <- sample(dat$Y)
z.perm <- geneset.score(dat$X, S.perm, Y.perm)
```

```{r fig.width=3.2, fig.height=2, echo=FALSE}
.qq.norm(z.perm, "permuted\nz-scores") | .pv.plot(z.perm)
```

:::
::: {.column width=.45}

* Repeat the permutation of case-control labels while
  preserving the same number of cases and controls

* Compute set-level z-scores (or a similar kind) and construct null distribution

* Calculate p-values by counting the frequency of observed $S_{k}^{*} > S_{k}^{\textsf{perm}}$

:::
::::::

## There is a smart way to permute genes/samples

