---
title: "GWAS: population structure"
author: |
    | Yongjin Park
    | University of British Columbia
date: "`r format(Sys.time(), '%d %B, %Y')`"
classoption: "aspectratio=169"
output:
    beamer_presentation:
        colortheme: "orchid"
        keep_tex: true
        latex_engine: xelatex
        slide_level: 2
header-includes:
  - \usepackage{cancel}
  - \usepackage{booktabs}
  - \usepackage{longtable}
  - \usepackage{array}
  - \usepackage{multirow}
  - \usepackage{wrapfig}
  - \usepackage{float}
  - \usepackage{colortbl}
  - \usepackage{pdflscape}
  - \usepackage{tabu}
  - \usepackage{threeparttable}
  - \usepackage{threeparttablex}
  - \usepackage[normalem]{ulem}
  - \usepackage{makecell}
  - \usepackage{xcolor}
  - \AtBeginSection[]{\begin{frame}\frametitle{Today's lecture}{\Large\tableofcontents[currentsection]}\end{frame}}
  - |
    \makeatletter
    \def\ps@titlepage{%
      \setbeamertemplate{footline}{}
    }
    \addtobeamertemplate{title page}{\thispagestyle{titlepage}}{}
    \makeatother
    \include{toc}
---

```{r setup, include=FALSE}
library(tidyverse)
library(data.table)
library(patchwork)
source("Util.R")
source("Setup.R")
fig.dir <- "Fig/"
setup.env(fig.dir)
dir.create("Data", showWarnings=FALSE)
theme_set(theme_classic())
```

```{r include = FALSE}
library(bigsnpr)
dir.create("Data/genotype/", recursive=TRUE, showWarnings=FALSE)
.bed.file <- "Data/genotype/1000G_phase3_common_norel.bed"
if(!file.exists(.bed.file)){
    download_1000G("Data/genotype/")
}
.bk.file <- "Data/genotype/1000G_phase3_common_norel.rds"
if(!file.exists(.bk.file)){
    BED <- snp_readBed(.bed.file)
}
X <- snp_attach(.bk.file)
```

```{r download_1kg_annotations, include=FALSE}
kg.url <-
  paste0("ftp://ftp.1000genomes.ebi.ac.uk/",
         "vol1/ftp/technical/working/",
         "20130606_sample_info/",
         "20130606_sample_info.xlsx")

kg.sample.file <- "Data/genotype/1KG.sample.xlsx"
if.needed(kg.sample.file,
{
    download.file(kg.url, destfile = kg.sample.file)
})
sample.info <- readxl::read_xlsx(kg.sample.file)
```

```{r parse_pop_information, include=FALSE}
cat.dt <- 
    data.table(Population =
                   c("GBR", "FIN", "IBS", "CEU", "TSI", 
                     "CHS", "CDX", "KHV", "CHB", "JPT", 
                     "PJL", "STU", "ITU", "GIH", "BEB",
                     "PUR", "CLM", "PEL", "MXL",
                     "MSL", "YRI", "LWK", "ASW", "ACB", "GWD", "ESN"),
               pop =
                   c(rep("eur", 5),
                     rep("asn", 5),
                     rep("ind", 5),
                     rep("lax", 4),
                     rep("afr", 7)))
sample.info <-
    sample.info %>% left_join(cat.dt)
```

## 1000 Genomes project data

:::::: {.columns}
::: {.column width=.5}

1KG contains whole genome sequencing data of `r num.int(nrow(X$genotypes))` individuals sampled from  `r length(unique(sample.info$Population))` groups based on the origins and geographical locations (as of 2013 phase3).

:::
::: {.column width=.5}

```{r fig.width=2, fig.height=2}
xx <- X$genotypes[1:10, 1:10]
.matshow(xx, .lab=3)
```

\tiny

First 10 individuals and 10 variants

:::
::::::


## Variant-level variation across individuals

Using the minor allele frequency (MAF), let $f_{j}$ be a minor allele frequency (MAF) of a variant $j$. In Binomial distribution,

\vfill

\large

\begin{itemize}
\item<1-> What is the mean of this variant?
\onslide<2->{$$\hat{\mathbb{E}}\!\left[X_{j}\right] = 2 f_{j}$$}
\item<3-> What is the variance of this variant?
\onslide<4>{$$\hat{\mathbb{V}}\!\left[X_{j}\right] = 2 f_{j} (1 - f_{j})$$}
\end{itemize}

\vfill

\small

*Remark*: Technically, the dosage (0,1,2) does not follow binomial distribution. Why? The underlying data generation process involves haplotypes (separating the maternal and paternal 0/1 counts) and dependency along the genomic axis.

## Variant-level variation across individuals

:::::: {.columns}
::: {.column width=.5}

We can easily calculate MAF using `bigsnpr`

```{r echo=TRUE, size="large"}
maf <- snp_MAF(X$genotypes, ncores=8)
```

$${}$$

What is your interpretation?

:::
::: {.column width=.5}

```{r fig.width=2.5, fig.height=2}
ggplot(data.table(MAF=maf), aes(MAF)) +
    geom_density(aes(y = ..count..), fill="gray")
```

:::
::::::


## Much of human genetics problems centre on two covariance matrices

For a standardized $n \times p$ genotype matrix $X$ ($n$: \#individuals, $p$: \#SNPs),

:::::: {.columns}
::: {.column width=.48}

::: {.block}
### 1. Genetic relatedness matrix (GRM)

individual by individual, $n \times n$ matrix

$$K \approx XX^{\top}/n$$

The matrix $K$ captures population structure/correlation across different individuals.

- Kinship matrix; population admixture

- Human migration history

:::

:::
::: {.column width=.48}

::: {.block}
### 2. Linkage disequilibrium (LD)

SNP by SNP, $p \times p$ matrix 

$$R \approx X^{\top}X/n$$

The matrix $R$ captures localized correlation patterns along the genomic axis within a chromosome.

- LD matrix

- The results of many, many recombination events

:::

:::
::::::


## SVD captures principal components

\Large
\onslide<1->{
$$X = U D V^{\top}$$
}

\large
\onslide<2->{
What is this?
$$\frac{1}{n} X^{\top}X = \frac{1}{n} V D U^{\top} U D V^{\top} = 
\underset{\textsf{\color{magenta} variant x variant}}{\frac{1}{n} V D^{2} V^{\top}}$$
}

\onslide<3->{
What is this?
$$\frac{1}{n} XX^{\top} = \frac{1}{n} U D V^{\top} V D U^{\top} = 
\underset{\textsf{\color{teal} sample x sample}}{\frac{1}{n} U D^{2} U^{\top}}$$
}

## Let's take top 1000 most frequent variants

```{r}
idx <-
  order(maf, decreasing=TRUE) %>%
  head(1000)

xx <- X$genotypes[,idx]

.svd <- rsvd::rsvd(xx,5)
u.dt <- data.table(.svd$u) %>%
    cbind(X$fam) %>%
    merge(sample.info,
          by.x = "sample.ID",
          by.y = "Sample")
```

PCA already teaches us something interesting...

:::::: {.columns}
::: {.column width=.5}

```{r echo=FALSE, fig.width=3, fig.height=2, onslide.plot="1-"}
.gg.plot(u.dt, aes(V1, V2, colour=pop)) +
    geom_point(stroke=0, alpha=.5) + xlab("PC1") + ylab("PC2") +
    scale_colour_brewer(palette = "Paired")
```

:::
::: {.column width=.5}

```{r echo=FALSE, fig.width=3, fig.height=2, onslide.plot=2}
.gg.plot(u.dt, aes(V2, V3, colour=pop)) +
    geom_point(stroke=0, alpha=.5) + xlab("PC2") + ylab("PC3") +
    scale_colour_brewer(palette = "Paired")
```

:::
::::::

## What about the 1000 least frequent variants?

```{r}
idx <- order(maf) %>% head(1000)
xx <- X$genotypes[,idx]

.svd <- rsvd::rsvd(xx,5)
u.dt <- data.table(.svd$u) %>%
    cbind(X$fam) %>%
    merge(sample.info,
          by.x = "sample.ID",
          by.y = "Sample")
```

:::::: {.columns}
::: {.column width=.5}

```{r echo=FALSE, fig.width=3, fig.height=2, onslide.plot="1-"}
.gg.plot(u.dt, aes(V1, V2, colour=pop)) +
    geom_point(stroke=0, alpha=.5) + xlab("PC1") + ylab("PC2") +
    scale_colour_brewer(palette = "Paired")
```

:::
::: {.column width=.5}

```{r echo=FALSE, fig.width=3, fig.height=2, onslide.plot=2}
.gg.plot(u.dt, aes(V2, V3, colour=pop)) +
    geom_point(stroke=0, alpha=.5) + xlab("PC2") + ylab("PC3") +
    scale_colour_brewer(palette = "Paired")
```

:::
::::::

## Why do we study population structures in human genetics?

\large

- If there is no mutation/variation, there is no genetic association. 

- Without knowing a macro-level dependency structures across cohorts, it is hard to dissect micro-level, perhaps disease-specific patterns.

- *Causal inference*: It also serves as a natural way to stratify/divide cohorts in a population genetics study to edify causal relationships that hold invariantly across multiple strata.

- *Precision health*: Characterization of population-invariant or specific variation is one of the first steps toward precision medicine.

\normalsize


## Population admixture learned from top 10k high MAF variants

```{r include = FALSE}
out.file <- "result_1kg_topic.RDS"
if.needed(out.file,
{
    library(torch)
    MY.DEV <- torch_device("cuda:0")
    torch_set_num_threads(16)
    idx <- order(maf) %>% tail(10000)
    xx <- torch_tensor(X$genotypes[,idx], device=MY.DEV)

    batch.size <- 128
    ntot <- nrow(xx)
    nbatch <- ceiling(ntot /batch.size)
    nn <- batch.size * nbatch

    source("torch_etm_train.R")
    llik.vec <- c()
    etm <- build.geno.etm(ncol(xx), ncol(xx), 9, d = c(128,128,16))
    etm$to(device=MY.DEV)
    opt <- optim_adam(etm$parameters, lr = 5e-3)
    for(tt in seq(0, 2000)){
        xx.t <- xx[sample(ntot, nn, TRUE), ]
        for(b in seq(0, nbatch-1)){
            .idx <- seq(1 + b * batch.size, (b+1) * batch.size)
            xx.b <- xx.t[.idx, ]
            opt$zero_grad()
            out <- etm(xx.b)
            kl <- kl.loss(out$z.mean, out$z.lnvar)
            llik <- etm.llik(xx.b, out$recon)
            loss <- torch_mean(kl - llik)
            loss$backward()
            opt$step();
            cat(tt, llik$mean()$item(), "\r")
            llik.vec <- c(llik.vec, llik$mean()$item())
        }
    }

    CPU <- torch_device("cpu")

    etm$eval()
    out <- etm(xx)
    Z <- as.matrix(out$z.mean$to(device=CPU))

    H <- 
        apply(Z, 1, function(zz) { exp(zz - max(zz)) }) %>%
        apply(2, function(pp) { pp/sum(pp) })

    W <- etm$dec$lbeta$to(device=CPU) %>% as.matrix()

    saveRDS(list(hidden=H, llik=llik.vec, weight=W),
            file=out.file)
})
etm.out <- readRDS(out.file)
```

A generative model:

:::::: {.columns}
::: {.column width=.6}

\small

- Sample each individual's topic proportion $H_{i}$
- Sample a topic membership for each variant $j$, say $Z_{ij}=k$ (could be implicitly handled)
- Genotype $X_{ij} | Z_{ij}=k$ $\sim$ topic-specific $\beta_{kj}$

:::
::: {.column width=.4}

\large
$$\mathcal{L} = \prod_{i=1}^{n}\prod_{j=1}^{\textsf{10k}} \left(\sum_{k=1}^{9} H_{ik} \beta_{kj}\right)^{X_{ij}}$$

:::
::::::

\vfill

```{r fig.width=5.5, fig.height=1.5}
colnames(etm.out$hidden) <- X$fam$sample.ID

.dt <- 
    reshape2::melt(etm.out$hidden) %>%
    as.data.table %>% 
    merge(sample.info,
          by.x = "Var2",
          by.y = "Sample")

.pop <-
    unique(sample.info %>% select(Population, pop)) %>%
    arrange(pop) %>%
    select(Population) %>%
    unlist

.dt[, Population := factor(`Population`, .pop)]

ggplot(.dt, aes(Var2, value, fill=as.factor(Var1))) +
    ggtitle("topic proportion H") +
    xlab(num.int(length(unique(.dt$Var2))) %&% " individuals") +
    ylab("topic proportion") +
    theme(title = element_text(size=8)) +
    theme(axis.title = element_text(size=4)) +
    theme(axis.text.y = element_text(size=2)) +
    theme(axis.text.x = element_blank()) +
    theme(axis.ticks.x = element_blank()) +
    theme(strip.text = element_text(size=4)) +
    theme(strip.background = element_blank()) +
    facet_grid(~Population, scales="free") +
    theme(panel.spacing = unit(.1, "lines")) +
    geom_bar(stat="identity") +
    scale_fill_brewer("", palette="Paired", guide="none", direction=-1)
```

\small

Related work: Pritchard, Stephens, Donnelly, *Genetics* (2000)


## What is the benefit of learning an admixture model in GWAS data?

\large

- Probabilistic interpretation of latent states

- Bayesian missing data imputation

- Potentially, a scalable approach for a biobank-scale data

\normalsize

