<!DOCTYPE html>
<html lang="en"><head>
<script src="lect11-epigenetics_files/libs/clipboard/clipboard.min.js"></script>
<script src="lect11-epigenetics_files/libs/quarto-html/tabby.min.js"></script>
<script src="lect11-epigenetics_files/libs/quarto-html/popper.min.js"></script>
<script src="lect11-epigenetics_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="lect11-epigenetics_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="lect11-epigenetics_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="lect11-epigenetics_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.2.313">

  <meta name="author" content="Keegan Korthauer">
  <meta name="dcterms.date" content="2023-02-16">
  <title>Statistical methods for epigenetics</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="lect11-epigenetics_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="lect11-epigenetics_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1.6em;
      vertical-align: middle;
    }
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="lect11-epigenetics_files/libs/revealjs/dist/theme/quarto.css" id="theme">
  <link href="lect11-epigenetics_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="lect11-epigenetics_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="lect11-epigenetics_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="lect11-epigenetics_files/libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="lect11-epigenetics_files/libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="lect11-epigenetics_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-captioned.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-captioned) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-captioned.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-captioned .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-captioned .callout-caption  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-captioned.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-captioned.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-caption {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-caption {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-caption {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-captioned .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-captioned .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-captioned) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-caption {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-caption {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-caption {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-caption {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-caption {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" data-background-color="#197aa0" data-background-image="https://github.com/STAT540-UBC/stat540-ubc.github.io/raw/main/images/stat540-logo-s.png" data-background-opacity="0.9" data-background-position="95% 90%" data-background-size="12%" class="quarto-title-block center">
  <h1 class="title">Statistical methods for epigenetics</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Keegan Korthauer 
</div>
</div>
</div>

  <p class="date">February 16, 2023</p>
</section>
<section id="learning-objectives-for-today" class="slide level2">
<h2>Learning objectives for today<sup>1</sup></h2>
<ul>
<li><p>Be familiar with the types of epigenetic measurements that involve <strong>“peak-calling”</strong>, and what a peak represents in each</p></li>
<li><p>Understand the main statistical test performed by the widely used peak-caller MACS2</p></li>
<li><p>Explain the similarities and differences (and understand their implications) between analysis of <strong>differential expression</strong> and <strong>differential methylation</strong>, with regard to:</p>
<ul>
<li><p>continuous models of microarray data vs count models of sequencing data (similarity)</p></li>
<li><p>methylation level represents a proportion vs expression level unbounded (difference)</p></li>
<li><p>grouping of sites into differentially methylated regions (difference)</p></li>
</ul></li>
</ul>
<aside><ol class="aside-footnotes"><li id="fn1"><p>We won’t have time to cover preprocessing / normalization steps</p></li></ol></aside></section>
<section id="epigenetics" class="slide level2">
<h2>Epigenetics</h2>

<img data-src="https://i1.abocdn.com/f8983e70d0e154ca2233646305eeafb7d3baa11d/resources/images/blog/epigenetics-schema.jpeg" class="r-stretch quarto-figure-center"><p class="caption"><a href="https://www.antibodies-online.com/areas/epigenetics/">Image source</a></p><p>Recall basics of DNA-binding and DNA methylation assays from <a href="https://github.com/STAT540-UBC/lectures/raw/main/lect02-technology.pdf">lecture 2</a></p>
</section>
<section id="open-chromatin-nucleosome-positioning" class="slide level2">
<h2>Open chromatin &amp; nucleosome positioning</h2>

<img data-src="img/atac.jpeg" class="r-stretch quarto-figure-center"><p class="caption"><a href="https://rockefelleruniversity.github.io/RU_ATAC_Workshop.html">Image source</a></p></section>
<section id="peak-calling" class="slide level2">
<h2>Peak calling</h2>
<ul>
<li><p>Sequencing after one of these selection/digestion techniques (IP, DNase, MNase, transposase) generally requires a peak-finding analysis</p></li>
<li><p>One of the most widely used methods is <a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2008-9-9-r137">MACS2</a>:</p></li>
</ul>

<img data-src="img/macs.png" class="r-stretch quarto-figure-center"><p class="caption">MACS2</p></section>
<section id="macs2-overview" class="slide level2">
<h2>MACS2 overview</h2>
<div class="columns">
<div class="column">
<ul>
<li>Shift reads to remove strand asymmetry: estimate <span class="math inline">\(d\)</span> from 1,000 regions</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://hbctraining.github.io/Intro-to-ChIPseq/img/plos_chipseq_arrow.png"></p>
<p></p><figcaption><a href="https://hbctraining.github.io/Intro-to-ChIPseq">images source</a></figcaption><p></p>
</figure>
</div>
</div><div class="column">
<ul>
<li>Let <span class="math inline">\(\lambda_{BG}\)</span> be the expected number of reads in each window</li>
<li>Let <span class="math inline">\(\lambda_{local} = max(\lambda_{BG}, \lambda_{1K}, \lambda_{5K}, \lambda_{10K})\)</span> be the expected number of reads in a window</li>
<li>P-value for <span class="math inline">\(k\)</span> observed reads = <span class="math inline">\(e^{-\lambda_{local}} \sum_{i=0}^k \frac{\lambda_{local}^i}{i!}\)</span></li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://hbctraining.github.io/Intro-to-ChIPseq/img/lambda.png"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="post-peak-calling" class="slide level2">
<h2>Post peak-calling</h2>
<p>Depending on the type of experiment and the hypotheses, next steps may include (but not limited to):</p>
<ul>
<li><a href="http://bioconductor.org/packages/release/bioc/vignettes/ChIPseeker/inst/doc/ChIPseeker.html">annotating peaks with genomic context</a></li>
<li><a href="https://www.bioconductor.org/packages/release/workflows/html/chipseqDB.html">differential binding analysis</a></li>
<li>enrichment analysis<sup>1</sup></li>
<li><a href="https://www.bioconductor.org/packages/release/workflows/html/generegulation.html">motif analysis</a></li>
<li>integration with other measurements (e.g.&nbsp;expression)</li>
</ul>
<aside><ol class="aside-footnotes"><li id="fn2"><p>we’ll discuss next lecture</p></li></ol></aside></section>
<section id="dna-methylation-5mc" class="slide level2">
<h2>DNA Methylation (5mC)</h2>
<ul>
<li>chemical modification to an individual DNA base (C = Cytosine), referred to as 5mC (5-methylcytosine)</li>
<li>doesn’t change coding sequence</li>
<li>modification can happen at any C, but in mammals occurs almost<sup>1</sup> exclusively at CpGs</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/5mC.jpg" style="height:3in"></p>
<p></p><figcaption><a href="https://www.epigentek.com">image source</a></figcaption><p></p>
</figure>
</div>
<aside><ol class="aside-footnotes"><li id="fn3"><p>other contexts observed in neuronal cells</p></li></ol></aside></section>
<section id="dna-methylation-at-cpgs" class="slide level2">
<h2>DNA Methylation at CpGs</h2>

<img data-src="img/5thbase.jpg" class="r-stretch quarto-figure-center"><p class="caption">~30 Million CpGs in human</p></section>
<section id="the-5th-base" class="slide level2">
<h2>The 5th base?</h2>
<ul>
<li>heritable upon cell division
<ul>
<li>5mC added to complementary strand by methyltransferases</li>
</ul></li>
<li>dynamic
<ul>
<li>changes during development and with age</li>
<li>can be altered by environmental exposure</li>
</ul></li>
</ul>

<img data-src="img/copymeth.jpg" class="r-stretch quarto-figure-center"><p class="caption"><a href="https://slideplayer.com/slide/10719924/">image source</a></p></section>
<section id="role-in-gene-regulation" class="slide level2">
<h2>Role in gene regulation</h2>

<img data-src="img/genereg.jpg" class="r-stretch quarto-figure-center"></section>
<section id="evolution-of-methylation-assays" class="slide level2">
<h2>Evolution of methylation assays</h2>

<img data-src="img/methassays.jpg" class="r-stretch quarto-figure-center"><p class="caption"><a href="https://doi.org/10.1038/nrg2732">Laird 2010</a></p></section>
<section id="distinguishing-methylated-and-unmethylated-sites" class="slide level2">
<h2>Distinguishing methylated and unmethylated sites</h2>
<div class="columns">
<div class="column">
<ol type="1">
<li>Methylated DNA immunoprecipitation (MeDIP)</li>
</ol>
<ul>
<li>treat fragmented DNA with antibody targeting 5mC</li>
<li>wash away unbound DNA</li>
<li>measure extent to which each cytosine is still present (methylated) with <strong>microarray or sequencing</strong></li>
</ul>
</div><div class="column">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/medip.jpg" style="height:7in"></p>
<p></p><figcaption><a href="https://en.wikipedia.org/wiki/Methylated_DNA_immunoprecipitation">image source</a></figcaption><p></p>
</figure>
</div>
</div>
</div>
</section>
<section id="distinguishing-methylated-and-unmethylated-sites-1" class="slide level2">
<h2>Distinguishing methylated and unmethylated sites</h2>
<div class="columns">
<div class="column">
<ol start="2" type="1">
<li>Bisulfite conversion</li>
</ol>
<ul>
<li>5mC <span class="math inline">\(\rightarrow\)</span> stays 5mC</li>
<li>C (not methylated) <span class="math inline">\(\rightarrow\)</span> converted to U</li>
<li>Sequencing:
<ul>
<li>5mC read as C</li>
<li>C read as T</li>
</ul></li>
<li>For each C measure fraction of signal corresponding to C vs T with <strong>microarray or sequencing</strong></li>
<li><em>Note that can’t distinguish between C <span class="math inline">\(\rightarrow\)</span> T mutation and unmethylated C</em></li>
</ul>
</div><div class="column">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/bisulfite-conversion.jpg"></p>
<p></p><figcaption><a href="https://www.diagenode.com/en/categories/bisulfite-conversion">image source</a></figcaption><p></p>
</figure>
</div>
</div>
</div>
</section>
<section id="analysis-of-medip" class="slide level2">
<h2>Analysis of MeDIP</h2>
<div class="columns">
<div class="column">
<ul>
<li>MeDIP targets <strong>only</strong> methylated DNA</li>
<li><strong>MeDIP-chip</strong>: much like gene expression microarrays</li>
<li><strong>MeDIP-seq</strong>: differential abundance of counts (similar to RNA-seq)
<ul>
<li>Complication: instead of genes or transcripts, need to define genomic “bins” to count reads over</li>
<li>One option is to use sliding windows: <a href="http://bioconductor.org/packages/MEDIPS/">MEDIPS R package</a> performs preprocessing and implements <code>edgeR</code> to test for differential abundance of each window</li>
</ul></li>
</ul>
</div><div class="column">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/medip.jpg" style="height:7in"></p>
<p></p><figcaption><a href="https://en.wikipedia.org/wiki/Methylated_DNA_immunoprecipitation">image source</a></figcaption><p></p>
</figure>
</div>
</div>
</div>
</section>
<section id="bisulfite-conversion-arrays-e.g.-infinum-450k-epic-850k" class="slide level2">
<h2>Bisulfite conversion arrays (e.g.&nbsp;Infinum 450K, Epic 850K)</h2>
<div class="columns">
<div class="column">
<ul>
<li>Most commonly used human methylation arrays use bisulfite conversion
<ul>
<li>450K / 850K refer to the number of sites probed</li>
</ul></li>
<li>Analysis has similarities to expression microarray analysis
<ul>
<li>Signal is <strong>continuous</strong> (intensity)</li>
<li>Moderate throughput: 1-16 samples per array</li>
</ul></li>
<li>Difference: focus on <strong>relative</strong> intensity of methylated vs unmethylated binding affinity</li>
</ul>
</div><div class="column">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/microarray.jpg"></p>
<p></p><figcaption><a href="https://en.wikipedia.org/wiki/Illumina_Methylation_Assay">image source</a></figcaption><p></p>
</figure>
</div>
</div>
</div>
</section>
<section id="beta-values-and-m-values" class="slide level2">
<h2>Beta values and M values</h2>
<p>Need some way to summarize the two relative measures for each probe:</p>
<ol type="1">
<li>methylated intensity</li>
<li>unmethylated intensity</li>
</ol>
</section>
<section id="beta-beta-values" class="slide level2">
<h2>Beta <span class="math inline">\((\beta)\)</span> values</h2>
<ul>
<li>Combine two measurements for each probe, and measure level of methylation: <span class="math display">\[\beta = \frac{\text{methylated intensity}}{(\text{methylated intensity} + \text{unmeth intensity})}\]</span></li>
<li>Range from 0 to 1</li>
<li>Easily interpretable as a methylation proportion</li>
<li>Best used for visualization</li>
</ul>
</section>
<section id="m-values" class="slide level2">
<h2>M-values</h2>
<ul>
<li>Transformed <span class="math inline">\(\beta\)</span> values: <span class="math display">\[M= log_2\big(\frac{\text{methylated intensity}}{\text{unmethylated intensity}}\big) = log_2(\beta/(1-\beta))\]</span></li>
<li>This type of transformation is referred to as the <em>logit</em>
<ul>
<li>maps a proportion onto real line</li>
</ul></li>
<li>alleviate heteroscedasticity of <span class="math inline">\(\beta\)</span> values (recall that Binomial(n,p) variance = np(1-p))</li>
</ul>

<img data-src="img/betavsmvals.png" class="r-stretch quarto-figure-center"><p class="caption"><a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-11-587">Du et al.&nbsp;(2010)</a></p></section>
<section id="methylation-array-analysis" class="slide level2">
<h2>Methylation array analysis</h2>
<ul>
<li>Analysis strategy: linear regression on M-values</li>
<li>R Packages:
<ul>
<li><a href="http://bioconductor.org/packages/minfi/">minfi</a> for preprocessing and normalization</li>
<li><a href="http://bioconductor.org/packages/limma/">limma</a> for individual cytosine analysis</li>
<li><a href="http://bioconductor.org/packages/bumphunter/">bumphunter</a> for region-level analysis</li>
<li>Workflow that combines these: <a href="https://bioconductor.org/packages/methylationArrayAnalysis/">methylationArrayAnalysis</a> - explore in Seminar 7</li>
</ul></li>
</ul>
</section>
<section id="example---dna-methylation-in-sorted-t-cells" class="slide level2">
<h2>Example - DNA Methylation in sorted T cells</h2>
<p>“Genome-wide DNA methylation analysis identifies hypomethylated genes regulated by FOXP3 in human regulatory T cells” from <a href="https://www.ncbi.nlm.nih.gov/pubmed/23974203">Zhang et al.&nbsp;(2013)</a></p>
<div class="columns">
<div class="column">
<blockquote>
<p>To explore the methylation landscape of nTreg, we analyzed genome-wide methylation in human naive nTreg (rTreg) and conventional naive CD4+ T cells (Naive)…. Activation changed the methylation status of …</p>
</blockquote>
</div><div class="column">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://www.ncbi.nlm.nih.gov/pmc/articles/instance/3798997/bin/2823f1.jpg" style="height:6.2in"></p>
<p></p><figcaption>Fig 1</figcaption><p></p>
</figure>
</div>
</div>
</div>
</section>
<section id="example---dna-methylation-in-sorted-t-cells-1" class="slide level2">
<h2>Example - DNA Methylation in sorted T cells</h2>
<p>Data from <a href="https://www.ncbi.nlm.nih.gov/pubmed/23974203">Zhang et al.&nbsp;(2013)</a></p>
<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE49667">GEO accession GSE49667</a></li>
<li>Here we use the Bioc objects downloaded from <a href="https://figshare.com/articles/dataset/methylAnalysisData_tar_gz/3471719">here</a></li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="co"># load libraries </span></span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="fu">library</span>(minfi)</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="fu">library</span>(limma)</span>
<span id="cb1-4"><a href="#cb1-4"></a></span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="co"># read in data</span></span>
<span id="cb1-6"><a href="#cb1-6"></a>mSet <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/mSet.rds"</span>)</span>
<span id="cb1-7"><a href="#cb1-7"></a>targets <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/targets.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>mSet</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: GenomicRatioSet 
dim: 439918 10 
metadata(0):
assays(2): M CN
rownames(439918): cg13869341 cg24669183 ... cg08265308 cg14273923
rowData names(0):
colnames(10): naive.1 rTreg.2 ... act_naive.9 act_rTreg.10
colData names(13): Sample_Name Sample_Well ... yMed predictedSex
Annotation
  array: IlluminaHumanMethylation450k
  annotation: ilmn12.hg19
Preprocessing
  Method: Raw (no normalization or bg correction)
  minfi version: 1.33.0
  Manifest version: 0.4.0</code></pre>
</div>
</div>
</section>
<section id="metadata-values" class="slide level2">
<h2>Metadata values</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="fu">head</span>(targets[<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Sample_Name Sample_Well Sample_Source Sample_Group Sample_Label
1           1          A1           M28        naive        naive
2           2          B1           M28        rTreg        rTreg
3           3          C1           M28    act_naive    act_naive
4           4          D1           M29        naive        naive
5           5          E1           M29    act_naive    act_naive
6           6          F1           M29    act_rTreg    act_rTreg</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="fu">table</span>(targets<span class="sc">$</span>Sample_Label)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
act_naive act_rTreg     naive     rTreg 
        3         2         3         2 </code></pre>
</div>
</div>
</section>
<section id="beta-values" class="slide level2">
<h2>Beta values</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a>bVals <span class="ot">&lt;-</span> <span class="fu">getBeta</span>(mSet)</span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="fu">dim</span>(bVals)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 439918     10</code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="fu">head</span>(bVals)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              naive.1    rTreg.2 act_naive.3    naive.4 act_naive.5 act_rTreg.6
cg13869341 0.84267937 0.85118462   0.8177504 0.82987650  0.81186174   0.8090798
cg24669183 0.81812908 0.82489238   0.8293297 0.75610281  0.81967323   0.8187838
cg15560884 0.77219626 0.74903910   0.7516263 0.77417882  0.77266205   0.7721528
cg01014490 0.08098986 0.06590459   0.0233755 0.04127262  0.04842397   0.0644404
cg17505339 0.89439216 0.93822870   0.9471357 0.90520570  0.92641305   0.9286016
cg11954957 0.74495496 0.79008516   0.7681146 0.84450764  0.75431167   0.8116911
             naive.7    rTreg.8 act_naive.9 act_rTreg.10
cg13869341 0.8891851 0.88537940  0.90916748   0.88334231
cg24669183 0.7903763 0.85304116  0.80930568   0.80979554
cg15560884 0.7658623 0.75909061  0.78099397   0.78569274
cg01014490 0.0245281 0.02832358  0.07740468   0.04640659
cg17505339 0.8889361 0.87205348  0.90099782   0.93508348
cg11954957 0.7832207 0.84929777  0.84719430   0.83350220</code></pre>
</div>
</div>
</section>
<section id="m-values-1" class="slide level2">
<h2>M values</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>mVals <span class="ot">&lt;-</span> <span class="fu">getM</span>(mSet)</span>
<span id="cb12-2"><a href="#cb12-2"></a><span class="fu">dim</span>(mVals)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 439918     10</code></pre>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="fu">head</span>(mVals)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             naive.1   rTreg.2 act_naive.3   naive.4 act_naive.5 act_rTreg.6
cg13869341  2.421276  2.515948    2.165745  2.286314    2.109441    2.083313
cg24669183  2.169414  2.235964    2.280734  1.632309    2.184435    2.175771
cg15560884  1.761176  1.577578    1.597503  1.777486    1.764999    1.760819
cg01014490 -3.504268 -3.825119   -5.384735 -4.537864   -4.296526   -3.859792
cg17505339  3.082191  3.924931    4.163206  3.255373    3.654134    3.701096
cg11954957  1.546401  1.912204    1.727910  2.441267    1.618331    2.107829
             naive.7   rTreg.8 act_naive.9 act_rTreg.10
cg13869341  3.004332  2.949430    3.323265     2.920691
cg24669183  1.914738  2.537203    2.085423     2.090007
cg15560884  1.709728  1.655782    1.834341     1.874285
cg01014490 -5.313593 -5.100400   -3.575205    -4.360973
cg17505339  3.000690  2.768876    3.185991     3.848438
cg11954957  1.853192  2.494570    2.470994     2.323683</code></pre>
</div>
</div>
</section>
<section id="comparing-beta-and-m-values" class="slide level2">
<h2>Comparing beta and M values</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a><span class="co"># graphics settings</span></span>
<span id="cb16-2"><a href="#cb16-2"></a>pal <span class="ot">&lt;-</span> <span class="fu">brewer.pal</span>(<span class="dv">8</span>,<span class="st">"Dark2"</span>)</span>
<span id="cb16-3"><a href="#cb16-3"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb16-4"><a href="#cb16-4"></a></span>
<span id="cb16-5"><a href="#cb16-5"></a><span class="co"># plot beta values</span></span>
<span id="cb16-6"><a href="#cb16-6"></a><span class="fu">densityPlot</span>(bVals, </span>
<span id="cb16-7"><a href="#cb16-7"></a>            <span class="at">sampGroups =</span> targets<span class="sc">$</span>Sample_Group, </span>
<span id="cb16-8"><a href="#cb16-8"></a>            <span class="at">main =</span> <span class="st">"Beta values"</span>, </span>
<span id="cb16-9"><a href="#cb16-9"></a>            <span class="at">legend =</span> <span class="cn">FALSE</span>, </span>
<span id="cb16-10"><a href="#cb16-10"></a>            <span class="at">xlab =</span> <span class="st">"Beta values"</span>)</span>
<span id="cb16-11"><a href="#cb16-11"></a><span class="fu">legend</span>(<span class="st">"top"</span>, </span>
<span id="cb16-12"><a href="#cb16-12"></a>       <span class="at">legend =</span> <span class="fu">levels</span>(<span class="fu">factor</span>(targets<span class="sc">$</span>Sample_Group)), </span>
<span id="cb16-13"><a href="#cb16-13"></a>       <span class="at">text.col =</span><span class="fu">brewer.pal</span>(<span class="dv">8</span>,<span class="st">"Dark2"</span>))</span>
<span id="cb16-14"><a href="#cb16-14"></a></span>
<span id="cb16-15"><a href="#cb16-15"></a><span class="co"># plot M values</span></span>
<span id="cb16-16"><a href="#cb16-16"></a><span class="fu">densityPlot</span>(mVals, </span>
<span id="cb16-17"><a href="#cb16-17"></a>            <span class="at">sampGroups =</span> targets<span class="sc">$</span>Sample_Group, </span>
<span id="cb16-18"><a href="#cb16-18"></a>            <span class="at">main =</span> <span class="st">"M-values"</span>, </span>
<span id="cb16-19"><a href="#cb16-19"></a>            <span class="at">legend =</span> <span class="cn">FALSE</span>, </span>
<span id="cb16-20"><a href="#cb16-20"></a>            <span class="at">xlab =</span> <span class="st">"M values"</span>)</span>
<span id="cb16-21"><a href="#cb16-21"></a><span class="fu">legend</span>(<span class="st">"topleft"</span>, </span>
<span id="cb16-22"><a href="#cb16-22"></a>       <span class="at">legend =</span> <span class="fu">levels</span>(<span class="fu">factor</span>(targets<span class="sc">$</span>Sample_Group)), </span>
<span id="cb16-23"><a href="#cb16-23"></a>       <span class="at">text.col =</span> <span class="fu">brewer.pal</span>(<span class="dv">8</span>,<span class="st">"Dark2"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="lect11-epigenetics_files/figure-revealjs/unnamed-chunk-7-1.png" width="960" class="r-stretch quarto-figure-center"></section>
<section id="comparing-beta-and-m-values-1" class="slide level2">
<h2>Comparing beta and M values</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>b <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>, <span class="at">by=</span><span class="fl">0.0001</span>)</span>
<span id="cb17-2"><a href="#cb17-2"></a>m <span class="ot">&lt;-</span> <span class="fu">log2</span>(b<span class="sc">/</span>(<span class="dv">1</span><span class="sc">-</span>b))</span>
<span id="cb17-3"><a href="#cb17-3"></a><span class="fu">plot</span>(b, m, <span class="at">type=</span><span class="st">"l"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="lect11-epigenetics_files/figure-revealjs/unnamed-chunk-8-1.png" width="672" class="r-stretch quarto-figure-center"><div class="fragment">
<div class="callout callout-tip callout-captioned callout-style-default">
<div class="callout-body">
<div class="callout-caption">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Question</strong></p>
</div>
<div class="callout-content">
<p>Why do we have beta values far from 0 or 1 if methylation is either present or not at each CpG?</p>
</div>
</div>
</div>
</div>
</section>
<section id="differentially-methylated-cytosines-dmc" class="slide level2">
<h2>Differentially methylated cytosines (DMC)</h2>
<p>Which cytosines/probes are differentially methylated between naive &amp; activated naive T cells?</p>
<div class="columns">
<div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="do">## this is the factor of interest</span></span>
<span id="cb18-2"><a href="#cb18-2"></a>cellType <span class="ot">&lt;-</span> <span class="fu">factor</span>(targets<span class="sc">$</span>Sample_Group)</span>
<span id="cb18-3"><a href="#cb18-3"></a></span>
<span id="cb18-4"><a href="#cb18-4"></a><span class="do">## this is the individual effect that we need to account for</span></span>
<span id="cb18-5"><a href="#cb18-5"></a>individual <span class="ot">&lt;-</span> <span class="fu">factor</span>(targets<span class="sc">$</span>Sample_Source) </span>
<span id="cb18-6"><a href="#cb18-6"></a></span>
<span id="cb18-7"><a href="#cb18-7"></a><span class="do">## use the above to create a design matrix</span></span>
<span id="cb18-8"><a href="#cb18-8"></a>design <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>( <span class="sc">~</span> cellType <span class="sc">+</span> individual, </span>
<span id="cb18-9"><a href="#cb18-9"></a>                       <span class="at">data =</span> targets)</span>
<span id="cb18-10"><a href="#cb18-10"></a></span>
<span id="cb18-11"><a href="#cb18-11"></a><span class="co"># rename columns to be more convenient</span></span>
<span id="cb18-12"><a href="#cb18-12"></a><span class="fu">colnames</span>(design) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">levels</span>(cellType),</span>
<span id="cb18-13"><a href="#cb18-13"></a>                      <span class="fu">levels</span>(individual)[<span class="sc">-</span><span class="dv">1</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div><div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>design</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   act_naive act_rTreg naive rTreg M29 M30
1          1         0     1     0   0   0
2          1         0     0     1   0   0
3          1         0     0     0   0   0
4          1         0     1     0   1   0
5          1         0     0     0   1   0
6          1         1     0     0   1   0
7          1         0     1     0   0   1
8          1         0     0     1   0   1
9          1         0     0     0   0   1
10         1         1     0     0   0   1
attr(,"assign")
[1] 0 1 1 1 2 2
attr(,"contrasts")
attr(,"contrasts")$cellType
[1] "contr.treatment"

attr(,"contrasts")$individual
[1] "contr.treatment"</code></pre>
</div>
</div>
</div>
</div>
</section>
<section id="differential-methylation-limma-on-m-values" class="slide level2">
<h2>Differential methylation: limma on M-values</h2>
<p>Here, the reference is <code>act_naive</code> cells</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a><span class="do">## fit the linear model (one for each probe)</span></span>
<span id="cb21-2"><a href="#cb21-2"></a>mfit <span class="ot">&lt;-</span> <span class="fu">lmFit</span>(mVals, design)</span>
<span id="cb21-3"><a href="#cb21-3"></a>cfit <span class="ot">&lt;-</span> <span class="fu">eBayes</span>(mfit)</span>
<span id="cb21-4"><a href="#cb21-4"></a></span>
<span id="cb21-5"><a href="#cb21-5"></a><span class="do">## look at top DM CpGs between naive and ref (act_naive)</span></span>
<span id="cb21-6"><a href="#cb21-6"></a><span class="fu">topTable</span>(cfit, <span class="at">coef=</span><span class="st">"naive"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               logFC     AveExpr         t      P.Value   adj.P.Val        B
cg15459165 -3.892074 -1.47895212 -22.09541 1.993958e-08 0.002793543 6.810420
cg19704755 -4.139885  0.09100138 -20.35779 3.787335e-08 0.002793543 6.594795
cg13531460  3.484448 -0.63132888  20.23020 3.978221e-08 0.002793543 6.577256
cg17137500 -3.007052 -2.47635924 -20.21569 4.000601e-08 0.002793543 6.575245
cg17048073 -3.650550 -0.87247048 -20.04737 4.270889e-08 0.002793543 6.551674
cg03899643  4.603864  0.02760746  19.86354 4.589787e-08 0.002793543 6.525407
cg19827923  3.316270  0.15932891  19.82438 4.661126e-08 0.002793543 6.519739
cg26800893 -3.080184 -2.39958457 -19.60718 5.080116e-08 0.002793543 6.487839
cg14688905 -3.119261 -1.31909419 -18.94774 6.635412e-08 0.003243375 6.385901
cg21837189 -3.717982 -1.08764112 -18.00052 9.895908e-08 0.003769603 6.224845</code></pre>
</div>
</div>
</section>
<section id="summary-for-all-coefficients" class="slide level2">
<h2>Summary for all coefficients</h2>
<p>Significant tests (BH-adjusted p-value less than 0.05) of the null hypotheses: each coefficient equal to zero</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a><span class="do">## look at the numbers of DM CpGs at FDR &lt; 0.05</span></span>
<span id="cb23-2"><a href="#cb23-2"></a><span class="fu">summary</span>(<span class="fu">decideTests</span>(cfit))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       act_naive act_rTreg  naive  rTreg    M29    M30
Down      184747       919    400    827   1367   1245
NotSig     14791    438440 439291 438570 437499 437716
Up        240380       559    227    521   1052    957</code></pre>
</div>
</div>
<p>Note that the first column tests whether the reference group (<code>act_naive</code>) has an M value of zero (not usually of interest)</p>
</section>
<section id="interpretation-of-parameters-and-effect-sizes" class="slide level2">
<h2>Interpretation of parameters and effect sizes</h2>
<ul>
<li>Our previous model gives parameter estimates for differences in M values <span class="math inline">\(\alpha\)</span></li>
<li>To interpret on the scale of methylation proportions, need to back-transform:
<ul>
<li>Recall: <span class="math inline">\(M= log_2\big(\frac{\text{methylated intensity}}{\text{unmethylated intensity}}\big) = log_2(\beta/(1-\beta))\)</span></li>
</ul></li>
<li>Solving for beta gives <span class="math inline">\(\beta = \frac{2^M}{(1+2^M)}\)</span></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a>(M <span class="ot">&lt;-</span> cfit<span class="sc">$</span>coefficients[<span class="st">"cg15459165"</span>, <span class="st">"naive"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] -3.892074</code></pre>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a><span class="dv">2</span><span class="sc">^</span>M<span class="sc">/</span>(<span class="dv">1</span><span class="sc">+</span><span class="dv">2</span><span class="sc">^</span>M)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.06310445</code></pre>
</div>
</div>
</section>
<section id="summary-of-array-dmc-analysis" class="slide level2">
<h2>Summary of array DMC analysis</h2>
<ul>
<li><p><strong>Just like for limma analysis of gene expression microarray</strong>: Can use additive models, models with interactions, continuous and categorical variables</p>
<ul>
<li><p><strong>interpretation of hypothesis tests, contrasts</strong> with <code>makeContrasts</code> (for other comparisons not represented by coefficients in our design matrix): follows as before</p></li>
<li><p><strong>interpretation of parameters/effect sizes</strong>: keep in mind that M-values are <em>transformed</em> proportions</p></li>
</ul></li>
</ul>
<div class="fragment">
<ul>
<li>Annotation packages (e.g.&nbsp;<code>IlluminaHumanMethylation450kanno.ilmn12.hg19</code>) allow for adding information about each CpG
<ul>
<li>chromosome and position</li>
<li>strand</li>
<li>whether in CpG Island</li>
</ul></li>
</ul>
</div>
</section>
<section id="bisulfite-sequencing" class="slide level2">
<h2>Bisulfite sequencing</h2>
<ul>
<li><p>Sequencing of bisulfite-converted DNA</p></li>
<li><p><strong>WGBS</strong> = whole-genome bisulfite sequencing</p></li>
<li><p><strong>RRBS</strong> = reduced representation bisulfite sequening (targets similar regions to methylation arrays e.g.&nbsp;islands)</p></li>
<li><p>Pros: base-pair resolution; WGBS measures all CpGs</p></li>
<li><p>Con: expensive (WGBS cost is on the order of whole genome DNA sequencing)</p></li>
</ul>
</section>
<section id="bisulfite-sequencing-1" class="slide level2">
<h2>Bisulfite sequencing</h2>

<img data-src="img/wgbscounts2.jpg" class="r-stretch quarto-figure-center"></section>
<section id="analysis-of-bisulfite-sequencing" class="slide level2">
<h2>Analysis of bisulfite sequencing</h2>
<div class="columns">
<div class="column" style="width:75%;">
<ul>
<li><p>BS assays (e.g.&nbsp;WGBS, RRBS) yield base-level counts of methylated <span class="math inline">\((M \text{counts})\)</span> and ununmethylated <span class="math inline">\((U \text{counts})\)</span> reads</p></li>
<li><p>Differential methylation analysis quantifies significance of differences in proportion of methylated reads <span class="math inline">\((\frac{M}{M+U})\)</span></p></li>
<li><p>R Packages:</p>
<ul>
<li><a href="http://bioconductor.org/packages/bsseq/">bsseq</a>: “BSmooth” - smoothed t-test</li>
<li><a href="http://bioconductor.org/packages/dss/">dss</a>: approximate beta-binomial regression</li>
<li><a href="http://bioconductor.org/packages/dmrseq/">dmrseq</a>: approximate beta-binomial regression over regions</li>
</ul></li>
</ul>
</div><div class="column" style="width:25%;">
<p><img data-src="img/bsseq.jpg" style="width:80.0%"> <img data-src="img/dmrseqhex.jpg" style="width:80.0%"></p>
</div>
</div>
</section>
<section id="bsmooth-t-test-on-smoothed-methylation-proportions" class="slide level2">
<h2>BSmooth: t-test on smoothed methylation proportions</h2>
<ul>
<li><p><strong>Key idea:</strong> methylation levels of nearby CpGs are correlated</p>
<ul>
<li><p>use locally weighted polynomial regression values (lowess) to obtain smoothed estimate of methylation difference</p></li>
<li><p>similarly, obtain smoothed location-dependent standard deviation</p></li>
</ul></li>
<li><p>Compute t-test using smoothed quantities</p></li>
</ul>

<img data-src="img/bsmooth.jpg" class="r-stretch quarto-figure-center"><p class="caption"><a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2012-13-10-r83#Sec8">Hansen, et al.&nbsp;2012</a></p><ul>
<li>Red line: 35x coverage</li>
<li>Black line: subsampled to 5x coverage</li>
</ul>
</section>
<section id="modeling-count-nature-of-bisulfite-reads" class="slide level2">
<h2>Modeling count nature of bisulfite reads</h2>
<ul>
<li><p>For cytosine <span class="math inline">\(i\)</span> and sample <span class="math inline">\(j\)</span> in condition <span class="math inline">\(s\)</span>, we have:</p>
<ul>
<li><p><span class="math inline">\(M_{ij}\)</span> reads corresponding to methylation</p></li>
<li><p><span class="math inline">\(N_{ij}\)</span> total reads</p></li>
</ul></li>
<li><p>Let <span class="math inline">\(p_s\)</span> be the methylation probability in condition <span class="math inline">\(s\)</span></p></li>
</ul>
<div class="fragment">
<ul>
<li>Let <span class="math inline">\(M_{ij} \sim Binom(N_{ij}, p_{s})\)</span> for <span class="math inline">\(j \in s\)</span></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>How to test whether <span class="math inline">\(p_1 = p_2\)</span> (and optionally adjust for covariates)?</li>
</ul>
</div>
</section>
<section id="binomial-logisitic-regression" class="slide level2">
<h2>Binomial (Logisitic) regression</h2>
<ul>
<li>Generalized linear model for probability of success <span class="math inline">\(p\)</span>: <span class="math display">\[logit(p)=log\Big(\frac{p}{1-p}\Big) = \boldsymbol{X\beta}\]</span></li>
</ul>
<div class="fragment">
<ul>
<li>Link function <span class="math inline">\(g(p) = log\big(\frac{p}{1-p}\big)\)</span> describes relationship between mean parameter of response and linear predictor</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>No closed form; fit with iterative ML estimation</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>Interpret coefficients on original scale with inverse link function <span class="math inline">\(g^{-1}: p = \frac{e^{\boldsymbol{X\beta}}}{1+e^{\boldsymbol{X\beta}}}\)</span></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>In R: <code>glm(cbind(successes, failures) ~ x, family="binomial")</code></li>
</ul>
</div>
</section>
<section id="other-link-functions-for-glm" class="slide level2">
<h2>Other link functions for GLM</h2>

<img data-src="img/link.jpg" class="r-stretch quarto-figure-center"><p class="caption">Thiele &amp; Markussen 2012 (https://doi.org/10.1079/PAVSNNR20127016)</p></section>
<section id="why-logit-transformation" class="slide level2">
<h2>Why logit transformation?</h2>

<img data-src="lect11-epigenetics_files/figure-revealjs/unnamed-chunk-14-1.png" width="672" class="r-stretch quarto-figure-center"><ul>
<li>Same idea as M-value (though M-value uses log base 2 instead of natural log by convention):
<ul>
<li>Transforms methylation proportions to the real line</li>
<li>Mitigate heteroscedasticity</li>
</ul></li>
</ul>
</section>
<section id="binomial-regression-with-overdispersion" class="slide level2">
<h2>Binomial regression with overdispersion</h2>
<ul>
<li>In ordinary logistic regression, <span class="math inline">\(p\)</span> is assumed to be constant for all samples with in group</li>
</ul>
<div class="fragment">
<ul>
<li>To model <em>overdispersion</em> (additional biological variability), allow this quantity to vary</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>For example, let <span class="math inline">\(p \sim Beta(\alpha, \beta)\)</span> and <span class="math inline">\(M \sim Binom(N, p)\)</span></li>
</ul>
</div>
<div class="fragment">
<ul>
<li><p>This is the <strong>Beta-Binomial</strong> distribution</p>
<ul>
<li><p>Mean of M <span class="math inline">\(= \frac{n\alpha}{\alpha + \beta} = n\pi\)</span></p></li>
<li><p>Variance of M <span class="math inline">\(= \pi(1-\pi)(1 + (N - 1)\phi)\)</span></p></li>
<li><p>where <span class="math inline">\(\phi\)</span> is the overdispersion parameter: <span class="math inline">\(\phi = \frac{1}{\alpha + \beta + 1}\)</span></p></li>
</ul></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>In R: <code>aod::betabin(cbind(successes,failures) ~ x, ~ x)</code></li>
</ul>
</div>
</section>
<section id="pitfalls-of-binomial-regression" class="slide level2">
<h2>Pitfalls of binomial regression</h2>
<ul>
<li>What happens to logit link function <span class="math inline">\(log\big(\frac{p}{1-p}\big)\)</span> if <span class="math inline">\(p=0\)</span> or <span class="math inline">\(1\)</span>?</li>
</ul>
<div class="fragment">
<div class="callout callout-caution callout-captioned callout-style-default">
<div class="callout-body">
<div class="callout-caption">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Caution</strong></p>
</div>
<div class="callout-content">
<ul>
<li><p>Binomial regression unstable for fully methylated or unmethylated cytosines</p></li>
<li><p>Computationally intensive to fit model at every cytosine</p></li>
</ul>
</div>
</div>
</div>
</div>
<div class="fragment">
<ul>
<li>DSS: <a href="https://doi.org/10.1093/bioinformatics/btw026">Park &amp; Wu 2016</a>
<ul>
<li>Differential methylation under general experimental design</li>
<li>Alternate link function: <span class="math inline">\(arcsine(2p-1)\)</span></li>
<li>Approximate fitting with Generalized Least Squares (GLS)</li>
</ul></li>
</ul>
</div>
</section>
<section id="generalized-least-squares-gls" class="slide level2">
<h2>Generalized Least Squares (GLS)</h2>
<ul>
<li><p>Hybrid of <strong>linear regression</strong> and <strong>generalized linear model</strong></p></li>
<li><p><strong>Pro</strong>: stable &amp; closed form estimates (fast)</p></li>
<li><p><strong>Con</strong>: approximate</p></li>
<li><p><strong>Key idea</strong>: flexible covariance structure allows for specification of approximate beta-binomial error</p></li>
</ul>
</section>
<section id="dmcs-vs-dmrs" class="slide level2">
<h2>DMCs vs DMRs</h2>
<ul>
<li><p>Depending on the biological question, it might not make sense to only test <em>individual</em> CpGs (DMCs)</p></li>
<li><p>If instead differences occur across groups of nearby CpGs, we call them differentially methylated regions (DMRs)</p>
<ul>
<li><p>Methylation of nearby CpGs is correlated</p></li>
<li><p>Individual CpGs are not thought to act independently (e.g.&nbsp;are regulated in tandem)</p></li>
</ul></li>
</ul>
</section>
<section id="individual-cytosine-differences" class="slide level2">
<h2>Individual cytosine differences</h2>

<img data-src="img/dmc1.jpg" class="r-stretch quarto-figure-center"></section>
<section id="cpg-1" class="slide level2">
<h2>CpG 1</h2>

<img data-src="img/dmc2.jpg" class="r-stretch quarto-figure-center"></section>
<section id="cpg-2" class="slide level2">
<h2>CpG 2</h2>

<img data-src="img/dmc3.jpg" class="r-stretch quarto-figure-center"></section>
<section id="previous-approaches-grouping-significant-cpgs" class="slide level2">
<h2>Previous approaches: grouping significant CpGs</h2>

<img data-src="img/fdr2.jpg" class="r-stretch quarto-figure-center"><p>Examples: BSmooth (<a href="https://doi.org/10.1186/gb-2012-13-10-r83">Hansen et al.&nbsp;2012</a>), DSS (<a href="https://doi.org/10.1093/bioinformatics/btw026">Park et al.&nbsp;2016</a>)</p>
</section>
<section id="fdr-at-the-region-level" class="slide level2">
<h2>FDR at the region level</h2>

<img data-src="img/fdr3.jpg" class="r-stretch quarto-figure-center"><p><span class="math display">\[ \text{False Discovery Rate (FDR)} = E\Big[\frac{FP}{FP + TP}\Big]\]</span></p>
<div class="fragment">
<ul>
<li><span class="math inline">\(FDR_{CpG} = 2/10 = 0.20\)</span></li>
</ul>
</div>
<div class="fragment">
<ul>
<li><span class="math inline">\(FDR_{DMR} = 1/2 = 0.50\)</span></li>
</ul>
</div>
</section>
<section id="accurate-inference-of-dmrs" class="slide level2">
<h2>Accurate inference of DMRs</h2>
<ul>
<li><p><a href="https://doi.org/10.1093/biostatistics/kxy007">Korthauer et al.&nbsp;2019</a></p></li>
<li><p>Key ideas:</p>
<ul>
<li>model methylation signal over region to account for correlated signal</li>
<li>permutation to acheive accurate FDR control</li>
</ul></li>
<li><p>Implemented in the <a href="http://bioconductor.org/packages/dmrseq/">dmrseq</a> Bioconductor package</p></li>
</ul>

<img data-src="img/dmrseqhex.jpg" class="r-stretch quarto-figure-center"></section>
<section id="dmrseq-2-stage-approach" class="slide level2">
<h2>dmrseq: 2-stage approach</h2>

<img data-src="img/2stage.jpg" class="r-stretch quarto-figure-center"></section>
<section id="dmrseq-output" class="slide level2">
<h2>dmrseq output</h2>

<img data-src="img/exdmr.jpg" class="r-stretch quarto-figure-center"><ul>
<li><em>de novo</em> regions with FDR estimate</li>
<li>ranked by effect size of methylation difference</li>
<li>can adjust for covariates</li>
</ul>
</section>
<section id="whats-changing-methylation-or-cell-type-composition" class="slide level2">
<h2>What’s changing: methylation or cell type composition?</h2>

<img data-src="img/celltypecomp2.jpg" class="r-stretch quarto-figure-center"><p class="caption"><a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r31">Jaffe &amp; Irizarry 2014</a></p></section>
<section id="cell-type-composition-confounding" class="slide level2">
<h2>Cell type composition confounding</h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/celltypecomp.jpg"></p>
<p></p><figcaption><a href="https://bioconductor.org/packages/methylationArrayAnalysis/">methylationArrayAnalysis workflow</a></figcaption><p></p>
</figure>
</div>
<p>Can estimate cell type composition with models built on purified cell type methylation data (not always available)</p>
<p><img src="https://github.com/STAT540-UBC/stat540-ubc.github.io/raw/main/images/stat540-logo-s.png" class="slide-logo"></p>
<div class="footer footer-default">

</div>
</section>

    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="lect11-epigenetics_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="lect11-epigenetics_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="lect11-epigenetics_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="lect11-epigenetics_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="lect11-epigenetics_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="lect11-epigenetics_files/libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="lect11-epigenetics_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="lect11-epigenetics_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="lect11-epigenetics_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="lect11-epigenetics_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="lect11-epigenetics_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1600,

        height: 900,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        target: function(trigger) {
          return trigger.previousElementSibling;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'quarto-reveal',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>