---
title: "Statistical methods for epigenetics"
title-slide-attributes:
  data-background-color: "#197aa0"
  data-background-opacity: "0.9"
  data-background-image: "https://github.com/STAT540-UBC/stat540-ubc.github.io/raw/main/images/stat540-logo-s.png"
  data-background-size: 12%
  data-background-position: 95% 90%
author: "Keegan Korthauer"
date: "16 February 2023"
date-format: long
format: 
  revealjs:
    chalkboard: true
    slide-number: c/t
    width: 1600
    height: 900
    logo: "https://github.com/STAT540-UBC/stat540-ubc.github.io/raw/main/images/stat540-logo-s.png"
    echo: true
    theme: [default, ../custom.scss]
    show-notes: false
---

```{r}
#| include: false
library(tidyverse)
library(gridExtra)
library(broom)
library(latex2exp)
library(limma)
library(RColorBrewer)
theme_set(theme_bw(base_size = 20))
```

## Learning objectives for today^[We won't have time to cover preprocessing / normalization steps]

* Be familiar with the types of epigenetic measurements that involve **"peak-calling"**, and what a peak represents in each

* Understand the main statistical test performed by the widely used peak-caller MACS2

* Explain the similarities and differences (and understand their implications) between analysis of **differential expression** and **differential methylation**, with regard to:

  * continuous models of microarray data vs count models of sequencing data (similarity)
  
  * methylation level represents a proportion vs expression level unbounded (difference)
  
  * grouping of sites into differentially methylated regions (difference)



## Epigenetics

![[Image source](https://www.antibodies-online.com/areas/epigenetics/)](https://i1.abocdn.com/f8983e70d0e154ca2233646305eeafb7d3baa11d/resources/images/blog/epigenetics-schema.jpeg)

Recall basics of DNA-binding and DNA methylation assays from [lecture 2](https://github.com/STAT540-UBC/lectures/raw/main/lect02-technology.pdf)


## Open chromatin & nucleosome positioning

![[Image source](https://rockefelleruniversity.github.io/RU_ATAC_Workshop.html)](img/atac.jpeg)





## Peak calling 

* Sequencing after one of these selection/digestion techniques (IP, DNase, MNase, transposase) generally requires a peak-finding analysis

* One of the most widely used methods is [MACS2](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2008-9-9-r137):

![MACS2](img/macs.png)



## MACS2 overview

::: columns
::: column

* Shift reads to remove strand asymmetry: estimate $d$ from 1,000 regions

![[images source](https://hbctraining.github.io/Intro-to-ChIPseq)](https://hbctraining.github.io/Intro-to-ChIPseq/img/plos_chipseq_arrow.png)

:::
::: column

* Let $\lambda_{BG}$ be the expected number of reads in each window
* Let $\lambda_{local} = max(\lambda_{BG}, \lambda_{1K}, \lambda_{5K}, \lambda_{10K})$ be the expected number of reads in a window
* P-value for $k$ observed reads = $e^{-\lambda_{local}} \sum_{i=0}^k \frac{\lambda_{local}^i}{i!}$  

![](https://hbctraining.github.io/Intro-to-ChIPseq/img/lambda.png){fig-align="center"}

:::
:::

## Post peak-calling

Depending on the type of experiment and the hypotheses, next steps may include (but not limited to):

* [annotating peaks with genomic context](http://bioconductor.org/packages/release/bioc/vignettes/ChIPseeker/inst/doc/ChIPseeker.html)
* [differential binding analysis](https://www.bioconductor.org/packages/release/workflows/html/chipseqDB.html)
* enrichment analysis^[we'll discuss next lecture]
* [motif analysis](https://www.bioconductor.org/packages/release/workflows/html/generegulation.html)
* integration with other measurements (e.g. expression)


## DNA Methylation (5mC) 

- chemical modification to an individual DNA base (C = Cytosine), referred to as 5mC (5-methylcytosine)
- doesn't change coding sequence
- modification can happen at any C, but in mammals occurs almost^[other contexts observed in neuronal cells] exclusively at CpGs 

![[image source](https://www.epigentek.com)](img/5mC.jpg){height=3in}

## DNA Methylation at CpGs

![~30 Million CpGs in human](img/5thbase.jpg)

## The 5th base?

* heritable upon cell division 
     - 5mC added to complementary strand by methyltransferases
* dynamic
     - changes during development and with age
     - can be altered by environmental exposure

![[image source](https://slideplayer.com/slide/10719924/)](img/copymeth.jpg)


## Role in gene regulation 

![](img/genereg.jpg){fig-align="center"}


## Evolution of methylation assays

![[Laird 2010](https://doi.org/10.1038/nrg2732)](img/methassays.jpg)


##  Distinguishing methylated and unmethylated sites

::: columns
::: column

1. Methylated DNA immunoprecipitation (MeDIP)
* treat fragmented DNA with antibody targeting 5mC
* wash away unbound DNA
* measure extent to which each cytosine is still present (methylated) with **microarray or sequencing**

:::
::: column
![[image source](https://en.wikipedia.org/wiki/Methylated_DNA_immunoprecipitation)](img/medip.jpg){height=7in}
:::
:::


##  Distinguishing methylated and unmethylated sites

::: columns
::: column
2. Bisulfite conversion
* 5mC $\rightarrow$ stays 5mC
* C (not methylated) $\rightarrow$ converted to U
* Sequencing:
    * 5mC read as C
    * C read as T 
* For each C measure fraction of signal corresponding to C vs T with **microarray or sequencing** 
* *Note that can't distinguish between C $\rightarrow$ T mutation and unmethylated C*
:::

::: column
![[image source](https://www.diagenode.com/en/categories/bisulfite-conversion)](img/bisulfite-conversion.jpg)
:::
:::



##  Analysis of MeDIP

::: columns
::: column
* MeDIP targets **only** methylated DNA
* **MeDIP-chip**: much like gene expression microarrays
* **MeDIP-seq**: differential abundance of counts (similar to RNA-seq)
  * Complication: instead of genes or transcripts, need to define genomic "bins" to count reads over 
  * One option is to use sliding windows: [MEDIPS R package](http://bioconductor.org/packages/MEDIPS/) performs preprocessing and implements `edgeR` to test for differential abundance of each window
:::


::: column
![[image source](https://en.wikipedia.org/wiki/Methylated_DNA_immunoprecipitation)](img/medip.jpg){height=7in}
:::
:::



##  Bisulfite conversion arrays (e.g. Infinum 450K, Epic 850K)

::: columns
::: column
* Most commonly used human methylation arrays use bisulfite conversion
  * 450K / 850K refer to the number of sites probed
* Analysis has similarities to expression microarray analysis
  * Signal is **continuous** (intensity)
  * Moderate throughput: 1-16 samples per array
* Difference: focus on **relative** intensity of methylated vs unmethylated binding affinity
:::

::: column

![[image source](https://en.wikipedia.org/wiki/Illumina_Methylation_Assay)](img/microarray.jpg)
:::
:::

##  Beta values and M values

Need some way to summarize the two relative measures for each probe:

1. methylated intensity
2. unmethylated intensity



##  Beta $(\beta)$ values 

* Combine two measurements for each probe, and measure level of methylation: $$\beta = \frac{\text{methylated intensity}}{(\text{methylated intensity} + \text{unmeth intensity})}$$
* Range from 0 to 1
* Easily interpretable as a methylation proportion
* Best used for visualization


##  M-values

* Transformed $\beta$ values: $$M= log_2\big(\frac{\text{methylated intensity}}{\text{unmethylated intensity}}\big) = log_2(\beta/(1-\beta))$$
* This type of transformation is referred to as the *logit*
  * maps a proportion onto real line
* alleviate heteroscedasticity of $\beta$ values (recall that Binomial(n,p) variance = np(1-p))
   
![[Du et al. (2010)](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-11-587)](img/betavsmvals.png)


## Methylation array analysis

* Analysis strategy: linear regression on M-values 
* R Packages: 
   * [minfi](http://bioconductor.org/packages/minfi/) for preprocessing and normalization
   * [limma](http://bioconductor.org/packages/limma/) for individual cytosine analysis
   * [bumphunter](http://bioconductor.org/packages/bumphunter/) for region-level analysis
   * Workflow that combines these: [methylationArrayAnalysis]( 	https://bioconductor.org/packages/methylationArrayAnalysis/) - explore in Seminar 7
    
##  Example - DNA Methylation in sorted T cells

"Genome-wide DNA methylation analysis identifies hypomethylated genes regulated by FOXP3 in human regulatory T cells" from [Zhang et al. (2013)](https://www.ncbi.nlm.nih.gov/pubmed/23974203)

::: columns
::: column
> To explore the methylation landscape of nTreg, we analyzed genome-wide methylation in human naive nTreg (rTreg) and conventional naive CD4+ T cells (Naive).... Activation changed the methylation status of ...

:::
::: column
![Fig 1](https://www.ncbi.nlm.nih.gov/pmc/articles/instance/3798997/bin/2823f1.jpg){height=6.2in}
:::
:::

##  Example - DNA Methylation in sorted T cells

Data from [Zhang et al. (2013)](https://www.ncbi.nlm.nih.gov/pubmed/23974203) 

* [GEO accession GSE49667](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE49667) 
* Here we use the Bioc objects downloaded from [here](https://figshare.com/articles/dataset/methylAnalysisData_tar_gz/3471719)

```{r}
#| code-fold: true
# load libraries 
library(minfi)
library(limma)

# read in data
mSet <- readRDS("data/mSet.rds")
targets <- readRDS("data/targets.rds")
```

```{r}
mSet
```



##  Metadata values

```{r}
head(targets[c(1:5)])
table(targets$Sample_Label)
```



##  Beta values

```{r}
bVals <- getBeta(mSet)
dim(bVals)
head(bVals)
```



##  M values

```{r}
mVals <- getM(mSet)
dim(mVals)
head(mVals)
```



##  Comparing beta and M values

```{r}
#| code-fold: true
#| fig-align: center
# graphics settings
pal <- brewer.pal(8,"Dark2")
par(mfrow=c(1,2))

# plot beta values
densityPlot(bVals, 
            sampGroups = targets$Sample_Group, 
            main = "Beta values", 
            legend = FALSE, 
            xlab = "Beta values")
legend("top", 
       legend = levels(factor(targets$Sample_Group)), 
       text.col =brewer.pal(8,"Dark2"))

# plot M values
densityPlot(mVals, 
            sampGroups = targets$Sample_Group, 
            main = "M-values", 
            legend = FALSE, 
            xlab = "M values")
legend("topleft", 
       legend = levels(factor(targets$Sample_Group)), 
       text.col = brewer.pal(8,"Dark2"))

```


##  Comparing beta and M values

```{r}
#| code-fold: true
#| fig-width: 7
#| fig-height: 7
#| fig-align: center
b <- seq(0,1, by=0.0001)
m <- log2(b/(1-b))
plot(b, m, type="l")
```


. . .  

:::{.callout-tip}
# Question
Why do we have beta values far from 0 or 1 if methylation is either present or not at each CpG?
:::


##  Differentially methylated cytosines (DMC)

Which cytosines/probes are differentially methylated between naive & activated naive T cells?

::: columns
::: column
```{r}
## this is the factor of interest
cellType <- factor(targets$Sample_Group)

## this is the individual effect that we need to account for
individual <- factor(targets$Sample_Source) 

## use the above to create a design matrix
design <- model.matrix( ~ cellType + individual, 
                       data = targets)

# rename columns to be more convenient
colnames(design) <- c(levels(cellType),
                      levels(individual)[-1])
```
:::
::: column

```{r}
design
```
:::
:::



##  Differential methylation: limma on M-values

Here, the reference is `act_naive` cells

```{r} 
## fit the linear model (one for each probe)
mfit <- lmFit(mVals, design)
cfit <- eBayes(mfit)

## look at top DM CpGs between naive and ref (act_naive)
topTable(cfit, coef="naive")
```



##  Summary for all coefficients 

Significant tests (BH-adjusted p-value less than 0.05) of the null hypotheses: each coefficient equal to zero

```{r}
## look at the numbers of DM CpGs at FDR < 0.05
summary(decideTests(cfit))
```

Note that the first column tests whether the reference group (`act_naive`) has an M value of zero (not usually of interest)



##  Interpretation of parameters and effect sizes

* Our previous model gives parameter estimates for differences in M values $\alpha$
* To interpret on the scale of methylation proportions, need to back-transform:
  * Recall: $M= log_2\big(\frac{\text{methylated intensity}}{\text{unmethylated intensity}}\big) = log_2(\beta/(1-\beta))$
 * Solving for beta gives $\beta = \frac{2^M}{(1+2^M)}$


```{r}
(M <- cfit$coefficients["cg15459165", "naive"])
2^M/(1+2^M)
```


##  Summary of array DMC analysis

  
* **Just like for limma analysis of gene expression microarray**: Can use additive models, models with interactions, continuous and categorical variables

  * **interpretation of hypothesis tests, contrasts** with `makeContrasts` (for other comparisons not represented by coefficients in our design matrix): follows as before
  
  * **interpretation of parameters/effect sizes**: keep in mind that M-values are *transformed* proportions


. . .  


* Annotation packages (e.g. `IlluminaHumanMethylation450kanno.ilmn12.hg19`) allow for adding information about each CpG 
  * chromosome and position
  * strand
  * whether in CpG Island



##  Bisulfite sequencing

* Sequencing of bisulfite-converted DNA

* **WGBS** = whole-genome bisulfite sequencing

* **RRBS** = reduced representation bisulfite sequening (targets similar regions to methylation arrays e.g. islands)

* Pros: base-pair resolution; WGBS measures all CpGs

* Con: expensive (WGBS cost is on the order of whole genome DNA sequencing)

 


##  Bisulfite sequencing

![](img/wgbscounts2.jpg){fig-align="center"}


##  Analysis of bisulfite sequencing

::: columns
:::{.column width="75%"}
* BS assays (e.g. WGBS, RRBS) yield base-level counts of methylated $(M \text{counts})$ and ununmethylated $(U \text{counts})$ reads

* Differential methylation analysis quantifies significance of differences in proportion of methylated reads $(\frac{M}{M+U})$

* R Packages: 
    * [bsseq](http://bioconductor.org/packages/bsseq/): "BSmooth" - smoothed t-test
    * [dss](http://bioconductor.org/packages/dss/): approximate beta-binomial regression
    * [dmrseq](http://bioconductor.org/packages/dmrseq/): approximate beta-binomial regression over regions
:::
:::{.column width="25%"}

![](img/bsseq.jpg){width=80%}
![](img/dmrseqhex.jpg){width=80%}

:::
:::


##  BSmooth: t-test on smoothed methylation proportions

* **Key idea:** methylation levels of nearby CpGs are correlated

  * use locally weighted polynomial regression values (lowess) to obtain smoothed estimate of methylation difference 
  
  * similarly, obtain smoothed location-dependent standard deviation  
  
* Compute t-test using smoothed quantities


![[Hansen, et al. 2012](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2012-13-10-r83#Sec8)](img/bsmooth.jpg)

  - Red line: 35x coverage
  - Black line: subsampled to 5x coverage



##  Modeling count nature of bisulfite reads 

* For cytosine $i$ and sample $j$ in condition $s$, we have:

  * $M_{ij}$ reads corresponding to methylation

  * $N_{ij}$ total reads 

* Let $p_s$ be the methylation probability in condition $s$


. . .  


* Let $M_{ij} \sim Binom(N_{ij}, p_{s})$ for $j \in s$


. . .  


* How to test whether $p_1 = p_2$ (and optionally adjust for covariates)?



##  Binomial (Logisitic) regression

- Generalized linear model for probability of success $p$: $$logit(p)=log\Big(\frac{p}{1-p}\Big) = \boldsymbol{X\beta}$$

. . .  


- Link function $g(p) = log\big(\frac{p}{1-p}\big)$ describes relationship between mean parameter of response and linear predictor


. . .  


- No closed form; fit with iterative ML estimation


. . .  


- Interpret coefficients on original scale with inverse link function $g^{-1}: p = \frac{e^{\boldsymbol{X\beta}}}{1+e^{\boldsymbol{X\beta}}}$


. . .  



- In R: `glm(cbind(successes, failures) ~ x, family="binomial")`



##  Other link functions for GLM

![Thiele & Markussen 2012 (https://doi.org/10.1079/PAVSNNR20127016)](img/link.jpg)


##  Why logit transformation?

```{r}
#| echo: false
#| fig-width: 7
#| fig-height: 7
#| fig-align: center
p <- sort(c(seq(0,0.01, by=1e-5), seq(0,1,by=0.01), seq(0.99,1, by=1e-5)))
plot(p, log(p/(1-p)), type = "l", ylab = "transformed", ylim=c(-5,5))
```

* Same idea as M-value (though M-value uses log base 2 instead of natural log by convention):
    * Transforms methylation proportions to the real line
    * Mitigate heteroscedasticity



##  Binomial regression with overdispersion

- In ordinary logistic regression, $p$ is assumed to be constant for all samples with in group


. . .  


- To model *overdispersion* (additional biological variability), allow this quantity to vary


. . .  


- For example, let $p \sim Beta(\alpha, \beta)$ and $M \sim Binom(N, p)$


. . .  


- This is the **Beta-Binomial** distribution

  - Mean of M $= \frac{n\alpha}{\alpha + \beta} = n\pi$
  
  - Variance of M $= \pi(1-\pi)(1 + (N - 1)\phi)$ 

  - where $\phi$ is the overdispersion parameter: $\phi = \frac{1}{\alpha + \beta + 1}$


. . .  


- In R: `aod::betabin(cbind(successes,failures) ~ x, ~ x)`
 



##  Pitfalls of binomial regression 

- What happens to logit link function $log\big(\frac{p}{1-p}\big)$ if $p=0$ or $1$?


. . .  

:::{.callout-caution}
# Caution

* Binomial regression unstable for fully methylated or unmethylated cytosines
    
* Computationally intensive to fit model at every cytosine
:::

. . .  


- DSS: [Park & Wu 2016](https://doi.org/10.1093/bioinformatics/btw026)
    - Differential methylation under general experimental design
    - Alternate link function: $arcsine(2p-1)$
    - Approximate fitting with Generalized Least Squares (GLS) 




##  Generalized Least Squares (GLS) 

- Hybrid of **linear regression** and **generalized linear model**

- **Pro**: stable & closed form estimates (fast)

- **Con**: approximate

- **Key idea**: flexible covariance structure allows for specification of approximate beta-binomial error



##  DMCs vs DMRs

* Depending on the biological question, it might not make sense to only test *individual* CpGs (DMCs)

* If instead differences occur across groups of nearby CpGs, we call them differentially methylated regions (DMRs)

  * Methylation of nearby CpGs is correlated
  
  * Individual CpGs are not thought to act independently (e.g. are regulated in tandem)




##  Individual cytosine differences

![](img/dmc1.jpg){fig-align="center"}

##  CpG 1 

![](img/dmc2.jpg){fig-align="center"}


##  CpG 2 

![](img/dmc3.jpg){fig-align="center"}


##  Previous approaches: grouping significant CpGs

![](img/fdr2.jpg){fig-align="center"}

Examples: BSmooth ([Hansen et al. 2012](https://doi.org/10.1186/gb-2012-13-10-r83)), DSS ([Park et al. 2016](https://doi.org/10.1093/bioinformatics/btw026))


##  FDR at the region level

![](img/fdr3.jpg){fig-align="center"}

$$ \text{False Discovery Rate (FDR)} = E\Big[\frac{FP}{FP + TP}\Big]$$

. . .  


- $FDR_{CpG} = 2/10 = 0.20$

. . .  


- $FDR_{DMR} = 1/2 = 0.50$




##  Accurate inference of DMRs

- [Korthauer et al. 2019](https://doi.org/10.1093/biostatistics/kxy007)

- Key ideas: 
    - model methylation signal over region to account for correlated signal
    - permutation to acheive accurate FDR control
    
- Implemented in the [dmrseq](http://bioconductor.org/packages/dmrseq/) Bioconductor package
    
![](img/dmrseqhex.jpg){fig-align="center"}



##  dmrseq: 2-stage approach

![](img/2stage.jpg){fig-align="center"}



##  dmrseq output

![](img/exdmr.jpg){fig-align="center"}

- *de novo* regions with FDR estimate
- ranked by effect size of methylation difference
- can adjust for covariates 



##  What's changing: methylation or cell type composition?

![[Jaffe & Irizarry 2014](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r31)](img/celltypecomp2.jpg)


##  Cell type composition confounding

![[methylationArrayAnalysis workflow](https://bioconductor.org/packages/methylationArrayAnalysis/)](img/celltypecomp.jpg)

Can estimate cell type composition with models built on purified cell type methylation data (not always available)
