<!DOCTYPE html>
<html lang="en"><head>
<script src="lect03-eda_files/libs/clipboard/clipboard.min.js"></script>
<script src="lect03-eda_files/libs/quarto-html/tabby.min.js"></script>
<script src="lect03-eda_files/libs/quarto-html/popper.min.js"></script>
<script src="lect03-eda_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="lect03-eda_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="lect03-eda_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="lect03-eda_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="lect03-eda_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.3.450">

  <meta name="author" content="Keegan Korthauer">
  <meta name="dcterms.date" content="2024-01-16">
  <title>Exploratory data analysis &amp; experimental design</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="lect03-eda_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="lect03-eda_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="lect03-eda_files/libs/revealjs/dist/theme/quarto.css">
  <link href="lect03-eda_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="lect03-eda_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="lect03-eda_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="lect03-eda_files/libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="lect03-eda_files/libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="lect03-eda_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" data-background-color="#197aa0" data-background-image="https://github.com/STAT540-UBC/stat540-ubc.github.io/raw/main/images/stat540-logo-s.png" data-background-opacity="0.9" data-background-position="95% 90%" data-background-size="12%" class="quarto-title-block center">
  <h1 class="title">Exploratory data analysis &amp; experimental design</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Keegan Korthauer 
</div>
</div>
</div>

  <p class="date">January 16, 2024</p>
</section>
<section id="today-getting-to-know-your-data-and-why-you-need-to-understand-where-it-came-from" class="slide level2">
<h2>Today: Getting to know your data, and why you need to understand where it came from</h2>
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Learning objectives:</strong></p>
</div>
<div class="callout-content">
<ol type="1">
<li><p>Understand general principles on organizing your data</p></li>
<li><p>Be familiar with the main components of exploratory data analysis (EDA)</p></li>
<li><p>Recognize and avoid common pitfalls of data visualization</p></li>
<li><p>Understand the impacts of data quality and experimental design</p></li>
</ol>
</div>
</div>
</div>
</section>
<section id="where-to-find-code-to-make-these-plots" class="slide level2">
<h2>Where to find code to make these plots</h2>
<ul>
<li><p>These slides were made using <a href="https://github.com/STAT540-UBC/lectures/blob/main/lect03-eda/lect03-eda.qmd">this Quarto Markdown file</a>; it contains the code used to generate the R plots shown here</p></li>
<li><p><a href="https://github.com/STAT540-UBC/resources/blob/main/exploration-examples/explore.md">The companion notes (GitHub Markdown document)</a> explore multiple ways of making these plots and more detailed info</p>
<ul>
<li>Source Rmd for these notes <a href="https://github.com/STAT540-UBC/resources/blob/main/exploration-examples/explore.Rmd">here</a></li>
</ul></li>
</ul>
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Instructions to create <code>.pdfs</code> from the <code>.html</code> slides:</strong></p>
</div>
<div class="callout-content">
<ol type="1">
<li>Open slides in Google Chrome or Chromium</li>
<li>Toggle into print view by pressing the <code>E</code> key</li>
<li>Open the print dialogue by pressing: <code>CTRL</code>/<code>CMD</code>+<code>P</code></li>
<li>Change the <strong>Destination</strong> setting to <strong>Save as PDF</strong></li>
<li>Change the <strong>Layout</strong> to <strong>Landscape</strong></li>
<li>Change the <strong>Margins</strong> to <strong>None</strong></li>
<li>Enable the <strong>Background graphics</strong> option</li>
<li>Click <strong>Save</strong></li>
</ol>
</div>
</div>
</div>
</section>
<section id="chd8-rna-seq-experiment" class="slide level2">
<h2>CHD8 RNA-seq experiment</h2>
<ul>
<li><p><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6008102/">Gompers et al.&nbsp;(Nature Neuroscience 2017)</a> performed RNA-seq on mice from two different genotypes (18 WT vs 26 CHD8 mutant) and 5 developmental stages</p></li>
<li><p>“Using a statistical model that accounted for sex, developmental stage and sequencing batch, we tested for differential expression across 11,936 genes that were robustly expressed”</p></li>
<li><p>We’ll use this dataset throughout this lecture to illustrate EDA</p></li>
</ul>
<div class="columns">
<div class="column">
<p><img data-src="img/GompersDesign.png"></p>
</div><div class="column">
<p><img data-src="img/GompersHeatmap.png" style="width:66.0%"></p>
</div>
</div>
</section>
<section id="data-organization" class="slide level2">
<h2>Data organization</h2>
<ul>
<li><p>As is often the case, this data was obtained in two separate files</p>
<ul>
<li><p><strong>Main data file</strong>: expression values - one per gene, per sample ( <span class="math inline">\(G\)</span> genes <span class="math inline">\(\times\)</span> <span class="math inline">\(N\)</span> samples)</p></li>
<li><p><strong>Metadata file</strong>: several covariates/experimental design variables for each sample ( <span class="math inline">\(N\)</span> samples <span class="math inline">\(\times\)</span> <span class="math inline">\(P\)</span> covariates)</p></li>
</ul></li>
<li><p>We read these into R (see <a href="https://github.com/STAT540-UBC/lectures/blob/main/lect03-eda/lect03-eda.Rmd">Rmd source code</a> and <a href="https://github.com/STAT540-UBC/resources/blob/main/exploration-examples/explore.md">companion notes</a>) as a <code>data.frame</code> or <code>tibble</code> - <strong>matrix-like</strong> objects that have column names, and variable types for each column:</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="co"># data (expression) matrix</span></span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="fu">dim</span>(d)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 12158    44</code></pre>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="co"># metadata</span></span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="fu">dim</span>(m)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 44  8</code></pre>
</div>
</div>
</section>
<section id="small-but-important-detail" class="slide level2">
<h2>Small but <strong>important</strong> detail</h2>
<p><strong>Columns of main data matrix should match the rows of the metadata matrix <em>exactly</em></strong></p>
<p>These two objects should represent the same set of samples (and be in the <strong>same order</strong>)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="fu">head</span>(<span class="fu">colnames</span>(d))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Sample_ANAN001A" "Sample_ANAN001B" "Sample_ANAN001C" "Sample_ANAN001D"
[5] "Sample_ANAN001E" "Sample_ANAN001F"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="fu">head</span>(m<span class="sc">$</span>Sample)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Sample_ANAN001A" "Sample_ANAN001B" "Sample_ANAN001C" "Sample_ANAN001D"
[5] "Sample_ANAN001E" "Sample_ANAN001F"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="fu">identical</span>(<span class="fu">colnames</span>(d), m<span class="sc">$</span>Sample)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
</section>
<section id="data-matrix" class="slide level2">
<h2>Data matrix</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="fu">head</span>(d)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              Sample_ANAN001A Sample_ANAN001B Sample_ANAN001C Sample_ANAN001D
0610007P14Rik        5.238647        5.042397        5.364602        5.207641
0610009B22Rik        4.251070        4.568908        4.625527        4.568743
0610009O20Rik        4.807282        4.743503        4.921915        4.813903
0610010F05Rik        4.830067        5.063055        4.861288        4.957009
0610010K14Rik        5.958641        5.810707        5.851939        6.043380
0610011F06Rik        4.619302        4.579443        4.900748        4.802624
              Sample_ANAN001E Sample_ANAN001F Sample_ANAN001G Sample_ANAN001H
0610007P14Rik        5.222326        5.129427        5.239697        5.127338
0610009B22Rik        4.555767        4.412691        4.253776        4.205859
0610009O20Rik        4.812471        4.807076        4.828430        4.708997
0610010F05Rik        5.065935        4.948113        5.067492        5.124294
0610010K14Rik        5.997013        5.979884        5.902530        5.846258
0610011F06Rik        4.756396        4.828523        4.575162        4.868060
              Chd8.e14.S12 Chd8.e14.S13 Chd8.e14.S14 Chd8.e14.S16 Chd8.e14.S17
0610007P14Rik     4.841569     5.197562     4.825154     5.181760     5.070708
0610009B22Rik     4.452122     4.746104     4.747871     4.789378     4.862351
0610009O20Rik     4.455875     4.649412     4.338314     4.555878     4.490673
0610010F05Rik     5.314438     5.387970     5.523043     5.308837     5.514546
0610010K14Rik     5.855191     6.148984     6.148754     6.098012     6.174872
0610011F06Rik     4.391892     4.890393     4.486821     4.836696     4.446990
              Chd8.e14.S18 Chd8.e14.S19 Chd8.e14.S20 Chd8.e14.S21
0610007P14Rik     5.237254     5.283928     5.196948     5.022812
0610009B22Rik     4.871463     5.062306     4.819321     4.719951
0610009O20Rik     4.474563     4.597433     4.578242     4.506912
0610010F05Rik     5.430677     5.185067     5.477901     5.488947
0610010K14Rik     6.253634     6.561689     6.133367     6.032934
0610011F06Rik     4.698659     5.281168     4.785467     4.610248
              Chd8.e17.5.1_S158 Chd8.e17.5.2_S159 Chd8.e17.5.4_S161
0610007P14Rik          5.504698          5.563476          5.469537
0610009B22Rik          3.880660          4.366783          4.187023
0610009O20Rik          4.367506          4.074845          4.016735
0610010F05Rik          5.352970          5.406458          5.437642
0610010K14Rik          6.020094          5.818114          6.029098
0610011F06Rik          4.222639          4.064907          4.237808
              Chd8.e17.5.5_S162 Chd8.e17.5.6_S163 Chd8.e17.5.1_S1
0610007P14Rik          5.210378          5.747802        5.360175
0610009B22Rik          3.896679          4.109132        4.083653
0610009O20Rik          4.260492          4.310498        4.171089
0610010F05Rik          5.270270          5.498813        5.237495
0610010K14Rik          5.830401          6.202582        5.907929
0610011F06Rik          4.255365          4.301042        4.208392
              Chd8.e17.5.2_S2 Chd8.e17.5.3_S3 Chd8.e17.5.4_S4 Chd8.e17.5.5_S5
0610007P14Rik        5.396869        5.480500        5.576968        5.792268
0610009B22Rik        4.012442        4.043540        4.023286        4.115983
0610009O20Rik        4.181214        4.075280        4.243173        4.396101
0610010F05Rik        5.359145        5.590551        5.466983        5.332507
0610010K14Rik        6.142755        6.067490        5.856817        6.354569
0610011F06Rik        4.355365        4.199820        4.108307        4.591746
              Chd8.P0.S1 Chd8.P0.S10 Chd8.P0.S11 Chd8.P0.S2 Chd8.P0.S3
0610007P14Rik   5.004910    5.618407    5.042920   5.516423   5.249834
0610009B22Rik   3.171498    3.817326    3.381746   3.585340   3.481041
0610009O20Rik   4.262572    4.276003    4.245660   4.234671   4.302060
0610010F05Rik   5.063260    5.308997    5.113254   5.273173   5.179340
0610010K14Rik   5.968141    6.273270    5.865349   6.197163   6.144795
0610011F06Rik   3.834333    4.442323    4.143605   4.277508   4.370633
              Chd8.P0.S4 Chd8.P0.S5 Chd8.P0.S6 Chd8.P0.S7 Chd8.P0.S8 Chd8.P0.S9
0610007P14Rik   5.490179   5.330979   5.089818   5.309648   5.207886   5.280196
0610009B22Rik   3.814849   3.634283   3.425465   3.657936   3.647259   3.641667
0610009O20Rik   4.262546   4.112654   4.331208   4.281473   4.390390   4.241503
0610010F05Rik   5.254348   5.272861   5.208265   5.232609   5.102297   5.227704
0610010K14Rik   6.153272   6.076559   6.136792   6.236681   6.217135   6.246569
0610011F06Rik   4.206831   4.138643   4.091530   4.309412   4.398383   4.260990
              Chd8.adult.S23 Chd8.adult.S24 Chd8.adult.S25 Chd8.adult.S28
0610007P14Rik       5.117776       5.105053       5.074071       5.274795
0610009B22Rik       3.736176       4.437970       4.150157       4.726862
0610009O20Rik       4.218049       4.409190       4.386743       4.494319
0610010F05Rik       4.391540       3.700743       3.745234       3.602500
0610010K14Rik       5.245851       4.827066       4.953091       4.971989
0610011F06Rik       4.562028       4.714897       4.696579       5.080126
              Chd8.adult.S29 Chd8.adult.S31
0610007P14Rik       5.051701       5.000228
0610009B22Rik       4.046562       4.485514
0610009O20Rik       4.402777       4.321922
0610010F05Rik       3.715847       3.912299
0610010K14Rik       4.910206       4.861016
0610011F06Rik       4.787186       4.849831</code></pre>
</div>
</div>
</section>
<section id="metadata" class="slide level2">
<h2>Metadata</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>m</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 44 × 8
   Number Sample            DPC Sex   Group  SeqRun MappedReads FeatureCounts
    &lt;dbl&gt; &lt;chr&gt;           &lt;dbl&gt; &lt;fct&gt; &lt;fct&gt;  &lt;fct&gt;        &lt;dbl&gt;         &lt;dbl&gt;
 1      1 Sample_ANAN001A  12.5 F     Mutant A         42452222      37655856
 2      2 Sample_ANAN001B  12.5 M     WT     A         54503162      47106938
 3      3 Sample_ANAN001C  12.5 M     WT     A         44978512      40448118
 4      4 Sample_ANAN001D  12.5 F     Mutant A         50099336      44993589
 5      5 Sample_ANAN001E  12.5 F     Mutant A         47163546      41840678
 6      6 Sample_ANAN001F  12.5 F     WT     A         43893480      39483622
 7      7 Sample_ANAN001G  12.5 F     Mutant A         53208684      47697298
 8      8 Sample_ANAN001H  12.5 F     WT     A         38925414      35056860
 9      9 Chd8.e14.S12     14.5 M     WT     B         23622816      19586363
10     10 Chd8.e14.S13     14.5 F     WT     B         23974703      21019829
# ℹ 34 more rows</code></pre>
</div>
</div>
</section>
<section id="organizing-your-data---two-main-issues" class="slide level2">
<h2>Organizing your data - two main issues</h2>
<ul>
<li><p>How you set up your input files for easy reading into R</p>
<ul>
<li><p>Easiest to work with are text files (e.g.&nbsp;tab-delimited <code>.tsv</code>)</p></li>
<li><p>Excel files not uncommon but not recommended</p></li>
<li><p>Almost always some data cleaning/wrangling involved (e.g.&nbsp;checking consistency, recoding, renaming variables)</p></li>
</ul></li>
<li><p>Within R, being able to refer to subsets of both data sets (main data &amp; metadata) and other manipulations that require “matching” the expression data with the sample information (“slicing and dicing”)</p>
<ul>
<li><p>For example: “Get me the expression level data for CHD8 in the female adult wild type mice” – this uses information from both sets</p></li>
<li><p>In practice, you may have to do it multiple ways to play nice with different R packages (e.g.&nbsp;one way for visualization, and another for downstream analysis)</p></li>
</ul></li>
</ul>
</section>
<section id="organizing-your-data-separated" class="slide level2">
<h2>Organizing your data: Separated</h2>
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Option 1 - “Separated”: Keep main data and metadata tables separate</strong></p>
</div>
<div class="callout-content">
<p>Pros:</p>
<ul>
<li><p>Minimal startup effort / extra code</p></li>
<li><p>Can be compatible with downstream analysis methods (e.g.&nbsp;<a href="https://bioconductor.org/">Bioconductor</a>)</p></li>
</ul>
<p>Cons:</p>
<ul>
<li><p><strong>Risky</strong>: easy to make a mistake when subsetting and/or reordering samples - extra sanity checks required</p></li>
<li><p>Not a convenient format for visualization since main data is separated from its metadata</p></li>
</ul>
<p>Overall: <span style="color: red;"><em><strong>not recommended</strong></em></span></p>
</div>
</div>
</div>
<p><img data-src="img/separateway.png" class="absolute" style="bottom: 0px; right: 50px; width: 950px; "></p>
</section>
<section id="organizing-your-data-tidy-way" class="slide level2">
<h2>Organizing your data: Tidy way</h2>
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Option 2 - “The tidy way”: Combine main data &amp; metadata into one ‘long’ table</strong></p>
</div>
<div class="callout-content">
<p>Pros:</p>
<ul>
<li><p>Unambiguous - keeps all data in one object with one row per observation (e.g.&nbsp;each sample/gene combination is one row, along with all its metadata)</p></li>
<li><p>Plays nice with tidyverse tools (e.g.&nbsp;dplyr manipulations, ggplot2 visualization)</p></li>
</ul>
<p>Cons:</p>
<ul>
<li><p>‘long’ format is inefficient data storage - sample information is repeated</p></li>
<li><p>Not compatible with many tools for downstream analysis (e.g.&nbsp;Bioconductor)</p></li>
</ul>
<p>Overall: <span style="color: red;"><em><strong>recommended for EDA/visualization</strong></em></span></p>
</div>
</div>
</div>
<p><img data-src="img/tidyway.png" class="absolute" style="bottom: 0px; right: 50px; width: 475px; "></p>
</section>
<section id="the-tidy-way" class="slide level2">
<h2>The Tidy way</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>d_long</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 534,952 × 10
   gene          Sample   Expression Number   DPC Sex   Group SeqRun MappedReads
   &lt;chr&gt;         &lt;chr&gt;         &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;        &lt;dbl&gt;
 1 0610007P14Rik Sample_…       5.24      1  12.5 F     Muta… A         42452222
 2 0610009B22Rik Sample_…       4.25      1  12.5 F     Muta… A         42452222
 3 0610009O20Rik Sample_…       4.81      1  12.5 F     Muta… A         42452222
 4 0610010F05Rik Sample_…       4.83      1  12.5 F     Muta… A         42452222
 5 0610010K14Rik Sample_…       5.96      1  12.5 F     Muta… A         42452222
 6 0610011F06Rik Sample_…       4.62      1  12.5 F     Muta… A         42452222
 7 0610012G03Rik Sample_…       4.03      1  12.5 F     Muta… A         42452222
 8 0610030E20Rik Sample_…       3.22      1  12.5 F     Muta… A         42452222
 9 0610031J06Rik Sample_…       4.38      1  12.5 F     Muta… A         42452222
10 0610037L13Rik Sample_…       4.23      1  12.5 F     Muta… A         42452222
# ℹ 534,942 more rows
# ℹ 1 more variable: FeatureCounts &lt;dbl&gt;</code></pre>
</div>
</div>
</section>
<section id="organizing-data-bioconductor-way" class="slide level2">
<h2>Organizing data: Bioconductor way</h2>
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Option 3 - “The <a href="https://bioconductor.org/">Bioconductor</a> way”: Combine main data &amp; metadata into one specially formatted object</strong></p>
</div>
<div class="callout-content">
<p>Pros:</p>
<ul>
<li><p>Unambiguous: keeps all data in one object with special slots that can be accessed with handy functions</p></li>
<li><p>Plays nice with Bioconductor tools</p></li>
<li><p>Efficient storage (no duplication of information like tidy way)</p></li>
</ul>
<p>Cons:</p>
<ul>
<li><p>Specific to Bioconductor</p></li>
<li><p>Not a compatible format for visualization (e.g.&nbsp;ggplot2)</p></li>
</ul>
<p>Overall: <span style="color: red;"><em><strong>recommended for downstream analysis (e.g.&nbsp;Differential Expression)</strong></em></span></p>
</div>
</div>
</div>
<p><img data-src="https://www.bioconductor.org/images/logo/jpg/bioconductor_logo_cmyk.jpg" class="absolute" style="left: 50px; bottom: 0px; width: 300px; "> <img data-src="img/summarizedexperiment.png" class="absolute" style="bottom: 0px; right: 50px; width: 475px; "></p>
</section>
<section id="the-bioconductor-way-summarizedexperiment" class="slide level2">
<h2>The Bioconductor way: SummarizedExperiment</h2>
<div class="columns">
<div class="column">
<ul>
<li><p><a href="https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html"><code>SummarizedExperiment</code></a>: A special object format that is designed to contain data &amp; metadata</p></li>
<li><p>Comes along with handy accessor functions</p></li>
<li><p>Many related / similar types of objects for specialized data types (e.g.&nbsp;<a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html#anatomy-of-a-summarizedexperiment"><code>RangedSummarizedExperiment</code></a>, <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html"><code>SingleCellExperiment</code></a>, <a href="https://rdrr.io/bioc/edgeR/man/DGEList.html"><code>DGEList</code></a>)</p></li>
</ul>
</div><div class="column">
<p><img data-src="img/summarizedexperiment.png"></p>
</div>
</div>
</section>
<section id="now-what-time-to-explore" class="slide level2">
<h2>Now what? Time to explore!</h2>
<ul>
<li><p>Understand / get a feel for the data</p></li>
<li><p>Formulate hypotheses / develop models</p></li>
<li><p>Identify problems</p></li>
</ul>
</section>
<section id="first-questions---sanity-checks" class="slide level2">
<h2>First questions - sanity checks</h2>
<ul>
<li><p>Is the file the expected size? Format?</p></li>
<li><p>Do we have the expected number of samples?</p></li>
<li><p>Do the sample names in both files match? (Do not assume same ordering!)</p></li>
<li><p>Are sample / feature names formatted correctly (e.g.&nbsp;no Excel conversion errors)?</p></li>
<li><p>What do features represent? (e.g.&nbsp;Gene names, probe identifiers, etc.)</p></li>
<li><p>Is the data numeric? Integers or decimal? Ratios (to what?)</p></li>
<li><p>Are the data on a log scale? If so what is the base?</p></li>
<li><p>Are there missing data points? What do they mean?</p></li>
<li><p>Do factors have the expected number of levels?</p></li>
<li><p>Do we have all the sample information we need?</p></li>
</ul>
</section>
<section id="these-questions-might-lead-to-concerns" class="slide level2">
<h2>These questions might lead to concerns</h2>
<ul>
<li><p>If you are the one generating the data, save yourself grief by paying attention to these issues up front - Document what you did and be consistent!</p></li>
<li><p>If you are the analyst, hopefully you were involved in the design stage so there will be fewer surprises</p></li>
</ul>
</section>
<section id="making-sense-of-the-data" class="slide level2">
<h2>Making sense of the data</h2>
<h4 id="data-is-what-we-observe-we-want-to-infer-something-about-where-it-came-from">Data is what we observe, we want to infer something about “where it came from”</h4>

<img data-src="img/statisticsdogma.png" class="r-stretch quarto-figure-center"></section>
<section id="model-of-gene-expression-data-generation" class="slide level2">
<h2>Model of gene expression data generation</h2>
<ul>
<li><p>The measured expression level of gene <span class="math inline">\(g\)</span> is the combination of many effects</p></li>
<li><p>Analysis goal is often to determine relative role of effects - separate signal from “noise”</p></li>
</ul>

<img data-src="img/geneexpmodel.png" class="r-stretch quarto-figure-center"></section>
<section id="variability-friend-and-foe" class="slide level2">
<h2>Variability: friend and foe</h2>
<ul>
<li><p>If there is no variability, you don’t have any information - key is controlling/understanding sources of <em>wanted vs.&nbsp;unwanted</em> variability</p></li>
<li><p>First line of defense: Know the enemy</p>
<ul>
<li><p>You can only “correct” for things you know about</p></li>
<li><p><strong>Keep track of potential sources of variance</strong>: Batches of reagents, slides, personnel, processing dates, etc.</p></li>
</ul></li>
<li><p><strong>Design experiments to minimize impact of technical variability</strong></p>
<ul>
<li>Don’t process all control samples on day 1 and all treated samples on day 10</li>
</ul></li>
<li><p>Ensure appropriate <strong>replication</strong></p>
<ul>
<li><p>Biological (important)</p></li>
<li><p>Technical (usually less important but might need to convince yourself)</p></li>
</ul></li>
</ul>
</section>
<section id="biggest-pitfall-in-high-dimensional-data-analysis" class="slide level2">
<h2>Biggest pitfall in (high-dimensional) data analysis</h2>
<div class="columns">
<div class="column">
<p>If you don’t <strong>look</strong> at the data, you are likely going to miss important things</p>
<ul>
<li><p>Not just at the beginning, but at every stage</p></li>
<li><p>That could mean making plots, or examining numerical patterns - probably both</p></li>
<li><p>“Sanity checks” should make up a lot of your early effort</p></li>
<li><p>Blindly following recipes/pipelines/vignettes/seminar code → trouble</p></li>
</ul>
</div><div class="column">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://bookdown.org/rdpeng/artofdatascience/images/epicycle.png" width="600"></p>
<figcaption>Epicycles of Analysis, from “Art of Data Science”</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="first-looks-at-the-chd8-expression-data" class="slide level2">
<h2>First looks at the CHD8 expression data</h2>
<div class="columns">
<div class="column">
<ul>
<li><p>What is the size of the data?</p></li>
<li><p>What is the range of the data?</p></li>
<li><p>Are there any missing values?</p></li>
<li><p>Are the data transformed in any way?</p></li>
</ul>
</div><div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>se</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SummarizedExperiment 
dim: 12158 44 
metadata(0):
assays(1): logrpkm
rownames(12158): 0610007P14Rik 0610009B22Rik ... Zzef1 Zzz3
rowData names(0):
colnames(44): Sample_ANAN001A Sample_ANAN001B ... Chd8.adult.S29
  Chd8.adult.S31
colData names(8): Number Sample ... MappedReads FeatureCounts</code></pre>
</div>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a><span class="fu">range</span>(<span class="fu">assays</span>(se)<span class="sc">$</span>logrpkm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] -7.87235 13.26689</code></pre>
</div>
</div>
</div>
</div>
</section>
<section id="examining-the-metadata" class="slide level2">
<h2>Examining the metadata</h2>
<div class="columns">
<div class="column">
<p>Sample info:</p>
<ul>
<li><p>Age (Days Post-conception, DPC)</p></li>
<li><p>Sex</p></li>
<li><p>Group (genotype)</p></li>
<li><p>Sequencing run (batch)</p></li>
<li><p>Number of mapped reads</p></li>
<li><p>feature counts</p></li>
</ul>
</div><div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a><span class="fu">table</span>(se<span class="sc">$</span>DPC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
12.5 14.5 17.5   21   77 
   8    9   10   11    6 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a><span class="fu">table</span>(se<span class="sc">$</span>Sex)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 M  F 
25 19 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a><span class="fu">table</span>(se<span class="sc">$</span>SeqRun)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 A  B  C  D  E  H 
 8  9  5  5 11  6 </code></pre>
</div>
</div>
</div>
</div>
</section>
<section id="how-to-look-at-the-rest-of-the-data" class="slide level2">
<h2>How to look at the rest of the data?</h2>

<img data-src="img/viz.png" class="r-stretch quarto-figure-center"></section>
<section id="exploratory-data-analysis-eda" class="slide level2">
<h2>Exploratory Data Analysis (EDA)</h2>
<ul>
<li><p>EDA is “compute a little and plot a lot”</p></li>
<li><p>Exploratory plots can be quick and dirty - making them publication quality takes a lot of time and effort</p></li>
<li><p>I’ll show a few simple approaches that are common in genomics</p></li>
<li><p>Reminder that code to generate the plots you see here is posted in the companion notes linked earlier</p></li>
</ul>
</section>
<section id="sanity-check-expression-of-chd8" class="slide level2">
<h2>Sanity check: expression of CHD8</h2>
<div class="columns">
<div class="column">
<ul>
<li><p>Paper reported that CHD8 went down over time, and is lower in the mutant: confirmed!</p></li>
<li><p>Note that we are not doing any formal “analysis” here nor trying to make this plot beautiful – keeping it very simple for our exploration</p></li>
</ul>
</div><div class="column">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="lect03-eda_files/figure-revealjs/unnamed-chunk-10-1.png" width="864"></p>
</div>
</div>
</div>
</div>
</section>
<section id="two-views-of-the-same-data-which-do-you-prefer" class="slide level2">
<h2>Two views of the same data: Which do you prefer?</h2>

<img data-src="lect03-eda_files/figure-revealjs/unnamed-chunk-11-1.png" width="1440" class="r-stretch"><p>How to best summarize patterns in the data?</p>
</section>
<section id="two-views-of-the-same-data-let-the-data-speak-for-itself" class="slide level2">
<h2>Two views of the same data: Let the data speak for itself!</h2>

<img data-src="lect03-eda_files/figure-revealjs/unnamed-chunk-12-1.png" width="1440" class="r-stretch"><ul>
<li><p>What is the sample size?</p></li>
<li><p>Is the distribution symmetrical, or skewed?</p></li>
<li><p>Are there any outliers?</p></li>
</ul>
</section>
<section id="comparing-sample-distributions-boxplots" class="slide level2">
<h2>Comparing sample distributions: Boxplots</h2>
<div class="columns">
<div class="column" style="width:30%;">
<ul>
<li><p>Quick and dirty; reasonable tool to summarize large amounts of data</p></li>
<li><p>Not ideal if the distribution is multimodal</p></li>
<li><p>Don’t use box plots (alone) when you have small numbers of points; <strong>show the points instead!</strong></p></li>
</ul>
</div><div class="column" style="width:70%;">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="lect03-eda_files/figure-revealjs/unnamed-chunk-13-1.png" width="1056"></p>
</div>
</div>
</div>
</div>
</section>
<section id="comparing-distributions-histograms-and-density-plots" class="slide level2">
<h2>Comparing distributions: Histograms and Density plots</h2>
<div class="columns">
<div class="column">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="lect03-eda_files/figure-revealjs/unnamed-chunk-14-1.png" width="672"></p>
</div>
</div>
</div><div class="column">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="lect03-eda_files/figure-revealjs/unnamed-chunk-15-1.png" width="672"></p>
</div>
</div>
</div>
</div>
</section>
<section id="dont-use-density-plots-for-bounded-data" class="slide level2">
<h2>Don’t use density plots for bounded data</h2>

<img data-src="lect03-eda_files/figure-revealjs/unnamed-chunk-16-1.png" width="1152" class="r-stretch"></section>
<section id="violin-plots" class="slide level2">
<h2>Violin plots</h2>
<h4 id="a-hybrid-of-density-plots-and-box-plots">A hybrid of density plots and box plots</h4>

<img data-src="lect03-eda_files/figure-revealjs/unnamed-chunk-17-1.png" width="1152" class="r-stretch"></section>
<section id="scatter-plots" class="slide level2">
<h2>Scatter plots</h2>
<h4 id="major-problem-is-over-plotting---use-transparency-or-2d-density">Major problem is ‘over-plotting’ - use transparency or 2D density</h4>

<img data-src="lect03-eda_files/figure-revealjs/unnamed-chunk-18-1.png" width="1344" class="r-stretch"></section>
<section id="transformations" class="slide level2">
<h2>Transformations</h2>
<ul>
<li>Many real data have very skewed distributions – in this case, most values are &lt;500, but there are a smattering up to 10,000</li>
<li>If your plots look like this, try taking the logarithm</li>
<li>If have non-positive values (typically 0) add a small constant “pseudocount”</li>
</ul>

<img data-src="lect03-eda_files/figure-revealjs/unnamed-chunk-19-1.png" width="1152" class="r-stretch"></section>
<section id="pairwise-scatter-plots-to-compare-samples" class="slide level2">
<h2>Pairwise (scatter) plots to compare samples</h2>
<div class="columns">
<div class="column" style="width:30%;">
<p>This is nice but unwieldy (or won’t work) for large data sets</p>
</div><div class="column" style="width:70%;">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="lect03-eda_files/figure-revealjs/unnamed-chunk-20-1.png" width="864"></p>
</div>
</div>
</div>
</div>
</section>
<section id="examining-the-metadata-more-globally" class="slide level2">
<h2>Examining the metadata more globally</h2>
<div class="columns">
<div class="column" style="width:30%;">
<ul>
<li><p>Sex is not that well balanced; all but one of the adults is male</p></li>
<li><p>There is a batch confound: The stages were run in different batches (except 17.5 was split in two)</p></li>
<li><p>Mapped reads varies with the batches (SeqRun)</p></li>
</ul>
</div><div class="column" style="width:70%;">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="lect03-eda_files/figure-revealjs/unnamed-chunk-21-1.png" width="1152"></p>
</div>
</div>
</div>
</div>
</section>
<section id="heatmaps" class="slide level2">
<h2>Heatmaps</h2>

<img data-src="img/heatmaps.png" class="r-stretch quarto-figure-center"></section>
<section id="r-options-for-making-heatmaps" class="slide level2">
<h2>R options for making heatmaps</h2>
<ul>
<li>Desired features
<ul>
<li><p>Doesn’t complain when handed slightly messy data and has sensible default behaviour</p></li>
<li><p>Allow easy control of layout such as annotations and scale bar</p></li>
<li><p>Allow easy control over clustering behaviour (row/column ordering)</p></li>
</ul></li>
<li>There are <em>many</em> functions/packages for making heatmaps; here are three
<ul>
<li><p><a href="https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/heatmap">base::heatmap</a> - ok for quick and dirty but otherwise very limited</p></li>
<li><p><a href="https://www.rdocumentation.org/packages/pheatmap/versions/1.0.12/topics/pheatmap">pheatmap</a> - Default colour schemes not ideal, otherwise good option (used in STAT 540 seminars)</p></li>
<li><p><a href="https://bioconductor.org/packages/release/bioc/html/ComplexHeatmap.html">ComplexHeatmaps</a> - most powerful/flexible, but a bit more complex to learn</p></li>
</ul></li>
</ul>
</section>
<section id="heatmap-of-50-random-rows" class="slide level2">
<h2>Heatmap of 50 random rows</h2>

<img data-src="lect03-eda_files/figure-revealjs/unnamed-chunk-22-1.png" width="768" class="r-stretch"></section>
<section id="revision-1-same-input-data-rows-are-scaled" class="slide level2">
<h2>Revision 1: Same input data; rows are scaled</h2>
<div class="columns">
<div class="column" style="width:70%;">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="lect03-eda_files/figure-revealjs/unnamed-chunk-23-1.png" width="768"></p>
</div>
</div>
</div><div class="column" style="width:30%;">
<ul>
<li><p>Rows are scaled to have mean 0 and variance 1 (z-scores)</p></li>
<li><p>Subtract the mean; divide by the standard deviation - use <code>scale()</code> on the data rows (<em>some packages will do this by default</em>)</p></li>
<li><p>It is now easier to compare the rows and see a bit of structure</p></li>
</ul>
</div>
</div>
</section>
<section id="revision-2-adjusting-contrast" class="slide level2">
<h2>Revision 2: adjusting contrast</h2>
<div class="columns">
<div class="column" style="width:70%;">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="lect03-eda_files/figure-revealjs/unnamed-chunk-24-1.png" width="768"></p>
</div>
</div>
</div><div class="column" style="width:30%;">
<ul>
<li><p>Range of values is <strong>clipped</strong> to (-2,2): aything more than two SDs from the row mean is set to 2</p></li>
<li><p>Limit values of 2 or 3 SDs are common</p></li>
</ul>
<h4 id="clipping-hides-outliers-but-allows-us-to-see-variation-in-the-bulk-of-the-data">Clipping hides outliers but allows us to see variation in the bulk of the data</h4>
</div>
</div>
</section>
<section id="plotting-too-much-data" class="slide level2">
<h2>Plotting too much data</h2>
<div class="columns">
<div class="column" style="width:30%;">
<p><img data-src="img/toomanyrows.png"></p>
</div><div class="column" style="width:70%;">
<ul>
<li><p>An entire data set (&gt;10k rows)</p></li>
<li><p>If the cells are less than 1 pixel, everything starts to turn to mush and can even be misleading</p></li>
<li><p>If your heatmap has too many rows to see labels (genes), make sure it is conveying useful information (what are you trying to show?)</p></li>
</ul>
</div>
</div>
</section>
<section id="choice-of-colours" class="slide level2">
<h2>Choice of colours</h2>

<img data-src="img/colour.png" class="r-stretch quarto-figure-center"><ul>
<li>Humans can’t really tell the difference between a 8- and 16-colour scale</li>
<li>Many R defaults, other common scales (green/red) <em>not colourblind friendly</em></li>
<li><a href="https://rdrr.io/cran/RColorBrewer/man/ColorBrewer.html"><code>RColorBrewer</code></a>: scales based on work of visualization experts</li>
<li>Greyscale: loss of dynamic range, but cheaper to publish!</li>
</ul>
</section>
<section id="divergent-vs.-sequential-maps" class="slide level2">
<h2>Divergent vs.&nbsp;sequential maps</h2>

<img data-src="img/seqdiv.png" class="r-stretch quarto-figure-center"><div class="columns">
<div class="column">
<p><strong>Divergent</strong>: colours pass through black or white at the midpoint (e.g.&nbsp;mean). Ideal if your original data are naturally “symmetric” around zero (or some other value) - Otherwise it might just be confusing</p>
</div><div class="columnn">
<p><strong>Sequential</strong>: colours go from light to dark. Darker colours mean “higher” by default in <code>RColorBrewer</code>. No defined midpoint.</p>
</div>
</div>
</section>
<section id="a-confusing-heat-map-matlab" class="slide level2">
<h2>A confusing heat map (Matlab)</h2>
<ul>
<li><p>Too many different colours to readily interpret relative ordering of values</p></li>
<li><p>Not recommended to use these types of scales for continuous values</p></li>
<li><p>Rainbows or scales with many distinct colours are better for factors / categorical variables</p></li>
</ul>

<img data-src="img/matlab.png" class="r-stretch quarto-figure-center"></section>
<section id="heatmap-recommendations" class="slide level2">
<h2>Heatmap recommendations</h2>
<div class="columns">
<div class="column">
<ul>
<li><p>Pick an appropriate colour scale</p></li>
<li><p>Show a scale bar (so don’t use <code>base::heatmap</code>)</p></li>
<li><p>Either cluster rows / columns, or order by something meaningful (e.g.&nbsp;sample info)</p></li>
<li><p>Add annotation bars of meaningful covariates</p></li>
<li><p>If you have missing data points, make sure it is obvious where they are (e.g.&nbsp;different colour)</p></li>
</ul>
</div><div class="column">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="lect03-eda_files/figure-revealjs/unnamed-chunk-25-1.png" width="768"></p>
</div>
</div>
</div>
</div>
</section>
<section id="eda-summary" class="slide level2">
<h2>EDA Summary</h2>
<p>Purpose: <strong>Evaluate whether data aligns with expectations</strong>:</p>
<ul>
<li>Sanity checks</li>
<li>Visualization to spot patterns and/or oddities</li>
</ul>
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Three principles of EDA:</strong></p>
</div>
<div class="callout-content">
<ol type="1">
<li><p><strong>Let the data speak for itself</strong> - avoid dynamite plots in favor of boxplots, overlayed with points if feasible number</p></li>
<li><p><strong>Avoid overplotting</strong> - use transparency or 2D density in scatterplots; make sure heatmap cells aren’t too small</p></li>
<li><p><strong>Consider transformations</strong> (e.g.&nbsp;log) for skewed distributions</p></li>
<li><p><strong>Use colour intentionally</strong> to convey information</p></li>
</ol>
</div>
</div>
</div>
<p>Additional exploratory techniques will be discussed later in the course (e.g.&nbsp;Clustering, PCA)</p>
</section>
<section id="experimental-design-and-quality" class="slide level2">
<h2>Experimental design and quality</h2>

<img data-src="https://media.springernature.com/lw685/springer-static/image/art%3A10.1038%2Fsj.ebd.6400436/MediaObjects/41432_2006_Article_BF6400436_Fig1_HTML.jpg?as=webp" class="r-stretch quarto-figure-center"><p class="caption">Fig 1, 10.1038/sj.ebd.6400436</p></section>
<section id="experimental-design-considerations" class="slide level2">
<h2>Experimental design considerations</h2>
<ul>
<li><strong>Will the experiment answer the question?</strong>
<ul>
<li>Consider properties of the measurement technology as well as the biology</li>
</ul></li>
<li><strong>How many individuals should I study?</strong>
<ul>
<li>For many experiments, this is hard to answer (<em>even though your grant reviewer might demand it (power calculations)!</em>)</li>
<li>Often settle for “enough power to find something useful” rather than “find everything”</li>
</ul></li>
<li><strong>Biological replicates are essential and more important than technical replicates</strong> (usually)
<ul>
<li>Doing a study with too few replicates is arguably worse than not doing the study</li>
</ul></li>
<li><strong>Beware of (and try to control for) unwanted variation</strong>
<ul>
<li>“Pooling” samples is not a substitute for replication</li>
<li>E.g. if your question is “do people respond to a drug”, the unit of analysis must be Person, not Group of people</li>
</ul></li>
</ul>
</section>
<section id="data-quality-considerations" class="slide level2">
<h2>Data quality considerations</h2>
<p>General types of issues:</p>
<ul>
<li><p>High technical variability</p></li>
<li><p>Outliers</p></li>
<li><p>Batch artifacts (or other systematic trends)</p></li>
</ul>
<p>Effects on the data:</p>
<ul>
<li><p>Some will yield false positives (confounds, “false signals”)</p></li>
<li><p>Others will yield false negatives (by decreasing signal-to-noise)</p></li>
</ul>
</section>
<section id="filtering-your-data" class="slide level2">
<h2>Filtering your data</h2>
<ul>
<li><p>Here I mean “Removing part of the data from a sample” and doing that to all samples</p></li>
<li><p>In many studies, especially gene expression, it is common to remove genes that have no or very low signal (“not expressed”)</p></li>
<li><p>Deciding what to remove is often not straighforward, but make a principled decision and stick with it (see next slide)</p></li>
</ul>
<div class="callout callout-warning callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Filters must be “unsupervised”</strong></p>
</div>
<div class="callout-content">
<ul>
<li><p>Filtering strategy should treat all samples the same</p></li>
<li><p>Filtering strategy should be decided up front</p></li>
</ul>
</div>
</div>
</div>
</section>
<section id="batch-effects" class="slide level2">
<h2>Batch effects</h2>
<h4 id="batch-effects-are-sub-groups-of-measurements-that-have-qualitatively-different-behaviour-across-conditions-and-are-unrelated-to-the-biological-or-scientific-variables-in-a-study"><strong>“Batch effects are sub-groups of measurements that have qualitatively different behaviour across conditions and are unrelated to the biological or scientific variables in a study”</strong></h4>
<p><strong>Leek et al.&nbsp;2010 Nature Rev.&nbsp;Genetics 11:733</strong></p>

<img data-src="img/batch.png" class="r-stretch quarto-figure-center"><ul>
<li><p>Magnitude of batch effects vary</p></li>
<li><p>Consider correcting for them if possible - batch artifact detection and correction will be covered in more detail in a later lecture</p></li>
</ul>
</section>
<section id="avoiding-batch-artifacts" class="slide level2">
<h2>Avoiding batch artifacts</h2>
<ul>
<li>Don’t samples run in batches - <strong>Confounding</strong> (can be hard to avoid)</li>
<li>Balance or randomize design with respect to batches</li>
<li>Avoid (or at least record) obvious potential sources of artifacts, such as a new tube of a reagent, or a different person doing the bench work</li>
<li>Run some technical replicates across your batches</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a><span class="fu">table</span>(se<span class="sc">$</span>DPC, se<span class="sc">$</span>SeqRun)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      
        A  B  C  D  E  H
  12.5  8  0  0  0  0  0
  14.5  0  9  0  0  0  0
  17.5  0  0  5  5  0  0
  21    0  0  0  0 11  0
  77    0  0  0  0  0  6</code></pre>
</div>
</div>
</section>
<section id="outliers" class="slide level2">
<h2>Outliers</h2>
<p>Hard to define</p>
<ul>
<li><p>“A sample that deviates significantly from the rest of the samples in its class”</p></li>
<li><p>“An observation differing widely from the rest of the data.”</p></li>
<li><p>“A data point notably further out from the central value than the others. Outliers invite explanation as observational errors, or intrusions from another set, or something of that sort.”</p></li>
<li><p>“… a value that lies 1.5 IQR beyond the upper or lower quartile”</p></li>
</ul>
<p>Features (e.g.&nbsp;genes) are not <strong>usually</strong> called “outliers” in the sense of “should remove from the data”</p>
</section>
<section id="identifying-outlier-samples" class="slide level2">
<h2>Identifying outlier samples</h2>
<ul>
<li><p>Relative vs.&nbsp;absolute quality is important</p></li>
<li><p>We might consider a sample an outlier if (relative to others):</p>
<ul>
<li>It has “very low” signals</li>
<li>“High” background</li>
<li>“Low” correlation with most (or all) other samples from the same group</li>
<li>There may be technology-specific factors</li>
</ul></li>
<li><p>If a sample is questionable, we might ask: Is there anything suspect in the experimental process? (e.g.&nbsp;“Sample dropped on floor”)<br>
– this will help justify decisions to remove a sample beyond arbitrary criteria like “&gt;1.5 IQR”</p></li>
</ul>
</section>
<section id="a-basic-but-effective-outlier-detection-method" class="slide level2">
<h2>A basic but effective outlier detection method</h2>
<div class="columns">
<div class="column">
<p>A <strong>heatmap of the sample-sample correlation</strong> matrix (generally a useful diagnostic plot)</p>
<p>Expect correlations to be tighter <strong>within</strong> experimental groups than <strong>across</strong> groups</p>
<ul>
<li><p>There are no firm guidelines for evaluating this … but you could apply a rule like “out of &gt;1.5 IQR”</p></li>
<li><p>In practice, outliers are often pretty obvious</p></li>
</ul>
</div><div class="column">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="lect03-eda_files/figure-revealjs/unnamed-chunk-27-1.png" width="864"></p>
</div>
</div>
</div>
</div>
<p><img src="https://github.com/STAT540-UBC/stat540-ubc.github.io/raw/main/images/stat540-logo-s.png" class="slide-logo"></p>
<div class="footer footer-default">

</div>
</section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="lect03-eda_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="lect03-eda_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="lect03-eda_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="lect03-eda_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="lect03-eda_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="lect03-eda_files/libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="lect03-eda_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="lect03-eda_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="lect03-eda_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="lect03-eda_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="lect03-eda_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1600,

        height: 900,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>