<!DOCTYPE html>
<html lang="en"><head>
<script src="lect08-continuous_files/libs/clipboard/clipboard.min.js"></script>
<script src="lect08-continuous_files/libs/quarto-html/tabby.min.js"></script>
<script src="lect08-continuous_files/libs/quarto-html/popper.min.js"></script>
<script src="lect08-continuous_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="lect08-continuous_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="lect08-continuous_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="lect08-continuous_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="lect08-continuous_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.3.450">

  <meta name="author" content="Keegan Korthauer">
  <meta name="dcterms.date" content="2025-02-04">
  <title>Continuous models and intro to limma</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="lect08-continuous_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="lect08-continuous_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="lect08-continuous_files/libs/revealjs/dist/theme/quarto.css">
  <link href="lect08-continuous_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="lect08-continuous_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="lect08-continuous_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="lect08-continuous_files/libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="lect08-continuous_files/libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="lect08-continuous_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" data-background-color="#197aa0" data-background-image="https://github.com/STAT540-UBC/stat540-ubc.github.io/raw/main/images/stat540-logo-s.png" data-background-opacity="0.9" data-background-position="95% 90%" data-background-size="12%" class="quarto-title-block center">
  <h1 class="title">Continuous models and intro to <code>limma</code></h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Keegan Korthauer 
</div>
</div>
</div>

  <p class="date">February 4, 2025</p>
</section>
<section id="laying-the-foundation-of-linear-models" class="slide level2">
<h2>Laying the foundation of linear models</h2>

<img data-src="https://blog.leightonbroadcasting.com/hs-fs/hubfs/Bricks-and-mortar.jpg?width=300&amp;name=Bricks-and-mortar.jpg" class="r-stretch quarto-figure-center"><ul>
<li><p><strong><em>t</em></strong> <strong>-tests</strong> can be used to test the equality of 2 population means</p></li>
<li><p><strong>ANOVA</strong> can be used to test the equality of more than 2 population means</p></li>
<li><p><strong>Linear regression</strong> provides a general framework for modeling the relationship between a response variable and different types of explanatory variables</p>
<ul>
<li><p><em>t</em>-tests can be used to test the significance of <em>individual</em> coefficients</p></li>
<li><p><em>F</em>-tests can be used to test the simultaneous significance of <em>multiple</em> coefficients (e.g.&nbsp;multiple levels of a single categorical factor, or multiple factors at once)</p></li>
<li><p><em>F</em>-tests can be used to compare nested models (<strong>overall</strong> effects or <strong>goodness of fit</strong>)</p></li>
</ul></li>
</ul>
</section>
<section id="learning-objectives-for-today" class="slide level2">
<h2>Learning objectives for today</h2>
<ol type="1">
<li><p>Understand how linear regression represents <strong>continuous</strong> variables:</p>
<ol type="i">
<li><p>Be familiar with the intuition behind how the regression line is estimated (<strong>Ordinary Least Squares</strong>)</p></li>
<li><p>Interpret parameters in a <strong>multiple linear regression</strong> model with continuous and factor variables</p></li>
</ol></li>
<li><p>Explain the motivation behind specialized regression models in high-dimensional settings</p>
<ol type="i">
<li>List the advantages of the Empirical Bayes techniques in <code>limma</code> compared to traditional linear regression models</li>
</ol></li>
</ol>
</section>
<section id="what-if-we-treat-age-as-a-continuous-variable" class="slide level2">
<h2>What if we treat age as a continuous variable?</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="co"># read in our dataset</span></span>
<span id="cb1-2"><a href="#cb1-2"></a>eset <span class="ot">&lt;-</span> <span class="fu">getGEO</span>(<span class="st">"GSE4051"</span>, <span class="at">getGPL =</span> <span class="cn">FALSE</span>)[[<span class="dv">1</span>]]</span>
<span id="cb1-3"><a href="#cb1-3"></a></span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="co"># recode time points</span></span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="fu">pData</span>(eset) <span class="ot">&lt;-</span> <span class="fu">pData</span>(eset) <span class="sc">%&gt;%</span></span>
<span id="cb1-6"><a href="#cb1-6"></a>  <span class="fu">mutate</span>(<span class="at">sample_id =</span> geo_accession) <span class="sc">%&gt;%</span></span>
<span id="cb1-7"><a href="#cb1-7"></a>  <span class="fu">mutate</span>(<span class="at">dev_stage =</span>  <span class="fu">case_when</span>(</span>
<span id="cb1-8"><a href="#cb1-8"></a>    <span class="fu">grepl</span>(<span class="st">"E16"</span>, title) <span class="sc">~</span> <span class="st">"E16"</span>,</span>
<span id="cb1-9"><a href="#cb1-9"></a>    <span class="fu">grepl</span>(<span class="st">"P2"</span>, title) <span class="sc">~</span> <span class="st">"P2"</span>,</span>
<span id="cb1-10"><a href="#cb1-10"></a>    <span class="fu">grepl</span>(<span class="st">"P6"</span>, title) <span class="sc">~</span> <span class="st">"P6"</span>,</span>
<span id="cb1-11"><a href="#cb1-11"></a>    <span class="fu">grepl</span>(<span class="st">"P10"</span>, title) <span class="sc">~</span> <span class="st">"P10"</span>,</span>
<span id="cb1-12"><a href="#cb1-12"></a>    <span class="fu">grepl</span>(<span class="st">"4 weeks"</span>, title) <span class="sc">~</span> <span class="st">"P28"</span></span>
<span id="cb1-13"><a href="#cb1-13"></a>  )) <span class="sc">%&gt;%</span></span>
<span id="cb1-14"><a href="#cb1-14"></a>  <span class="fu">mutate</span>(<span class="at">genotype =</span> <span class="fu">case_when</span>(</span>
<span id="cb1-15"><a href="#cb1-15"></a>    <span class="fu">grepl</span>(<span class="st">"Nrl-ko"</span>, title) <span class="sc">~</span> <span class="st">"NrlKO"</span>,</span>
<span id="cb1-16"><a href="#cb1-16"></a>    <span class="fu">grepl</span>(<span class="st">"wt"</span>, title) <span class="sc">~</span> <span class="st">"WT"</span></span>
<span id="cb1-17"><a href="#cb1-17"></a>  ))</span>
<span id="cb1-18"><a href="#cb1-18"></a></span>
<span id="cb1-19"><a href="#cb1-19"></a><span class="co"># reorder factor levels, add continous age variable</span></span>
<span id="cb1-20"><a href="#cb1-20"></a><span class="fu">pData</span>(eset) <span class="ot">&lt;-</span> <span class="fu">pData</span>(eset) <span class="sc">%&gt;%</span></span>
<span id="cb1-21"><a href="#cb1-21"></a>  <span class="fu">mutate</span>(<span class="at">dev_stage =</span> <span class="fu">fct_relevel</span>(dev_stage, <span class="st">"E16"</span>, <span class="st">"P2"</span>, <span class="st">"P6"</span>, <span class="st">"P10"</span>, <span class="st">"P28"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb1-22"><a href="#cb1-22"></a>  <span class="fu">mutate</span>(<span class="at">genotype =</span> <span class="fu">as.factor</span>(genotype)) <span class="sc">%&gt;%</span></span>
<span id="cb1-23"><a href="#cb1-23"></a>  <span class="fu">mutate</span>(<span class="at">genotype =</span> <span class="fu">fct_relevel</span>(genotype, <span class="st">"WT"</span>, <span class="st">"NrlKO"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb1-24"><a href="#cb1-24"></a>  <span class="fu">mutate</span>(<span class="at">age =</span> <span class="fu">ifelse</span>(dev_stage <span class="sc">==</span> <span class="st">"E16"</span>, <span class="sc">-</span><span class="dv">4</span>,</span>
<span id="cb1-25"><a href="#cb1-25"></a>                            <span class="fu">ifelse</span>(dev_stage <span class="sc">==</span> <span class="st">"P2"</span>, <span class="dv">2</span>, </span>
<span id="cb1-26"><a href="#cb1-26"></a>                                   <span class="fu">ifelse</span>(dev_stage <span class="sc">==</span> <span class="st">"P6"</span>, <span class="dv">6</span>, </span>
<span id="cb1-27"><a href="#cb1-27"></a>                                          <span class="fu">ifelse</span>(dev_stage <span class="sc">==</span> <span class="st">"P10"</span>, <span class="dv">10</span>, <span class="dv">28</span>)))))</span>
<span id="cb1-28"><a href="#cb1-28"></a></span>
<span id="cb1-29"><a href="#cb1-29"></a><span class="co"># function to return tidy data that merges expression matrix and metadata</span></span>
<span id="cb1-30"><a href="#cb1-30"></a>toLongerMeta <span class="ot">&lt;-</span> <span class="cf">function</span>(expset) {</span>
<span id="cb1-31"><a href="#cb1-31"></a>    <span class="fu">stopifnot</span>(<span class="fu">class</span>(expset) <span class="sc">==</span> <span class="st">"ExpressionSet"</span>)</span>
<span id="cb1-32"><a href="#cb1-32"></a>    </span>
<span id="cb1-33"><a href="#cb1-33"></a>    expressionMatrix <span class="ot">&lt;-</span> lonExpressionressionMatrix <span class="ot">&lt;-</span> <span class="fu">exprs</span>(expset) <span class="sc">%&gt;%</span> </span>
<span id="cb1-34"><a href="#cb1-34"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb1-35"><a href="#cb1-35"></a>    <span class="fu">rownames_to_column</span>(<span class="st">"gene"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb1-36"><a href="#cb1-36"></a>    <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">!</span>gene, </span>
<span id="cb1-37"><a href="#cb1-37"></a>                 <span class="at">values_to =</span> <span class="st">"expression"</span>,</span>
<span id="cb1-38"><a href="#cb1-38"></a>                 <span class="at">names_to =</span> <span class="st">"sample_id"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb1-39"><a href="#cb1-39"></a>    <span class="fu">left_join</span>(<span class="fu">pData</span>(expset) <span class="sc">%&gt;%</span> <span class="fu">select</span>(sample_id, dev_stage, age, genotype),</span>
<span id="cb1-40"><a href="#cb1-40"></a>            <span class="at">by =</span> <span class="st">"sample_id"</span>)</span>
<span id="cb1-41"><a href="#cb1-41"></a>  <span class="fu">return</span>(expressionMatrix)</span>
<span id="cb1-42"><a href="#cb1-42"></a>}</span>
<span id="cb1-43"><a href="#cb1-43"></a></span>
<span id="cb1-44"><a href="#cb1-44"></a><span class="co"># pull out two genes of interest</span></span>
<span id="cb1-45"><a href="#cb1-45"></a>twoGenes <span class="ot">&lt;-</span> <span class="fu">toLongerMeta</span>(eset) <span class="sc">%&gt;%</span> </span>
<span id="cb1-46"><a href="#cb1-46"></a>  <span class="fu">filter</span>(gene <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"1456341_a_at"</span>, <span class="st">"1441811_x_at"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb1-47"><a href="#cb1-47"></a>  <span class="fu">mutate</span>(<span class="at">gene =</span> <span class="fu">ifelse</span>(gene <span class="sc">==</span> <span class="st">"1456341_a_at"</span>, <span class="st">"Klf9"</span>, <span class="st">"Tmem176a"</span>)) </span>
<span id="cb1-48"><a href="#cb1-48"></a></span>
<span id="cb1-49"><a href="#cb1-49"></a><span class="co"># make some plots - first with age continuous</span></span>
<span id="cb1-50"><a href="#cb1-50"></a>Klf9_C <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"Klf9"</span>), </span>
<span id="cb1-51"><a href="#cb1-51"></a>                  <span class="fu">aes</span>(<span class="at">x =</span> age, <span class="at">y =</span> expression)) <span class="sc">+</span> </span>
<span id="cb1-52"><a href="#cb1-52"></a>             <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb1-53"><a href="#cb1-53"></a>             <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Klf9"</span>) <span class="sc">+</span></span>
<span id="cb1-54"><a href="#cb1-54"></a>             <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>) <span class="sc">+</span></span>
<span id="cb1-55"><a href="#cb1-55"></a>             <span class="fu">ylim</span>(<span class="dv">5</span>, <span class="dv">11</span>) <span class="sc">+</span></span>
<span id="cb1-56"><a href="#cb1-56"></a>             <span class="fu">xlab</span>(<span class="st">"age (days)"</span>) </span>
<span id="cb1-57"><a href="#cb1-57"></a></span>
<span id="cb1-58"><a href="#cb1-58"></a>Tmem176a_C <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"Tmem176a"</span>), </span>
<span id="cb1-59"><a href="#cb1-59"></a>                 <span class="fu">aes</span>(<span class="at">x =</span> age, <span class="at">y =</span> expression)) <span class="sc">+</span> </span>
<span id="cb1-60"><a href="#cb1-60"></a>             <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb1-61"><a href="#cb1-61"></a>             <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Tmem176a"</span>) <span class="sc">+</span></span>
<span id="cb1-62"><a href="#cb1-62"></a>             <span class="fu">ylim</span>(<span class="dv">5</span>, <span class="dv">11</span>) <span class="sc">+</span></span>
<span id="cb1-63"><a href="#cb1-63"></a>             <span class="fu">xlab</span>(<span class="st">"age (days)"</span>) </span>
<span id="cb1-64"><a href="#cb1-64"></a></span>
<span id="cb1-65"><a href="#cb1-65"></a><span class="co"># next with age categorical</span></span>
<span id="cb1-66"><a href="#cb1-66"></a>Klf9 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"Klf9"</span>), </span>
<span id="cb1-67"><a href="#cb1-67"></a>                  <span class="fu">aes</span>(<span class="at">x =</span> dev_stage, <span class="at">y =</span> expression)) <span class="sc">+</span> </span>
<span id="cb1-68"><a href="#cb1-68"></a>             <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb1-69"><a href="#cb1-69"></a>             <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>) <span class="sc">+</span></span>
<span id="cb1-70"><a href="#cb1-70"></a>             <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Klf9"</span>) <span class="sc">+</span></span>
<span id="cb1-71"><a href="#cb1-71"></a>             <span class="fu">ylim</span>(<span class="dv">5</span>, <span class="dv">11</span>) <span class="sc">+</span></span>
<span id="cb1-72"><a href="#cb1-72"></a>             <span class="fu">xlab</span>(<span class="st">"Developmental Stage"</span>)</span>
<span id="cb1-73"><a href="#cb1-73"></a></span>
<span id="cb1-74"><a href="#cb1-74"></a>Tmem176a <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"Tmem176a"</span>), </span>
<span id="cb1-75"><a href="#cb1-75"></a>                 <span class="fu">aes</span>(<span class="at">x =</span> dev_stage, <span class="at">y =</span> expression)) <span class="sc">+</span> </span>
<span id="cb1-76"><a href="#cb1-76"></a>             <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb1-77"><a href="#cb1-77"></a>             <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Tmem176a"</span>) <span class="sc">+</span></span>
<span id="cb1-78"><a href="#cb1-78"></a>             <span class="fu">ylim</span>(<span class="dv">5</span>, <span class="dv">11</span>) <span class="sc">+</span></span>
<span id="cb1-79"><a href="#cb1-79"></a>             <span class="fu">xlab</span>(<span class="st">"Developmental Stage"</span>) </span>
<span id="cb1-80"><a href="#cb1-80"></a></span>
<span id="cb1-81"><a href="#cb1-81"></a><span class="fu">grid.arrange</span>(Klf9, </span>
<span id="cb1-82"><a href="#cb1-82"></a>             Tmem176a <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">""</span>), </span>
<span id="cb1-83"><a href="#cb1-83"></a>             Klf9_C, </span>
<span id="cb1-84"><a href="#cb1-84"></a>             Tmem176a_C <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">""</span>), </span>
<span id="cb1-85"><a href="#cb1-85"></a>             <span class="at">nrow =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="lect08-continuous_files/figure-revealjs/unnamed-chunk-2-1.png" width="1344" class="r-stretch quarto-figure-center"></section>
<section id="linear-model-with-age-as-continuous-covariate" class="slide level2">
<h2>Linear model with age as continuous covariate</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="fu">grid.arrange</span>(Klf9_C  <span class="sc">+</span> <span class="fu">geom_smooth</span>(<span class="at">method=</span><span class="st">'lm'</span>, <span class="at">colour =</span> <span class="st">"grey"</span>, <span class="at">se =</span> <span class="cn">FALSE</span>), </span>
<span id="cb2-2"><a href="#cb2-2"></a>             Tmem176a_C <span class="sc">+</span> <span class="fu">geom_smooth</span>(<span class="at">method=</span><span class="st">'lm'</span>, <span class="at">colour =</span> <span class="st">"grey"</span>, <span class="at">se =</span> <span class="cn">FALSE</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">""</span>), </span>
<span id="cb2-3"><a href="#cb2-3"></a>             <span class="at">nrow =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="lect08-continuous_files/figure-revealjs/unnamed-chunk-3-1.png" width="1344" class="r-stretch quarto-figure-center"><ul>
<li><p>Linear looks reasonable for gene Tmem176a, but not so much for Klf9</p></li>
<li><p>For now, assume linear is reasonable</p></li>
</ul>
</section>
<section id="simple-linear-regression-matrix-form" class="slide level2">
<h2>Simple Linear Regression (Matrix form)</h2>
<p><span class="math display">\[ \mathbf{Y = X \boldsymbol\alpha + \boldsymbol\varepsilon}\]</span></p>
<p>For 1 continuous/quantitative covariate:</p>
<p><span class="math display">\[\mathbf{Y} = \begin{bmatrix}
  y_{1} \\
  y_{2} \\
  \vdots \\
  y_{n} \\
\end{bmatrix}, \hspace{1em}
\mathbf{X} = \begin{bmatrix}
  1 &amp; x_{1} \\
  1 &amp; x_{2} \\
  \vdots &amp; \vdots \\
  1 &amp; x_{n} \\
\end{bmatrix}, \hspace{1em}
\boldsymbol\alpha = \begin{bmatrix}
  \alpha_{0} \\
  \alpha_{1} \\
\end{bmatrix}, \hspace{1em}
\boldsymbol\varepsilon=\begin{bmatrix}
  \varepsilon_{1} \\
  \varepsilon_{2} \\
  \vdots \\
  \varepsilon_{n} \\
\end{bmatrix}\]</span></p>
<ul>
<li><p><span class="math inline">\(\alpha_0=\)</span> the <strong>intercept</strong> (expected value of <span class="math inline">\(y\)</span> when <span class="math inline">\(x\)</span> is equal to zero)</p></li>
<li><p><span class="math inline">\(\alpha_1=\)</span> the <strong>slope</strong> (expected change in <span class="math inline">\(y\)</span> for every one-unit increase in <span class="math inline">\(x\)</span>)</p></li>
</ul>
</section>
<section id="simple-linear-regression-matrix-form-1" class="slide level2">
<h2>Simple Linear Regression (Matrix form)</h2>
<p><span class="math display">\[ \mathbf{Y = X \boldsymbol\alpha + \boldsymbol\varepsilon}\]</span></p>
<p>Remember / convince yourself that the matrix algebra yields simple linear equations: <span class="math display">\[\begin{bmatrix}
  y_{1} \\
  y_{2} \\
  \vdots \\
  y_{n} \\
\end{bmatrix}=\begin{bmatrix}
  1 &amp; x_{1} \\
  1 &amp; x_{2} \\
  \vdots &amp; \vdots \\
  1 &amp; x_{n} \\
\end{bmatrix}\begin{bmatrix}
  \alpha_{0} \\
  \alpha_{1} \\
\end{bmatrix}+\begin{bmatrix}
  \varepsilon_{1} \\
  \varepsilon_{2} \\
  \vdots \\
  \varepsilon_{n} \\
\end{bmatrix}=\begin{bmatrix}
  1*\alpha_0 + x_{1}\alpha_{1} \\
  1*\alpha_0 + x_{2}\alpha_{1} \\
   \vdots \\
  1*\alpha_0 + x_{n}\alpha_{1} \\
\end{bmatrix}+\begin{bmatrix}
  \varepsilon_{1} \\
  \varepsilon_{2} \\
  \vdots \\
  \varepsilon_{n} \\
\end{bmatrix}\]</span></p>
<p><span class="math display">\[=\begin{bmatrix}
  \alpha_0 + x_{1}\alpha_{1} + \varepsilon_{1} \\
  \alpha_0 + x_{2}\alpha_{1} + \varepsilon_{2}\\
   \vdots \\
  \alpha_0 + x_{n}\alpha_{1} + \varepsilon_{n} \\
\end{bmatrix}\]</span> <span class="math display">\[\Rightarrow y_i = \alpha_0 + x_i\alpha_1 + \varepsilon_i\]</span></p>
</section>
<section id="slr-with-continuous-age-covariate" class="slide level2">
<h2>SLR with continuous age covariate</h2>
<div class="columns">
<div class="column">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="lect08-continuous_files/figure-revealjs/unnamed-chunk-4-1.png" width="768"></p>
</div>
</div>
</div><div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>Tmem_fit <span class="ot">&lt;-</span> <span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"Tmem176a"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb3-2"><a href="#cb3-2"></a>  <span class="fu">lm</span>(expression <span class="sc">~</span> age, <span class="at">data =</span> .)</span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="fu">tidy</span>(Tmem_fit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 5
  term        estimate std.error statistic  p.value
  &lt;chr&gt;          &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)   8.10     0.114       71.1  3.58e-41
2 age          -0.0614   0.00821     -7.47 6.74e- 9</code></pre>
</div>
</div>
<p>Interpretation of <strong>intercept</strong>:</p>
<p><span class="math inline">\(H_0: \alpha_0 = 0\)</span> tests the null hypothesis that the intercept is zero - usually, not of interest</p>
</div>
</div>
</section>
<section id="slr-with-continuous-age-covariate-1" class="slide level2">
<h2>SLR with continuous age covariate</h2>
<div class="columns">
<div class="column">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="lect08-continuous_files/figure-revealjs/unnamed-chunk-6-1.png" width="768"></p>
</div>
</div>
</div><div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="fu">tidy</span>(Tmem_fit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 5
  term        estimate std.error statistic  p.value
  &lt;chr&gt;          &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)   8.10     0.114       71.1  3.58e-41
2 age          -0.0614   0.00821     -7.47 6.74e- 9</code></pre>
</div>
</div>
<p>Interpretation of <strong>slope</strong>:</p>
<p><span class="math inline">\(H_0: \alpha_1 = 0\)</span> tests the null hypothesis that there is no association between gene expression and age - usually of interest</p>
</div>
</div>
</section>
<section id="how-do-we-estimate-the-intercept-and-slope" class="slide level2">
<h2>How do we estimate the intercept and slope?</h2>
<p>Why is this the <strong>optimal</strong> line?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="fu">tidy</span>(Tmem_fit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 5
  term        estimate std.error statistic  p.value
  &lt;chr&gt;          &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)   8.10     0.114       71.1  3.58e-41
2 age          -0.0614   0.00821     -7.47 6.74e- 9</code></pre>
</div>
</div>
</section>
<section id="which-one-is-the-best-line" class="slide level2">
<h2>Which one is the <em>best</em> line?</h2>

<img data-src="lect08-continuous_files/figure-revealjs/unnamed-chunk-9-1.png" width="768" class="r-stretch quarto-figure-center"></section>
<section id="ordinary-least-squares" class="slide level2">
<h2>Ordinary Least Squares</h2>
<div class="columns">
<div class="column">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="lect08-continuous_files/figure-revealjs/unnamed-chunk-10-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</div><div class="column">
<ul>
<li><p><strong>Ordinary Least Squares (OLS)</strong> regression: parameter estimates minimize the sum of squared errors</p></li>
<li><p><strong>Error</strong>: vertical <span class="math inline">\((y)\)</span> distance between the true regression line (unobserved) and the real observation</p></li>
<li><p><strong>Residual</strong>: vertical <span class="math inline">\((y)\)</span> distance between the fitted regression line and the real observation (estimated error)</p></li>
</ul>
</div>
</div>
</section>
<section id="ols-estimator-for-simple-linear-regression-1-covariate" class="slide level2">
<h2>OLS Estimator for Simple Linear Regression (1 covariate)</h2>
<ul>
<li><p>Mathematically: <span class="math inline">\(\varepsilon_i\)</span> represents the error: <span class="math inline">\(\varepsilon_i = y_i - \alpha_0 - \alpha_1x_i, i = 1, ..., n\)</span></p></li>
<li><p>We want to find the line (i.e.&nbsp;intercept and slope) that minimizes the <strong>sum of squared errors</strong>: <span class="math display">\[S(\alpha_0, \alpha_1)= \sum_{i=1}^n (y_i - \alpha_0 - \alpha_1 x_i)^2\]</span></p>
<ul>
<li><p><span class="math inline">\(S(\alpha_0, \alpha_1)\)</span> is called an <em>objective function</em></p></li>
<li><p>the observed sum of squared errors is referred to as <strong>Residual Sum of Squares (RSS)</strong></p></li>
</ul></li>
</ul>
<div class="fragment">
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Error vs Residual</strong></p>
</div>
<div class="callout-content">
<ul>
<li><p><strong>Error</strong> refers to the deviation of the observed value to the (underlying) true quantity of interest</p></li>
<li><p><strong>Residual</strong> refers to the difference between the observed and estimated quantity of interest</p></li>
</ul>
</div>
</div>
</div>
<ul>
<li>How to obtain estimates <span class="math inline">\((\hat{\alpha}_0, \hat{\alpha}_1)\)</span> ? Let’s look at a more general case</li>
</ul>
</div>
</section>
<section id="ols-for-multiple-linear-regression-p-covariates" class="slide level2">
<h2>OLS for Multiple Linear Regression (p covariates)</h2>
<p><span class="math display">\[S(\alpha_0, \alpha_1, \alpha_2, ..., \alpha_p) = \sum_{i=1}^n (y_i - \alpha_0 - \alpha_1 x_{1i} - \alpha_2 x_{2i} - ... - \alpha_p x_{pi})^2\]</span> <span class="math display">\[=(\mathbf{y}-\mathbf{X}\boldsymbol\alpha)^T(\mathbf{y}-\mathbf{X}\boldsymbol\alpha)\]</span></p>
<ul>
<li>We need to find values of <span class="math inline">\(\boldsymbol\alpha=(\alpha_0, \alpha_1, ..., \alpha_p)\)</span> that minimize the sum of squares <span class="math inline">\(S\)</span></li>
</ul>
<div class="fragment">
<ul>
<li>Take partial derivatives with respect to each coeff, set to zero, solve system of equations:</li>
</ul>
<div class="columns">
<div class="column">
<p><span class="math display">\[\small \frac{\partial S}{\partial\boldsymbol\alpha}=\begin{bmatrix}
  \frac{\partial S}{\partial\alpha_0} \\
  \frac{\partial S}{\partial\alpha_1} \\
  \vdots\\
  \frac{\partial S}{\partial\alpha_p} \\
\end{bmatrix}=\begin{bmatrix}
  0 \\
  0 \\
  \vdots\\
  0 \\
\end{bmatrix}\]</span></p>
</div><div class="column">
<p><img data-src="https://xaktly.com/Images/Mathematics/TheDerivative/DerivativeOfXSquared.png"></p>
</div>
</div>
</div>
</section>
<section id="sums-of-squares-for-tmem176a" class="slide level2">
<h2>Sums of squares for Tmem176a</h2>
<p>Fixing the intercept at 8.1000793, let’s plot the RSS values for a range of possible slope values.</p>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>rss <span class="ot">&lt;-</span> <span class="cf">function</span>(intercept, slope, x, y, data){</span>
<span id="cb9-2"><a href="#cb9-2"></a>  resid <span class="ot">&lt;-</span> <span class="fu">pull</span>(data, y) <span class="sc">-</span> (intercept <span class="sc">+</span> slope<span class="sc">*</span><span class="fu">pull</span>(data, x))</span>
<span id="cb9-3"><a href="#cb9-3"></a>  <span class="fu">return</span>(<span class="fu">sum</span>(resid<span class="sc">^</span><span class="dv">2</span>))</span>
<span id="cb9-4"><a href="#cb9-4"></a>}</span>
<span id="cb9-5"><a href="#cb9-5"></a></span>
<span id="cb9-6"><a href="#cb9-6"></a>slopes <span class="ot">=</span> <span class="fu">seq</span>(<span class="sc">-</span><span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="at">len =</span> <span class="dv">400</span>)</span>
<span id="cb9-7"><a href="#cb9-7"></a>results <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">slope =</span> slopes,</span>
<span id="cb9-8"><a href="#cb9-8"></a>                      <span class="at">RSS =</span> <span class="fu">sapply</span>(slopes, </span>
<span id="cb9-9"><a href="#cb9-9"></a>                                   rss, </span>
<span id="cb9-10"><a href="#cb9-10"></a>                                   <span class="at">intercept =</span> Tmem_fit<span class="sc">$</span>coefficients[<span class="dv">1</span>],</span>
<span id="cb9-11"><a href="#cb9-11"></a>                                   <span class="at">x =</span> <span class="st">"age"</span>, </span>
<span id="cb9-12"><a href="#cb9-12"></a>                                   <span class="at">y =</span> <span class="st">"expression"</span>,</span>
<span id="cb9-13"><a href="#cb9-13"></a>                                   <span class="at">data =</span> <span class="fu">filter</span>(twoGenes, gene <span class="sc">==</span> <span class="st">"Tmem176a"</span>)))</span>
<span id="cb9-14"><a href="#cb9-14"></a>best.i <span class="ot">&lt;-</span> <span class="fu">which.min</span>(results<span class="sc">$</span>RSS)</span>
<span id="cb9-15"><a href="#cb9-15"></a><span class="fu">ggplot</span>(results, <span class="fu">aes</span>(slope, RSS)) <span class="sc">+</span> </span>
<span id="cb9-16"><a href="#cb9-16"></a>  <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb9-17"><a href="#cb9-17"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> results[best.i,], <span class="at">colour =</span> <span class="st">"red"</span>, <span class="at">size =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="lect08-continuous_files/figure-revealjs/unnamed-chunk-11-1.png" width="960" class="r-stretch quarto-figure-center"></section>
<section id="ols-interactive-demo" class="slide level2">
<h2><a href="http://setosa.io/ev/ordinary-least-squares-regression/" data-preview-link="true">OLS interactive demo</a></h2>
<p><a href="http://setosa.io/ev/ordinary-least-squares-regression/" data-preview-link="true">Launch demo</a></p>
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Explore</strong></p>
</div>
<div class="callout-content">
<ol type="1">
<li><p>In the first plot, drag individual points around and observe what changes in second plot</p></li>
<li><p>In the second plot, adjust the slope and intercept dials - what happens to the total area of the squares in the second plot when you modify the slope and intercept from the default values?</p>
<ul>
<li>Note that you can reset to default values by refreshing the page</li>
</ul></li>
</ol>
</div>
</div>
</div>
</section>
<section id="properties-of-ols-regression" class="slide level2">
<h2>Properties of OLS regression</h2>
<p><strong>Regression model</strong>: <span class="math inline">\(\mathbf{Y = X \boldsymbol\alpha + \boldsymbol\varepsilon}\)</span></p>
<p><strong>OLS estimator</strong>: <span class="math inline">\(\hat{\boldsymbol\alpha} =(\mathbf{X}^T\mathbf{X})^{-1}\mathbf{X}^T\mathbf{y}\)</span></p>
<p><strong>Fitted/predicted values</strong>: <span class="math inline">\(\hat{\mathbf{y}} = \mathbf{X} \hat{\boldsymbol\alpha} = \mathbf{X} (\mathbf{X}^T\mathbf{X})^{-1}\mathbf{X}^T\mathbf{y} = \mathbf{H}\mathbf{y}\)</span></p>
<p>where <span class="math inline">\(\mathbf{H}=\mathbf{X} (\mathbf{X}^T\mathbf{X})^{-1}\mathbf{X}^T\)</span> is called the “hat” or projection matrix</p>
<div class="fragment">
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Assumptions of OLS Regression</strong></p>
</div>
<div class="callout-content">
<ol type="1">
<li><p><span class="math inline">\(\boldsymbol\varepsilon\)</span> have mean zero</p></li>
<li><p><span class="math inline">\(\boldsymbol\varepsilon\)</span> are iid (implies constant variance)</p></li>
<li><p><em>(only required for hypothesis testing in small sample settings)</em> <span class="math inline">\(\boldsymbol\varepsilon\)</span> are Normally distributed</p></li>
</ol>
</div>
</div>
</div>
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Connection to other estimators</strong></p>
</div>
<div class="callout-content">
<p>If <span class="math inline">\(\boldsymbol\varepsilon\)</span> are iid <strong>Normal</strong>, then OLS estimator is also MLE (Maximum Likelihood Estimator)</p>
</div>
</div>
</div>
</div>
</section>
<section id="properties-of-ols-regression-contd" class="slide level2">
<h2>Properties of OLS regression (cont’d)</h2>
<p><strong>Residuals</strong>: (note NOT the same as errors <span class="math inline">\(\boldsymbol\varepsilon\)</span>) <span class="math display">\[\hat{\boldsymbol\varepsilon} = \mathbf{y} - \hat{\mathbf{y}} = \mathbf{y} -  \mathbf{X} \hat{\boldsymbol\alpha}\]</span></p>
<p><strong>Estimated error (residual) variance</strong>: <span class="math display">\[\hat{\sigma}^2 =  \frac{1}{n-p-1} \hat{\boldsymbol\varepsilon}^T \hat{\boldsymbol\varepsilon}\]</span></p>
<p><strong>Estimated covariance matrix of</strong> <span class="math inline">\(\hat{\boldsymbol\alpha}\)</span>: <span class="math display">\[\hat{Var}(\hat{\boldsymbol\alpha}) = \hat{\sigma}^2(\mathbf{X}^T\mathbf{X})^{-1}\]</span></p>
<p><strong>Estimated standard errors for estimated regression coefficients</strong>: <span class="math inline">\(\hat{se}(\hat{\alpha}_j)\)</span>, obtained by taking the square root of the diagonal elements of <span class="math inline">\(\hat{Var}(\hat{\boldsymbol\alpha})\)</span></p>
</section>
<section id="inference-in-regression-normal-iid-errors" class="slide level2">
<h2>Inference in Regression (normal iid errors)</h2>
<p>How to test <span class="math inline">\(H_0: \alpha_j = 0\)</span>?</p>
<p>With a <strong><em>t</em>-test</strong>!</p>
<p>Under <span class="math inline">\(H_0\)</span>,</p>
<p><span class="math display">\[\frac{\hat{\alpha}_j}{\hat{se}(\hat{\alpha}_j)} \sim t_{n-p-1}\]</span></p>
<p>So a <em>p</em>-value is obtained by computing a tail probability for the observed value of <span class="math inline">\(\hat{\alpha}_j\)</span> from a <span class="math inline">\(t_{n-p-1}\)</span> distribution</p>
</section>
<section id="inference---what-if-we-dont-assume-normal-errors" class="slide level2">
<h2>Inference - what if we don’t assume Normal errors?</h2>
<div class="fragment">
<p>How to test <span class="math inline">\(H_0: \alpha_j = 0\)</span>?</p>
<p>Assuming large enough sample size, with a <strong><em>t</em>-test</strong>!</p>
<p>Under <span class="math inline">\(H_0\)</span>, the _____________ tells us that as <span class="math inline">\(n\)</span> gets large, even if errors aren’t normal:</p>
<p><span class="math display">\[\frac{\hat{\alpha}_j}{\hat{se}(\hat{\alpha}_j)} \sim t_{n-p-1}\]</span></p>
<p>So <em>with a large enough sample size</em> a <em>p</em>-value for this hypothesis test is obtained by computing a tail probability for the observed value of <span class="math inline">\(\hat{\alpha}_j\)</span> from a <span class="math inline">\(t_{n-p-1}\)</span> distribution</p>
</div>
</section>
<section id="diagnostics-plots" class="slide level2">
<h2>Diagnostics plots</h2>
<div class="columns">
<div class="column" style="width:35%;">
<p>Do our assumptions hold?</p>
<ul>
<li>Constant variance</li>
<li>iid errors</li>
<li>Normality of errors</li>
</ul>
</div><div class="column" style="width:65%;">
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="co"># requires package 'ggResidpanel'</span></span>
<span id="cb10-2"><a href="#cb10-2"></a>ggResidpanel<span class="sc">::</span><span class="fu">resid_panel</span>(Tmem_fit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img data-src="lect08-continuous_files/figure-revealjs/unnamed-chunk-12-1.png" width="960"></p>
</div>
</div>
</div>
</div>
</section>
<section id="linear-regression" class="slide level2">
<h2><em>Linear</em> regression</h2>
<p>The nature of the regression function <span class="math inline">\(y=f(x|\boldsymbol\alpha)\)</span> is one of the defining characteristics of a regression model</p>
<ol type="1">
<li><p>If <span class="math inline">\(f\)</span> is not linear in <span class="math inline">\(\boldsymbol\alpha \Rightarrow\)</span> <strong>nonlinear model</strong></p>
<ul>
<li>For example, consider nonlinear parametric regression: <span class="math display">\[ y_i = \frac{1}{1 + e^{\alpha_0 + \alpha_1 x_i}} + \varepsilon_i\]</span></li>
</ul></li>
<li><p>If <span class="math inline">\(f\)</span> is linear in <span class="math inline">\(\boldsymbol\alpha \Rightarrow\)</span> <strong>linear model</strong></p>
<ul>
<li><p>We just examined simple linear regression (a linear model): <span class="math inline">\(y_i = \alpha_0 + \alpha_1x_i + \varepsilon_i\)</span></p></li>
<li><p>What we could do instead: polynomial regression (also a linear model) <span class="math display">\[y_i = \alpha_0 + \alpha_1x_i + \alpha_2x_i^2 + \varepsilon_i\]</span></p></li>
</ul></li>
</ol>
</section>
<section id="polynomial-regression" class="slide level2">
<h2>Polynomial regression</h2>
<div class="panel-tabset">
<ul id="tabset-1" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-1-1"><code>lm</code> output</a></li><li><a href="#tabset-1-2">Plot</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1">
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>oneGene <span class="ot">&lt;-</span> <span class="fu">toLongerMeta</span>(eset) <span class="sc">%&gt;%</span> </span>
<span id="cb11-2"><a href="#cb11-2"></a>  <span class="fu">filter</span>(gene <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"1427275_at"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb11-3"><a href="#cb11-3"></a>  <span class="fu">mutate</span>(<span class="at">gene =</span> <span class="st">"Smc4"</span>)</span>
<span id="cb11-4"><a href="#cb11-4"></a><span class="fu">lm</span>(expression <span class="sc">~</span> age <span class="sc">+</span> <span class="fu">I</span>(age<span class="sc">^</span><span class="dv">2</span>), <span class="at">data =</span> oneGene) <span class="sc">%&gt;%</span> <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 5
  term        estimate std.error statistic  p.value
  &lt;chr&gt;          &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)  8.48      0.161       52.7  1.11e-35
2 age         -0.147     0.0326      -4.52 6.52e- 5
3 I(age^2)     0.00501   0.00116      4.30 1.23e- 4</code></pre>
</div>
</div>
</div>
<div id="tabset-1-2">
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>Smc4 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(oneGene, <span class="fu">aes</span>(<span class="at">x =</span> age, <span class="at">y =</span> expression)) <span class="sc">+</span> </span>
<span id="cb13-2"><a href="#cb13-2"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb13-3"><a href="#cb13-3"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Smc4"</span>) <span class="sc">+</span></span>
<span id="cb13-4"><a href="#cb13-4"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>) <span class="sc">+</span></span>
<span id="cb13-5"><a href="#cb13-5"></a>  <span class="fu">ylim</span>(<span class="dv">6</span>, <span class="dv">11</span>) <span class="sc">+</span></span>
<span id="cb13-6"><a href="#cb13-6"></a>  <span class="fu">xlab</span>(<span class="st">"age (days)"</span>) <span class="sc">+</span> </span>
<span id="cb13-7"><a href="#cb13-7"></a>  <span class="fu">stat_smooth</span>(<span class="at">method=</span><span class="st">"lm"</span>, <span class="at">se=</span><span class="cn">FALSE</span>, <span class="at">fill=</span><span class="cn">NA</span>, </span>
<span id="cb13-8"><a href="#cb13-8"></a>              <span class="at">formula=</span>y <span class="sc">~</span> <span class="fu">poly</span>(x, <span class="dv">2</span>),<span class="at">colour=</span><span class="st">"blue"</span>)</span>
<span id="cb13-9"><a href="#cb13-9"></a></span>
<span id="cb13-10"><a href="#cb13-10"></a>Smc4</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="lect08-continuous_files/figure-revealjs/unnamed-chunk-14-1.png" width="768"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="callout callout-important callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Important</strong></p>
</div>
<div class="callout-content">
<p>Note that <strong>this is still a linear model</strong>, because it is linear in the <span class="math inline">\(\alpha_j\)</span>.</p>
<p>Hint: think about what the design matrix looks like in this case.</p>
</div>
</div>
</div>
</section>
<section id="putting-it-all-together-continuous-categorical-variables" class="slide level2">
<h2>Putting it all together (continuous + categorical variables)</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>Tmem176a_B <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">filter</span>(twoGenes, gene<span class="sc">==</span><span class="st">"Tmem176a"</span>), </span>
<span id="cb14-2"><a href="#cb14-2"></a>                  <span class="fu">aes</span>(<span class="at">x =</span> age, <span class="at">y =</span> expression, <span class="at">colour =</span> genotype)) <span class="sc">+</span> </span>
<span id="cb14-3"><a href="#cb14-3"></a>             <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb14-4"><a href="#cb14-4"></a>             <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Tmem176a"</span>) <span class="sc">+</span></span>
<span id="cb14-5"><a href="#cb14-5"></a>             <span class="fu">ylim</span>(<span class="dv">5</span>, <span class="dv">10</span>) <span class="sc">+</span></span>
<span id="cb14-6"><a href="#cb14-6"></a>             <span class="fu">xlab</span>(<span class="st">"age (days)"</span>) <span class="sc">+</span> </span>
<span id="cb14-7"><a href="#cb14-7"></a>             <span class="fu">stat_smooth</span>(<span class="at">method=</span><span class="st">"lm"</span>, <span class="at">se=</span><span class="cn">FALSE</span>)</span>
<span id="cb14-8"><a href="#cb14-8"></a>Tmem176a_B </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="lect08-continuous_files/figure-revealjs/unnamed-chunk-15-1.png" width="864" class="r-stretch quarto-figure-center"></section>
<section id="interaction-between-continuous-and-categorical-variables" class="slide level2">
<h2>Interaction between continuous and categorical variables</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a><span class="fu">lm</span>(expression <span class="sc">~</span> age<span class="sc">*</span>genotype, <span class="at">data =</span> <span class="fu">filter</span>(twoGenes, gene<span class="sc">==</span><span class="st">"Tmem176a"</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb15-2"><a href="#cb15-2"></a>  <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 5
  term              estimate std.error statistic  p.value
  &lt;chr&gt;                &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)        8.03       0.157     51.3   1.57e-34
2 age               -0.0665     0.0114    -5.82  1.33e- 6
3 genotypeNrlKO      0.142      0.228      0.623 5.37e- 1
4 age:genotypeNrlKO  0.00987    0.0164     0.600 5.52e- 1</code></pre>
</div>
</div>
<p><code>(Intercept)</code>: Intercept of reference line (_____)</p>
<p><code>age</code>: Slope of reference line (_____)</p>
<p><code>genotypeNrlKO</code>: Difference in intercepts (_____ vs _____)</p>
<p><code>age:genotypeNrlKO</code>: Difference in slopes (_____ vs _____)</p>
</section>
<section id="interpreting-the-intercept" class="slide level2">
<h2>Interpreting the Intercept</h2>

<img data-src="lect08-continuous_files/figure-revealjs/unnamed-chunk-17-1.png" width="960" class="r-stretch quarto-figure-center"><div class="callout callout-important callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Important</strong></p>
</div>
<div class="callout-content">
<p>Intercept terms refer to the estimates when the continuous covariate is equal to zero. This is not usually very interesting on its own</p>
</div>
</div>
</div>
</section>
<section id="interaction-between-continuous-and-categorical-variables-1" class="slide level2">
<h2>Interaction between continuous and categorical variables</h2>
<p><span class="math display">\[y_{ij} = \alpha_{0} + \tau_{KO}x_{ij, KO} + \tau_{Age} x_{ij, Age} + \tau_{KO:Age} x_{ij, KO}x_{ij, Age}\]</span></p>
<p>where</p>
<ul>
<li><span class="math inline">\(j \in \{ WT, NrlKO\}\)</span>, <span class="math inline">\(i = 1,2,...,n_j\)</span></li>
<li><span class="math inline">\(x_{ij, KO}\)</span> is the indicator for WT vs KO ( <span class="math inline">\(x_{ij, KO}=1\)</span> for <span class="math inline">\(j=NrlKO\)</span> and 0 for <span class="math inline">\(j=WT\)</span> )</li>
<li><span class="math inline">\(x_{ij, Age}\)</span> is the continuous age covariate</li>
</ul>
<p>Interpretation of parameters:</p>
<ul>
<li><span class="math inline">\(\alpha_0\)</span> is the expected expression <em>in WT</em> for age = ____</li>
<li>The “intercept” for the knockouts is:</li>
<li>The expected increase in expression <em>in WT</em> for every 1 day increase in age is:</li>
<li>The slope for the knockouts is:</li>
</ul>
<aside class="notes">
<p><span class="math inline">\(\alpha_0 + \tau_{KO}\)</span> <span class="math inline">\(\tau_{Age}\)</span> <span class="math inline">\(\tau_{Age} + \tau_{KO:Age}\)</span></p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="nested-models" class="slide level2">
<h2>Nested models</h2>
<p>As always, you can assess the relevance of several terms at once (e.g.&nbsp;everything involving genotype) with an <strong><em>F</em>-test</strong>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>Klf9dat <span class="ot">&lt;-</span> <span class="fu">filter</span>(twoGenes, gene<span class="sc">==</span><span class="st">"Klf9"</span>)</span>
<span id="cb17-2"><a href="#cb17-2"></a><span class="fu">anova</span>(<span class="fu">lm</span>(expression <span class="sc">~</span> age<span class="sc">*</span>genotype, <span class="at">data =</span> Klf9dat),</span>
<span id="cb17-3"><a href="#cb17-3"></a>      <span class="fu">lm</span>(expression <span class="sc">~</span> age, <span class="at">data =</span> Klf9dat))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Variance Table

Model 1: expression ~ age * genotype
Model 2: expression ~ age
  Res.Df    RSS Df Sum of Sq      F Pr(&gt;F)
1     35 45.948                           
2     37 45.984 -2 -0.036415 0.0139 0.9862</code></pre>
</div>
</div>
<div class="fragment">
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Conclusion</strong></p>
</div>
<div class="callout-content">
<p>We don’t have evidence that genotype affects the intercept or the slope</p>
</div>
</div>
</div>
</div>
</section>
<section id="f-tests-in-regression" class="slide level2">
<h2><em>F</em>-tests in regression</h2>
<table>
<colgroup>
<col style="width: 15%">
<col style="width: 37%">
<col style="width: 18%">
<col style="width: 18%">
<col style="width: 9%">
</colgroup>
<thead>
<tr class="header">
<th>Model</th>
<th>Example</th>
<th># predictors</th>
<th># coefs (df)</th>
<th>RSS</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Reduced</td>
<td>expression ~ age</td>
<td><span class="math inline">\(p_{Red}=1\)</span></td>
<td><span class="math inline">\(p_{Red}+1=2\)</span></td>
<td><span class="math inline">\(RSS_{Red}\)</span></td>
</tr>
<tr class="even">
<td>Full</td>
<td>expression ~ age * genotype</td>
<td><span class="math inline">\(p_{Full}=3\)</span></td>
<td><span class="math inline">\(p_{Full}+1 = 4\)</span></td>
<td><span class="math inline">\(RSS_{Full}\)</span></td>
</tr>
</tbody>
</table>
<p><strong>Full:</strong> <span class="math inline">\(y_{ij} = \alpha_{0} + \tau_{KO}x_{ij, KO} + \tau_{Age} x_{ij, Age} + \tau_{KO:Age} x_{ij, KO}x_{ij, Age}\)</span></p>
<p><strong>Reduced:</strong> <span class="math inline">\(y_{ij} = \alpha_{0} + \tau_{Age} x_{ij, Age}\)</span></p>
<p>Under <span class="math inline">\(H_0:\)</span> the reduced model explains the same amount variation in the outcome as the full,</p>
<p><span class="math display">\[F = \frac{\frac{RSS_{Red}-RSS_{Full}}{p_{Full}-p_{Red}}}{\frac{RSS_{Full}}{n-p_{Full}-1}} \sim \text{  } F_{p_{Fill}-p_{Red},\text{ } n-{p_{Full}-1}}\]</span> A significant <em>F</em>-test means we reject the null; we have evidence that the full model explains significantly more variation in the outcome than the reduced.</p>
</section>
<section id="collinearity-confounding" class="slide level2">
<h2>Collinearity &amp; confounding</h2>
<ul>
<li>If there are problems in the experimental design, we may not be able to estimate effects of interest</li>
</ul>
<div class="callout callout-caution callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Caution</strong></p>
</div>
<div class="callout-content">
<p>The technical definition of <strong>collinearity</strong> is that a column of the design matrix can be obtained (or <em>accurately approximated</em>) as a linear combination of other columns.</p>
<p>You can think of this as one column (or variable) not containing <em>unique information</em></p>
</div>
</div>
</div>
<ul>
<li><p>This phenomenon can occur if our design is confounded</p></li>
<li><p>As an example, let’s pretend we know that all the E16 and P2 mice are female and the rest are male</p></li>
</ul>
</section>
<section id="example-e16p2-mice-are-female-and-the-rest-are-male" class="slide level2">
<h2>Example: E16/P2 mice are female and the rest are male</h2>
<div class="panel-tabset">
<ul id="tabset-2" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-2-1">Dataset</a></li><li><a href="#tabset-2-2">Design matrix</a></li><li><a href="#tabset-2-3">Linear combination</a></li><li><a href="#tabset-2-4"><code>lm</code> output</a></li></ul>
<div class="tab-content">
<div id="tabset-2-1">
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>Klf9dat_conf <span class="ot">&lt;-</span> Klf9dat <span class="sc">%&gt;%</span></span>
<span id="cb19-2"><a href="#cb19-2"></a>  <span class="fu">mutate</span>(<span class="at">sex =</span> <span class="fu">ifelse</span>(dev_stage <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"E16"</span>, <span class="st">"P2"</span>), <span class="st">"female"</span>, <span class="st">"male"</span>))</span>
<span id="cb19-3"><a href="#cb19-3"></a><span class="fu">table</span>(Klf9dat_conf<span class="sc">$</span>sex, Klf9dat_conf<span class="sc">$</span>dev_stage)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        
         E16 P2 P6 P10 P28
  female   7  8  0   0   0
  male     0  0  8   8   8</code></pre>
</div>
</div>
</div>
<div id="tabset-2-2">
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a>(mm <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>( <span class="sc">~</span> sex <span class="sc">+</span> dev_stage, <span class="at">data =</span> Klf9dat_conf))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   (Intercept) sexmale dev_stageP2 dev_stageP6 dev_stageP10 dev_stageP28
1            1       1           0           0            0            1
2            1       1           0           0            0            1
3            1       1           0           0            0            1
4            1       1           0           0            0            1
5            1       0           0           0            0            0
6            1       0           0           0            0            0
7            1       0           0           0            0            0
8            1       1           0           0            1            0
9            1       1           0           0            1            0
10           1       1           0           0            1            0
11           1       1           0           0            1            0
12           1       0           1           0            0            0
13           1       0           1           0            0            0
14           1       0           1           0            0            0
15           1       0           1           0            0            0
16           1       1           0           1            0            0
17           1       1           0           1            0            0
18           1       1           0           1            0            0
19           1       1           0           1            0            0
20           1       1           0           0            0            1
21           1       1           0           0            0            1
22           1       1           0           0            0            1
23           1       1           0           0            0            1
24           1       0           0           0            0            0
25           1       0           0           0            0            0
26           1       0           0           0            0            0
27           1       0           0           0            0            0
28           1       1           0           0            1            0
29           1       1           0           0            1            0
30           1       1           0           0            1            0
31           1       1           0           0            1            0
32           1       0           1           0            0            0
33           1       0           1           0            0            0
34           1       0           1           0            0            0
35           1       0           1           0            0            0
36           1       1           0           1            0            0
37           1       1           0           1            0            0
38           1       1           0           1            0            0
39           1       1           0           1            0            0
attr(,"assign")
[1] 0 1 2 2 2 2
attr(,"contrasts")
attr(,"contrasts")$sex
[1] "contr.treatment"

attr(,"contrasts")$dev_stage
[1] "contr.treatment"</code></pre>
</div>
</div>
</div>
<div id="tabset-2-3">
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a><span class="co"># pull out column corresponding to male sex indicator variable</span></span>
<span id="cb23-2"><a href="#cb23-2"></a>mm[,<span class="st">"sexmale"</span>] </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 
 1  1  1  1  0  0  0  1  1  1  1  0  0  0  0  1  1  1  1  1  1  1  1  0  0  0 
27 28 29 30 31 32 33 34 35 36 37 38 39 
 0  1  1  1  1  0  0  0  0  1  1  1  1 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a><span class="co"># construct linear combination from columns corresponding to male timepoint indicator variables</span></span>
<span id="cb25-2"><a href="#cb25-2"></a>mm[,<span class="st">"dev_stageP6"</span>] <span class="sc">+</span> mm[,<span class="st">"dev_stageP10"</span>] <span class="sc">+</span> mm[,<span class="st">"dev_stageP28"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 
 1  1  1  1  0  0  0  1  1  1  1  0  0  0  0  1  1  1  1  1  1  1  1  0  0  0 
27 28 29 30 31 32 33 34 35 36 37 38 39 
 0  1  1  1  1  0  0  0  0  1  1  1  1 </code></pre>
</div>
</div>
</div>
<div id="tabset-2-4">
<p><code>lm</code> can’t estimate all the parameters in this collinear/confounded design.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a><span class="fu">lm</span>(expression <span class="sc">~</span> sex <span class="sc">+</span> dev_stage, <span class="at">data =</span> Klf9dat_conf) <span class="sc">%&gt;%</span></span>
<span id="cb27-2"><a href="#cb27-2"></a>  <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 5
  term         estimate std.error statistic   p.value
  &lt;chr&gt;           &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
1 (Intercept)     7.14      0.240    29.7    6.66e-26
2 sexmale         2.85      0.329     8.64   4.25e-10
3 dev_stageP2    -0.815     0.329    -2.48   1.84e- 2
4 dev_stageP6    -1.20      0.318    -3.76   6.44e- 4
5 dev_stageP10   -0.225     0.318    -0.707  4.84e- 1
6 dev_stageP28   NA        NA        NA     NA       </code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="hidden-confounding" class="slide level2">
<h2>Hidden confounding</h2>
<div class="columns">
<div class="column" style="width:30%;">
<div class="callout callout-caution callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Caution</strong></p>
</div>
<div class="callout-content">
<p>Hidden <strong>confounders</strong> are variables with associations to both the outcome and predictor variable(s), but are not measured (or not included in the model).</p>
<p>They can cause spurious correlations, and illustrate why <em>correlation does not imply causation</em>.</p>
</div>
</div>
</div>
</div><div class="column" style="width:70%;">
<div class="panel-tabset">
<ul id="tabset-3" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-3-1">Hidden</a></li><li><a href="#tabset-3-2">Revealed</a></li><li><a href="#tabset-3-3">Simpson’s paradox</a></li></ul>
<div class="tab-content">
<div id="tabset-3-1">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="lect08-continuous_files/figure-revealjs/unnamed-chunk-23-1.png" width="768"></p>
</div>
</div>
</div>
<div id="tabset-3-2">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="lect08-continuous_files/figure-revealjs/unnamed-chunk-24-1.png" width="921"></p>
</div>
</div>
</div>
<div id="tabset-3-3">
<p>Extreme example: trend reverses with/without confounder</p>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://raw.githubusercontent.com/simplystats/simplystats.github.io/master/_images/simpsons-paradox.gif" width="600"></p>
</figure>
</div>
<p><a href="https://simplystatistics.org/posts/2017-08-08-code-for-my-educational-gifs/">image source: Simply Statistics</a></p>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="linear-regression-summary" class="slide level2">
<h2>Linear regression summary</h2>
<ul>
<li><p>linear model framework is extremely general</p></li>
<li><p>one extreme (simple): two-sample common variance <em>t</em>-test</p></li>
<li><p>another extreme (flexible): a polynomial, potentially different for each level of some factor</p>
<ul>
<li><p>dichotomous predictor? 👍</p></li>
<li><p>categorical predictor? 👍</p></li>
<li><p>quantitative predictor? 👍</p></li>
<li><p>various combinations of the above? 👍</p></li>
</ul></li>
<li><p>Don’t be afraid to build models with more than 1 covariate</p></li>
<li><p>later, we’ll talk about extensions to discrete outcomes (e.g.&nbsp;dichotomous or counts) via <em>generalized linear models</em></p></li>
</ul>
</section>
<section id="what-about-the-other-45-thousand-probesets" class="slide level2">
<h2>What about the other 45 thousand probesets??</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a>eset</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ExpressionSet (storageMode: lockedEnvironment)
assayData: 45101 features, 39 samples 
  element names: exprs 
protocolData: none
phenoData
  sampleNames: GSM92610 GSM92611 ... GSM92648 (39 total)
  varLabels: title geo_accession ... age (40 total)
  varMetadata: labelDescription
featureData: none
experimentData: use 'experimentData(object)'
  pubMedIds: 16505381 
Annotation: GPL1261 </code></pre>
</div>
</div>
</section>
<section id="linear-regression-of-many-genes" class="slide level2">
<h2>Linear regression of many genes</h2>
<p><span class="math display">\[\Large \mathbf{Y}_g = \mathbf{X}_g \boldsymbol\alpha_g + \boldsymbol\varepsilon_g\]</span></p>
<ul>
<li><p>The g in the subscript reminds us that we’ll be fitting a model like this <em>for each gene g</em> that we have measured for all samples</p></li>
<li><p>Most of the time, the design matrices <span class="math inline">\(\mathbf{X}_g\)</span> are, in fact, the same for all <span class="math inline">\(g\)</span>. This means we can just use <span class="math inline">\(\mathbf{X}\)</span></p></li>
<li><p>Note this means the residual degrees of freedom are also the same for all <span class="math inline">\(g\)</span> <span class="math display">\[d_g = d = n - \text{dimension of } \boldsymbol\alpha = n-p-1\]</span></p></li>
</ul>
</section>
<section id="linear-regression-of-many-genes-contd" class="slide level2">
<h2>Linear regression of many genes (cont’d)</h2>
<ul>
<li><p>Data model: <span class="math inline">\(\large\mathbf{Y}_g = \mathbf{X} \boldsymbol\alpha_g + \boldsymbol\varepsilon_g\)</span></p></li>
<li><p>Unknown error variance: <span class="math inline">\(\large Var(\boldsymbol\varepsilon_g) = \sigma^2_g\)</span></p></li>
<li><p>Estimated error variance: <span class="math inline">\(\large\hat{\sigma}^2_g = s^2_g = \frac{1}{n-p-1}\hat{\boldsymbol\varepsilon_g}^T\hat{\boldsymbol\varepsilon_g}\)</span></p></li>
<li><p>Estimated variance of parameter estimates: <span class="math inline">\(\large\hat{Var}(\hat{\boldsymbol\alpha_g}) = s^2_g (\mathbf{X}^T\mathbf{X})^{-1} = s^2_g \mathbf{V}\)</span></p>
<ul>
<li><p><span class="math inline">\(\bf{V}\)</span> is the “unscaled covariance” matrix, and is the same for all genes!</p></li>
<li><p>Estimated standard errors for estimated regression coefficients: <span class="math inline">\(\large\hat{se}(\hat{\alpha}_{jg})\)</span> obtained by taking the square root of the <span class="math inline">\(j^{th}\)</span> diagonal element of <span class="math inline">\(\hat{Var}(\hat{\boldsymbol\alpha}_g)\)</span>, which is <span class="math inline">\(s_g\sqrt{v_{jj}}\)</span></p></li>
</ul></li>
</ul>
</section>
<section id="whats-the-big-deal" class="slide level2">
<h2>What’s the big deal?</h2>
<p>So far, nothing is new - these are the “regular” <em>t</em> statistics for gene <em>g</em> and parameter <em>j</em>:</p>
<p><span class="math display">\[t_{gj} = \frac{\hat{\alpha}_{gj}}{s_g \sqrt{v_{jj}}} \sim t_{d} \text{ under } H_0\]</span></p>
<p>But there are <em>so</em> many of them!! 😲</p>
</section>
<section id="observed-i.e.-empirical-issues-with-the-standard-t-test-approach-for-assessing-differential-expression" class="slide level2">
<h2>Observed (i.e.&nbsp;empirical) issues with the “standard” <em>t</em>-test approach for assessing differential expression</h2>
<div class="columns">
<div class="column" style="width:70%;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/Volcano_eg.jpg" width="800"></p>
</figure>
</div>
</div><div class="column" style="width:30%;">
<div class="callout callout-important callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Important</strong></p>
</div>
<div class="callout-content">
<p>Some genes with very small p-values (i.e.&nbsp;large -log10 p-values) are not <em>biologically meaningful</em> (small effect size, e.g.&nbsp;fold change)</p>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="how-do-we-end-up-with-small-p-values-but-subtle-effects" class="slide level2">
<h2>How do we end up with small p-values but subtle effects?</h2>
<p><span class="math display">\[\large t_{gj} = \frac{\hat{\alpha}_{gj}}{\hat{se}(\hat{\alpha}_{gj})} = \frac{\hat{\alpha}_{gj}}{s_g \sqrt{v_{jj}}} \sim t_d \text{ under } H_0\]</span></p>
<div class="fragment">
<ul>
<li>Small variance estimate <span class="math inline">\(s_g\)</span> leads to large <em>t</em> statistic <span class="math inline">\(\rightarrow\)</span> small <em>p</em>-value</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>Recall: estimates of sample variance from small sample sizes tend to under-estimate the true variance!</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>This has led to the development of specialized methodology for assessing genome-wide differential expression</li>
</ul>
</div>
</section>
<section id="empirical-bayesian-techniques-limma" class="slide level2">
<h2>Empirical Bayesian techniques: <code>limma</code></h2>
<div class="columns">
<div class="column">
<p><img data-src="img/limma_paper.png"> <a href="http://www.statsci.org/smyth/pubs/ebayes.pdf">Smyth 2004</a></p>
</div><div class="column">
<p><img data-src="img/limma_bioc.png"></p>
</div>
</div>
</section>
<section id="why-use-limma-instead-of-regular-t-tests" class="slide level2">
<h2>Why use <code>limma</code> instead of regular <em>t</em>-tests?</h2>
<div class="columns">
<div class="column" style="width:25%;">
<p><img data-src="https://raw.githubusercontent.com/Bioconductor/BiocStickers/master/limma/limma.png"></p>
</div><div class="column" style="width:75%;">
<ul>
<li><p><strong>Borrows information</strong> from all genes to get a better estimate of the variance (especially in smaller sample size settings)</p></li>
<li><p>Efficiently fits many regression models <strong>without replicating unnecessary calculations</strong>!</p></li>
<li><p>Arranges output in a convenient way to ease further analysis, visualization, and interpretation</p></li>
</ul>
</div>
</div>
</section>
<section id="how-does-empirical-bayes-work" class="slide level2">
<h2>How does Empirical Bayes work?</h2>
<div class="columns">
<div class="column">
<ul>
<li><p><strong>Empirical</strong>: observed</p></li>
<li><p><strong>Bayesian</strong>: incorporate ‘prior’ information</p></li>
<li><p>Intuition: estimate prior information from data; <em>shrink</em> (nudge) all estimates toward the consensus</p></li>
</ul>
</div><div class="column">
<p><strong>Shrinkage = borrowing information across all genes</strong></p>
<p><img data-src="img/shrink.jpg" style="width:100.0%"></p>
</div>
</div>
</section>
<section id="genome-wide-ols-fits" class="slide level2">
<h2>Genome-wide OLS fits</h2>
<ul>
<li><p>Gene by gene:</p>
<ul>
<li><p><code>lm(y ~ x, data = gene)</code> for each gene</p></li>
<li><p>For example, using <code>dplyr::group_modify</code> and <code>broom::tidy</code></p></li>
</ul></li>
</ul>
<div class="fragment">
<ul>
<li><p>All genes at once, using <code>limma</code>:</p>
<ul>
<li><p><code>lmFit(object, design)</code></p></li>
<li><p><code>object</code> matrix-like object with expression values for all genes</p></li>
<li><p><code>design</code> is a specially formatted design matrix (more on this later)</p></li>
<li><p>Note that <code>object</code> can be a Bioconductor container, such as an <a href="https://kasperdanielhansen.github.io/genbioconductor/html/ExpressionSet.html"><code>ExpressionSet</code></a> object</p></li>
</ul></li>
</ul>
</div>
</section>
<section id="industrial-scale-model-fitting-is-good-because-computations-involving-just-the-design-matrix-mathbfx-are-not-repeated-tens-of-thousands-of-times-unnecessarily" class="slide level2">
<h2>‘Industrial scale’ model fitting is good, because computations involving just the design matrix <span class="math inline">\(\mathbf{X}\)</span> are not repeated tens of thousands of times unnecessarily:</h2>
<ul>
<li><p><strong>OLS estimator</strong>: <span class="math display">\[\hat{\boldsymbol\alpha}_g =(\mathbf{X}^T\mathbf{X})^{-1}\mathbf{X}^T\mathbf{y}_g\]</span></p></li>
<li><p><strong>Fitted/predicted values</strong>: <span class="math display">\[\hat{\mathbf{y}}_g = \mathbf{X} (\mathbf{X}^T\mathbf{X})^{-1}\mathbf{X}^T\mathbf{y}_g = \mathbf{H}\mathbf{y}_g\]</span></p></li>
</ul>
</section>
<section id="ols-of-first-2000-genes-using-lm-gene-by-gene" class="slide level2">
<h2>OLS of first 2000 genes, using <code>lm</code> gene by gene</h2>
<div class="columns">
<div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a>allGenes <span class="ot">&lt;-</span> <span class="fu">toLongerMeta</span>(eset) </span>
<span id="cb31-2"><a href="#cb31-2"></a>allGenes <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 6
   gene       sample_id expression dev_stage   age genotype
   &lt;chr&gt;      &lt;chr&gt;          &lt;dbl&gt; &lt;fct&gt;     &lt;dbl&gt; &lt;fct&gt;   
 1 1415670_at GSM92610        7.11 P28          28 NrlKO   
 2 1415670_at GSM92611        7.32 P28          28 NrlKO   
 3 1415670_at GSM92612        7.42 P28          28 NrlKO   
 4 1415670_at GSM92613        7.35 P28          28 NrlKO   
 5 1415670_at GSM92614        7.24 E16          -4 NrlKO   
 6 1415670_at GSM92615        7.34 E16          -4 NrlKO   
 7 1415670_at GSM92616        7.38 E16          -4 NrlKO   
 8 1415670_at GSM92617        7.22 P10          10 NrlKO   
 9 1415670_at GSM92618        7.22 P10          10 NrlKO   
10 1415670_at GSM92619        7.12 P10          10 NrlKO   </code></pre>
</div>
</div>
</div><div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1"></a>(t.ols <span class="ot">&lt;-</span> <span class="fu">system.time</span>(lmfits <span class="ot">&lt;-</span> allGenes <span class="sc">%&gt;%</span></span>
<span id="cb33-2"><a href="#cb33-2"></a>    <span class="fu">filter</span>(gene <span class="sc">%in%</span> <span class="fu">unique</span>(allGenes<span class="sc">$</span>gene)[<span class="dv">1</span><span class="sc">:</span><span class="dv">2000</span>]) <span class="sc">%&gt;%</span></span>
<span id="cb33-3"><a href="#cb33-3"></a>    <span class="fu">group_by</span>(gene) <span class="sc">%&gt;%</span> </span>
<span id="cb33-4"><a href="#cb33-4"></a>    <span class="fu">group_modify</span>(<span class="sc">~</span> <span class="fu">tidy</span>(<span class="fu">lm</span>(expression <span class="sc">~</span> age <span class="sc">+</span> genotype, </span>
<span id="cb33-5"><a href="#cb33-5"></a>                           <span class="at">data =</span> .x))) <span class="sc">%&gt;%</span></span>
<span id="cb33-6"><a href="#cb33-6"></a>    <span class="fu">select</span>(gene, term, estimate) <span class="sc">%&gt;%</span></span>
<span id="cb33-7"><a href="#cb33-7"></a>    <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> term, </span>
<span id="cb33-8"><a href="#cb33-8"></a>                <span class="at">values_from =</span> estimate)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   user  system elapsed 
  1.240   0.007   1.248 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1"></a>lmfits <span class="sc">%&gt;%</span> <span class="fu">head</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
# Groups:   gene [6]
  gene         `(Intercept)`       age genotypeNrlKO
  &lt;chr&gt;                &lt;dbl&gt;     &lt;dbl&gt;         &lt;dbl&gt;
1 1415670_at            7.22  0.000623     -0.000286
2 1415671_at            9.32 -0.00184       0.145   
3 1415672_at            9.76 -0.00393      -0.0422  
4 1415673_at            8.40  0.00398      -0.0436  
5 1415674_a_at          8.52 -0.00598       0.0192  
6 1415675_at            9.67 -0.00642       0.133   </code></pre>
</div>
</div>
</div>
</div>
</section>
<section id="ols-of-all-genes-at-once-using-limma" class="slide level2">
<h2>OLS of <strong>all</strong> genes at once, using <code>limma</code>:</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1"></a>(t.limma <span class="ot">&lt;-</span> <span class="fu">system.time</span>( limmafits <span class="ot">&lt;-</span> </span>
<span id="cb37-2"><a href="#cb37-2"></a>  <span class="fu">lmFit</span>(eset, <span class="fu">model.matrix</span>(<span class="sc">~</span> age <span class="sc">+</span> genotype, <span class="at">data =</span> <span class="fu">pData</span>(eset)))))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   user  system elapsed 
  0.034   0.002   0.036 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1"></a>limmafits<span class="sc">$</span>coefficients <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             (Intercept)           age genotypeNrlKO
1415670_at      7.217851  0.0006225228 -0.0002861005
1415671_at      9.320083 -0.0018405479  0.1446053811
1415672_at      9.759959 -0.0039281143 -0.0421705559
1415673_at      8.404053  0.0039777804 -0.0436443351
1415674_a_at    8.517675 -0.0059840405  0.0192159017
1415675_at      9.665691 -0.0064185855  0.1330272055</code></pre>
</div>
</div>
<div class="fragment">
<p>So far, no shrinkage.</p>
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Tip</strong></p>
</div>
<div class="callout-content">
<p>Avoiding repetitive calculations involving <span class="math inline">\(\mathbf{X}\)</span> leads to ~930X faster computation in this example</p>
</div>
</div>
</div>
</div>
</section>
<section id="how-can-we-better-estimate-the-se" class="slide level2">
<h2>How can we better estimate the SE?</h2>
<p><span class="math display">\[\large t_{gj} = \frac{\hat{\alpha}_{gj}}{\hat{se}(\hat{\alpha}_{gj})} = \frac{\hat{\alpha}_{gj}}{s_g \sqrt{v_{jj}}} \sim t_d \text{ under } H_0\]</span></p>
<div class="callout callout-important callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Important</strong></p>
</div>
<div class="callout-content">
<p>Under-estimated variance leads to overly large t statistic, which leads to artificially small p-value</p>
</div>
</div>
</div>
</section>
<section id="modeling-in-limma" class="slide level2">
<h2>Modeling in <code>limma</code></h2>
<p>limma assumes that for each gene <span class="math inline">\(g\)</span> we have the following sampling distributions of the coefficient estimates and sample variances:</p>
<p><span class="math display">\[\hat{\alpha}_{gj} \,|\,\alpha_{gj}, \sigma_g^2 \sim N(\alpha_{gj}, \sigma_g^2 v_{jj})\]</span> <span class="math display">\[s^2_g \,|\, \sigma_g^2 \sim \frac{\sigma_g^2}{d}\chi^2_d\]</span> These are the same as the usual assumptions about ordinary <span class="math inline">\(t\)</span>-statistics:</p>
<p><span class="math display">\[\large t_{gj} = \frac{\hat{\alpha}_{gj}}{\hat{se}(\hat{\alpha}_{gj})} = \frac{\hat{\alpha}_{gj}}{s_g \sqrt{v_{jj}}} \sim t_d \text{ under } H_0\]</span></p>
<p>So far, nothing new…</p>
</section>
<section id="modeling-in-limma---shrinkage" class="slide level2">
<h2>Modeling in <code>limma</code> - shrinkage</h2>
<ul>
<li>limma imposes a hierarchical model for how the gene-wise <span class="math inline">\(\alpha_{gj}\)</span> and <span class="math inline">\(\sigma^2_g\)</span> vary <strong>across genes</strong></li>
</ul>
<div class="callout callout-important callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Important</strong></p>
</div>
<div class="callout-content">
<p>Under the limma framework, we are no longer considering genes in isolation. We will leverage information across genes to obtain improved estimates.</p>
</div>
</div>
</div>
<ul>
<li><p>this is done by assuming a <strong>prior distribution</strong> for those quantities</p></li>
<li><p>Prior distribution for <strong>gene-specific variances</strong> <span class="math inline">\(\sigma^2_g\)</span>: inverse <span class="math inline">\(\chi^2\)</span> with mean <span class="math inline">\(s_0^2\)</span>, and <span class="math inline">\(d_0\)</span> df:</p></li>
</ul>
<p><span class="math display">\[\frac{1}{\sigma^2_g} \sim \frac{1}{d_0s_0^2} \chi^2_{d_0}\]</span></p>
<div class="fragment">
<ul>
<li>this should feel a bit funny compared to previous lectures… <span class="math inline">\(\sigma^2_g\)</span> is no longer a <strong>fixed</strong> quantity! (i.e.&nbsp;this is <strong>Bayesian</strong>)</li>
</ul>
</div>
</section>
<section id="how-does-this-help-us-better-estimate-the-variance" class="slide level2">
<h2>How does this help us better estimate the variance?</h2>
<ul>
<li><p>The <strong>posterior distribution</strong> is an updated version of the observed likelihood based on incorporating the prior information</p></li>
<li><p>The posterior mean for gene-specific variance:</p></li>
</ul>
<p><span class="math display">\[\tilde{s}^2_g = \frac{d_0s_0^2 + ds^2_g}{d_0 + d}\]</span></p>
<div class="fragment">
<ul>
<li>Rewrite as a weighted mean of the <em>prior</em> and the <em>observed</em> gene-specific variances:</li>
</ul>
<p><span class="math display">\[\tilde{s}^2_g = \frac{d_0}{d_0 + d}s_0^2 + \frac{d}{d_0 + d}s^2_g\]</span></p>
</div>
<div class="fragment">
<ul>
<li>More simply: “shrinking” the observed gene-specific variance towards the “typical” variance implied by the prior</li>
</ul>
</div>
</section>
<section id="illustration" class="slide level2">
<h2>Illustration</h2>

<img data-src="https://media.springernature.com/lw685/springer-static/image/art%3A10.1186%2Fs12859-019-3248-9/MediaObjects/12859_2019_3248_Fig5_HTML.png?as=webp" class="r-stretch quarto-figure-center"><p>Figure 5 from <a href="https://doi.org/10.1186/s12859-019-3248-9">https://doi.org/10.1186/s12859-019-3248-9</a></p>
</section>
<section id="moderated-t-statistic" class="slide level2">
<h2>Moderated <em>t</em> statistic</h2>
<ul>
<li>plug in this posterior mean estimate to obtain a ‘moderated’ <em>t</em> statistic:</li>
</ul>
<p><span class="math display">\[\large \tilde{t}_{gj} = \frac{\hat{\alpha}_{gj}}{\tilde{s}_g \sqrt{v_{jj}}}\]</span></p>
<ul>
<li><p>Under limma’s assumptions, we know the null distribution for <span class="math inline">\(\tilde{t}_{gj}\)</span> under <span class="math inline">\(H_0\)</span>: <span class="math display">\[\tilde{t}_{gj} \sim t_{d_0+d}\]</span></p></li>
<li><p>parameters from the prior <span class="math inline">\(d_0\)</span> and <span class="math inline">\(s_0^2\)</span> are estimated from the data</p></li>
</ul>
<div class="fragment">
<div class="columns">
<div class="column" style="width:80%;">
<ul>
<li>This is how limma is a <strong>hybrid</strong> of frequentist (<em>t</em>-statistic) and Bayesian (hierarchical model) approaches (i.e.&nbsp;empirical Bayes)</li>
</ul>
</div><div class="column" style="width:20%;">
<p><img data-src="img/hybrid.jpg"></p>
</div>
</div>
</div>
</section>
<section id="side-by-side-comparison-of-key-quantities-and-results" class="slide level2">
<h2>Side-by-side comparison of key quantities and results</h2>
<table style="width:100%;">
<colgroup>
<col style="width: 42%">
<col style="width: 28%">
<col style="width: 28%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th style="text-align: center;">OLS</th>
<th style="text-align: center;">limma</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Estimated gene-wise residual variance<sup>1</sup><sup>2</sup>:</td>
<td style="text-align: center;"><span class="math inline">\({s}_g^2 = \frac{1}{d} \hat{\boldsymbol\varepsilon}^T \hat{\boldsymbol\varepsilon}\)</span></td>
<td style="text-align: center;"><span class="math inline">\(\tilde{s}^2_g = \frac{d_0s_0^2 + ds^2_g}{d_0 + d}\)</span></td>
</tr>
<tr class="even">
<td><em>t</em>-statistic for <span class="math inline">\(H_0: \alpha_{gj} = 0\)</span>:</td>
<td style="text-align: center;"><span class="math inline">\({t}_{gj} = \frac{\hat{\alpha}_{gj}}{s_g \sqrt{v_{jj}}}\)</span></td>
<td style="text-align: center;"><span class="math inline">\(\tilde{t}_{gj} = \frac{\hat{\alpha}_{gj}}{\tilde{s}_g \sqrt{v_{jj}}}\)</span></td>
</tr>
<tr class="odd">
<td>Distribution of <em>t</em>-statistic under <span class="math inline">\(H_0\)</span>:</td>
<td style="text-align: center;"><span class="math inline">\({t}_{gj} \sim t_{d}\)</span></td>
<td style="text-align: center;"><span class="math inline">\(\tilde{t}_{gj} \sim t_{d_0+d}\)</span></td>
</tr>
</tbody>
</table>
<aside><ol class="aside-footnotes"><li id="fn1"><p>Not shown: estimation formulas for prior parameters <span class="math inline">\(d_0\)</span> and <span class="math inline">\(s_0^2\)</span></p></li><li id="fn2"><p>Here the degrees of freedom of our linear model are <span class="math inline">\(d = n - p - 1\)</span>, where <span class="math inline">\(n\)</span> is the sample size and <span class="math inline">\(p\)</span> is the number of covariates</p></li></ol></aside></section>
<section id="moderated-vs-traditional-tests" class="slide level2">
<h2>Moderated vs traditional tests</h2>
<ul>
<li><strong>moderated variances</strong> will be <em>“shrunk” toward the typical gene-wise variance</em>, relative to to raw sample residual variances</li>
</ul>
<div class="fragment">
<ul>
<li><strong>degrees of freedom</strong> for null distribution <em>increase</em> relative to default <span class="math inline">\((d \text{ vs } d_0 + d)\)</span>
<ul>
<li><span class="math inline">\(\rightarrow\)</span> makes it closer to a standard normal</li>
<li><span class="math inline">\(\rightarrow\)</span> makes tail probabilities (p-values) smaller</li>
<li><span class="math inline">\(\rightarrow\)</span> easier to reject the null</li>
</ul></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>overall, when all is well, limma will deliver statistical results that are <strong>more stable</strong> and <strong>more powerful</strong></li>
</ul>
</div>
</section>
<section id="preview-limma-workflow" class="slide level2">
<h2>Preview: <code>limma</code> workflow</h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/limma_flow.png" style="width:68.0%"></p>
</figure>
</div>
<p><img src="https://github.com/STAT540-UBC/stat540-ubc.github.io/raw/main/images/stat540-logo-s.png" class="slide-logo"></p>
<div class="footer footer-default">

</div>
</section>

    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="lect08-continuous_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="lect08-continuous_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="lect08-continuous_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="lect08-continuous_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="lect08-continuous_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="lect08-continuous_files/libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="lect08-continuous_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="lect08-continuous_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="lect08-continuous_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="lect08-continuous_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="lect08-continuous_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1600,

        height: 900,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>